{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a8d87909-472c-46de-8df9-986ba981be4e",
   "metadata": {},
   "source": [
    "# 待办事项"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbe949a6-f1ce-4730-b92a-b6bd1583fec2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-31T13:29:31.844149Z",
     "iopub.status.busy": "2022-10-31T13:29:31.843886Z",
     "iopub.status.idle": "2022-10-31T13:29:31.848102Z",
     "shell.execute_reply": "2022-10-31T13:29:31.847341Z",
     "shell.execute_reply.started": "2022-10-31T13:29:31.844128Z"
    }
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "834e5693-79fe-4e0d-b00f-927245ecfc4a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 依赖包和环境设置"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b880507-7d5a-432f-ba5f-5cacb058a8d4",
   "metadata": {},
   "source": [
    "import python packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef64d1d6-60ca-48f8-8be8-dfef3ce0136e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-07T12:53:44.079439Z",
     "iopub.status.busy": "2022-12-07T12:53:44.078181Z",
     "iopub.status.idle": "2022-12-07T12:53:44.084914Z",
     "shell.execute_reply": "2022-12-07T12:53:44.083611Z",
     "shell.execute_reply.started": "2022-12-07T12:53:44.079380Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# !pip install matplotlib matplotlib_venn pandas numpy seaborn pandarallel pybedtools plotly upsetplotly --upgrade\n",
    "# !pip install git+https://github.com/ponnhide/pyCircos.git --upgrade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "2ccbf30b-41ed-4ca8-b6d6-296308872dd3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-07T12:53:46.024488Z",
     "iopub.status.busy": "2022-12-07T12:53:46.023258Z",
     "iopub.status.idle": "2022-12-07T12:54:25.148912Z",
     "shell.execute_reply": "2022-12-07T12:54:25.147350Z",
     "shell.execute_reply.started": "2022-12-07T12:53:46.024424Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import collections\n",
    "import os\n",
    "from glob import glob\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2\n",
    "from pybedtools import BedTool\n",
    "from pycircos import (  # pip install git+https://github.com/ponnhide/pyCircos.git\n",
    "    Garc,\n",
    "    Gcircle,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "eeb30344-f693-4482-9a7b-46d9b07517d4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:22:59.522528Z",
     "iopub.status.busy": "2022-11-29T14:22:59.521666Z",
     "iopub.status.idle": "2022-11-29T14:22:59.583249Z",
     "shell.execute_reply": "2022-11-29T14:22:59.582372Z",
     "shell.execute_reply.started": "2022-11-29T14:22:59.522467Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 20 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "from pandarallel import pandarallel\n",
    "\n",
    "pd.set_option(\"max_colwidth\", 35)  # column最大宽度\n",
    "pd.set_option(\"display.width\", 200)  # dataframe宽度\n",
    "pd.set_option(\"display.max_columns\", None)  # column最大显示数\n",
    "pd.set_option(\"display.max_rows\", 50)  # row最大显示数\n",
    "pandarallel.initialize()  # 多线程设置，默认使用全部核心 nb_workers=24"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e8e66af-b297-4ee4-90eb-07ae2eec3b93",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-27T08:20:10.504290Z",
     "iopub.status.busy": "2022-10-27T08:20:10.504004Z",
     "iopub.status.idle": "2022-10-27T08:20:10.509189Z",
     "shell.execute_reply": "2022-10-27T08:20:10.508390Z",
     "shell.execute_reply.started": "2022-10-27T08:20:10.504271Z"
    }
   },
   "source": [
    "enable rpy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "2625a77e-0c70-494d-80e5-924b6b108702",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:23:09.960133Z",
     "iopub.status.busy": "2022-11-29T14:23:09.959321Z",
     "iopub.status.idle": "2022-11-29T14:23:11.486019Z",
     "shell.execute_reply": "2022-11-29T14:23:11.484888Z",
     "shell.execute_reply.started": "2022-11-29T14:23:09.960070Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# enables the gaR magic,  not neceasary  if you’ve already done this\n",
    "# %load_ext rpy2.ipython\n",
    "# %reload_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0bd5e40-c367-4566-af46-e7a64198cf2d",
   "metadata": {},
   "source": [
    "import R libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e98275bb",
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# install.packages('ggpubr')\n",
    "# install.packages('ggrepel')\n",
    "# install.packages('ggupset')\n",
    "# install.packages('UpSetR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "b771c8c3-77fd-4b5b-b194-2c69022fa58e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:45:52.336028Z",
     "iopub.status.busy": "2022-11-29T14:45:52.335591Z",
     "iopub.status.idle": "2022-11-29T14:45:52.432302Z",
     "shell.execute_reply": "2022-11-29T14:45:52.430919Z",
     "shell.execute_reply.started": "2022-11-29T14:45:52.336001Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# 可以再%%R后面放光标 cmd + i\n",
    "# library(tidyverse)\n",
    "# library(ggpubr)\n",
    "# library(ggrepel)\n",
    "# library(ggupset)\n",
    "# library(UpSetR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8e2e93e-1bb7-4304-9a41-226db2b6c142",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# QC 步骤\n",
    "查看测序和 Mapping 质量"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8b4cc92-2f21-4a47-a47c-c96b61564f7f",
   "metadata": {},
   "source": [
    "## 搜集 MultiQC 信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "91fc94d3-2a4d-49ab-830f-38aff5662be5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:45:55.859260Z",
     "iopub.status.busy": "2022-11-29T14:45:55.858762Z",
     "iopub.status.idle": "2022-11-29T14:45:57.241651Z",
     "shell.execute_reply": "2022-11-29T14:45:57.240861Z",
     "shell.execute_reply.started": "2022-11-29T14:45:55.859233Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>% Dups</th>\n",
       "      <th>% GC</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>% Failed</th>\n",
       "      <th>M Seqs &lt;raw fq&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>62.80</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>568.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>66.20</td>\n",
       "      <td>41.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>612.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>55.95</td>\n",
       "      <td>40.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>583.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-1</td>\n",
       "      <td>60.15</td>\n",
       "      <td>43.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>521.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-2</td>\n",
       "      <td>49.70</td>\n",
       "      <td>43.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>355.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>57.40</td>\n",
       "      <td>41.5</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>528.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>40.75</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>427.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test</td>\n",
       "      <td>8.75</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Sample Name  % Dups  % GC  Read Length  % Failed  M Seqs <raw fq>\n",
       "0   DetectSeq_ATP8-DddA11_REP-1   62.80  42.0        300.0      10.0            568.4\n",
       "1    DetectSeq_ATP8-DddA6_REP-1   66.20  41.0        300.0      10.0            612.8\n",
       "2   DetectSeq_ATP8-DddAwt_REP-1   55.95  40.0        300.0      10.0            583.0\n",
       "3   DetectSeq_JAK2-DddA11_REP-1   60.15  43.0        300.0      10.0            521.4\n",
       "4   DetectSeq_JAK2-DddA11_REP-2   49.70  43.0        300.0      15.0            355.8\n",
       "5  DetectSeq_SIRT6-DddA11_REP-1   57.40  41.5        300.0      10.0            528.0\n",
       "6  DetectSeq_SIRT6-DddA11_REP-2   40.75  42.0        300.0       5.0            427.0\n",
       "7                          test    8.75  42.0        300.0       0.0              0.0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_qc = pd.read_html('../qc/multiqc/multiqc_report.html')[0]\n",
    "# df_qc\n",
    "df_qc['Sample Name'] = df_qc['Sample Name'].str[:-3]\n",
    "df_qc['% Dups'] = df_qc['% Dups'].str[:-1].astype(float)\n",
    "df_qc['% GC'] = df_qc['% GC'].str[:-1].astype(float)\n",
    "df_qc['Read Length'] = df_qc['Read Length'].str[:-3].astype(int)\n",
    "df_qc['% Failed'] = df_qc['% Failed'].str[:-1].astype(float)\n",
    "df_qc = df_qc.groupby('Sample Name').agg(np.mean)\n",
    "df_qc['Read Length'] = df_qc['Read Length'] * 2\n",
    "df_qc['M Seqs'] = df_qc['M Seqs'] * 2\n",
    "df_qc.reset_index(inplace=True)\n",
    "df_qc.rename(columns={'M Seqs': 'M Seqs <raw fq>'}, inplace=True)\n",
    "df_qc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "034853aa-a79c-4904-9ff3-43b901ab087e",
   "metadata": {},
   "source": [
    "## 搜集 Hisat-3N Mapping 信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "db938818-b808-4c75-842d-d83ed6ccc4b0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:46:03.836896Z",
     "iopub.status.busy": "2022-11-29T14:46:03.836503Z",
     "iopub.status.idle": "2022-11-29T14:46:04.058108Z",
     "shell.execute_reply": "2022-11-29T14:46:04.057278Z",
     "shell.execute_reply.started": "2022-11-29T14:46:03.836865Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>% Hisat-3n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>66.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>71.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>79.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-1</td>\n",
       "      <td>61.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-2</td>\n",
       "      <td>61.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>69.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>74.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test</td>\n",
       "      <td>66.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Sample Name  % Hisat-3n\n",
       "0   DetectSeq_ATP8-DddA11_REP-1       66.90\n",
       "1    DetectSeq_ATP8-DddA6_REP-1       71.75\n",
       "2   DetectSeq_ATP8-DddAwt_REP-1       79.26\n",
       "3   DetectSeq_JAK2-DddA11_REP-1       61.50\n",
       "4   DetectSeq_JAK2-DddA11_REP-2       61.79\n",
       "5  DetectSeq_SIRT6-DddA11_REP-1       69.31\n",
       "6  DetectSeq_SIRT6-DddA11_REP-2       74.18\n",
       "7                          test       66.62"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls = sorted(glob('../bam/*_hisat3n.hisat3n.log'))\n",
    "tmpls = []\n",
    "\n",
    "for file in ls:\n",
    "    with open(file, 'rt') as f:\n",
    "        sname = file.split('/')[-1].replace('_hisat3n.hisat3n.log', '')\n",
    "        ratio = float(f.readlines()[-1].split('%')[0])\n",
    "        tmpls.append([sname, ratio])\n",
    "df_3n = pd.DataFrame(tmpls, columns=['Sample Name', '% Hisat-3n'])\n",
    "df_3n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa764f8-28d0-4e2d-8be4-28b3342189a4",
   "metadata": {},
   "source": [
    "## 搜集 Final Mapping 信息 （Hisat3N mapping + BWA remapping）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b48f8f9d-14db-443e-a087-067602acde31",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:46:06.044885Z",
     "iopub.status.busy": "2022-11-29T14:46:06.044499Z",
     "iopub.status.idle": "2022-11-29T14:46:08.297907Z",
     "shell.execute_reply": "2022-11-29T14:46:08.296663Z",
     "shell.execute_reply.started": "2022-11-29T14:46:06.044855Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>M Seqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>../bam/DetectSeq_ATP8-DddA11_RE...</td>\n",
       "      <td>176.744M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>../bam/DetectSeq_ATP8-DddA6_REP...</td>\n",
       "      <td>163.108M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>../bam/DetectSeq_ATP8-DddAwt_RE...</td>\n",
       "      <td>219.36M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>../bam/DetectSeq_JAK2-DddA11_RE...</td>\n",
       "      <td>163.946M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>../bam/DetectSeq_JAK2-DddA11_RE...</td>\n",
       "      <td>143.23M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>../bam/DetectSeq_SIRT6-DddA11_R...</td>\n",
       "      <td>180.379M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>../bam/DetectSeq_SIRT6-DddA11_R...</td>\n",
       "      <td>215.754M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>../bam/test_final_rmdup.bam</td>\n",
       "      <td>0.030129M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Sample Name     M Seqs\n",
       "0  ../bam/DetectSeq_ATP8-DddA11_RE...   176.744M\n",
       "1  ../bam/DetectSeq_ATP8-DddA6_REP...   163.108M\n",
       "2  ../bam/DetectSeq_ATP8-DddAwt_RE...    219.36M\n",
       "3  ../bam/DetectSeq_JAK2-DddA11_RE...   163.946M\n",
       "4  ../bam/DetectSeq_JAK2-DddA11_RE...    143.23M\n",
       "5  ../bam/DetectSeq_SIRT6-DddA11_R...   180.379M\n",
       "6  ../bam/DetectSeq_SIRT6-DddA11_R...   215.754M\n",
       "7         ../bam/test_final_rmdup.bam  0.030129M"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 法1：\n",
    "# ls = sorted(glob('../bam/*_final_rmdup.bam.flagstats.tsv'))\n",
    "# tmpls = []\n",
    "\n",
    "# for file in ls:\n",
    "#     with open(file, 'rt') as f:\n",
    "#         sname = file.split('/')[-1].replace('_final_rmdup.bam.flagstats.tsv', '')\n",
    "#         ratio = float(f.readlines()[6].split('\\t')[0])\n",
    "#         tmpls.append([sname, ratio])\n",
    "# df_final_map_flagstats = pd.DataFrame(tmpls, columns=['Sample Name', 'Seqs'])\n",
    "# df_final_map_flagstats['']\n",
    "# df_final_map_flagstats\n",
    "\n",
    "# 法2：\n",
    "text = \"\"\"\\\n",
    "touch ../bam/all_final_mapped_reads.txt\n",
    "rm ../bam/all_final_mapped_reads.txt\n",
    "\n",
    "for i in `ls ../bam/*_rmdup.bam`\n",
    "    samtools idxstats $i | \\\n",
    "    \tawk '{sum += $3} END {print sum/1000000 \"M reads\"}' | \\\n",
    "    \txargs echo \"$i\" \\\n",
    "    \t>> ../bam/all_final_mapped_reads.txt\n",
    "\"\"\"\n",
    "with open('../bam/all_final_mapped_reads.sh', 'wt') as f:\n",
    "    f.write(text)\n",
    "\n",
    "assert os.system('zsh ../bam/all_final_mapped_reads.sh') == 0\n",
    "\n",
    "\n",
    "df_final_map = pd.read_csv('../bam/all_final_mapped_reads.txt', sep=' ', header=None, names=['Sample Name', 'M Seqs', '_'], usecols=[0, 1])\n",
    "df_final_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "40c815ea-d434-4089-9246-2b88a3801406",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:46:08.313086Z",
     "iopub.status.busy": "2022-11-29T14:46:08.312836Z",
     "iopub.status.idle": "2022-11-29T14:46:08.326128Z",
     "shell.execute_reply": "2022-11-29T14:46:08.325033Z",
     "shell.execute_reply.started": "2022-11-29T14:46:08.313063Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_18829/498917256.py:1: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  df_final_map['Sample Name'] = df_final_map['Sample Name'].str.replace('../bam/', '').str.replace('_final_rmdup.bam', '')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>M Seqs &lt;final mapped rm dup&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>176.744M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>163.108M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>219.36M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-1</td>\n",
       "      <td>163.946M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-2</td>\n",
       "      <td>143.23M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>180.379M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>215.754M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test</td>\n",
       "      <td>0.030129M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Sample Name M Seqs <final mapped rm dup>\n",
       "0   DetectSeq_ATP8-DddA11_REP-1                     176.744M\n",
       "1    DetectSeq_ATP8-DddA6_REP-1                     163.108M\n",
       "2   DetectSeq_ATP8-DddAwt_REP-1                      219.36M\n",
       "3   DetectSeq_JAK2-DddA11_REP-1                     163.946M\n",
       "4   DetectSeq_JAK2-DddA11_REP-2                      143.23M\n",
       "5  DetectSeq_SIRT6-DddA11_REP-1                     180.379M\n",
       "6  DetectSeq_SIRT6-DddA11_REP-2                     215.754M\n",
       "7                          test                    0.030129M"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_map['Sample Name'] = df_final_map['Sample Name'].str.replace('../bam/', '').str.replace('_final_rmdup.bam', '')\n",
    "df_final_map.rename(columns={'M Seqs': 'M Seqs <final mapped rm dup>'}, inplace=True)\n",
    "df_final_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "880a2a61-34aa-442d-bfb3-db13b8f89502",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-24T17:17:55.775720Z",
     "iopub.status.busy": "2022-11-24T17:17:55.775475Z",
     "iopub.status.idle": "2022-11-24T17:17:55.785193Z",
     "shell.execute_reply": "2022-11-24T17:17:55.784625Z",
     "shell.execute_reply.started": "2022-11-24T17:17:55.775704Z"
    },
    "tags": []
   },
   "source": [
    "## 汇总 QC 信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "5701d3ca-d378-4af4-9833-7c6b63f56d54",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:46:15.154915Z",
     "iopub.status.busy": "2022-11-29T14:46:15.154366Z",
     "iopub.status.idle": "2022-11-29T14:46:15.389717Z",
     "shell.execute_reply": "2022-11-29T14:46:15.387935Z",
     "shell.execute_reply.started": "2022-11-29T14:46:15.154886Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>% Dups</th>\n",
       "      <th>% GC</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>% Failed</th>\n",
       "      <th>M Seqs &lt;raw fq&gt;</th>\n",
       "      <th>% Hisat-3n</th>\n",
       "      <th>M Seqs &lt;final mapped rm dup&gt;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>62.80</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>568.4</td>\n",
       "      <td>66.90</td>\n",
       "      <td>176.744M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>66.20</td>\n",
       "      <td>41.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>612.8</td>\n",
       "      <td>71.75</td>\n",
       "      <td>163.108M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>55.95</td>\n",
       "      <td>40.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>583.0</td>\n",
       "      <td>79.26</td>\n",
       "      <td>219.36M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-1</td>\n",
       "      <td>60.15</td>\n",
       "      <td>43.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>521.4</td>\n",
       "      <td>61.50</td>\n",
       "      <td>163.946M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-2</td>\n",
       "      <td>49.70</td>\n",
       "      <td>43.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>355.8</td>\n",
       "      <td>61.79</td>\n",
       "      <td>143.23M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>57.40</td>\n",
       "      <td>41.5</td>\n",
       "      <td>300.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>528.0</td>\n",
       "      <td>69.31</td>\n",
       "      <td>180.379M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>40.75</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>74.18</td>\n",
       "      <td>215.754M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test</td>\n",
       "      <td>8.75</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.62</td>\n",
       "      <td>0.030129M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Sample Name  % Dups  % GC  Read Length  % Failed  M Seqs <raw fq>  % Hisat-3n M Seqs <final mapped rm dup>\n",
       "0   DetectSeq_ATP8-DddA11_REP-1   62.80  42.0        300.0      10.0            568.4       66.90                     176.744M\n",
       "1    DetectSeq_ATP8-DddA6_REP-1   66.20  41.0        300.0      10.0            612.8       71.75                     163.108M\n",
       "2   DetectSeq_ATP8-DddAwt_REP-1   55.95  40.0        300.0      10.0            583.0       79.26                      219.36M\n",
       "3   DetectSeq_JAK2-DddA11_REP-1   60.15  43.0        300.0      10.0            521.4       61.50                     163.946M\n",
       "4   DetectSeq_JAK2-DddA11_REP-2   49.70  43.0        300.0      15.0            355.8       61.79                      143.23M\n",
       "5  DetectSeq_SIRT6-DddA11_REP-1   57.40  41.5        300.0      10.0            528.0       69.31                     180.379M\n",
       "6  DetectSeq_SIRT6-DddA11_REP-2   40.75  42.0        300.0       5.0            427.0       74.18                     215.754M\n",
       "7                          test    8.75  42.0        300.0       0.0              0.0       66.62                    0.030129M"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_qc_all = df_qc.merge(df_3n).merge(df_final_map)\n",
    "df_qc_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "e3307b6f-f790-4e6a-98a2-693f7b904390",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:46:18.222172Z",
     "iopub.status.busy": "2022-11-29T14:46:18.221747Z",
     "iopub.status.idle": "2022-11-29T14:46:18.248412Z",
     "shell.execute_reply": "2022-11-29T14:46:18.247615Z",
     "shell.execute_reply.started": "2022-11-29T14:46:18.222145Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>% Dups</th>\n",
       "      <th>% GC</th>\n",
       "      <th>Read Length</th>\n",
       "      <th>M Seqs &lt;raw fq&gt;</th>\n",
       "      <th>M Seqs &lt;Hisat-3n mapped&gt;</th>\n",
       "      <th>M Seqs &lt;final mapped rm dup&gt;</th>\n",
       "      <th>% Effective Seqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>62.80</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300</td>\n",
       "      <td>568.4</td>\n",
       "      <td>380.26</td>\n",
       "      <td>176.74</td>\n",
       "      <td>31.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>66.20</td>\n",
       "      <td>41.0</td>\n",
       "      <td>300</td>\n",
       "      <td>612.8</td>\n",
       "      <td>439.68</td>\n",
       "      <td>163.11</td>\n",
       "      <td>26.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>55.95</td>\n",
       "      <td>40.0</td>\n",
       "      <td>300</td>\n",
       "      <td>583.0</td>\n",
       "      <td>462.09</td>\n",
       "      <td>219.36</td>\n",
       "      <td>37.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-1</td>\n",
       "      <td>60.15</td>\n",
       "      <td>43.0</td>\n",
       "      <td>300</td>\n",
       "      <td>521.4</td>\n",
       "      <td>320.66</td>\n",
       "      <td>163.95</td>\n",
       "      <td>31.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_JAK2-DddA11_REP-2</td>\n",
       "      <td>49.70</td>\n",
       "      <td>43.0</td>\n",
       "      <td>300</td>\n",
       "      <td>355.8</td>\n",
       "      <td>219.85</td>\n",
       "      <td>143.23</td>\n",
       "      <td>40.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>57.40</td>\n",
       "      <td>41.5</td>\n",
       "      <td>300</td>\n",
       "      <td>528.0</td>\n",
       "      <td>365.96</td>\n",
       "      <td>180.38</td>\n",
       "      <td>34.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>40.75</td>\n",
       "      <td>42.0</td>\n",
       "      <td>300</td>\n",
       "      <td>427.0</td>\n",
       "      <td>316.75</td>\n",
       "      <td>215.75</td>\n",
       "      <td>50.53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Sample Name  % Dups  % GC  Read Length  M Seqs <raw fq>  M Seqs <Hisat-3n mapped>  M Seqs <final mapped rm dup>  % Effective Seqs\n",
       "0   DetectSeq_ATP8-DddA11_REP-1   62.80  42.0          300            568.4                    380.26                        176.74             31.09\n",
       "1    DetectSeq_ATP8-DddA6_REP-1   66.20  41.0          300            612.8                    439.68                        163.11             26.62\n",
       "2   DetectSeq_ATP8-DddAwt_REP-1   55.95  40.0          300            583.0                    462.09                        219.36             37.63\n",
       "3   DetectSeq_JAK2-DddA11_REP-1   60.15  43.0          300            521.4                    320.66                        163.95             31.44\n",
       "4   DetectSeq_JAK2-DddA11_REP-2   49.70  43.0          300            355.8                    219.85                        143.23             40.26\n",
       "5  DetectSeq_SIRT6-DddA11_REP-1   57.40  41.5          300            528.0                    365.96                        180.38             34.16\n",
       "6  DetectSeq_SIRT6-DddA11_REP-2   40.75  42.0          300            427.0                    316.75                        215.75             50.53"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_qc_all['Read Length'] = df_qc_all['Read Length'].astype(int)\n",
    "df_qc_all['M Seqs <Hisat-3n mapped>'] = df_qc_all['M Seqs <raw fq>'] * df_qc_all['% Hisat-3n'] / 100\n",
    "df_qc_all['M Seqs <Hisat-3n mapped>'] = df_qc_all['M Seqs <Hisat-3n mapped>'].map(lambda x: round(x, 2))\n",
    "df_qc_all['M Seqs <final mapped rm dup>'] = df_qc_all['M Seqs <final mapped rm dup>'].str[:-1].astype(float)\n",
    "df_qc_all['M Seqs <final mapped rm dup>'] = df_qc_all['M Seqs <final mapped rm dup>'].map(lambda x: round(x, 2))\n",
    "df_qc_all = df_qc_all.drop(columns=['% Failed', '% Hisat-3n'])\n",
    "df_qc_all = df_qc_all.iloc[:, [0, 1, 2, 3, 4, 6, 5]].copy()\n",
    "df_qc_all = df_qc_all[df_qc_all['Sample Name'] != 'test'].copy()\n",
    "df_qc_all['% Effective Seqs'] = df_qc_all['M Seqs <final mapped rm dup>'] / df_qc_all['M Seqs <raw fq>'] * 100\n",
    "df_qc_all['% Effective Seqs'] = df_qc_all['% Effective Seqs'].map(lambda x: round(x, 2))\n",
    "df_qc_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2261ff8-5150-4993-849d-0efe514f9e76",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# [TEST] 测试 call 点\n",
    "搞一个老的 TAS-independent 的 mpmat 文件\n",
    "\n",
    "generate old TAS-independent off-target mpmat to call regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "613f27e9-52a2-4d35-a3c2-e83b1ea6587f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:49:50.099951Z",
     "iopub.status.busy": "2022-11-29T14:49:50.099525Z",
     "iopub.status.idle": "2022-11-29T14:49:50.231617Z",
     "shell.execute_reply": "2022-11-29T14:49:50.230101Z",
     "shell.execute_reply.started": "2022-11-29T14:49:50.099923Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>id_ND4</th>\n",
       "      <th>id_ND5.1</th>\n",
       "      <th>id_ND6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_1471366_1471410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_1693068_1693084</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_2044988_2044998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_5806517_5806533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_5983069_5983085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            mpmat_index id_ND4         id_ND5.1 id_ND6\n",
       "0  chr1_1471366_1471410    NaN              NaN    NaN\n",
       "1  chr1_1693068_1693084    NaN              NaN    NaN\n",
       "2  chr1_2044988_2044998    NaN  ND5.1-TAS.IND-1    NaN\n",
       "3  chr1_5806517_5806533    NaN              NaN    NaN\n",
       "4  chr1_5983069_5983085    NaN              NaN    NaN"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_final_list = pd.read_csv(\n",
    "    \"../tables/20220312-DdCBE-off_target_type.FinallistV4.CheckPrimer.AddV4ID.tsv\",\n",
    "    sep='\\t',\n",
    "    header=0,\n",
    "    usecols=['region_id', 'off_target_id.V4.ND4', 'off_target_id.V4.ND5.1', 'off_target_id.V4.ND6'],\n",
    ")\n",
    "df_old_final_list.columns = ['mpmat_index', 'id_ND4', 'id_ND5.1', 'id_ND6']\n",
    "df_old_final_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "1795a1cc-3fcf-4bfa-9d47-c873da3e8138",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:49:53.273375Z",
     "iopub.status.busy": "2022-11-29T14:49:53.272749Z",
     "iopub.status.idle": "2022-11-29T14:49:54.927897Z",
     "shell.execute_reply": "2022-11-29T14:49:54.926736Z",
     "shell.execute_reply.started": "2022-11-29T14:49:53.273347Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "650"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def query_ind(x):\n",
    "    x = x.fillna('')\n",
    "    if 'IND' in x['id_ND4']:\n",
    "        return True\n",
    "    elif 'IND' in x['id_ND5.1']:\n",
    "        return True\n",
    "    elif 'IND' in x['id_ND6']:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "\n",
    "df_old_final_list.apply(query_ind, axis=1).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d160c059-b463-4f2a-a1ef-002732b9dfa8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:49:59.092892Z",
     "iopub.status.busy": "2022-11-29T14:49:59.092207Z",
     "iopub.status.idle": "2022-11-29T14:50:00.832515Z",
     "shell.execute_reply": "2022-11-29T14:50:00.831610Z",
     "shell.execute_reply.started": "2022-11-29T14:49:59.092836Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>id_ND4</th>\n",
       "      <th>id_ND5.1</th>\n",
       "      <th>id_ND6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_2044988_2044998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_9702414_9702449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-2</td>\n",
       "      <td>ND6-TAS.IND-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_12618165_12618174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_13019998_13020019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_13354705_13354738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>chrX_150683735_150683758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-452</td>\n",
       "      <td>ND6-TAS.IND-540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-453</td>\n",
       "      <td>ND6-TAS.IND-541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-454</td>\n",
       "      <td>ND6-TAS.IND-542</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>650 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mpmat_index id_ND4           id_ND5.1           id_ND6\n",
       "0        chr1_2044988_2044998    NaN    ND5.1-TAS.IND-1              NaN\n",
       "1        chr1_9702414_9702449    NaN    ND5.1-TAS.IND-2    ND6-TAS.IND-1\n",
       "2      chr1_12618165_12618174    NaN    ND5.1-TAS.IND-3              NaN\n",
       "3      chr1_13019998_13020019    NaN                NaN    ND6-TAS.IND-2\n",
       "4      chr1_13354705_13354738    NaN                NaN    ND6-TAS.IND-3\n",
       "..                        ...    ...                ...              ...\n",
       "645  chrX_136925330_136925338    NaN                NaN  ND6-TAS.IND-538\n",
       "646  chrX_138128211_138128237    NaN                NaN  ND6-TAS.IND-539\n",
       "647  chrX_150683735_150683758    NaN  ND5.1-TAS.IND-452  ND6-TAS.IND-540\n",
       "648  chrX_153669436_153669453    NaN  ND5.1-TAS.IND-453  ND6-TAS.IND-541\n",
       "649  chrX_155049205_155049215    NaN  ND5.1-TAS.IND-454  ND6-TAS.IND-542\n",
       "\n",
       "[650 rows x 4 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_share = df_old_final_list[df_old_final_list.apply(query_ind, axis=1)].reset_index(drop=True)\n",
    "df_old_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "fa1e6097-5822-44ae-9dec-c7c72aee1e35",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:50:00.859278Z",
     "iopub.status.busy": "2022-11-29T14:50:00.858935Z",
     "iopub.status.idle": "2022-11-29T14:50:01.771490Z",
     "shell.execute_reply": "2022-11-29T14:50:01.770439Z",
     "shell.execute_reply.started": "2022-11-29T14:50:00.859250Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>mpmat_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>48137</td>\n",
       "      <td>48144</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_48137_CT,chr1_48139_CT,chr...</td>\n",
       "      <td>4,2,4</td>\n",
       "      <td>19,19,19</td>\n",
       "      <td>0.21053,0.10526,0.21053</td>\n",
       "      <td>False,False,False</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>Pass,Filter,Pass</td>\n",
       "      <td>chr1_48137_48144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>127550</td>\n",
       "      <td>127558</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_127550_CT,chr1_127558_CT</td>\n",
       "      <td>5,5</td>\n",
       "      <td>22,21</td>\n",
       "      <td>0.22727,0.2381</td>\n",
       "      <td>False,False</td>\n",
       "      <td>0,0</td>\n",
       "      <td>Pass,Pass</td>\n",
       "      <td>chr1_127550_127558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>261533</td>\n",
       "      <td>261543</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_261533_GA,chr1_261543_GA</td>\n",
       "      <td>4,4</td>\n",
       "      <td>14,17</td>\n",
       "      <td>0.28571,0.23529</td>\n",
       "      <td>False,False</td>\n",
       "      <td>0,0</td>\n",
       "      <td>Pass,Pass</td>\n",
       "      <td>chr1_261533_261543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629090</td>\n",
       "      <td>629115</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_629090_CT,chr1_629091_CT,c...</td>\n",
       "      <td>27,3,5,23,14,8,24,2,90</td>\n",
       "      <td>716,807,1057,1448,1633,1884,209...</td>\n",
       "      <td>0.03771,0.00372,0.00473,0.01588...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Filter,Filter,Pass,Filter,...</td>\n",
       "      <td>chr1_629090_629115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629136</td>\n",
       "      <td>629148</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_629136_CT,chr1_629138_CT,c...</td>\n",
       "      <td>100,10,19,0,9,68</td>\n",
       "      <td>3579,3571,3591,3604,3614,3617</td>\n",
       "      <td>0.02794,0.0028,0.00529,0.0,0.00...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0</td>\n",
       "      <td>Pass,Filter,Filter,Filter,Filte...</td>\n",
       "      <td>chr1_629136_629148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103310</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16218</td>\n",
       "      <td>16261</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM_16218_CT,chrM_16221_CT,chr...</td>\n",
       "      <td>142,102,109,61,179,94,147,120,1...</td>\n",
       "      <td>7817,7756,7847,7891,7860,7832,7...</td>\n",
       "      <td>0.01817,0.01315,0.01389,0.00773...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0...</td>\n",
       "      <td>Pass,Pass,Pass,Filter,Pass,Pass...</td>\n",
       "      <td>chrM_16218_16261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103311</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16255</td>\n",
       "      <td>16274</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>chrM_16255_GA,chrM_16273_GA,chr...</td>\n",
       "      <td>110,145,337</td>\n",
       "      <td>7536,7620,7678</td>\n",
       "      <td>0.0146,0.01903,0.04389</td>\n",
       "      <td>False,False,False</td>\n",
       "      <td>0,0,0</td>\n",
       "      <td>Pass,Pass,Pass</td>\n",
       "      <td>chrM_16255_16274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103312</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16259</td>\n",
       "      <td>16279</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>chrM_16259_CT,chrM_16260_CT,chr...</td>\n",
       "      <td>104,85,87,58,129,107,93,103,712...</td>\n",
       "      <td>7715,7730,7801,7811,7836,7758,7...</td>\n",
       "      <td>0.01348,0.011,0.01115,0.00743,0...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Filter,Pass,Pass...</td>\n",
       "      <td>chrM_16259_16279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103313</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16329</td>\n",
       "      <td>16373</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>chrM_16329_GA,chrM_16336_GA,chr...</td>\n",
       "      <td>131,147,246,421,466,656,106</td>\n",
       "      <td>7673,7755,7746,7663,7797,7764,7859</td>\n",
       "      <td>0.01707,0.01896,0.03176,0.05494...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass,Pass,Pass</td>\n",
       "      <td>chrM_16329_16373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103314</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16470</td>\n",
       "      <td>16558</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>chrM_16470_GA,chrM_16471_GA,chr...</td>\n",
       "      <td>154,130,150,138,138,202,229,294...</td>\n",
       "      <td>7223,7148,7187,7190,7127,7119,7...</td>\n",
       "      <td>0.02132,0.01819,0.02087,0.01919...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0...</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass,Pass,P...</td>\n",
       "      <td>chrM_16470_16558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103315 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0       1       2   3   4  5                                   6                                   7                                   8                                   9  \\\n",
       "0       chr1   48137   48144   3   2  0  chr1_48137_CT,chr1_48139_CT,chr...                               4,2,4                            19,19,19             0.21053,0.10526,0.21053   \n",
       "1       chr1  127550  127558   2   2  0       chr1_127550_CT,chr1_127558_CT                                 5,5                               22,21                      0.22727,0.2381   \n",
       "2       chr1  261533  261543   2   2  0       chr1_261533_GA,chr1_261543_GA                                 4,4                               14,17                     0.28571,0.23529   \n",
       "3       chr1  629090  629115   9   4  0  chr1_629090_CT,chr1_629091_CT,c...              27,3,5,23,14,8,24,2,90  716,807,1057,1448,1633,1884,209...  0.03771,0.00372,0.00473,0.01588...   \n",
       "4       chr1  629136  629148   6   2  0  chr1_629136_CT,chr1_629138_CT,c...                    100,10,19,0,9,68       3579,3571,3591,3604,3614,3617  0.02794,0.0028,0.00529,0.0,0.00...   \n",
       "...      ...     ...     ...  ..  .. ..                                 ...                                 ...                                 ...                                 ...   \n",
       "103310  chrM   16218   16261  20  18  1  chrM_16218_CT,chrM_16221_CT,chr...  142,102,109,61,179,94,147,120,1...  7817,7756,7847,7891,7860,7832,7...  0.01817,0.01315,0.01389,0.00773...   \n",
       "103311  chrM   16255   16274   3   3  0  chrM_16255_GA,chrM_16273_GA,chr...                         110,145,337                      7536,7620,7678              0.0146,0.01903,0.04389   \n",
       "103312  chrM   16259   16279  11   8  1  chrM_16259_CT,chrM_16260_CT,chr...  104,85,87,58,129,107,93,103,712...  7715,7730,7801,7811,7836,7758,7...  0.01348,0.011,0.01115,0.00743,0...   \n",
       "103313  chrM   16329   16373   7   7  0  chrM_16329_GA,chrM_16336_GA,chr...         131,147,246,421,466,656,106  7673,7755,7746,7663,7797,7764,7859  0.01707,0.01896,0.03176,0.05494...   \n",
       "103314  chrM   16470   16558  19  13  0  chrM_16470_GA,chrM_16471_GA,chr...  154,130,150,138,138,202,229,294...  7223,7148,7187,7190,7127,7119,7...  0.02132,0.01819,0.02087,0.01919...   \n",
       "\n",
       "                                        10                                  11                                  12         mpmat_index  \n",
       "0                        False,False,False                               0,0,0                    Pass,Filter,Pass    chr1_48137_48144  \n",
       "1                              False,False                                 0,0                           Pass,Pass  chr1_127550_127558  \n",
       "2                              False,False                                 0,0                           Pass,Pass  chr1_261533_261543  \n",
       "3       False,False,False,False,False,F...                   0,0,0,0,0,0,0,0,0  Pass,Filter,Filter,Pass,Filter,...  chr1_629090_629115  \n",
       "4       False,False,False,False,False,F...                         0,0,0,0,0,0  Pass,Filter,Filter,Filter,Filte...  chr1_629136_629148  \n",
       "...                                    ...                                 ...                                 ...                 ...  \n",
       "103310  False,False,False,False,False,F...  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0...  Pass,Pass,Pass,Filter,Pass,Pass...    chrM_16218_16261  \n",
       "103311                   False,False,False                               0,0,0                      Pass,Pass,Pass    chrM_16255_16274  \n",
       "103312  False,False,False,False,False,F...               0,0,0,0,0,0,0,0,0,0,0  Pass,Pass,Pass,Filter,Pass,Pass...    chrM_16259_16279  \n",
       "103313  False,False,False,False,False,F...                       0,0,0,0,0,0,0  Pass,Pass,Pass,Pass,Pass,Pass,Pass    chrM_16329_16373  \n",
       "103314  False,False,False,False,False,F...  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0...  Pass,Pass,Pass,Pass,Pass,Pass,P...    chrM_16470_16558  \n",
       "\n",
       "[103315 rows x 14 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 拿 650 个 old share 的点做一个 mpmat 出来去对着这五个 sample call 一下相关位置\n",
    "df_mpmat = pd.read_csv(\n",
    "    '../tables/293T-DdCBE-merge_hg38.MAPQ20.C6_M4_R1_T10.sort.V4.mpmat.gz',\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    ")\n",
    "df_mpmat['mpmat_index'] = df_mpmat[0] + '_' + df_mpmat[1].astype(str) + '_' + df_mpmat[2].astype(str)\n",
    "df_mpmat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1a41a3dd-bab6-493f-9cd7-d7d709bf1b6c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:50:02.614052Z",
     "iopub.status.busy": "2022-11-29T14:50:02.613361Z",
     "iopub.status.idle": "2022-11-29T14:50:02.781163Z",
     "shell.execute_reply": "2022-11-29T14:50:02.779777Z",
     "shell.execute_reply.started": "2022-11-29T14:50:02.613993Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_2044988_CT,chr1_2044989_CT...</td>\n",
       "      <td>14,11,7,11,0,8,5</td>\n",
       "      <td>25,25,25,23,22,22,23</td>\n",
       "      <td>0.56,0.44,0.28,0.47826,0.0,0.36...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Filter,Pass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_9702414_GA,chr1_9702415_GA...</td>\n",
       "      <td>4,1,3,8,7,7,18,10,20,12,32</td>\n",
       "      <td>103,100,96,100,94,97,91,91,93,9...</td>\n",
       "      <td>0.03883,0.01,0.03125,0.08,0.074...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Filter,Filter,Pass,Pass,Pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_12618165_CT,chr1_12618166_...</td>\n",
       "      <td>11,4,11,8,7</td>\n",
       "      <td>26,29,29,31,26</td>\n",
       "      <td>0.42308,0.13793,0.37931,0.25806...</td>\n",
       "      <td>False,False,False,False,False</td>\n",
       "      <td>0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_13019998_GA,chr1_13020006_...</td>\n",
       "      <td>7,2,7,5,5,11,7,7,19</td>\n",
       "      <td>38,42,40,42,42,41,42,41,42</td>\n",
       "      <td>0.18421,0.04762,0.175,0.11905,0...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Filter,Pass,Pass,Pass,Pass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_13354705_CT,chr1_13354706_...</td>\n",
       "      <td>37,9,28,23,13,16,12,4,10,4,5</td>\n",
       "      <td>63,61,66,68,67,68,65,68,68,71,69</td>\n",
       "      <td>0.5873,0.14754,0.42424,0.33824,...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass,Pass,P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>chrX_136925330_CT,chrX_13692533...</td>\n",
       "      <td>10,6,15,8,9</td>\n",
       "      <td>31,31,30,32,31</td>\n",
       "      <td>0.32258,0.19355,0.5,0.25,0.29032</td>\n",
       "      <td>False,False,False,False,False</td>\n",
       "      <td>0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>chrX_138128211_GA,chrX_13812821...</td>\n",
       "      <td>5,4,3,4,4,3,16,9,26</td>\n",
       "      <td>37,38,38,38,38,34,37,37,35</td>\n",
       "      <td>0.13514,0.10526,0.07895,0.10526...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Filter,Pass,Pass,Filt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150683735</td>\n",
       "      <td>150683758</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>chrX_150683735_GA,chrX_15068373...</td>\n",
       "      <td>8,12,13,15,35,41,75,57,52,90,9</td>\n",
       "      <td>217,211,211,209,207,204,211,203...</td>\n",
       "      <td>0.03687,0.05687,0.06161,0.07177...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass,Pass,P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>chrX_153669436_CT,chrX_15366943...</td>\n",
       "      <td>28,22,25,11,4,7,3,7,3,7</td>\n",
       "      <td>62,64,61,66,64,66,67,66,65,67</td>\n",
       "      <td>0.45161,0.34375,0.40984,0.16667...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass,Pass,F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>chrX_155049205_GA,chrX_15504920...</td>\n",
       "      <td>8,39,39,38,19,38,85</td>\n",
       "      <td>137,131,129,126,131,130,124</td>\n",
       "      <td>0.05839,0.29771,0.30233,0.30159...</td>\n",
       "      <td>False,False,False,False,False,F...</td>\n",
       "      <td>0,0,0,0,0,0,0</td>\n",
       "      <td>Pass,Pass,Pass,Pass,Pass,Pass,Pass</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>650 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0          1          2   3   4  5                                   6                               7                                   8                                   9  \\\n",
       "0    chr1    2044988    2044998   7   6  0  chr1_2044988_CT,chr1_2044989_CT...                14,11,7,11,0,8,5                25,25,25,23,22,22,23  0.56,0.44,0.28,0.47826,0.0,0.36...   \n",
       "1    chr1    9702414    9702449  11   9  0  chr1_9702414_GA,chr1_9702415_GA...      4,1,3,8,7,7,18,10,20,12,32  103,100,96,100,94,97,91,91,93,9...  0.03883,0.01,0.03125,0.08,0.074...   \n",
       "2    chr1   12618165   12618174   5   5  0  chr1_12618165_CT,chr1_12618166_...                     11,4,11,8,7                      26,29,29,31,26  0.42308,0.13793,0.37931,0.25806...   \n",
       "3    chr1   13019998   13020019   9   8  0  chr1_13019998_GA,chr1_13020006_...             7,2,7,5,5,11,7,7,19          38,42,40,42,42,41,42,41,42  0.18421,0.04762,0.175,0.11905,0...   \n",
       "4    chr1   13354705   13354738  11  11  0  chr1_13354705_CT,chr1_13354706_...    37,9,28,23,13,16,12,4,10,4,5    63,61,66,68,67,68,65,68,68,71,69  0.5873,0.14754,0.42424,0.33824,...   \n",
       "..    ...        ...        ...  ..  .. ..                                 ...                             ...                                 ...                                 ...   \n",
       "645  chrX  136925330  136925338   5   5  0  chrX_136925330_CT,chrX_13692533...                     10,6,15,8,9                      31,31,30,32,31    0.32258,0.19355,0.5,0.25,0.29032   \n",
       "646  chrX  138128211  138128237   9   7  0  chrX_138128211_GA,chrX_13812821...             5,4,3,4,4,3,16,9,26          37,38,38,38,38,34,37,37,35  0.13514,0.10526,0.07895,0.10526...   \n",
       "647  chrX  150683735  150683758  11  11  0  chrX_150683735_GA,chrX_15068373...  8,12,13,15,35,41,75,57,52,90,9  217,211,211,209,207,204,211,203...  0.03687,0.05687,0.06161,0.07177...   \n",
       "648  chrX  153669436  153669453  10   8  0  chrX_153669436_CT,chrX_15366943...         28,22,25,11,4,7,3,7,3,7       62,64,61,66,64,66,67,66,65,67  0.45161,0.34375,0.40984,0.16667...   \n",
       "649  chrX  155049205  155049215   7   7  0  chrX_155049205_GA,chrX_15504920...             8,39,39,38,19,38,85         137,131,129,126,131,130,124  0.05839,0.29771,0.30233,0.30159...   \n",
       "\n",
       "                                     10                     11                                  12  \n",
       "0    False,False,False,False,False,F...          0,0,0,0,0,0,0  Pass,Pass,Pass,Pass,Filter,Pass...  \n",
       "1    False,False,False,False,False,F...  0,0,0,0,0,0,0,0,0,0,0  Pass,Filter,Filter,Pass,Pass,Pa...  \n",
       "2         False,False,False,False,False              0,0,0,0,0            Pass,Pass,Pass,Pass,Pass  \n",
       "3    False,False,False,False,False,F...      0,0,0,0,0,0,0,0,0  Pass,Filter,Pass,Pass,Pass,Pass...  \n",
       "4    False,False,False,False,False,F...  0,0,0,0,0,0,0,0,0,0,0  Pass,Pass,Pass,Pass,Pass,Pass,P...  \n",
       "..                                  ...                    ...                                 ...  \n",
       "645       False,False,False,False,False              0,0,0,0,0            Pass,Pass,Pass,Pass,Pass  \n",
       "646  False,False,False,False,False,F...      0,0,0,0,0,0,0,0,0  Pass,Pass,Filter,Pass,Pass,Filt...  \n",
       "647  False,False,False,False,False,F...  0,0,0,0,0,0,0,0,0,0,0  Pass,Pass,Pass,Pass,Pass,Pass,P...  \n",
       "648  False,False,False,False,False,F...    0,0,0,0,0,0,0,0,0,0  Pass,Pass,Pass,Pass,Pass,Pass,F...  \n",
       "649  False,False,False,False,False,F...          0,0,0,0,0,0,0  Pass,Pass,Pass,Pass,Pass,Pass,Pass  \n",
       "\n",
       "[650 rows x 13 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge = pd.merge(left=df_mpmat, right=df_old_share, on='mpmat_index', how='right')\n",
    "\n",
    "df_aim_mpmat = df_merge.iloc[:, :-4]\n",
    "df_aim_mpmat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "f4f7dbc2-9619-4dad-a39e-08c25036742a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:50:03.308950Z",
     "iopub.status.busy": "2022-11-29T14:50:03.308456Z",
     "iopub.status.idle": "2022-11-29T14:50:03.376886Z",
     "shell.execute_reply": "2022-11-29T14:50:03.375825Z",
     "shell.execute_reply.started": "2022-11-29T14:50:03.308907Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "os.makedirs('../mpmat', exist_ok=True)\n",
    "df_aim_mpmat.to_csv('../mpmat/2022-10-21_nature_4_5.1_6_share_650-off-targets.mpmat', header=None, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e15b4a3b-0857-4c10-8831-640728cfc840",
   "metadata": {
    "tags": []
   },
   "source": [
    "## analysis with old TAS-independent list (650 sites) -> df_aim_mpmat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5aab32b2-bf6b-44ad-a38e-ecbc9fb9fa31",
   "metadata": {},
   "source": [
    "### Detect-seq signal comparation for DddA\\_wt/6/11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "7d5f2bdb-e1ff-44b3-9900-64be6a9fc2f3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:31.786217Z",
     "iopub.status.busy": "2022-11-29T14:47:31.785799Z",
     "iopub.status.idle": "2022-11-29T14:47:31.884100Z",
     "shell.execute_reply": "2022-11-29T14:47:31.883242Z",
     "shell.execute_reply.started": "2022-11-29T14:47:31.786191Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "      <td>chr1_2044988_2044998</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chr1_2044988_CT,chr1_2044989_CT...</td>\n",
       "      <td>N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_2044989_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.022784</td>\n",
       "      <td>0.038561</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019281</td>\n",
       "      <td>0,1,2,3,4,5,6,7 9,0,0,0,0,0,0,0...</td>\n",
       "      <td>0.759147</td>\n",
       "      <td>1.251110</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.153850e-01</td>\n",
       "      <td>4.105579e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "      <td>chr1_9702414_9702449</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_9702414_GA,chr1_9702415_GA...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_9702414_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.045567</td>\n",
       "      <td>0.083549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012854</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 16,2,...</td>\n",
       "      <td>0.874624</td>\n",
       "      <td>0.666147</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.559715e-01</td>\n",
       "      <td>5.418769e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "      <td>chr1_12618165_12618174</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1_12618165_CT,chr1_12618166_...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chr1_12618166_CT</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025315</td>\n",
       "      <td>0.051415</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012854</td>\n",
       "      <td>0,1,2,3,4,5 10,0,0,0,0,0 4,2,2,...</td>\n",
       "      <td>1.022181</td>\n",
       "      <td>0.666147</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.559715e-01</td>\n",
       "      <td>5.418769e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "      <td>chr1_13019998_13020019</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>chr1_13019998_GA,chr1_13020006_...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_13019998_GA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.010126</td>\n",
       "      <td>0.025707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 4,0,0,0,0,0...</td>\n",
       "      <td>1.344109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>8.274602e-01</td>\n",
       "      <td>8.274602e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "      <td>chr1_13354705_13354738</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_13354705_CT,chr1_13354706_...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_13354705_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.060756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 24,0,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>8.274602e-01</td>\n",
       "      <td>8.274602e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrX_136925330_CT,chrX_13692533...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chrX_136925330_CT</td>\n",
       "      <td>42</td>\n",
       "      <td>103</td>\n",
       "      <td>0.407767</td>\n",
       "      <td>43</td>\n",
       "      <td>129</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>0.108855</td>\n",
       "      <td>0.717403</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.328114</td>\n",
       "      <td>0,1,2,3,4,5 43,0,0,0,0,0 27,43,...</td>\n",
       "      <td>2.720374</td>\n",
       "      <td>5.608487</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.433039e-20</td>\n",
       "      <td>9.702301e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3246</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>chrX_138128211_GA,chrX_13812821...</td>\n",
       "      <td>N-N-N-B-N-N-N-N-N</td>\n",
       "      <td>chrX_138128237_GA</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>33</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>0.205767</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.072296</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 32,0,1,0,0,...</td>\n",
       "      <td>1.300471</td>\n",
       "      <td>4.835851</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.889936e-04</td>\n",
       "      <td>4.053103e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3247</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>150683735</td>\n",
       "      <td>150683758</td>\n",
       "      <td>chrX_150683735_150683758</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chrX_150683735_GA,chrX_15068373...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_150683756_GA</td>\n",
       "      <td>93</td>\n",
       "      <td>200</td>\n",
       "      <td>0.465000</td>\n",
       "      <td>37</td>\n",
       "      <td>269</td>\n",
       "      <td>2</td>\n",
       "      <td>188</td>\n",
       "      <td>0.093666</td>\n",
       "      <td>1.495979</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>1.045517</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 35,0,...</td>\n",
       "      <td>3.997421</td>\n",
       "      <td>7.690000</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.003205e-70</td>\n",
       "      <td>1.626736e-68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>chrX_153669436_CT,chrX_15366943...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_153669436_CT</td>\n",
       "      <td>7</td>\n",
       "      <td>22</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>22</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0.055693</td>\n",
       "      <td>0.144593</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061174</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10 22,0,0,0...</td>\n",
       "      <td>1.376420</td>\n",
       "      <td>3.185275</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.185269e-03</td>\n",
       "      <td>1.578739e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_155049205_GA,chrX_15504920...</td>\n",
       "      <td>N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_155049215_GA</td>\n",
       "      <td>69</td>\n",
       "      <td>118</td>\n",
       "      <td>0.584746</td>\n",
       "      <td>25</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>0.063288</td>\n",
       "      <td>0.834189</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.600616</td>\n",
       "      <td>0,1,2,3,4,5,6,7 24,0,0,1,0,0,0,...</td>\n",
       "      <td>3.720374</td>\n",
       "      <td>7.890299</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>6.779916e-39</td>\n",
       "      <td>8.012628e-38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3250 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          <sample> chr_name  region_start  region_end               mpmat_index  region_site_num  region_block_site_num  region_mut_site_num                   region_site_index  \\\n",
       "0       DetectSeq_ATP8-DddA6_REP-1     chr1       2044988     2044998      chr1_2044988_2044998                7                      0                    7  chr1_2044988_CT,chr1_2044989_CT...   \n",
       "1       DetectSeq_ATP8-DddA6_REP-1     chr1       9702414     9702449      chr1_9702414_9702449               11                      0                   11  chr1_9702414_GA,chr1_9702415_GA...   \n",
       "2       DetectSeq_ATP8-DddA6_REP-1     chr1      12618165    12618174    chr1_12618165_12618174                5                      0                    5  chr1_12618165_CT,chr1_12618166_...   \n",
       "3       DetectSeq_ATP8-DddA6_REP-1     chr1      13019998    13020019    chr1_13019998_13020019                9                      0                    9  chr1_13019998_GA,chr1_13020006_...   \n",
       "4       DetectSeq_ATP8-DddA6_REP-1     chr1      13354705    13354738    chr1_13354705_13354738               11                      0                   11  chr1_13354705_CT,chr1_13354706_...   \n",
       "...                            ...      ...           ...         ...                       ...              ...                    ...                  ...                                 ...   \n",
       "3245  DetectSeq_SIRT6-DddA11_REP-1     chrX     136925330   136925338  chrX_136925330_136925338                5                      0                    5  chrX_136925330_CT,chrX_13692533...   \n",
       "3246  DetectSeq_SIRT6-DddA11_REP-1     chrX     138128211   138128237  chrX_138128211_138128237                9                      1                    8  chrX_138128211_GA,chrX_13812821...   \n",
       "3247  DetectSeq_SIRT6-DddA11_REP-1     chrX     150683735   150683758  chrX_150683735_150683758               11                      0                   11  chrX_150683735_GA,chrX_15068373...   \n",
       "3248  DetectSeq_SIRT6-DddA11_REP-1     chrX     153669436   153669453  chrX_153669436_153669453               10                      0                   10  chrX_153669436_CT,chrX_15366943...   \n",
       "3249  DetectSeq_SIRT6-DddA11_REP-1     chrX     155049205   155049215  chrX_155049205_155049215                7                      0                    7  chrX_155049205_GA,chrX_15504920...   \n",
       "\n",
       "         region_block_state region_highest_site_index  region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  \\\n",
       "0             N-N-N-N-N-N-N           chr1_2044989_CT                            4                              6                       0.666667           9            6               0   \n",
       "1     N-N-N-N-N-N-N-N-N-N-N           chr1_9702414_GA                            1                             10                       0.100000          18           13               0   \n",
       "2                 N-N-N-N-N          chr1_12618166_CT                            2                              8                       0.250000          10            8               0   \n",
       "3         N-N-N-N-N-N-N-N-N          chr1_13019998_GA                            0                              2                       0.000000           4            4               0   \n",
       "4     N-N-N-N-N-N-N-N-N-N-N          chr1_13354705_CT                            0                              0                       0.000000          24            0               0   \n",
       "...                     ...                       ...                          ...                            ...                            ...         ...          ...             ...   \n",
       "3245              N-N-N-N-N         chrX_136925330_CT                           42                            103                       0.407767          43          129               0   \n",
       "3246      N-N-N-B-N-N-N-N-N         chrX_138128237_GA                           12                             28                       0.428571          33           37               1   \n",
       "3247  N-N-N-N-N-N-N-N-N-N-N         chrX_150683756_GA                           93                            200                       0.465000          37          269               2   \n",
       "3248    N-N-N-N-N-N-N-N-N-N         chrX_153669436_CT                            7                             22                       0.318182          22           26               0   \n",
       "3249          N-N-N-N-N-N-N         chrX_155049215_GA                           69                            118                       0.584746          25          150               1   \n",
       "\n",
       "      treat_mut_count  ctrl_count.norm  treat_count.norm  ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state       p_value           FDR  \n",
       "0                   3         0.022784          0.038561             0.000000              0.019281  0,1,2,3,4,5,6,7 9,0,0,0,0,0,0,0...  0.759147     1.251110     TestOK  2.153850e-01  4.105579e-01  \n",
       "1                   2         0.045567          0.083549             0.000000              0.012854  0,1,2,3,4,5,6,7,8,9,10,11 16,2,...  0.874624     0.666147     TestOK  3.559715e-01  5.418769e-01  \n",
       "2                   2         0.025315          0.051415             0.000000              0.012854  0,1,2,3,4,5 10,0,0,0,0,0 4,2,2,...  1.022181     0.666147     TestOK  3.559715e-01  5.418769e-01  \n",
       "3                   0         0.010126          0.025707             0.000000              0.000000  0,1,2,3,4,5,6,7,8,9 4,0,0,0,0,0...  1.344109          NaN     TestOK  8.274602e-01  8.274602e-01  \n",
       "4                   0         0.060756          0.000000             0.000000              0.000000  0,1,2,3,4,5,6,7,8,9,10,11 24,0,...       NaN          NaN     TestOK  8.274602e-01  8.274602e-01  \n",
       "...               ...              ...               ...                  ...                   ...                                 ...       ...          ...        ...           ...           ...  \n",
       "3245               59         0.108855          0.717403             0.000000              0.328114  0,1,2,3,4,5 43,0,0,0,0,0 27,43,...  2.720374     5.608487     TestOK  2.433039e-20  9.702301e-20  \n",
       "3246               13         0.083540          0.205767             0.002532              0.072296  0,1,2,3,4,5,6,7,8,9 32,0,1,0,0,...  1.300471     4.835851     TestOK  2.889936e-04  4.053103e-04  \n",
       "3247              188         0.093666          1.495979             0.005063              1.045517  0,1,2,3,4,5,6,7,8,9,10,11 35,0,...  3.997421     7.690000     TestOK  3.003205e-70  1.626736e-68  \n",
       "3248               11         0.055693          0.144593             0.000000              0.061174  0,1,2,3,4,5,6,7,8,9,10 22,0,0,0...  1.376420     3.185275     TestOK  1.185269e-03  1.578739e-03  \n",
       "3249              108         0.063288          0.834189             0.002532              0.600616  0,1,2,3,4,5,6,7 24,0,0,1,0,0,0,...  3.720374     7.890299     TestOK  6.779916e-39  8.012628e-38  \n",
       "\n",
       "[3250 rows x 28 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 650 mpmat to get 650 poisson_res\n",
    "df_old_site_new_signal = pd.read_csv(\n",
    "    '../poisson_res_use_old650/poisson_res_all_use650.tsv.gz',\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    sep='\\t'\n",
    ")\n",
    "df_old_site_new_signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "f3d166bf-bfcb-45b9-a996-0c748ab32203",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:50.796929Z",
     "iopub.status.busy": "2022-11-29T14:47:50.796497Z",
     "iopub.status.idle": "2022-11-29T14:47:50.851258Z",
     "shell.execute_reply": "2022-11-29T14:47:50.850402Z",
     "shell.execute_reply.started": "2022-11-29T14:47:50.796902Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 3250 entries, 0 to 3249\n",
      "Data columns (total 28 columns):\n",
      " #   Column                         Non-Null Count  Dtype  \n",
      "---  ------                         --------------  -----  \n",
      " 0   <sample>                       3250 non-null   object \n",
      " 1   chr_name                       3250 non-null   object \n",
      " 2   region_start                   3250 non-null   int64  \n",
      " 3   region_end                     3250 non-null   int64  \n",
      " 4   mpmat_index                    3250 non-null   object \n",
      " 5   region_site_num                3250 non-null   int64  \n",
      " 6   region_block_site_num          3250 non-null   int64  \n",
      " 7   region_mut_site_num            3250 non-null   int64  \n",
      " 8   region_site_index              3250 non-null   object \n",
      " 9   region_block_state             3250 non-null   object \n",
      " 10  region_highest_site_index      3250 non-null   object \n",
      " 11  region_highest_site_mut_num    3250 non-null   int64  \n",
      " 12  region_highest_site_cover_num  3250 non-null   int64  \n",
      " 13  region_highest_site_mut_ratio  3250 non-null   float64\n",
      " 14  ctrl_count                     3250 non-null   int64  \n",
      " 15  treat_count                    3250 non-null   int64  \n",
      " 16  ctrl_mut_count                 3250 non-null   int64  \n",
      " 17  treat_mut_count                3250 non-null   int64  \n",
      " 18  ctrl_count.norm                3250 non-null   float64\n",
      " 19  treat_count.norm               3250 non-null   float64\n",
      " 20  ctrl_mut_count.norm            3250 non-null   float64\n",
      " 21  treat_mut_count.norm           3250 non-null   float64\n",
      " 22  count_info                     3250 non-null   object \n",
      " 23  log2_FC                        3248 non-null   float64\n",
      " 24  log2_FC_mut                    2882 non-null   float64\n",
      " 25  test_state                     3250 non-null   object \n",
      " 26  p_value                        3250 non-null   float64\n",
      " 27  FDR                            3250 non-null   float64\n",
      "dtypes: float64(9), int64(11), object(8)\n",
      "memory usage: 711.1+ KB\n"
     ]
    }
   ],
   "source": [
    "df_old_site_new_signal.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "97ec1531-4398-40fe-88bd-be0eeabc8675",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:51.538446Z",
     "iopub.status.busy": "2022-11-29T14:47:51.538195Z",
     "iopub.status.idle": "2022-11-29T14:47:51.549343Z",
     "shell.execute_reply": "2022-11-29T14:47:51.548503Z",
     "shell.execute_reply.started": "2022-11-29T14:47:51.538424Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "370"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_site_new_signal.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "e57c3672-6095-45b6-946b-ba63b4c1eb54",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:51.922122Z",
     "iopub.status.busy": "2022-11-29T14:47:51.921764Z",
     "iopub.status.idle": "2022-11-29T14:47:51.957694Z",
     "shell.execute_reply": "2022-11-29T14:47:51.956978Z",
     "shell.execute_reply.started": "2022-11-29T14:47:51.922098Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "      <td>chr1_2044988_2044998</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chr1_2044988_CT,chr1_2044989_CT...</td>\n",
       "      <td>N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_2044989_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.022784</td>\n",
       "      <td>0.038561</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019281</td>\n",
       "      <td>0,1,2,3,4,5,6,7 9,0,0,0,0,0,0,0...</td>\n",
       "      <td>0.759147</td>\n",
       "      <td>1.251110</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.153850e-01</td>\n",
       "      <td>4.105579e-01</td>\n",
       "      <td>chr1_2044988_2044998_highest_ch...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "      <td>chr1_9702414_9702449</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_9702414_GA,chr1_9702415_GA...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_9702414_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.045567</td>\n",
       "      <td>0.083549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012854</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 16,2,...</td>\n",
       "      <td>0.874624</td>\n",
       "      <td>0.666147</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.559715e-01</td>\n",
       "      <td>5.418769e-01</td>\n",
       "      <td>chr1_9702414_9702449_highest_ch...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "      <td>chr1_12618165_12618174</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1_12618165_CT,chr1_12618166_...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chr1_12618166_CT</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.025315</td>\n",
       "      <td>0.051415</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012854</td>\n",
       "      <td>0,1,2,3,4,5 10,0,0,0,0,0 4,2,2,...</td>\n",
       "      <td>1.022181</td>\n",
       "      <td>0.666147</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.559715e-01</td>\n",
       "      <td>5.418769e-01</td>\n",
       "      <td>chr1_12618165_12618174_highest_...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "      <td>chr1_13019998_13020019</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>chr1_13019998_GA,chr1_13020006_...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_13019998_GA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.010126</td>\n",
       "      <td>0.025707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 4,0,0,0,0,0...</td>\n",
       "      <td>1.344109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>8.274602e-01</td>\n",
       "      <td>8.274602e-01</td>\n",
       "      <td>chr1_13019998_13020019_highest_...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "      <td>chr1_13354705_13354738</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_13354705_CT,chr1_13354706_...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_13354705_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.060756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 24,0,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>8.274602e-01</td>\n",
       "      <td>8.274602e-01</td>\n",
       "      <td>chr1_13354705_13354738_highest_...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrX_136925330_CT,chrX_13692533...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chrX_136925330_CT</td>\n",
       "      <td>42</td>\n",
       "      <td>103</td>\n",
       "      <td>0.407767</td>\n",
       "      <td>43</td>\n",
       "      <td>129</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>0.108855</td>\n",
       "      <td>0.717403</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.328114</td>\n",
       "      <td>0,1,2,3,4,5 43,0,0,0,0,0 27,43,...</td>\n",
       "      <td>2.720374</td>\n",
       "      <td>5.608487</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.433039e-20</td>\n",
       "      <td>9.702301e-20</td>\n",
       "      <td>chrX_136925330_136925338_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3246</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>chrX_138128211_GA,chrX_13812821...</td>\n",
       "      <td>N-N-N-B-N-N-N-N-N</td>\n",
       "      <td>chrX_138128237_GA</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>33</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>0.205767</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.072296</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 32,0,1,0,0,...</td>\n",
       "      <td>1.300471</td>\n",
       "      <td>4.835851</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.889936e-04</td>\n",
       "      <td>4.053103e-04</td>\n",
       "      <td>chrX_138128211_138128237_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3247</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>150683735</td>\n",
       "      <td>150683758</td>\n",
       "      <td>chrX_150683735_150683758</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chrX_150683735_GA,chrX_15068373...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_150683756_GA</td>\n",
       "      <td>93</td>\n",
       "      <td>200</td>\n",
       "      <td>0.465000</td>\n",
       "      <td>37</td>\n",
       "      <td>269</td>\n",
       "      <td>2</td>\n",
       "      <td>188</td>\n",
       "      <td>0.093666</td>\n",
       "      <td>1.495979</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>1.045517</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 35,0,...</td>\n",
       "      <td>3.997421</td>\n",
       "      <td>7.690000</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.003205e-70</td>\n",
       "      <td>1.626736e-68</td>\n",
       "      <td>chrX_150683735_150683758_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>chrX_153669436_CT,chrX_15366943...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_153669436_CT</td>\n",
       "      <td>7</td>\n",
       "      <td>22</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>22</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0.055693</td>\n",
       "      <td>0.144593</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061174</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10 22,0,0,0...</td>\n",
       "      <td>1.376420</td>\n",
       "      <td>3.185275</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.185269e-03</td>\n",
       "      <td>1.578739e-03</td>\n",
       "      <td>chrX_153669436_153669453_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_155049205_GA,chrX_15504920...</td>\n",
       "      <td>N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_155049215_GA</td>\n",
       "      <td>69</td>\n",
       "      <td>118</td>\n",
       "      <td>0.584746</td>\n",
       "      <td>25</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>0.063288</td>\n",
       "      <td>0.834189</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.600616</td>\n",
       "      <td>0,1,2,3,4,5,6,7 24,0,0,1,0,0,0,...</td>\n",
       "      <td>3.720374</td>\n",
       "      <td>7.890299</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>6.779916e-39</td>\n",
       "      <td>8.012628e-38</td>\n",
       "      <td>chrX_155049205_155049215_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3250 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          <sample> chr_name  region_start  region_end               mpmat_index  region_site_num  region_block_site_num  region_mut_site_num                   region_site_index  \\\n",
       "0       DetectSeq_ATP8-DddA6_REP-1     chr1       2044988     2044998      chr1_2044988_2044998                7                      0                    7  chr1_2044988_CT,chr1_2044989_CT...   \n",
       "1       DetectSeq_ATP8-DddA6_REP-1     chr1       9702414     9702449      chr1_9702414_9702449               11                      0                   11  chr1_9702414_GA,chr1_9702415_GA...   \n",
       "2       DetectSeq_ATP8-DddA6_REP-1     chr1      12618165    12618174    chr1_12618165_12618174                5                      0                    5  chr1_12618165_CT,chr1_12618166_...   \n",
       "3       DetectSeq_ATP8-DddA6_REP-1     chr1      13019998    13020019    chr1_13019998_13020019                9                      0                    9  chr1_13019998_GA,chr1_13020006_...   \n",
       "4       DetectSeq_ATP8-DddA6_REP-1     chr1      13354705    13354738    chr1_13354705_13354738               11                      0                   11  chr1_13354705_CT,chr1_13354706_...   \n",
       "...                            ...      ...           ...         ...                       ...              ...                    ...                  ...                                 ...   \n",
       "3245  DetectSeq_SIRT6-DddA11_REP-1     chrX     136925330   136925338  chrX_136925330_136925338                5                      0                    5  chrX_136925330_CT,chrX_13692533...   \n",
       "3246  DetectSeq_SIRT6-DddA11_REP-1     chrX     138128211   138128237  chrX_138128211_138128237                9                      1                    8  chrX_138128211_GA,chrX_13812821...   \n",
       "3247  DetectSeq_SIRT6-DddA11_REP-1     chrX     150683735   150683758  chrX_150683735_150683758               11                      0                   11  chrX_150683735_GA,chrX_15068373...   \n",
       "3248  DetectSeq_SIRT6-DddA11_REP-1     chrX     153669436   153669453  chrX_153669436_153669453               10                      0                   10  chrX_153669436_CT,chrX_15366943...   \n",
       "3249  DetectSeq_SIRT6-DddA11_REP-1     chrX     155049205   155049215  chrX_155049205_155049215                7                      0                    7  chrX_155049205_GA,chrX_15504920...   \n",
       "\n",
       "         region_block_state region_highest_site_index  region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  \\\n",
       "0             N-N-N-N-N-N-N           chr1_2044989_CT                            4                              6                       0.666667           9            6               0   \n",
       "1     N-N-N-N-N-N-N-N-N-N-N           chr1_9702414_GA                            1                             10                       0.100000          18           13               0   \n",
       "2                 N-N-N-N-N          chr1_12618166_CT                            2                              8                       0.250000          10            8               0   \n",
       "3         N-N-N-N-N-N-N-N-N          chr1_13019998_GA                            0                              2                       0.000000           4            4               0   \n",
       "4     N-N-N-N-N-N-N-N-N-N-N          chr1_13354705_CT                            0                              0                       0.000000          24            0               0   \n",
       "...                     ...                       ...                          ...                            ...                            ...         ...          ...             ...   \n",
       "3245              N-N-N-N-N         chrX_136925330_CT                           42                            103                       0.407767          43          129               0   \n",
       "3246      N-N-N-B-N-N-N-N-N         chrX_138128237_GA                           12                             28                       0.428571          33           37               1   \n",
       "3247  N-N-N-N-N-N-N-N-N-N-N         chrX_150683756_GA                           93                            200                       0.465000          37          269               2   \n",
       "3248    N-N-N-N-N-N-N-N-N-N         chrX_153669436_CT                            7                             22                       0.318182          22           26               0   \n",
       "3249          N-N-N-N-N-N-N         chrX_155049215_GA                           69                            118                       0.584746          25          150               1   \n",
       "\n",
       "      treat_mut_count  ctrl_count.norm  treat_count.norm  ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state       p_value           FDR  \\\n",
       "0                   3         0.022784          0.038561             0.000000              0.019281  0,1,2,3,4,5,6,7 9,0,0,0,0,0,0,0...  0.759147     1.251110     TestOK  2.153850e-01  4.105579e-01   \n",
       "1                   2         0.045567          0.083549             0.000000              0.012854  0,1,2,3,4,5,6,7,8,9,10,11 16,2,...  0.874624     0.666147     TestOK  3.559715e-01  5.418769e-01   \n",
       "2                   2         0.025315          0.051415             0.000000              0.012854  0,1,2,3,4,5 10,0,0,0,0,0 4,2,2,...  1.022181     0.666147     TestOK  3.559715e-01  5.418769e-01   \n",
       "3                   0         0.010126          0.025707             0.000000              0.000000  0,1,2,3,4,5,6,7,8,9 4,0,0,0,0,0...  1.344109          NaN     TestOK  8.274602e-01  8.274602e-01   \n",
       "4                   0         0.060756          0.000000             0.000000              0.000000  0,1,2,3,4,5,6,7,8,9,10,11 24,0,...       NaN          NaN     TestOK  8.274602e-01  8.274602e-01   \n",
       "...               ...              ...               ...                  ...                   ...                                 ...       ...          ...        ...           ...           ...   \n",
       "3245               59         0.108855          0.717403             0.000000              0.328114  0,1,2,3,4,5 43,0,0,0,0,0 27,43,...  2.720374     5.608487     TestOK  2.433039e-20  9.702301e-20   \n",
       "3246               13         0.083540          0.205767             0.002532              0.072296  0,1,2,3,4,5,6,7,8,9 32,0,1,0,0,...  1.300471     4.835851     TestOK  2.889936e-04  4.053103e-04   \n",
       "3247              188         0.093666          1.495979             0.005063              1.045517  0,1,2,3,4,5,6,7,8,9,10,11 35,0,...  3.997421     7.690000     TestOK  3.003205e-70  1.626736e-68   \n",
       "3248               11         0.055693          0.144593             0.000000              0.061174  0,1,2,3,4,5,6,7,8,9,10 22,0,0,0...  1.376420     3.185275     TestOK  1.185269e-03  1.578739e-03   \n",
       "3249              108         0.063288          0.834189             0.002532              0.600616  0,1,2,3,4,5,6,7 24,0,0,1,0,0,0,...  3.720374     7.890299     TestOK  6.779916e-39  8.012628e-38   \n",
       "\n",
       "                                bed_name strand  \n",
       "0     chr1_2044988_2044998_highest_ch...      .  \n",
       "1     chr1_9702414_9702449_highest_ch...      .  \n",
       "2     chr1_12618165_12618174_highest_...      .  \n",
       "3     chr1_13019998_13020019_highest_...      .  \n",
       "4     chr1_13354705_13354738_highest_...      .  \n",
       "...                                  ...    ...  \n",
       "3245  chrX_136925330_136925338_highes...      .  \n",
       "3246  chrX_138128211_138128237_highes...      .  \n",
       "3247  chrX_150683735_150683758_highes...      .  \n",
       "3248  chrX_153669436_153669453_highes...      .  \n",
       "3249  chrX_155049205_155049215_highes...      .  \n",
       "\n",
       "[3250 rows x 30 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_site_new_signal = df_old_site_new_signal.assign(\n",
    "    bed_name=df_old_site_new_signal.mpmat_index + '_highest_' + df_old_site_new_signal.region_highest_site_index,\n",
    "    strand='.'\n",
    ")\n",
    "df_old_site_new_signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "c4cbfb36-a9fb-44c0-a0ee-6057733fd9b5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:52.204588Z",
     "iopub.status.busy": "2022-11-29T14:47:52.204178Z",
     "iopub.status.idle": "2022-11-29T14:47:52.841084Z",
     "shell.execute_reply": "2022-11-29T14:47:52.840303Z",
     "shell.execute_reply.started": "2022-11-29T14:47:52.204551Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "617\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_start</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_end</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_site_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_block_site_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_mut_site_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_highest_site_mut_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_highest_site_cover_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_highest_site_mut_ratio</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_mut_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_mut_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_mut_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_mut_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">log2_FC</th>\n",
       "      <th colspan=\"8\" halign=\"left\">log2_FC_mut</th>\n",
       "      <th colspan=\"8\" halign=\"left\">p_value</th>\n",
       "      <th colspan=\"8\" halign=\"left\">FDR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA11_REP-1</th>\n",
       "      <td>31.0</td>\n",
       "      <td>8.600734e+07</td>\n",
       "      <td>4.994655e+07</td>\n",
       "      <td>2220193.0</td>\n",
       "      <td>50522662.50</td>\n",
       "      <td>78617702.0</td>\n",
       "      <td>1.133686e+08</td>\n",
       "      <td>201232409.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>8.600737e+07</td>\n",
       "      <td>4.994655e+07</td>\n",
       "      <td>2220214.0</td>\n",
       "      <td>50522683.0</td>\n",
       "      <td>78617723.0</td>\n",
       "      <td>1.133686e+08</td>\n",
       "      <td>201232430.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>10.838710</td>\n",
       "      <td>3.327630</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.00</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.00</td>\n",
       "      <td>18.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.064516</td>\n",
       "      <td>0.249731</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>10.774194</td>\n",
       "      <td>3.263022</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.00</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.00</td>\n",
       "      <td>18.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>21.225806</td>\n",
       "      <td>12.227591</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>24.50</td>\n",
       "      <td>68.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>50.645161</td>\n",
       "      <td>30.638264</td>\n",
       "      <td>25.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>59.00</td>\n",
       "      <td>174.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.422565</td>\n",
       "      <td>0.053080</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>0.378354</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.453463</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>31.0</td>\n",
       "      <td>22.580645</td>\n",
       "      <td>10.760341</td>\n",
       "      <td>9.0</td>\n",
       "      <td>14.00</td>\n",
       "      <td>20.0</td>\n",
       "      <td>28.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>66.290323</td>\n",
       "      <td>40.377546</td>\n",
       "      <td>32.0</td>\n",
       "      <td>40.50</td>\n",
       "      <td>48.0</td>\n",
       "      <td>80.50</td>\n",
       "      <td>230.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>0.340777</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>35.290323</td>\n",
       "      <td>16.925707</td>\n",
       "      <td>20.0</td>\n",
       "      <td>23.00</td>\n",
       "      <td>27.0</td>\n",
       "      <td>45.5</td>\n",
       "      <td>88.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.057163</td>\n",
       "      <td>0.027240</td>\n",
       "      <td>0.022784</td>\n",
       "      <td>0.035441</td>\n",
       "      <td>0.050630</td>\n",
       "      <td>0.072148</td>\n",
       "      <td>0.124044</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.395146</td>\n",
       "      <td>0.240684</td>\n",
       "      <td>0.190747</td>\n",
       "      <td>0.241414</td>\n",
       "      <td>0.286121</td>\n",
       "      <td>0.479848</td>\n",
       "      <td>1.370995</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.000863</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.210360</td>\n",
       "      <td>0.100892</td>\n",
       "      <td>0.119217</td>\n",
       "      <td>0.137099</td>\n",
       "      <td>0.160943</td>\n",
       "      <td>0.271219</td>\n",
       "      <td>0.524554</td>\n",
       "      <td>31.0</td>\n",
       "      <td>2.763185</td>\n",
       "      <td>0.879470</td>\n",
       "      <td>1.369818</td>\n",
       "      <td>2.050195</td>\n",
       "      <td>2.735088</td>\n",
       "      <td>3.345861</td>\n",
       "      <td>5.031376</td>\n",
       "      <td>31.0</td>\n",
       "      <td>5.091027</td>\n",
       "      <td>0.768367</td>\n",
       "      <td>3.879483</td>\n",
       "      <td>4.619327</td>\n",
       "      <td>4.952822</td>\n",
       "      <td>5.559368</td>\n",
       "      <td>7.444970</td>\n",
       "      <td>31.0</td>\n",
       "      <td>4.192761e-07</td>\n",
       "      <td>8.614629e-07</td>\n",
       "      <td>1.866343e-31</td>\n",
       "      <td>2.807248e-15</td>\n",
       "      <td>5.548434e-09</td>\n",
       "      <td>1.776817e-07</td>\n",
       "      <td>3.152999e-06</td>\n",
       "      <td>31.0</td>\n",
       "      <td>4.793318e-06</td>\n",
       "      <td>9.426111e-06</td>\n",
       "      <td>1.213123e-28</td>\n",
       "      <td>1.330349e-13</td>\n",
       "      <td>1.092873e-07</td>\n",
       "      <td>2.531712e-06</td>\n",
       "      <td>0.000033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA6_REP-1</th>\n",
       "      <td>14.0</td>\n",
       "      <td>7.320022e+07</td>\n",
       "      <td>5.423443e+07</td>\n",
       "      <td>1056219.0</td>\n",
       "      <td>35163961.75</td>\n",
       "      <td>76116851.5</td>\n",
       "      <td>9.922139e+07</td>\n",
       "      <td>201232409.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7.320025e+07</td>\n",
       "      <td>5.423443e+07</td>\n",
       "      <td>1056232.0</td>\n",
       "      <td>35163998.0</td>\n",
       "      <td>76116884.5</td>\n",
       "      <td>9.922141e+07</td>\n",
       "      <td>201232430.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.214286</td>\n",
       "      <td>3.400873</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.25</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.75</td>\n",
       "      <td>18.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.363137</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.071429</td>\n",
       "      <td>3.269069</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.25</td>\n",
       "      <td>10.5</td>\n",
       "      <td>13.75</td>\n",
       "      <td>17.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15.357143</td>\n",
       "      <td>5.865245</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.5</td>\n",
       "      <td>13.5</td>\n",
       "      <td>18.75</td>\n",
       "      <td>31.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>35.285714</td>\n",
       "      <td>14.440480</td>\n",
       "      <td>24.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>31.0</td>\n",
       "      <td>37.25</td>\n",
       "      <td>80.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.438128</td>\n",
       "      <td>0.066487</td>\n",
       "      <td>0.360000</td>\n",
       "      <td>0.406539</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>0.450779</td>\n",
       "      <td>0.633333</td>\n",
       "      <td>14.0</td>\n",
       "      <td>20.571429</td>\n",
       "      <td>5.720601</td>\n",
       "      <td>12.0</td>\n",
       "      <td>17.25</td>\n",
       "      <td>20.5</td>\n",
       "      <td>25.5</td>\n",
       "      <td>29.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>49.500000</td>\n",
       "      <td>19.606710</td>\n",
       "      <td>32.0</td>\n",
       "      <td>36.25</td>\n",
       "      <td>43.0</td>\n",
       "      <td>53.75</td>\n",
       "      <td>103.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.363137</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>27.857143</td>\n",
       "      <td>8.908571</td>\n",
       "      <td>20.0</td>\n",
       "      <td>21.25</td>\n",
       "      <td>24.5</td>\n",
       "      <td>31.5</td>\n",
       "      <td>48.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.052077</td>\n",
       "      <td>0.014482</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.043669</td>\n",
       "      <td>0.051896</td>\n",
       "      <td>0.064554</td>\n",
       "      <td>0.073414</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.318129</td>\n",
       "      <td>0.126009</td>\n",
       "      <td>0.205659</td>\n",
       "      <td>0.232973</td>\n",
       "      <td>0.276354</td>\n",
       "      <td>0.345443</td>\n",
       "      <td>0.661965</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.000362</td>\n",
       "      <td>0.000919</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.179033</td>\n",
       "      <td>0.057254</td>\n",
       "      <td>0.128537</td>\n",
       "      <td>0.136570</td>\n",
       "      <td>0.157458</td>\n",
       "      <td>0.202446</td>\n",
       "      <td>0.308488</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2.582972</td>\n",
       "      <td>0.705075</td>\n",
       "      <td>1.759147</td>\n",
       "      <td>2.125168</td>\n",
       "      <td>2.384784</td>\n",
       "      <td>2.869809</td>\n",
       "      <td>4.445647</td>\n",
       "      <td>14.0</td>\n",
       "      <td>4.982211</td>\n",
       "      <td>0.677993</td>\n",
       "      <td>4.324449</td>\n",
       "      <td>4.423941</td>\n",
       "      <td>4.875736</td>\n",
       "      <td>5.258613</td>\n",
       "      <td>6.701661</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2.372577e-07</td>\n",
       "      <td>3.271692e-07</td>\n",
       "      <td>1.269741e-16</td>\n",
       "      <td>1.086194e-09</td>\n",
       "      <td>7.676918e-08</td>\n",
       "      <td>4.191620e-07</td>\n",
       "      <td>1.064301e-06</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7.773420e-06</td>\n",
       "      <td>9.629707e-06</td>\n",
       "      <td>8.253317e-14</td>\n",
       "      <td>8.472959e-08</td>\n",
       "      <td>3.623942e-06</td>\n",
       "      <td>1.412051e-05</td>\n",
       "      <td>0.000030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddAwt_REP-1</th>\n",
       "      <td>11.0</td>\n",
       "      <td>6.947866e+07</td>\n",
       "      <td>5.711713e+07</td>\n",
       "      <td>15721070.0</td>\n",
       "      <td>27271003.50</td>\n",
       "      <td>45187694.0</td>\n",
       "      <td>1.018704e+08</td>\n",
       "      <td>201232409.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>6.947868e+07</td>\n",
       "      <td>5.711713e+07</td>\n",
       "      <td>15721094.0</td>\n",
       "      <td>27271029.5</td>\n",
       "      <td>45187712.0</td>\n",
       "      <td>1.018704e+08</td>\n",
       "      <td>201232430.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.181818</td>\n",
       "      <td>3.429816</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.00</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.00</td>\n",
       "      <td>18.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.301511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.090909</td>\n",
       "      <td>3.238967</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.00</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.00</td>\n",
       "      <td>17.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>18.636364</td>\n",
       "      <td>5.661673</td>\n",
       "      <td>11.0</td>\n",
       "      <td>15.5</td>\n",
       "      <td>18.0</td>\n",
       "      <td>21.00</td>\n",
       "      <td>29.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>41.181818</td>\n",
       "      <td>10.571832</td>\n",
       "      <td>31.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>48.00</td>\n",
       "      <td>63.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.452721</td>\n",
       "      <td>0.079768</td>\n",
       "      <td>0.352941</td>\n",
       "      <td>0.381818</td>\n",
       "      <td>0.460317</td>\n",
       "      <td>0.494444</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>11.0</td>\n",
       "      <td>25.272727</td>\n",
       "      <td>9.696297</td>\n",
       "      <td>12.0</td>\n",
       "      <td>19.50</td>\n",
       "      <td>26.0</td>\n",
       "      <td>28.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>55.272727</td>\n",
       "      <td>12.554608</td>\n",
       "      <td>40.0</td>\n",
       "      <td>45.00</td>\n",
       "      <td>54.0</td>\n",
       "      <td>61.00</td>\n",
       "      <td>82.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.404520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>29.545455</td>\n",
       "      <td>6.137811</td>\n",
       "      <td>22.0</td>\n",
       "      <td>24.00</td>\n",
       "      <td>30.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.063978</td>\n",
       "      <td>0.024546</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.049365</td>\n",
       "      <td>0.065819</td>\n",
       "      <td>0.072148</td>\n",
       "      <td>0.124044</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.257347</td>\n",
       "      <td>0.058454</td>\n",
       "      <td>0.186238</td>\n",
       "      <td>0.209518</td>\n",
       "      <td>0.251422</td>\n",
       "      <td>0.284013</td>\n",
       "      <td>0.381788</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.001024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.137562</td>\n",
       "      <td>0.028577</td>\n",
       "      <td>0.102431</td>\n",
       "      <td>0.111743</td>\n",
       "      <td>0.139679</td>\n",
       "      <td>0.153647</td>\n",
       "      <td>0.186238</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.064729</td>\n",
       "      <td>0.611259</td>\n",
       "      <td>0.787929</td>\n",
       "      <td>1.865594</td>\n",
       "      <td>2.108559</td>\n",
       "      <td>2.396889</td>\n",
       "      <td>3.201005</td>\n",
       "      <td>11.0</td>\n",
       "      <td>4.662551</td>\n",
       "      <td>0.658438</td>\n",
       "      <td>3.724677</td>\n",
       "      <td>4.401319</td>\n",
       "      <td>4.533561</td>\n",
       "      <td>4.790711</td>\n",
       "      <td>6.127005</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.719746e-07</td>\n",
       "      <td>3.022964e-07</td>\n",
       "      <td>8.904081e-13</td>\n",
       "      <td>1.325083e-10</td>\n",
       "      <td>1.429707e-09</td>\n",
       "      <td>1.884336e-07</td>\n",
       "      <td>8.404122e-07</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8.711978e-06</td>\n",
       "      <td>1.450839e-05</td>\n",
       "      <td>2.893826e-10</td>\n",
       "      <td>1.722608e-08</td>\n",
       "      <td>1.327586e-07</td>\n",
       "      <td>1.125794e-05</td>\n",
       "      <td>0.000039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_JAK2-DddA11_REP-1</th>\n",
       "      <td>274.0</td>\n",
       "      <td>8.185866e+07</td>\n",
       "      <td>5.691111e+07</td>\n",
       "      <td>2044988.0</td>\n",
       "      <td>34589019.50</td>\n",
       "      <td>73660869.5</td>\n",
       "      <td>1.215376e+08</td>\n",
       "      <td>246696638.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>8.185868e+07</td>\n",
       "      <td>5.691111e+07</td>\n",
       "      <td>2044998.0</td>\n",
       "      <td>34589037.5</td>\n",
       "      <td>73660888.5</td>\n",
       "      <td>1.215376e+08</td>\n",
       "      <td>246696654.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>9.277372</td>\n",
       "      <td>2.459131</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.00</td>\n",
       "      <td>18.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.040146</td>\n",
       "      <td>0.196661</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>9.237226</td>\n",
       "      <td>2.450671</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.00</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.75</td>\n",
       "      <td>18.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>39.218978</td>\n",
       "      <td>27.998356</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>47.00</td>\n",
       "      <td>196.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>79.452555</td>\n",
       "      <td>50.681650</td>\n",
       "      <td>20.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>66.5</td>\n",
       "      <td>96.00</td>\n",
       "      <td>337.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.486151</td>\n",
       "      <td>0.092111</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>0.413949</td>\n",
       "      <td>0.472490</td>\n",
       "      <td>0.549172</td>\n",
       "      <td>0.885714</td>\n",
       "      <td>274.0</td>\n",
       "      <td>22.354015</td>\n",
       "      <td>9.339479</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.00</td>\n",
       "      <td>22.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>106.054745</td>\n",
       "      <td>69.047581</td>\n",
       "      <td>27.0</td>\n",
       "      <td>59.00</td>\n",
       "      <td>86.0</td>\n",
       "      <td>132.75</td>\n",
       "      <td>443.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.105839</td>\n",
       "      <td>0.308195</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>61.357664</td>\n",
       "      <td>42.646642</td>\n",
       "      <td>20.0</td>\n",
       "      <td>31.00</td>\n",
       "      <td>49.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>263.0</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.056590</td>\n",
       "      <td>0.023643</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>0.040504</td>\n",
       "      <td>0.055693</td>\n",
       "      <td>0.068351</td>\n",
       "      <td>0.141765</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.649492</td>\n",
       "      <td>0.422856</td>\n",
       "      <td>0.165351</td>\n",
       "      <td>0.361323</td>\n",
       "      <td>0.526674</td>\n",
       "      <td>0.812977</td>\n",
       "      <td>2.712986</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.000780</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>274.0</td>\n",
       "      <td>0.375762</td>\n",
       "      <td>0.261173</td>\n",
       "      <td>0.122482</td>\n",
       "      <td>0.189848</td>\n",
       "      <td>0.300082</td>\n",
       "      <td>0.447061</td>\n",
       "      <td>1.610644</td>\n",
       "      <td>274.0</td>\n",
       "      <td>3.406059</td>\n",
       "      <td>0.876955</td>\n",
       "      <td>1.011468</td>\n",
       "      <td>2.809669</td>\n",
       "      <td>3.305550</td>\n",
       "      <td>3.934290</td>\n",
       "      <td>6.213102</td>\n",
       "      <td>274.0</td>\n",
       "      <td>5.757354</td>\n",
       "      <td>0.960699</td>\n",
       "      <td>4.055972</td>\n",
       "      <td>4.957687</td>\n",
       "      <td>5.704771</td>\n",
       "      <td>6.353241</td>\n",
       "      <td>9.022695</td>\n",
       "      <td>274.0</td>\n",
       "      <td>4.608038e-08</td>\n",
       "      <td>1.849436e-07</td>\n",
       "      <td>2.087440e-100</td>\n",
       "      <td>7.771788e-26</td>\n",
       "      <td>1.587117e-16</td>\n",
       "      <td>2.028121e-10</td>\n",
       "      <td>1.617993e-06</td>\n",
       "      <td>274.0</td>\n",
       "      <td>8.688948e-08</td>\n",
       "      <td>3.436207e-07</td>\n",
       "      <td>1.356836e-97</td>\n",
       "      <td>5.874026e-25</td>\n",
       "      <td>5.997826e-16</td>\n",
       "      <td>4.901360e-10</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_SIRT6-DddA11_REP-1</th>\n",
       "      <td>287.0</td>\n",
       "      <td>8.015725e+07</td>\n",
       "      <td>5.744566e+07</td>\n",
       "      <td>242415.0</td>\n",
       "      <td>33921669.50</td>\n",
       "      <td>70133094.0</td>\n",
       "      <td>1.193109e+08</td>\n",
       "      <td>246696638.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>8.015727e+07</td>\n",
       "      <td>5.744566e+07</td>\n",
       "      <td>242448.0</td>\n",
       "      <td>33921685.5</td>\n",
       "      <td>70133130.0</td>\n",
       "      <td>1.193109e+08</td>\n",
       "      <td>246696654.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>9.219512</td>\n",
       "      <td>2.561940</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.50</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.00</td>\n",
       "      <td>19.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.034843</td>\n",
       "      <td>0.183703</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>9.184669</td>\n",
       "      <td>2.541414</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.50</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.00</td>\n",
       "      <td>19.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>44.634146</td>\n",
       "      <td>31.882355</td>\n",
       "      <td>9.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>54.50</td>\n",
       "      <td>269.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>88.874564</td>\n",
       "      <td>57.708136</td>\n",
       "      <td>23.0</td>\n",
       "      <td>51.5</td>\n",
       "      <td>72.0</td>\n",
       "      <td>108.50</td>\n",
       "      <td>503.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.495395</td>\n",
       "      <td>0.091579</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>0.421683</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.550716</td>\n",
       "      <td>0.934783</td>\n",
       "      <td>287.0</td>\n",
       "      <td>21.867596</td>\n",
       "      <td>9.096793</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.00</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>117.261324</td>\n",
       "      <td>77.076916</td>\n",
       "      <td>31.0</td>\n",
       "      <td>68.00</td>\n",
       "      <td>93.0</td>\n",
       "      <td>143.00</td>\n",
       "      <td>669.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.108014</td>\n",
       "      <td>0.310940</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>68.397213</td>\n",
       "      <td>47.691415</td>\n",
       "      <td>20.0</td>\n",
       "      <td>36.00</td>\n",
       "      <td>54.0</td>\n",
       "      <td>84.5</td>\n",
       "      <td>341.0</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.055358</td>\n",
       "      <td>0.023029</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>0.040504</td>\n",
       "      <td>0.053162</td>\n",
       "      <td>0.068351</td>\n",
       "      <td>0.141765</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.652121</td>\n",
       "      <td>0.428645</td>\n",
       "      <td>0.172399</td>\n",
       "      <td>0.378166</td>\n",
       "      <td>0.517197</td>\n",
       "      <td>0.795260</td>\n",
       "      <td>3.720484</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.000787</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>287.0</td>\n",
       "      <td>0.380375</td>\n",
       "      <td>0.265224</td>\n",
       "      <td>0.111225</td>\n",
       "      <td>0.200205</td>\n",
       "      <td>0.300308</td>\n",
       "      <td>0.469927</td>\n",
       "      <td>1.896390</td>\n",
       "      <td>287.0</td>\n",
       "      <td>3.451020</td>\n",
       "      <td>0.829845</td>\n",
       "      <td>1.800227</td>\n",
       "      <td>2.847506</td>\n",
       "      <td>3.415520</td>\n",
       "      <td>4.009866</td>\n",
       "      <td>6.004406</td>\n",
       "      <td>287.0</td>\n",
       "      <td>5.775047</td>\n",
       "      <td>0.969127</td>\n",
       "      <td>3.849767</td>\n",
       "      <td>5.047696</td>\n",
       "      <td>5.658974</td>\n",
       "      <td>6.392643</td>\n",
       "      <td>8.651111</td>\n",
       "      <td>287.0</td>\n",
       "      <td>5.600235e-08</td>\n",
       "      <td>2.546377e-07</td>\n",
       "      <td>8.941195e-130</td>\n",
       "      <td>3.717371e-29</td>\n",
       "      <td>3.908722e-18</td>\n",
       "      <td>8.759697e-12</td>\n",
       "      <td>1.938486e-06</td>\n",
       "      <td>287.0</td>\n",
       "      <td>9.222598e-08</td>\n",
       "      <td>4.131651e-07</td>\n",
       "      <td>5.811777e-127</td>\n",
       "      <td>2.598162e-28</td>\n",
       "      <td>1.365208e-17</td>\n",
       "      <td>1.997826e-11</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             region_start                                                                                             region_end                                                      \\\n",
       "                                    count          mean           std         min          25%         50%           75%          max      count          mean           std         min         25%   \n",
       "<sample>                                                                                                                                                                                               \n",
       "DetectSeq_ATP8-DddA11_REP-1          31.0  8.600734e+07  4.994655e+07   2220193.0  50522662.50  78617702.0  1.133686e+08  201232409.0       31.0  8.600737e+07  4.994655e+07   2220214.0  50522683.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1           14.0  7.320022e+07  5.423443e+07   1056219.0  35163961.75  76116851.5  9.922139e+07  201232409.0       14.0  7.320025e+07  5.423443e+07   1056232.0  35163998.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1          11.0  6.947866e+07  5.711713e+07  15721070.0  27271003.50  45187694.0  1.018704e+08  201232409.0       11.0  6.947868e+07  5.711713e+07  15721094.0  27271029.5   \n",
       "DetectSeq_JAK2-DddA11_REP-1         274.0  8.185866e+07  5.691111e+07   2044988.0  34589019.50  73660869.5  1.215376e+08  246696638.0      274.0  8.185868e+07  5.691111e+07   2044998.0  34589037.5   \n",
       "DetectSeq_SIRT6-DddA11_REP-1        287.0  8.015725e+07  5.744566e+07    242415.0  33921669.50  70133094.0  1.193109e+08  246696638.0      287.0  8.015727e+07  5.744566e+07    242448.0  33921685.5   \n",
       "\n",
       "                                                                    region_site_num                                                    region_block_site_num                                          \\\n",
       "                                     50%           75%          max           count       mean       std  min   25%   50%    75%   max                 count      mean       std  min  25%  50%  75%   \n",
       "<sample>                                                                                                                                                                                               \n",
       "DetectSeq_ATP8-DddA11_REP-1   78617723.0  1.133686e+08  201232430.0            31.0  10.838710  3.327630  4.0  9.00  11.0  13.00  18.0                  31.0  0.064516  0.249731  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    76116884.5  9.922141e+07  201232430.0            14.0  11.214286  3.400873  6.0  8.25  11.0  13.75  18.0                  14.0  0.142857  0.363137  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   45187712.0  1.018704e+08  201232430.0            11.0  11.181818  3.429816  6.0  9.00  11.0  13.00  18.0                  11.0  0.090909  0.301511  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   73660888.5  1.215376e+08  246696654.0           274.0   9.277372  2.459131  4.0  8.00   9.0  11.00  18.0                 274.0  0.040146  0.196661  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  70133130.0  1.193109e+08  246696654.0           287.0   9.219512  2.561940  4.0  7.50   9.0  11.00  19.0                 287.0  0.034843  0.183703  0.0  0.0  0.0  0.0   \n",
       "\n",
       "                                  region_mut_site_num                                                    region_highest_site_mut_num                                                        \\\n",
       "                              max               count       mean       std  min   25%   50%    75%   max                       count       mean        std   min   25%   50%    75%    max   \n",
       "<sample>                                                                                                                                                                                     \n",
       "DetectSeq_ATP8-DddA11_REP-1   1.0                31.0  10.774194  3.263022  4.0  9.00  11.0  13.00  18.0                        31.0  21.225806  12.227591  10.0  13.0  18.0  24.50   68.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    1.0                14.0  11.071429  3.269069  6.0  8.25  10.5  13.75  17.0                        14.0  15.357143   5.865245   9.0  11.5  13.5  18.75   31.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   1.0                11.0  11.090909  3.238967  6.0  9.00  11.0  13.00  17.0                        11.0  18.636364   5.661673  11.0  15.5  18.0  21.00   29.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   1.0               274.0   9.237226  2.450671  4.0  8.00   9.0  10.75  18.0                       274.0  39.218978  27.998356   8.0  21.0  32.0  47.00  196.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1.0               287.0   9.184669  2.541414  4.0  7.50   9.0  11.00  19.0                       287.0  44.634146  31.882355   9.0  24.0  36.0  54.50  269.0   \n",
       "\n",
       "                             region_highest_site_cover_num                                                        region_highest_site_mut_ratio                                                    \\\n",
       "                                                     count       mean        std   min   25%   50%     75%    max                         count      mean       std       min       25%       50%   \n",
       "<sample>                                                                                                                                                                                            \n",
       "DetectSeq_ATP8-DddA11_REP-1                           31.0  50.645161  30.638264  25.0  31.0  41.0   59.00  174.0                          31.0  0.422565  0.053080  0.350000  0.378354  0.411765   \n",
       "DetectSeq_ATP8-DddA6_REP-1                            14.0  35.285714  14.440480  24.0  27.5  31.0   37.25   80.0                          14.0  0.438128  0.066487  0.360000  0.406539  0.416667   \n",
       "DetectSeq_ATP8-DddAwt_REP-1                           11.0  41.181818  10.571832  31.0  33.0  36.0   48.00   63.0                          11.0  0.452721  0.079768  0.352941  0.381818  0.460317   \n",
       "DetectSeq_JAK2-DddA11_REP-1                          274.0  79.452555  50.681650  20.0  45.0  66.5   96.00  337.0                         274.0  0.486151  0.092111  0.350000  0.413949  0.472490   \n",
       "DetectSeq_SIRT6-DddA11_REP-1                         287.0  88.874564  57.708136  23.0  51.5  72.0  108.50  503.0                         287.0  0.495395  0.091579  0.350000  0.421683  0.500000   \n",
       "\n",
       "                                                 ctrl_count                                                      treat_count                                                          ctrl_mut_count  \\\n",
       "                                   75%       max      count       mean        std   min    25%   50%   75%   max       count        mean        std   min    25%   50%     75%    max          count   \n",
       "<sample>                                                                                                                                                                                               \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.453463  0.545455       31.0  22.580645  10.760341   9.0  14.00  20.0  28.5  49.0        31.0   66.290323  40.377546  32.0  40.50  48.0   80.50  230.0           31.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.450779  0.633333       14.0  20.571429   5.720601  12.0  17.25  20.5  25.5  29.0        14.0   49.500000  19.606710  32.0  36.25  43.0   53.75  103.0           14.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.494444  0.606061       11.0  25.272727   9.696297  12.0  19.50  26.0  28.5  49.0        11.0   55.272727  12.554608  40.0  45.00  54.0   61.00   82.0           11.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.549172  0.885714      274.0  22.354015   9.339479   5.0  16.00  22.0  27.0  56.0       274.0  106.054745  69.047581  27.0  59.00  86.0  132.75  443.0          274.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.550716  0.934783      287.0  21.867596   9.096793   5.0  16.00  21.0  27.0  56.0       287.0  117.261324  77.076916  31.0  68.00  93.0  143.00  669.0          287.0   \n",
       "\n",
       "                                                                          treat_mut_count                                                       ctrl_count.norm                                \\\n",
       "                                  mean       std  min  25%  50%  75%  max           count       mean        std   min    25%   50%   75%    max           count      mean       std       min   \n",
       "<sample>                                                                                                                                                                                        \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.129032  0.340777  0.0  0.0  0.0  0.0  1.0            31.0  35.290323  16.925707  20.0  23.00  27.0  45.5   88.0            31.0  0.057163  0.027240  0.022784   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.142857  0.363137  0.0  0.0  0.0  0.0  1.0            14.0  27.857143   8.908571  20.0  21.25  24.5  31.5   48.0            14.0  0.052077  0.014482  0.030378   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.181818  0.404520  0.0  0.0  0.0  0.0  1.0            11.0  29.545455   6.137811  22.0  24.00  30.0  33.0   40.0            11.0  0.063978  0.024546  0.030378   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.105839  0.308195  0.0  0.0  0.0  0.0  1.0           274.0  61.357664  42.646642  20.0  31.00  49.0  73.0  263.0           274.0  0.056590  0.023643  0.012658   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.108014  0.310940  0.0  0.0  0.0  0.0  1.0           287.0  68.397213  47.691415  20.0  36.00  54.0  84.5  341.0           287.0  0.055358  0.023029  0.012658   \n",
       "\n",
       "                                                                     treat_count.norm                                                                       ctrl_mut_count.norm                      \\\n",
       "                                   25%       50%       75%       max            count      mean       std       min       25%       50%       75%       max               count      mean       std   \n",
       "<sample>                                                                                                                                                                                              \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.035441  0.050630  0.072148  0.124044             31.0  0.395146  0.240684  0.190747  0.241414  0.286121  0.479848  1.370995                31.0  0.000327  0.000863   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.043669  0.051896  0.064554  0.073414             14.0  0.318129  0.126009  0.205659  0.232973  0.276354  0.345443  0.661965                14.0  0.000362  0.000919   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.049365  0.065819  0.072148  0.124044             11.0  0.257347  0.058454  0.186238  0.209518  0.251422  0.284013  0.381788                11.0  0.000460  0.001024   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.040504  0.055693  0.068351  0.141765            274.0  0.649492  0.422856  0.165351  0.361323  0.526674  0.812977  2.712986               274.0  0.000268  0.000780   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.040504  0.053162  0.068351  0.141765            287.0  0.652121  0.428645  0.172399  0.378166  0.517197  0.795260  3.720484               287.0  0.000273  0.000787   \n",
       "\n",
       "                                                           treat_mut_count.norm                                                                       log2_FC                                          \\\n",
       "                              min  25%  50%  75%       max                count      mean       std       min       25%       50%       75%       max   count      mean       std       min       25%   \n",
       "<sample>                                                                                                                                                                                                \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.0  0.0  0.0  0.0  0.002532                 31.0  0.210360  0.100892  0.119217  0.137099  0.160943  0.271219  0.524554    31.0  2.763185  0.879470  1.369818  2.050195   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.0  0.0  0.0  0.0  0.002532                 14.0  0.179033  0.057254  0.128537  0.136570  0.157458  0.202446  0.308488    14.0  2.582972  0.705075  1.759147  2.125168   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.0  0.0  0.0  0.0  0.002532                 11.0  0.137562  0.028577  0.102431  0.111743  0.139679  0.153647  0.186238    11.0  2.064729  0.611259  0.787929  1.865594   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.0  0.0  0.0  0.0  0.002532                274.0  0.375762  0.261173  0.122482  0.189848  0.300082  0.447061  1.610644   274.0  3.406059  0.876955  1.011468  2.809669   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.0  0.0  0.0  0.0  0.002532                287.0  0.380375  0.265224  0.111225  0.200205  0.300308  0.469927  1.896390   287.0  3.451020  0.829845  1.800227  2.847506   \n",
       "\n",
       "                                                           log2_FC_mut                                                                       p_value                                             \\\n",
       "                                   50%       75%       max       count      mean       std       min       25%       50%       75%       max   count          mean           std            min   \n",
       "<sample>                                                                                                                                                                                          \n",
       "DetectSeq_ATP8-DddA11_REP-1   2.735088  3.345861  5.031376        31.0  5.091027  0.768367  3.879483  4.619327  4.952822  5.559368  7.444970    31.0  4.192761e-07  8.614629e-07   1.866343e-31   \n",
       "DetectSeq_ATP8-DddA6_REP-1    2.384784  2.869809  4.445647        14.0  4.982211  0.677993  4.324449  4.423941  4.875736  5.258613  6.701661    14.0  2.372577e-07  3.271692e-07   1.269741e-16   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   2.108559  2.396889  3.201005        11.0  4.662551  0.658438  3.724677  4.401319  4.533561  4.790711  6.127005    11.0  1.719746e-07  3.022964e-07   8.904081e-13   \n",
       "DetectSeq_JAK2-DddA11_REP-1   3.305550  3.934290  6.213102       274.0  5.757354  0.960699  4.055972  4.957687  5.704771  6.353241  9.022695   274.0  4.608038e-08  1.849436e-07  2.087440e-100   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  3.415520  4.009866  6.004406       287.0  5.775047  0.969127  3.849767  5.047696  5.658974  6.392643  8.651111   287.0  5.600235e-08  2.546377e-07  8.941195e-130   \n",
       "\n",
       "                                                                                        FDR                                                                                                 \n",
       "                                       25%           50%           75%           max  count          mean           std            min           25%           50%           75%       max  \n",
       "<sample>                                                                                                                                                                                    \n",
       "DetectSeq_ATP8-DddA11_REP-1   2.807248e-15  5.548434e-09  1.776817e-07  3.152999e-06   31.0  4.793318e-06  9.426111e-06   1.213123e-28  1.330349e-13  1.092873e-07  2.531712e-06  0.000033  \n",
       "DetectSeq_ATP8-DddA6_REP-1    1.086194e-09  7.676918e-08  4.191620e-07  1.064301e-06   14.0  7.773420e-06  9.629707e-06   8.253317e-14  8.472959e-08  3.623942e-06  1.412051e-05  0.000030  \n",
       "DetectSeq_ATP8-DddAwt_REP-1   1.325083e-10  1.429707e-09  1.884336e-07  8.404122e-07   11.0  8.711978e-06  1.450839e-05   2.893826e-10  1.722608e-08  1.327586e-07  1.125794e-05  0.000039  \n",
       "DetectSeq_JAK2-DddA11_REP-1   7.771788e-26  1.587117e-16  2.028121e-10  1.617993e-06  274.0  8.688948e-08  3.436207e-07   1.356836e-97  5.874026e-25  5.997826e-16  4.901360e-10  0.000003  \n",
       "DetectSeq_SIRT6-DddA11_REP-1  3.717371e-29  3.908722e-18  8.759697e-12  1.938486e-06  287.0  9.222598e-08  4.131651e-07  5.811777e-127  2.598162e-28  1.365208e-17  1.997826e-11  0.000003  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_strict = (df_old_site_new_signal\n",
    "                  .query('FDR <= 0.0001')  # 10911\n",
    "                  .query('log2_FC_mut >= 2')  # 10911\n",
    "                  .query('ctrl_mut_count <= 1')  # 9551\n",
    "                  .query('`treat_mut_count.norm` * 100 >= 10')  # 9551\n",
    "                  .query('treat_mut_count >= 20')  # 9551\n",
    "                  .query('treat_mut_count / treat_count >= 0.15')  # 6080\n",
    "                  .query('region_block_site_num <= 1')  # 5720\n",
    "                  .query('region_highest_site_mut_ratio >= 0.35')  # 2929\n",
    "                  )\n",
    "\n",
    "print(df_sign_strict.shape[0])\n",
    "df_sign_strict.groupby('<sample>').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "a2cffc45-6f3a-41f8-a08c-b053caeab06a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:53.006339Z",
     "iopub.status.busy": "2022-11-29T14:47:53.006071Z",
     "iopub.status.idle": "2022-11-29T14:47:53.459822Z",
     "shell.execute_reply": "2022-11-29T14:47:53.459005Z",
     "shell.execute_reply.started": "2022-11-29T14:47:53.006315Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "938\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_start</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_end</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_site_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_block_site_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_mut_site_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_highest_site_mut_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_highest_site_cover_num</th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_highest_site_mut_ratio</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_mut_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_mut_count</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ctrl_mut_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">treat_mut_count.norm</th>\n",
       "      <th colspan=\"8\" halign=\"left\">log2_FC</th>\n",
       "      <th colspan=\"8\" halign=\"left\">log2_FC_mut</th>\n",
       "      <th colspan=\"8\" halign=\"left\">p_value</th>\n",
       "      <th colspan=\"8\" halign=\"left\">FDR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA11_REP-1</th>\n",
       "      <td>95.0</td>\n",
       "      <td>8.606697e+07</td>\n",
       "      <td>5.376701e+07</td>\n",
       "      <td>1056219.0</td>\n",
       "      <td>43608049.50</td>\n",
       "      <td>76681403.0</td>\n",
       "      <td>1.188599e+08</td>\n",
       "      <td>226608903.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>8.606699e+07</td>\n",
       "      <td>5.376701e+07</td>\n",
       "      <td>1056232.0</td>\n",
       "      <td>43608069.00</td>\n",
       "      <td>76681420.0</td>\n",
       "      <td>118859940.0</td>\n",
       "      <td>226608925.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>10.168421</td>\n",
       "      <td>2.762410</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.50</td>\n",
       "      <td>18.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.031579</td>\n",
       "      <td>0.175804</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>10.136842</td>\n",
       "      <td>2.731263</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.50</td>\n",
       "      <td>18.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>14.189474</td>\n",
       "      <td>9.837288</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>16.00</td>\n",
       "      <td>68.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>37.073684</td>\n",
       "      <td>25.218721</td>\n",
       "      <td>11.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>42.50</td>\n",
       "      <td>174.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.383037</td>\n",
       "      <td>0.062045</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.370370</td>\n",
       "      <td>0.419872</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>95.0</td>\n",
       "      <td>22.252632</td>\n",
       "      <td>9.492403</td>\n",
       "      <td>6.0</td>\n",
       "      <td>17.00</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>47.852632</td>\n",
       "      <td>32.786533</td>\n",
       "      <td>15.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>56.5</td>\n",
       "      <td>230.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.136842</td>\n",
       "      <td>0.345504</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>23.715789</td>\n",
       "      <td>15.158498</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.50</td>\n",
       "      <td>18.0</td>\n",
       "      <td>27.50</td>\n",
       "      <td>88.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.056333</td>\n",
       "      <td>0.024030</td>\n",
       "      <td>0.015189</td>\n",
       "      <td>0.043036</td>\n",
       "      <td>0.053162</td>\n",
       "      <td>0.068351</td>\n",
       "      <td>0.141765</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.285242</td>\n",
       "      <td>0.195435</td>\n",
       "      <td>0.089413</td>\n",
       "      <td>0.172865</td>\n",
       "      <td>0.214590</td>\n",
       "      <td>0.336788</td>\n",
       "      <td>1.370995</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000875</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.141366</td>\n",
       "      <td>0.090357</td>\n",
       "      <td>0.059608</td>\n",
       "      <td>0.080471</td>\n",
       "      <td>0.107295</td>\n",
       "      <td>0.163923</td>\n",
       "      <td>0.524554</td>\n",
       "      <td>95.0</td>\n",
       "      <td>2.248123</td>\n",
       "      <td>0.871279</td>\n",
       "      <td>0.598087</td>\n",
       "      <td>1.584011</td>\n",
       "      <td>2.083514</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>5.031376</td>\n",
       "      <td>95.0</td>\n",
       "      <td>4.430105</td>\n",
       "      <td>0.873532</td>\n",
       "      <td>3.016986</td>\n",
       "      <td>3.694221</td>\n",
       "      <td>4.336251</td>\n",
       "      <td>5.034626</td>\n",
       "      <td>7.444970</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.000314</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>1.866343e-31</td>\n",
       "      <td>4.502563e-09</td>\n",
       "      <td>6.046509e-06</td>\n",
       "      <td>2.415574e-04</td>\n",
       "      <td>0.002439</td>\n",
       "      <td>95.0</td>\n",
       "      <td>0.001436</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>1.213123e-28</td>\n",
       "      <td>9.088318e-08</td>\n",
       "      <td>6.046509e-05</td>\n",
       "      <td>1.388242e-03</td>\n",
       "      <td>0.009909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA6_REP-1</th>\n",
       "      <td>45.0</td>\n",
       "      <td>7.500217e+07</td>\n",
       "      <td>5.349980e+07</td>\n",
       "      <td>819163.0</td>\n",
       "      <td>34500821.00</td>\n",
       "      <td>70144413.0</td>\n",
       "      <td>1.080633e+08</td>\n",
       "      <td>219416797.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>7.500219e+07</td>\n",
       "      <td>5.349980e+07</td>\n",
       "      <td>819192.0</td>\n",
       "      <td>34500856.00</td>\n",
       "      <td>70144447.0</td>\n",
       "      <td>108063381.0</td>\n",
       "      <td>219416815.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>10.222222</td>\n",
       "      <td>2.770698</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>12.00</td>\n",
       "      <td>18.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.044444</td>\n",
       "      <td>0.208409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>10.177778</td>\n",
       "      <td>2.707640</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>12.00</td>\n",
       "      <td>17.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>11.200000</td>\n",
       "      <td>4.827007</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.00</td>\n",
       "      <td>31.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>27.311111</td>\n",
       "      <td>11.908405</td>\n",
       "      <td>12.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>31.00</td>\n",
       "      <td>80.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.419022</td>\n",
       "      <td>0.091167</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.407407</td>\n",
       "      <td>0.450000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>45.0</td>\n",
       "      <td>21.666667</td>\n",
       "      <td>8.546663</td>\n",
       "      <td>6.0</td>\n",
       "      <td>17.00</td>\n",
       "      <td>21.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>37.466667</td>\n",
       "      <td>16.435272</td>\n",
       "      <td>17.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>103.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.088889</td>\n",
       "      <td>0.287799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>20.111111</td>\n",
       "      <td>8.296281</td>\n",
       "      <td>11.0</td>\n",
       "      <td>14.00</td>\n",
       "      <td>18.0</td>\n",
       "      <td>22.00</td>\n",
       "      <td>48.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.054849</td>\n",
       "      <td>0.021636</td>\n",
       "      <td>0.015189</td>\n",
       "      <td>0.043036</td>\n",
       "      <td>0.053162</td>\n",
       "      <td>0.068351</td>\n",
       "      <td>0.124044</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.240792</td>\n",
       "      <td>0.105627</td>\n",
       "      <td>0.109256</td>\n",
       "      <td>0.173525</td>\n",
       "      <td>0.218513</td>\n",
       "      <td>0.282781</td>\n",
       "      <td>0.661965</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.129251</td>\n",
       "      <td>0.053319</td>\n",
       "      <td>0.070695</td>\n",
       "      <td>0.089976</td>\n",
       "      <td>0.115683</td>\n",
       "      <td>0.141391</td>\n",
       "      <td>0.308488</td>\n",
       "      <td>45.0</td>\n",
       "      <td>2.138880</td>\n",
       "      <td>0.758834</td>\n",
       "      <td>0.908010</td>\n",
       "      <td>1.528534</td>\n",
       "      <td>2.040717</td>\n",
       "      <td>2.692033</td>\n",
       "      <td>4.445647</td>\n",
       "      <td>45.0</td>\n",
       "      <td>4.375650</td>\n",
       "      <td>0.688075</td>\n",
       "      <td>3.251110</td>\n",
       "      <td>3.860816</td>\n",
       "      <td>4.293545</td>\n",
       "      <td>4.710541</td>\n",
       "      <td>6.701661</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.000218</td>\n",
       "      <td>1.269741e-16</td>\n",
       "      <td>2.221841e-07</td>\n",
       "      <td>4.900206e-06</td>\n",
       "      <td>9.911734e-05</td>\n",
       "      <td>0.000883</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.001184</td>\n",
       "      <td>0.002270</td>\n",
       "      <td>8.253317e-14</td>\n",
       "      <td>9.026229e-06</td>\n",
       "      <td>1.137548e-04</td>\n",
       "      <td>1.370772e-03</td>\n",
       "      <td>0.008970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddAwt_REP-1</th>\n",
       "      <td>30.0</td>\n",
       "      <td>8.786307e+07</td>\n",
       "      <td>5.704240e+07</td>\n",
       "      <td>8410441.0</td>\n",
       "      <td>40895779.00</td>\n",
       "      <td>76076125.5</td>\n",
       "      <td>1.193665e+08</td>\n",
       "      <td>219416797.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>8.786310e+07</td>\n",
       "      <td>5.704240e+07</td>\n",
       "      <td>8410462.0</td>\n",
       "      <td>40895795.50</td>\n",
       "      <td>76076167.0</td>\n",
       "      <td>119366554.0</td>\n",
       "      <td>219416815.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.566667</td>\n",
       "      <td>2.920597</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.5</td>\n",
       "      <td>12.75</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.182574</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>10.533333</td>\n",
       "      <td>2.837353</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.5</td>\n",
       "      <td>12.75</td>\n",
       "      <td>17.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>5.699365</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>16.25</td>\n",
       "      <td>29.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>29.666667</td>\n",
       "      <td>11.771776</td>\n",
       "      <td>12.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>33.00</td>\n",
       "      <td>63.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.443145</td>\n",
       "      <td>0.089441</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.443198</td>\n",
       "      <td>0.522059</td>\n",
       "      <td>0.606061</td>\n",
       "      <td>30.0</td>\n",
       "      <td>22.600000</td>\n",
       "      <td>10.591473</td>\n",
       "      <td>8.0</td>\n",
       "      <td>14.75</td>\n",
       "      <td>21.5</td>\n",
       "      <td>27.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>39.433333</td>\n",
       "      <td>15.475639</td>\n",
       "      <td>18.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>45.5</td>\n",
       "      <td>82.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.379049</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>20.600000</td>\n",
       "      <td>8.084041</td>\n",
       "      <td>13.0</td>\n",
       "      <td>14.00</td>\n",
       "      <td>17.5</td>\n",
       "      <td>24.50</td>\n",
       "      <td>40.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.057212</td>\n",
       "      <td>0.026812</td>\n",
       "      <td>0.020252</td>\n",
       "      <td>0.037340</td>\n",
       "      <td>0.054428</td>\n",
       "      <td>0.068351</td>\n",
       "      <td>0.124044</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.183600</td>\n",
       "      <td>0.072054</td>\n",
       "      <td>0.083807</td>\n",
       "      <td>0.135023</td>\n",
       "      <td>0.162958</td>\n",
       "      <td>0.211846</td>\n",
       "      <td>0.381788</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0.000960</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.095913</td>\n",
       "      <td>0.037639</td>\n",
       "      <td>0.060527</td>\n",
       "      <td>0.065183</td>\n",
       "      <td>0.081479</td>\n",
       "      <td>0.114071</td>\n",
       "      <td>0.186238</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.734376</td>\n",
       "      <td>0.640308</td>\n",
       "      <td>0.608988</td>\n",
       "      <td>1.186891</td>\n",
       "      <td>1.879077</td>\n",
       "      <td>2.142457</td>\n",
       "      <td>3.201005</td>\n",
       "      <td>30.0</td>\n",
       "      <td>4.061864</td>\n",
       "      <td>0.731220</td>\n",
       "      <td>3.008470</td>\n",
       "      <td>3.521385</td>\n",
       "      <td>3.860675</td>\n",
       "      <td>4.568028</td>\n",
       "      <td>6.127005</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>8.904081e-13</td>\n",
       "      <td>1.279882e-07</td>\n",
       "      <td>1.958503e-05</td>\n",
       "      <td>2.776883e-04</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.002414</td>\n",
       "      <td>0.003014</td>\n",
       "      <td>2.893826e-10</td>\n",
       "      <td>8.068019e-06</td>\n",
       "      <td>5.551199e-04</td>\n",
       "      <td>5.220160e-03</td>\n",
       "      <td>0.008736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_JAK2-DddA11_REP-1</th>\n",
       "      <td>362.0</td>\n",
       "      <td>8.030659e+07</td>\n",
       "      <td>5.749051e+07</td>\n",
       "      <td>819163.0</td>\n",
       "      <td>34041683.00</td>\n",
       "      <td>69810892.5</td>\n",
       "      <td>1.215376e+08</td>\n",
       "      <td>246696638.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>8.030661e+07</td>\n",
       "      <td>5.749051e+07</td>\n",
       "      <td>819192.0</td>\n",
       "      <td>34041698.00</td>\n",
       "      <td>69810913.0</td>\n",
       "      <td>121537589.5</td>\n",
       "      <td>246696654.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>9.138122</td>\n",
       "      <td>2.477094</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.00</td>\n",
       "      <td>20.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.035912</td>\n",
       "      <td>0.186327</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>9.102210</td>\n",
       "      <td>2.465956</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.00</td>\n",
       "      <td>20.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>33.121547</td>\n",
       "      <td>27.035867</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>25.5</td>\n",
       "      <td>41.00</td>\n",
       "      <td>196.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>69.566298</td>\n",
       "      <td>49.873050</td>\n",
       "      <td>10.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>86.75</td>\n",
       "      <td>337.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.464163</td>\n",
       "      <td>0.103841</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.380238</td>\n",
       "      <td>0.450490</td>\n",
       "      <td>0.538316</td>\n",
       "      <td>0.885714</td>\n",
       "      <td>362.0</td>\n",
       "      <td>21.616022</td>\n",
       "      <td>8.973588</td>\n",
       "      <td>5.0</td>\n",
       "      <td>15.00</td>\n",
       "      <td>21.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>92.734807</td>\n",
       "      <td>67.660316</td>\n",
       "      <td>15.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>443.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.110497</td>\n",
       "      <td>0.313942</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>52.591160</td>\n",
       "      <td>41.301858</td>\n",
       "      <td>10.0</td>\n",
       "      <td>25.25</td>\n",
       "      <td>40.0</td>\n",
       "      <td>65.75</td>\n",
       "      <td>263.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.054721</td>\n",
       "      <td>0.022717</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>0.037973</td>\n",
       "      <td>0.053162</td>\n",
       "      <td>0.065819</td>\n",
       "      <td>0.141765</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.567919</td>\n",
       "      <td>0.414360</td>\n",
       "      <td>0.091862</td>\n",
       "      <td>0.281710</td>\n",
       "      <td>0.447061</td>\n",
       "      <td>0.704274</td>\n",
       "      <td>2.712986</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.000280</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.322075</td>\n",
       "      <td>0.252938</td>\n",
       "      <td>0.061241</td>\n",
       "      <td>0.154634</td>\n",
       "      <td>0.244965</td>\n",
       "      <td>0.402661</td>\n",
       "      <td>1.610644</td>\n",
       "      <td>362.0</td>\n",
       "      <td>3.191582</td>\n",
       "      <td>0.927139</td>\n",
       "      <td>0.933465</td>\n",
       "      <td>2.530578</td>\n",
       "      <td>3.156033</td>\n",
       "      <td>3.776270</td>\n",
       "      <td>6.213102</td>\n",
       "      <td>362.0</td>\n",
       "      <td>5.460010</td>\n",
       "      <td>1.106517</td>\n",
       "      <td>2.918468</td>\n",
       "      <td>4.722598</td>\n",
       "      <td>5.382253</td>\n",
       "      <td>6.163651</td>\n",
       "      <td>9.022695</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>2.087440e-100</td>\n",
       "      <td>7.959148e-23</td>\n",
       "      <td>2.506066e-13</td>\n",
       "      <td>2.201052e-08</td>\n",
       "      <td>0.003405</td>\n",
       "      <td>362.0</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>1.356836e-97</td>\n",
       "      <td>4.703133e-22</td>\n",
       "      <td>7.657920e-13</td>\n",
       "      <td>4.534517e-08</td>\n",
       "      <td>0.005030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_SIRT6-DddA11_REP-1</th>\n",
       "      <td>406.0</td>\n",
       "      <td>8.120440e+07</td>\n",
       "      <td>5.850608e+07</td>\n",
       "      <td>242415.0</td>\n",
       "      <td>33993359.75</td>\n",
       "      <td>70429299.5</td>\n",
       "      <td>1.220325e+08</td>\n",
       "      <td>246696638.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>8.120442e+07</td>\n",
       "      <td>5.850608e+07</td>\n",
       "      <td>242448.0</td>\n",
       "      <td>33993378.75</td>\n",
       "      <td>70429323.5</td>\n",
       "      <td>122032497.5</td>\n",
       "      <td>246696654.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>9.024631</td>\n",
       "      <td>2.481414</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.00</td>\n",
       "      <td>19.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.039409</td>\n",
       "      <td>0.194806</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>8.985222</td>\n",
       "      <td>2.472521</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.00</td>\n",
       "      <td>19.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>36.724138</td>\n",
       "      <td>30.225780</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>28.5</td>\n",
       "      <td>46.75</td>\n",
       "      <td>269.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>77.411330</td>\n",
       "      <td>55.624853</td>\n",
       "      <td>12.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>97.75</td>\n",
       "      <td>503.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.460399</td>\n",
       "      <td>0.105733</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.444878</td>\n",
       "      <td>0.534861</td>\n",
       "      <td>0.934783</td>\n",
       "      <td>406.0</td>\n",
       "      <td>21.842365</td>\n",
       "      <td>9.103779</td>\n",
       "      <td>5.0</td>\n",
       "      <td>16.00</td>\n",
       "      <td>21.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>101.763547</td>\n",
       "      <td>74.138930</td>\n",
       "      <td>16.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>669.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.125616</td>\n",
       "      <td>0.331825</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>57.278325</td>\n",
       "      <td>45.868560</td>\n",
       "      <td>10.0</td>\n",
       "      <td>26.00</td>\n",
       "      <td>44.0</td>\n",
       "      <td>74.00</td>\n",
       "      <td>341.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.055294</td>\n",
       "      <td>0.023046</td>\n",
       "      <td>0.012658</td>\n",
       "      <td>0.040504</td>\n",
       "      <td>0.053162</td>\n",
       "      <td>0.065819</td>\n",
       "      <td>0.141765</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.565934</td>\n",
       "      <td>0.412306</td>\n",
       "      <td>0.088980</td>\n",
       "      <td>0.289186</td>\n",
       "      <td>0.450462</td>\n",
       "      <td>0.711841</td>\n",
       "      <td>3.720484</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.318540</td>\n",
       "      <td>0.255087</td>\n",
       "      <td>0.055613</td>\n",
       "      <td>0.144593</td>\n",
       "      <td>0.244695</td>\n",
       "      <td>0.411533</td>\n",
       "      <td>1.896390</td>\n",
       "      <td>406.0</td>\n",
       "      <td>3.183387</td>\n",
       "      <td>0.930393</td>\n",
       "      <td>0.971913</td>\n",
       "      <td>2.532717</td>\n",
       "      <td>3.166664</td>\n",
       "      <td>3.806076</td>\n",
       "      <td>6.004406</td>\n",
       "      <td>406.0</td>\n",
       "      <td>5.438947</td>\n",
       "      <td>1.137137</td>\n",
       "      <td>2.773504</td>\n",
       "      <td>4.628070</td>\n",
       "      <td>5.419245</td>\n",
       "      <td>6.213994</td>\n",
       "      <td>8.651111</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000350</td>\n",
       "      <td>8.941195e-130</td>\n",
       "      <td>8.315924e-26</td>\n",
       "      <td>1.286622e-14</td>\n",
       "      <td>2.341810e-08</td>\n",
       "      <td>0.004223</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>5.811777e-127</td>\n",
       "      <td>4.547063e-25</td>\n",
       "      <td>3.587809e-14</td>\n",
       "      <td>4.290270e-08</td>\n",
       "      <td>0.005458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             region_start                                                                                            region_end                                                      \\\n",
       "                                    count          mean           std        min          25%         50%           75%          max      count          mean           std        min          25%   \n",
       "<sample>                                                                                                                                                                                              \n",
       "DetectSeq_ATP8-DddA11_REP-1          95.0  8.606697e+07  5.376701e+07  1056219.0  43608049.50  76681403.0  1.188599e+08  226608903.0       95.0  8.606699e+07  5.376701e+07  1056232.0  43608069.00   \n",
       "DetectSeq_ATP8-DddA6_REP-1           45.0  7.500217e+07  5.349980e+07   819163.0  34500821.00  70144413.0  1.080633e+08  219416797.0       45.0  7.500219e+07  5.349980e+07   819192.0  34500856.00   \n",
       "DetectSeq_ATP8-DddAwt_REP-1          30.0  8.786307e+07  5.704240e+07  8410441.0  40895779.00  76076125.5  1.193665e+08  219416797.0       30.0  8.786310e+07  5.704240e+07  8410462.0  40895795.50   \n",
       "DetectSeq_JAK2-DddA11_REP-1         362.0  8.030659e+07  5.749051e+07   819163.0  34041683.00  69810892.5  1.215376e+08  246696638.0      362.0  8.030661e+07  5.749051e+07   819192.0  34041698.00   \n",
       "DetectSeq_SIRT6-DddA11_REP-1        406.0  8.120440e+07  5.850608e+07   242415.0  33993359.75  70429299.5  1.220325e+08  246696638.0      406.0  8.120442e+07  5.850608e+07   242448.0  33993378.75   \n",
       "\n",
       "                                                                   region_site_num                                                   region_block_site_num                                          \\\n",
       "                                     50%          75%          max           count       mean       std  min  25%   50%    75%   max                 count      mean       std  min  25%  50%  75%   \n",
       "<sample>                                                                                                                                                                                             \n",
       "DetectSeq_ATP8-DddA11_REP-1   76681420.0  118859940.0  226608925.0            95.0  10.168421  2.762410  4.0  8.0  10.0  11.50  18.0                  95.0  0.031579  0.175804  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    70144447.0  108063381.0  219416815.0            45.0  10.222222  2.770698  3.0  8.0  10.0  12.00  18.0                  45.0  0.044444  0.208409  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   76076167.0  119366554.0  219416815.0            30.0  10.566667  2.920597  6.0  9.0  10.5  12.75  18.0                  30.0  0.033333  0.182574  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   69810913.0  121537589.5  246696654.0           362.0   9.138122  2.477094  3.0  7.0   9.0  11.00  20.0                 362.0  0.035912  0.186327  0.0  0.0  0.0  0.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  70429323.5  122032497.5  246696654.0           406.0   9.024631  2.481414  3.0  7.0   9.0  11.00  19.0                 406.0  0.039409  0.194806  0.0  0.0  0.0  0.0   \n",
       "\n",
       "                                  region_mut_site_num                                                   region_highest_site_mut_num                                                       \\\n",
       "                              max               count       mean       std  min  25%   50%    75%   max                       count       mean        std  min   25%   50%    75%    max   \n",
       "<sample>                                                                                                                                                                                   \n",
       "DetectSeq_ATP8-DddA11_REP-1   1.0                95.0  10.136842  2.731263  4.0  8.0  10.0  11.50  18.0                        95.0  14.189474   9.837288  4.0   8.0  11.0  16.00   68.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    1.0                45.0  10.177778  2.707640  3.0  8.0  10.0  12.00  17.0                        45.0  11.200000   4.827007  5.0   9.0  10.0  13.00   31.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   1.0                30.0  10.533333  2.837353  6.0  9.0  10.5  12.75  17.0                        30.0  13.000000   5.699365  5.0   9.0  12.0  16.25   29.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   1.0               362.0   9.102210  2.465956  3.0  7.0   9.0  10.00  20.0                       362.0  33.121547  27.035867  4.0  15.0  25.5  41.00  196.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1.0               406.0   8.985222  2.472521  3.0  7.0   9.0  10.00  19.0                       406.0  36.724138  30.225780  5.0  17.0  28.5  46.75  269.0   \n",
       "\n",
       "                             region_highest_site_cover_num                                                       region_highest_site_mut_ratio                                                         \\\n",
       "                                                     count       mean        std   min   25%   50%    75%    max                         count      mean       std  min       25%       50%       75%   \n",
       "<sample>                                                                                                                                                                                                \n",
       "DetectSeq_ATP8-DddA11_REP-1                           95.0  37.073684  25.218721  11.0  23.0  29.0  42.50  174.0                          95.0  0.383037  0.062045  0.3  0.333333  0.370370  0.419872   \n",
       "DetectSeq_ATP8-DddA6_REP-1                            45.0  27.311111  11.908405  12.0  20.0  24.0  31.00   80.0                          45.0  0.419022  0.091167  0.3  0.357143  0.407407  0.450000   \n",
       "DetectSeq_ATP8-DddAwt_REP-1                           30.0  29.666667  11.771776  12.0  22.0  27.5  33.00   63.0                          30.0  0.443145  0.089441  0.3  0.363636  0.443198  0.522059   \n",
       "DetectSeq_JAK2-DddA11_REP-1                          362.0  69.566298  49.873050  10.0  36.0  56.0  86.75  337.0                         362.0  0.464163  0.103841  0.3  0.380238  0.450490  0.538316   \n",
       "DetectSeq_SIRT6-DddA11_REP-1                         406.0  77.411330  55.624853  12.0  41.0  61.0  97.75  503.0                         406.0  0.460399  0.105733  0.3  0.375000  0.444878  0.534861   \n",
       "\n",
       "                                       ctrl_count                                                     treat_count                                                        ctrl_mut_count            \\\n",
       "                                   max      count       mean        std  min    25%   50%   75%   max       count        mean        std   min   25%   50%    75%    max          count      mean   \n",
       "<sample>                                                                                                                                                                                            \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.545455       95.0  22.252632   9.492403  6.0  17.00  21.0  27.0  56.0        95.0   47.852632  32.786533  15.0  29.0  36.0   56.5  230.0           95.0  0.136842   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.666667       45.0  21.666667   8.546663  6.0  17.00  21.0  27.0  49.0        45.0   37.466667  16.435272  17.0  27.0  34.0   44.0  103.0           45.0  0.088889   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.606061       30.0  22.600000  10.591473  8.0  14.75  21.5  27.0  49.0        30.0   39.433333  15.475639  18.0  29.0  35.0   45.5   82.0           30.0  0.166667   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.885714      362.0  21.616022   8.973588  5.0  15.00  21.0  26.0  56.0       362.0   92.734807  67.660316  15.0  46.0  73.0  115.0  443.0          362.0  0.110497   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.934783      406.0  21.842365   9.103779  5.0  16.00  21.0  26.0  56.0       406.0  101.763547  74.138930  16.0  52.0  81.0  128.0  669.0          406.0  0.125616   \n",
       "\n",
       "                                                                treat_mut_count                                                        ctrl_count.norm                                          \\\n",
       "                                   std  min  25%  50%  75%  max           count       mean        std   min    25%   50%    75%    max           count      mean       std       min       25%   \n",
       "<sample>                                                                                                                                                                                         \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.345504  0.0  0.0  0.0  0.0  1.0            95.0  23.715789  15.158498  10.0  13.50  18.0  27.50   88.0            95.0  0.056333  0.024030  0.015189  0.043036   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.287799  0.0  0.0  0.0  0.0  1.0            45.0  20.111111   8.296281  11.0  14.00  18.0  22.00   48.0            45.0  0.054849  0.021636  0.015189  0.043036   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.379049  0.0  0.0  0.0  0.0  1.0            30.0  20.600000   8.084041  13.0  14.00  17.5  24.50   40.0            30.0  0.057212  0.026812  0.020252  0.037340   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.313942  0.0  0.0  0.0  0.0  1.0           362.0  52.591160  41.301858  10.0  25.25  40.0  65.75  263.0           362.0  0.054721  0.022717  0.012658  0.037973   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.331825  0.0  0.0  0.0  0.0  1.0           406.0  57.278325  45.868560  10.0  26.00  44.0  74.00  341.0           406.0  0.055294  0.023046  0.012658  0.040504   \n",
       "\n",
       "                                                           treat_count.norm                                                                       ctrl_mut_count.norm                                \\\n",
       "                                   50%       75%       max            count      mean       std       min       25%       50%       75%       max               count      mean       std  min  25%   \n",
       "<sample>                                                                                                                                                                                              \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.053162  0.068351  0.141765             95.0  0.285242  0.195435  0.089413  0.172865  0.214590  0.336788  1.370995                95.0  0.000346  0.000875  0.0  0.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.053162  0.068351  0.124044             45.0  0.240792  0.105627  0.109256  0.173525  0.218513  0.282781  0.661965                45.0  0.000225  0.000729  0.0  0.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.054428  0.068351  0.124044             30.0  0.183600  0.072054  0.083807  0.135023  0.162958  0.211846  0.381788                30.0  0.000422  0.000960  0.0  0.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.053162  0.065819  0.141765            362.0  0.567919  0.414360  0.091862  0.281710  0.447061  0.704274  2.712986               362.0  0.000280  0.000795  0.0  0.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.053162  0.065819  0.141765            406.0  0.565934  0.412306  0.088980  0.289186  0.450462  0.711841  3.720484               406.0  0.000318  0.000840  0.0  0.0   \n",
       "\n",
       "                                                 treat_mut_count.norm                                                                       log2_FC                                                    \\\n",
       "                              50%  75%       max                count      mean       std       min       25%       50%       75%       max   count      mean       std       min       25%       50%   \n",
       "<sample>                                                                                                                                                                                                \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.0  0.0  0.002532                 95.0  0.141366  0.090357  0.059608  0.080471  0.107295  0.163923  0.524554    95.0  2.248123  0.871279  0.598087  1.584011  2.083514   \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.0  0.0  0.002532                 45.0  0.129251  0.053319  0.070695  0.089976  0.115683  0.141391  0.308488    45.0  2.138880  0.758834  0.908010  1.528534  2.040717   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.0  0.0  0.002532                 30.0  0.095913  0.037639  0.060527  0.065183  0.081479  0.114071  0.186238    30.0  1.734376  0.640308  0.608988  1.186891  1.879077   \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.0  0.0  0.002532                362.0  0.322075  0.252938  0.061241  0.154634  0.244965  0.402661  1.610644   362.0  3.191582  0.927139  0.933465  2.530578  3.156033   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.0  0.0  0.002532                406.0  0.318540  0.255087  0.055613  0.144593  0.244695  0.411533  1.896390   406.0  3.183387  0.930393  0.971913  2.532717  3.166664   \n",
       "\n",
       "                                                 log2_FC_mut                                                                       p_value                                                   \\\n",
       "                                   75%       max       count      mean       std       min       25%       50%       75%       max   count      mean       std            min           25%   \n",
       "<sample>                                                                                                                                                                                      \n",
       "DetectSeq_ATP8-DddA11_REP-1   2.820479  5.031376        95.0  4.430105  0.873532  3.016986  3.694221  4.336251  5.034626  7.444970    95.0  0.000314  0.000650   1.866343e-31  4.502563e-09   \n",
       "DetectSeq_ATP8-DddA6_REP-1    2.692033  4.445647        45.0  4.375650  0.688075  3.251110  3.860816  4.293545  4.710541  6.701661    45.0  0.000101  0.000218   1.269741e-16  2.221841e-07   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   2.142457  3.201005        30.0  4.061864  0.731220  3.008470  3.521385  3.860675  4.568028  6.127005    30.0  0.000147  0.000200   8.904081e-13  1.279882e-07   \n",
       "DetectSeq_JAK2-DddA11_REP-1   3.776270  6.213102       362.0  5.460010  1.106517  2.918468  4.722598  5.382253  6.163651  9.022695   362.0  0.000074  0.000358  2.087440e-100  7.959148e-23   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  3.806076  6.004406       406.0  5.438947  1.137137  2.773504  4.628070  5.419245  6.213994  8.651111   406.0  0.000064  0.000350  8.941195e-130  8.315924e-26   \n",
       "\n",
       "                                                                      FDR                                                                                         \n",
       "                                       50%           75%       max  count      mean       std            min           25%           50%           75%       max  \n",
       "<sample>                                                                                                                                                          \n",
       "DetectSeq_ATP8-DddA11_REP-1   6.046509e-06  2.415574e-04  0.002439   95.0  0.001436  0.002767   1.213123e-28  9.088318e-08  6.046509e-05  1.388242e-03  0.009909  \n",
       "DetectSeq_ATP8-DddA6_REP-1    4.900206e-06  9.911734e-05  0.000883   45.0  0.001184  0.002270   8.253317e-14  9.026229e-06  1.137548e-04  1.370772e-03  0.008970  \n",
       "DetectSeq_ATP8-DddAwt_REP-1   1.958503e-05  2.776883e-04  0.000605   30.0  0.002414  0.003014   2.893826e-10  8.068019e-06  5.551199e-04  5.220160e-03  0.008736  \n",
       "DetectSeq_JAK2-DddA11_REP-1   2.506066e-13  2.201052e-08  0.003405  362.0  0.000112  0.000534   1.356836e-97  4.703133e-22  7.657920e-13  4.534517e-08  0.005030  \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1.286622e-14  2.341810e-08  0.004223  406.0  0.000086  0.000458  5.811777e-127  4.547063e-25  3.587809e-14  4.290270e-08  0.005458  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_lenient = (df_old_site_new_signal\n",
    "                   .query('FDR <= 0.01')  # 25118\n",
    "                   .query('log2_FC_mut >= 2')  # 25118\n",
    "                   .query('ctrl_mut_count <= 1')  # 22009\n",
    "                   .query('`treat_mut_count.norm` * 100 >= 5')  # 22009\n",
    "                   .query('treat_mut_count >= 10')  # 22009\n",
    "                   .query('treat_mut_count / treat_count >= 0.15')  # 14476\n",
    "                   .query('region_block_site_num <= 1')  # 13522\n",
    "                   .query('region_highest_site_mut_ratio >= 0.30')  # 7001\n",
    "                   )\n",
    "\n",
    "print(df_sign_lenient.shape[0])\n",
    "df_sign_lenient.groupby('<sample>').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "6b0fc442-f6ab-4c97-b03c-383c59f42b30",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:53.944172Z",
     "iopub.status.busy": "2022-11-29T14:47:53.943748Z",
     "iopub.status.idle": "2022-11-29T14:47:53.977346Z",
     "shell.execute_reply": "2022-11-29T14:47:53.976635Z",
     "shell.execute_reply.started": "2022-11-29T14:47:53.944148Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18907111</td>\n",
       "      <td>18907133</td>\n",
       "      <td>chr1_18907111_18907133</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>chr1_18907111_GA,chr1_18907119_...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_18907133_GA</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>0.611111</td>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0.070882</td>\n",
       "      <td>0.154244</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.109256</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 28,0,0,0,0,...</td>\n",
       "      <td>1.121717</td>\n",
       "      <td>3.753610</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.206580e-05</td>\n",
       "      <td>4.395162e-04</td>\n",
       "      <td>chr1_18907111_18907133_highest_...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67532557</td>\n",
       "      <td>67532593</td>\n",
       "      <td>chr1_67532557_67532593</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>chr1_67532557_GA,chr1_67532565_...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_67532592_GA</td>\n",
       "      <td>15</td>\n",
       "      <td>44</td>\n",
       "      <td>0.340909</td>\n",
       "      <td>49</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>0.124044</td>\n",
       "      <td>0.385610</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.212086</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10 46,3,0,0...</td>\n",
       "      <td>1.636290</td>\n",
       "      <td>4.710541</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.009651e-10</td>\n",
       "      <td>9.375327e-09</td>\n",
       "      <td>chr1_67532557_67532593_highest_...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>153223231</td>\n",
       "      <td>153223253</td>\n",
       "      <td>chr1_153223231_153223253</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr1_153223231_CT,chr1_15322323...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_153223231_CT</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>23</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0.058225</td>\n",
       "      <td>0.109256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.077122</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 20...</td>\n",
       "      <td>0.908010</td>\n",
       "      <td>3.251110</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>7.761196e-04</td>\n",
       "      <td>8.270126e-03</td>\n",
       "      <td>chr1_153223231_153223253_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>161209983</td>\n",
       "      <td>161210006</td>\n",
       "      <td>chr1_161209983_161210006</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr1_161209983_GA,chr1_16120998...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_161210002_GA</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>0.588235</td>\n",
       "      <td>24</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0.060756</td>\n",
       "      <td>0.141391</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.102829</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 19...</td>\n",
       "      <td>1.218579</td>\n",
       "      <td>5.344109</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.566141e-05</td>\n",
       "      <td>7.933340e-04</td>\n",
       "      <td>chr1_161209983_161210006_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>219416797</td>\n",
       "      <td>219416815</td>\n",
       "      <td>chr1_219416797_219416815</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>chr1_219416797_CT,chr1_21941679...</td>\n",
       "      <td>N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_219416798_CT</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>0.347826</td>\n",
       "      <td>12</td>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.199232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102829</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8 12,0,0,0,0,0,...</td>\n",
       "      <td>2.713343</td>\n",
       "      <td>3.666147</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.566141e-05</td>\n",
       "      <td>7.933340e-04</td>\n",
       "      <td>chr1_219416797_219416815_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3243</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>132292961</td>\n",
       "      <td>132292983</td>\n",
       "      <td>chrX_132292961_132292983</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>chrX_132292961_GA,chrX_13229296...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_132292983_GA</td>\n",
       "      <td>8</td>\n",
       "      <td>26</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>0.222450</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.077858</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 39,0,1,0,0,...</td>\n",
       "      <td>1.135412</td>\n",
       "      <td>4.942767</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.408338e-04</td>\n",
       "      <td>2.007500e-04</td>\n",
       "      <td>chrX_132292961_132292983_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrX_136925330_CT,chrX_13692533...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chrX_136925330_CT</td>\n",
       "      <td>42</td>\n",
       "      <td>103</td>\n",
       "      <td>0.407767</td>\n",
       "      <td>43</td>\n",
       "      <td>129</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "      <td>0.108855</td>\n",
       "      <td>0.717403</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.328114</td>\n",
       "      <td>0,1,2,3,4,5 43,0,0,0,0,0 27,43,...</td>\n",
       "      <td>2.720374</td>\n",
       "      <td>5.608487</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.433039e-20</td>\n",
       "      <td>9.702301e-20</td>\n",
       "      <td>chrX_136925330_136925338_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3246</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>chrX_138128211_GA,chrX_13812821...</td>\n",
       "      <td>N-N-N-B-N-N-N-N-N</td>\n",
       "      <td>chrX_138128237_GA</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>33</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>0.205767</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.072296</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 32,0,1,0,0,...</td>\n",
       "      <td>1.300471</td>\n",
       "      <td>4.835851</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.889936e-04</td>\n",
       "      <td>4.053103e-04</td>\n",
       "      <td>chrX_138128211_138128237_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>chrX_153669436_CT,chrX_15366943...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_153669436_CT</td>\n",
       "      <td>7</td>\n",
       "      <td>22</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>22</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0.055693</td>\n",
       "      <td>0.144593</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.061174</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10 22,0,0,0...</td>\n",
       "      <td>1.376420</td>\n",
       "      <td>3.185275</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.185269e-03</td>\n",
       "      <td>1.578739e-03</td>\n",
       "      <td>chrX_153669436_153669453_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_155049205_GA,chrX_15504920...</td>\n",
       "      <td>N-N-N-N-N-N-N</td>\n",
       "      <td>chrX_155049215_GA</td>\n",
       "      <td>69</td>\n",
       "      <td>118</td>\n",
       "      <td>0.584746</td>\n",
       "      <td>25</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>0.063288</td>\n",
       "      <td>0.834189</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.600616</td>\n",
       "      <td>0,1,2,3,4,5,6,7 24,0,0,1,0,0,0,...</td>\n",
       "      <td>3.720374</td>\n",
       "      <td>7.890299</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>6.779916e-39</td>\n",
       "      <td>8.012628e-38</td>\n",
       "      <td>chrX_155049205_155049215_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>938 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          <sample> chr_name  region_start  region_end               mpmat_index  region_site_num  region_block_site_num  region_mut_site_num                   region_site_index  \\\n",
       "8       DetectSeq_ATP8-DddA6_REP-1     chr1      18907111    18907133    chr1_18907111_18907133                9                      0                    9  chr1_18907111_GA,chr1_18907119_...   \n",
       "36      DetectSeq_ATP8-DddA6_REP-1     chr1      67532557    67532593    chr1_67532557_67532593               10                      0                   10  chr1_67532557_GA,chr1_67532565_...   \n",
       "56      DetectSeq_ATP8-DddA6_REP-1     chr1     153223231   153223253  chr1_153223231_153223253               12                      0                   12  chr1_153223231_CT,chr1_15322323...   \n",
       "60      DetectSeq_ATP8-DddA6_REP-1     chr1     161209983   161210006  chr1_161209983_161210006               12                      0                   12  chr1_161209983_GA,chr1_16120998...   \n",
       "75      DetectSeq_ATP8-DddA6_REP-1     chr1     219416797   219416815  chr1_219416797_219416815                8                      0                    8  chr1_219416797_CT,chr1_21941679...   \n",
       "...                            ...      ...           ...         ...                       ...              ...                    ...                  ...                                 ...   \n",
       "3243  DetectSeq_SIRT6-DddA11_REP-1     chrX     132292961   132292983  chrX_132292961_132292983                9                      0                    9  chrX_132292961_GA,chrX_13229296...   \n",
       "3245  DetectSeq_SIRT6-DddA11_REP-1     chrX     136925330   136925338  chrX_136925330_136925338                5                      0                    5  chrX_136925330_CT,chrX_13692533...   \n",
       "3246  DetectSeq_SIRT6-DddA11_REP-1     chrX     138128211   138128237  chrX_138128211_138128237                9                      1                    8  chrX_138128211_GA,chrX_13812821...   \n",
       "3248  DetectSeq_SIRT6-DddA11_REP-1     chrX     153669436   153669453  chrX_153669436_153669453               10                      0                   10  chrX_153669436_CT,chrX_15366943...   \n",
       "3249  DetectSeq_SIRT6-DddA11_REP-1     chrX     155049205   155049215  chrX_155049205_155049215                7                      0                    7  chrX_155049205_GA,chrX_15504920...   \n",
       "\n",
       "           region_block_state region_highest_site_index  region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  \\\n",
       "8           N-N-N-N-N-N-N-N-N          chr1_18907133_GA                           11                             18                       0.611111          28           24               0   \n",
       "36        N-N-N-N-N-N-N-N-N-N          chr1_67532592_GA                           15                             44                       0.340909          49           60               0   \n",
       "56    N-N-N-N-N-N-N-N-N-N-N-N         chr1_153223231_CT                            8                             12                       0.666667          23           17               0   \n",
       "60    N-N-N-N-N-N-N-N-N-N-N-N         chr1_161210002_GA                           10                             17                       0.588235          24           22               1   \n",
       "75            N-N-N-N-N-N-N-N         chr1_219416798_CT                            8                             23                       0.347826          12           31               0   \n",
       "...                       ...                       ...                          ...                            ...                            ...         ...          ...             ...   \n",
       "3243        N-N-N-N-N-N-N-N-N         chrX_132292983_GA                            8                             26                       0.307692          40           40               1   \n",
       "3245                N-N-N-N-N         chrX_136925330_CT                           42                            103                       0.407767          43          129               0   \n",
       "3246        N-N-N-B-N-N-N-N-N         chrX_138128237_GA                           12                             28                       0.428571          33           37               1   \n",
       "3248      N-N-N-N-N-N-N-N-N-N         chrX_153669436_CT                            7                             22                       0.318182          22           26               0   \n",
       "3249            N-N-N-N-N-N-N         chrX_155049215_GA                           69                            118                       0.584746          25          150               1   \n",
       "\n",
       "      treat_mut_count  ctrl_count.norm  treat_count.norm  ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state       p_value           FDR  \\\n",
       "8                  17         0.070882          0.154244             0.000000              0.109256  0,1,2,3,4,5,6,7,8,9 28,0,0,0,0,...  1.121717     3.753610     TestOK  2.206580e-05  4.395162e-04   \n",
       "36                 33         0.124044          0.385610             0.000000              0.212086  0,1,2,3,4,5,6,7,8,9,10 46,3,0,0...  1.636290     4.710541     TestOK  1.009651e-10  9.375327e-09   \n",
       "56                 12         0.058225          0.109256             0.000000              0.077122  0,1,2,3,4,5,6,7,8,9,10,11,12 20...  0.908010     3.251110     TestOK  7.761196e-04  8.270126e-03   \n",
       "60                 16         0.060756          0.141391             0.002532              0.102829  0,1,2,3,4,5,6,7,8,9,10,11,12 19...  1.218579     5.344109     TestOK  4.566141e-05  7.933340e-04   \n",
       "75                 16         0.030378          0.199232             0.000000              0.102829  0,1,2,3,4,5,6,7,8 12,0,0,0,0,0,...  2.713343     3.666147     TestOK  4.566141e-05  7.933340e-04   \n",
       "...               ...              ...               ...                  ...                   ...                                 ...       ...          ...        ...           ...           ...   \n",
       "3243               14         0.101261          0.222450             0.002532              0.077858  0,1,2,3,4,5,6,7,8,9 39,0,1,0,0,...  1.135412     4.942767     TestOK  1.408338e-04  2.007500e-04   \n",
       "3245               59         0.108855          0.717403             0.000000              0.328114  0,1,2,3,4,5 43,0,0,0,0,0 27,43,...  2.720374     5.608487     TestOK  2.433039e-20  9.702301e-20   \n",
       "3246               13         0.083540          0.205767             0.002532              0.072296  0,1,2,3,4,5,6,7,8,9 32,0,1,0,0,...  1.300471     4.835851     TestOK  2.889936e-04  4.053103e-04   \n",
       "3248               11         0.055693          0.144593             0.000000              0.061174  0,1,2,3,4,5,6,7,8,9,10 22,0,0,0...  1.376420     3.185275     TestOK  1.185269e-03  1.578739e-03   \n",
       "3249              108         0.063288          0.834189             0.002532              0.600616  0,1,2,3,4,5,6,7 24,0,0,1,0,0,0,...  3.720374     7.890299     TestOK  6.779916e-39  8.012628e-38   \n",
       "\n",
       "                                bed_name strand  \n",
       "8     chr1_18907111_18907133_highest_...      .  \n",
       "36    chr1_67532557_67532593_highest_...      .  \n",
       "56    chr1_153223231_153223253_highes...      .  \n",
       "60    chr1_161209983_161210006_highes...      .  \n",
       "75    chr1_219416797_219416815_highes...      .  \n",
       "...                                  ...    ...  \n",
       "3243  chrX_132292961_132292983_highes...      .  \n",
       "3245  chrX_136925330_136925338_highes...      .  \n",
       "3246  chrX_138128211_138128237_highes...      .  \n",
       "3248  chrX_153669436_153669453_highes...      .  \n",
       "3249  chrX_155049205_155049215_highes...      .  \n",
       "\n",
       "[938 rows x 30 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_lenient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "448d2e0c-84fe-4f4d-b5da-b09e30110689",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:54.523250Z",
     "iopub.status.busy": "2022-11-29T14:47:54.522778Z",
     "iopub.status.idle": "2022-11-29T14:47:54.532263Z",
     "shell.execute_reply": "2022-11-29T14:47:54.531210Z",
     "shell.execute_reply.started": "2022-11-29T14:47:54.523224Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count                        938\n",
       "unique                       439\n",
       "top       chr1_18907111_18907133\n",
       "freq                           5\n",
       "Name: mpmat_index, dtype: object"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_lenient.mpmat_index.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c4417eba-dad4-448c-9c4b-ce9525844e50",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:55.050187Z",
     "iopub.status.busy": "2022-11-29T14:47:55.049932Z",
     "iopub.status.idle": "2022-11-29T14:47:55.054242Z",
     "shell.execute_reply": "2022-11-29T14:47:55.053264Z",
     "shell.execute_reply.started": "2022-11-29T14:47:55.050165Z"
    }
   },
   "outputs": [],
   "source": [
    "df_old_439 = df_sign_lenient"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af52c383-319c-4dbd-a1d5-60d971335989",
   "metadata": {},
   "source": [
    "### data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "ace8d92c-c4af-4646-a35b-a7b0aac2aa2f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:57.050225Z",
     "iopub.status.busy": "2022-11-29T14:47:57.049823Z",
     "iopub.status.idle": "2022-11-29T14:47:57.088127Z",
     "shell.execute_reply": "2022-11-29T14:47:57.087187Z",
     "shell.execute_reply.started": "2022-11-29T14:47:57.050198Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>sample</th>\n",
       "      <th>detect_seq_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1_18907111_18907133</td>\n",
       "      <td>0.109256</td>\n",
       "      <td>0.154244</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.054818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1_67532557_67532593</td>\n",
       "      <td>0.212086</td>\n",
       "      <td>0.385610</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.064156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1_153223231_153223253</td>\n",
       "      <td>0.077122</td>\n",
       "      <td>0.109256</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.038428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1_161209983_161210006</td>\n",
       "      <td>0.102829</td>\n",
       "      <td>0.141391</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.054389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1_219416797_219416815</td>\n",
       "      <td>0.102829</td>\n",
       "      <td>0.199232</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.027393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3243</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX_132292961_132292983</td>\n",
       "      <td>0.077858</td>\n",
       "      <td>0.222450</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.009538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>0.328114</td>\n",
       "      <td>0.717403</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.068636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3246</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>0.072296</td>\n",
       "      <td>0.205767</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.008925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>0.061174</td>\n",
       "      <td>0.144593</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.010950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>0.600616</td>\n",
       "      <td>0.834189</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.311359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>938 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          <sample>               mpmat_index  treat_mut_count.norm  treat_count.norm                        sample  detect_seq_score\n",
       "8       DetectSeq_ATP8-DddA6_REP-1    chr1_18907111_18907133              0.109256          0.154244    DetectSeq_ATP8-DddA6_REP-1          0.054818\n",
       "36      DetectSeq_ATP8-DddA6_REP-1    chr1_67532557_67532593              0.212086          0.385610    DetectSeq_ATP8-DddA6_REP-1          0.064156\n",
       "56      DetectSeq_ATP8-DddA6_REP-1  chr1_153223231_153223253              0.077122          0.109256    DetectSeq_ATP8-DddA6_REP-1          0.038428\n",
       "60      DetectSeq_ATP8-DddA6_REP-1  chr1_161209983_161210006              0.102829          0.141391    DetectSeq_ATP8-DddA6_REP-1          0.054389\n",
       "75      DetectSeq_ATP8-DddA6_REP-1  chr1_219416797_219416815              0.102829          0.199232    DetectSeq_ATP8-DddA6_REP-1          0.027393\n",
       "...                            ...                       ...                   ...               ...                           ...               ...\n",
       "3243  DetectSeq_SIRT6-DddA11_REP-1  chrX_132292961_132292983              0.077858          0.222450  DetectSeq_SIRT6-DddA11_REP-1          0.009538\n",
       "3245  DetectSeq_SIRT6-DddA11_REP-1  chrX_136925330_136925338              0.328114          0.717403  DetectSeq_SIRT6-DddA11_REP-1          0.068636\n",
       "3246  DetectSeq_SIRT6-DddA11_REP-1  chrX_138128211_138128237              0.072296          0.205767  DetectSeq_SIRT6-DddA11_REP-1          0.008925\n",
       "3248  DetectSeq_SIRT6-DddA11_REP-1  chrX_153669436_153669453              0.061174          0.144593  DetectSeq_SIRT6-DddA11_REP-1          0.010950\n",
       "3249  DetectSeq_SIRT6-DddA11_REP-1  chrX_155049205_155049215              0.600616          0.834189  DetectSeq_SIRT6-DddA11_REP-1          0.311359\n",
       "\n",
       "[938 rows x 6 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_439 = df_old_439[['<sample>', 'mpmat_index',\n",
    "                         'treat_mut_count.norm', 'treat_count.norm']].copy()\n",
    "df_old_439 = (\n",
    "    df_old_439.\n",
    "    assign(\n",
    "        sample=df_old_439['<sample>'],\n",
    "        detect_seq_score=df_old_439.apply(lambda x: ((x['treat_mut_count.norm'] / x['treat_count.norm']) ** 2) * x['treat_mut_count.norm'], axis=1),\n",
    "    )\n",
    ")\n",
    "\n",
    "df_old_439"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "79563879-d42a-4d07-974c-e6efb114f38d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:47:59.648869Z",
     "iopub.status.busy": "2022-11-29T14:47:59.648447Z",
     "iopub.status.idle": "2022-11-29T14:47:59.663828Z",
     "shell.execute_reply": "2022-11-29T14:47:59.662926Z",
     "shell.execute_reply.started": "2022-11-29T14:47:59.648842Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>sample</th>\n",
       "      <th>detect_seq_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1_18907111_18907133</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.054818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1_67532557_67532593</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.064156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr1_153223231_153223253</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.038428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>chr1_161209983_161210006</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.054389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr1_219416797_219416815</td>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>0.027393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3243</th>\n",
       "      <td>chrX_132292961_132292983</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.009538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.068636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3246</th>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.008925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3248</th>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.010950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249</th>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>0.311359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>938 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   mpmat_index                        sample  detect_seq_score\n",
       "8       chr1_18907111_18907133    DetectSeq_ATP8-DddA6_REP-1          0.054818\n",
       "36      chr1_67532557_67532593    DetectSeq_ATP8-DddA6_REP-1          0.064156\n",
       "56    chr1_153223231_153223253    DetectSeq_ATP8-DddA6_REP-1          0.038428\n",
       "60    chr1_161209983_161210006    DetectSeq_ATP8-DddA6_REP-1          0.054389\n",
       "75    chr1_219416797_219416815    DetectSeq_ATP8-DddA6_REP-1          0.027393\n",
       "...                        ...                           ...               ...\n",
       "3243  chrX_132292961_132292983  DetectSeq_SIRT6-DddA11_REP-1          0.009538\n",
       "3245  chrX_136925330_136925338  DetectSeq_SIRT6-DddA11_REP-1          0.068636\n",
       "3246  chrX_138128211_138128237  DetectSeq_SIRT6-DddA11_REP-1          0.008925\n",
       "3248  chrX_153669436_153669453  DetectSeq_SIRT6-DddA11_REP-1          0.010950\n",
       "3249  chrX_155049205_155049215  DetectSeq_SIRT6-DddA11_REP-1          0.311359\n",
       "\n",
       "[938 rows x 3 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_detect_seq_score = df_old_439[['mpmat_index', 'sample', 'detect_seq_score']].copy()\n",
    "df_share_old = df_detect_seq_score[['mpmat_index', 'sample', 'detect_seq_score']].copy()\n",
    "df_share_old"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec1b0974-6822-45c1-93f4-168b41c928f4",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 初步过滤 poisson_res -> df_pois 并保存备用\n",
    "- log2_FC: log2(treat_count.norm/ctrl_count.norm), 如果ctrl_count.norm不存在，就用 chr 突变背景\n",
    "- log2_FC_mut: log2(treat_mut_count.norm/ctrl_mut_count.norm), 如果ctrl_mut_count.norm不存在，就用 chr 突变背景\n",
    "- region_block_state: B-Blocked, S-SNV, N-Non-SNV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2b37dba-c645-4656-bf26-069d91c6253e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "013f4366-febf-4dff-b7de-171b51bc2099",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:50:17.937372Z",
     "iopub.status.busy": "2022-11-29T14:50:17.936469Z",
     "iopub.status.idle": "2022-11-29T14:50:17.944446Z",
     "shell.execute_reply": "2022-11-29T14:50:17.942913Z",
     "shell.execute_reply.started": "2022-11-29T14:50:17.937310Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "file = '../poisson_res/poisson_res_all.tsv.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cba03e93-ccb4-447f-bacd-58b2f19216be",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:50:21.010495Z",
     "iopub.status.busy": "2022-11-29T14:50:21.009794Z",
     "iopub.status.idle": "2022-11-29T14:51:42.134111Z",
     "shell.execute_reply": "2022-11-29T14:51:42.133235Z",
     "shell.execute_reply.started": "2022-11-29T14:50:21.010434Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16395</td>\n",
       "      <td>16418</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_16395_CT,chr1_16397_C.,chr...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_16395_CT</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>57</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.144296</td>\n",
       "      <td>0.113256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 53,4,...</td>\n",
       "      <td>-0.349446</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>16540</td>\n",
       "      <td>chr1_16534_16540</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_16534_CT,chr1_16538_CT,chr...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>chr1_16538_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091135</td>\n",
       "      <td>0.047687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 36,0,0,0,0 8,0,0,0,0</td>\n",
       "      <td>-0.934408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20272</td>\n",
       "      <td>20284</td>\n",
       "      <td>chr1_20272_20284</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1_20272_GA,chr1_20275_GA,chr...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chr1_20283_GA</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.053648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...</td>\n",
       "      <td>0.820479</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31029</td>\n",
       "      <td>31037</td>\n",
       "      <td>chr1_31029_31037</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_31029_GA,chr1_31030_G.,chr...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>chr1_31031_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>0.035765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 2,0,0,0,0 5,0,1,0,0</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54043</td>\n",
       "      <td>54047</td>\n",
       "      <td>chr1_54043_54047</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_54043_CT,chr1_54044_C.,chr...</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>chr1_54043_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 0,0,0,0 0,4,0,0</td>\n",
       "      <td>-1.456561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      <sample> chr_name  region_start  region_end       mpmat_index  region_site_num  region_block_site_num  region_mut_site_num                   region_site_index  \\\n",
       "0  DetectSeq_ATP8-DddA11_REP-1     chr1         16395       16418  chr1_16395_16418               11                      0                   11  chr1_16395_CT,chr1_16397_C.,chr...   \n",
       "1  DetectSeq_ATP8-DddA11_REP-1     chr1         16534       16540  chr1_16534_16540                4                      1                    3  chr1_16534_CT,chr1_16538_CT,chr...   \n",
       "2  DetectSeq_ATP8-DddA11_REP-1     chr1         20272       20284  chr1_20272_20284                5                      0                    5  chr1_20272_GA,chr1_20275_GA,chr...   \n",
       "3  DetectSeq_ATP8-DddA11_REP-1     chr1         31029       31037  chr1_31029_31037                4                      1                    3  chr1_31029_GA,chr1_31030_G.,chr...   \n",
       "4  DetectSeq_ATP8-DddA11_REP-1     chr1         54043       54047  chr1_54043_54047                3                      0                    3  chr1_54043_CT,chr1_54044_C.,chr...   \n",
       "\n",
       "      region_block_state region_highest_site_index  region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  \\\n",
       "0  N-N-N-N-N-N-N-N-N-N-N             chr1_16395_CT                            3                             16                       0.187500          57           19               0   \n",
       "1                B-N-N-N             chr1_16538_CT                            0                              7                       0.000000          36            8               0   \n",
       "2              N-N-N-N-N             chr1_20283_GA                            2                              7                       0.285714          12            9               0   \n",
       "3                B-N-N-N             chr1_31031_GA                            1                              6                       0.166667           2            6               0   \n",
       "4                  N-N-N             chr1_54043_CT                            4                              4                       1.000000           0            4               0   \n",
       "\n",
       "   treat_mut_count  ctrl_count.norm  treat_count.norm  ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state   p_value       FDR  \n",
       "0                2         0.144296          0.113256                  0.0              0.011922  0,1,2,3,4,5,6,7,8,9,10,11 53,4,... -0.349446     0.557555     TestOK  0.377620  0.842234  \n",
       "1                0         0.091135          0.047687                  0.0              0.000000      0,1,2,3,4 36,0,0,0,0 8,0,0,0,0 -0.934408          NaN     TestOK  0.841752  0.842234  \n",
       "2                2         0.030378          0.053648                  0.0              0.011922  0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...  0.820479     0.557555     TestOK  0.377620  0.842234  \n",
       "3                1         0.005063          0.035765                  0.0              0.005961       0,1,2,3,4 2,0,0,0,0 5,0,1,0,0  2.820479    -0.442445     TestOK  0.582932  0.842234  \n",
       "4                0         0.000000          0.023843                  0.0              0.000000             0,1,2,3 0,0,0,0 0,4,0,0 -1.456561          NaN     TestOK  0.841752  0.842234  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16395</td>\n",
       "      <td>16418</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_16395_CT,chr1_16397_C.,chr...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>chr1_16395_CT</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>57</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.144296</td>\n",
       "      <td>0.113256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 53,4,...</td>\n",
       "      <td>-0.349446</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>16540</td>\n",
       "      <td>chr1_16534_16540</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_16534_CT,chr1_16538_CT,chr...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>chr1_16538_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091135</td>\n",
       "      <td>0.047687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 36,0,0,0,0 8,0,0,0,0</td>\n",
       "      <td>-0.934408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20272</td>\n",
       "      <td>20284</td>\n",
       "      <td>chr1_20272_20284</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1_20272_GA,chr1_20275_GA,chr...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>chr1_20283_GA</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.053648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...</td>\n",
       "      <td>0.820479</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31029</td>\n",
       "      <td>31037</td>\n",
       "      <td>chr1_31029_31037</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_31029_GA,chr1_31030_G.,chr...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>chr1_31031_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>0.035765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 2,0,0,0,0 5,0,1,0,0</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54043</td>\n",
       "      <td>54047</td>\n",
       "      <td>chr1_54043_54047</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_54043_CT,chr1_54044_C.,chr...</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>chr1_54043_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 0,0,0,0 0,4,0,0</td>\n",
       "      <td>-1.456561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      <sample> chr_name  region_start  region_end       mpmat_index  region_site_num  region_block_site_num  region_mut_site_num                   region_site_index  \\\n",
       "0  DetectSeq_ATP8-DddA11_REP-1     chr1         16395       16418  chr1_16395_16418               11                      0                   11  chr1_16395_CT,chr1_16397_C.,chr...   \n",
       "1  DetectSeq_ATP8-DddA11_REP-1     chr1         16534       16540  chr1_16534_16540                4                      1                    3  chr1_16534_CT,chr1_16538_CT,chr...   \n",
       "2  DetectSeq_ATP8-DddA11_REP-1     chr1         20272       20284  chr1_20272_20284                5                      0                    5  chr1_20272_GA,chr1_20275_GA,chr...   \n",
       "3  DetectSeq_ATP8-DddA11_REP-1     chr1         31029       31037  chr1_31029_31037                4                      1                    3  chr1_31029_GA,chr1_31030_G.,chr...   \n",
       "4  DetectSeq_ATP8-DddA11_REP-1     chr1         54043       54047  chr1_54043_54047                3                      0                    3  chr1_54043_CT,chr1_54044_C.,chr...   \n",
       "\n",
       "      region_block_state region_highest_site_index  region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  \\\n",
       "0  N-N-N-N-N-N-N-N-N-N-N             chr1_16395_CT                            3                             16                       0.187500          57           19               0   \n",
       "1                B-N-N-N             chr1_16538_CT                            0                              7                       0.000000          36            8               0   \n",
       "2              N-N-N-N-N             chr1_20283_GA                            2                              7                       0.285714          12            9               0   \n",
       "3                B-N-N-N             chr1_31031_GA                            1                              6                       0.166667           2            6               0   \n",
       "4                  N-N-N             chr1_54043_CT                            4                              4                       1.000000           0            4               0   \n",
       "\n",
       "   treat_mut_count  ctrl_count.norm  treat_count.norm  ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state   p_value       FDR  \n",
       "0                2         0.144296          0.113256                  0.0              0.011922  0,1,2,3,4,5,6,7,8,9,10,11 53,4,... -0.349446     0.557555     TestOK  0.377620  0.842234  \n",
       "1                0         0.091135          0.047687                  0.0              0.000000      0,1,2,3,4 36,0,0,0,0 8,0,0,0,0 -0.934408          NaN     TestOK  0.841752  0.842234  \n",
       "2                2         0.030378          0.053648                  0.0              0.011922  0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...  0.820479     0.557555     TestOK  0.377620  0.842234  \n",
       "3                1         0.005063          0.035765                  0.0              0.005961       0,1,2,3,4 2,0,0,0,0 5,0,1,0,0  2.820479    -0.442445     TestOK  0.582932  0.842234  \n",
       "4                0         0.000000          0.023843                  0.0              0.000000             0,1,2,3 0,0,0,0 0,4,0,0 -1.456561          NaN     TestOK  0.841752  0.842234  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\n",
    "        file,\n",
    "        header=0,\n",
    "        index_col=None,\n",
    "        sep='\\t'\n",
    "    )\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "27e091f0-9ab6-4a47-958e-d095bd39728f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:57:03.687751Z",
     "iopub.status.busy": "2022-11-29T14:57:03.687007Z",
     "iopub.status.idle": "2022-11-29T14:57:03.698880Z",
     "shell.execute_reply": "2022-11-29T14:57:03.697381Z",
     "shell.execute_reply.started": "2022-11-29T14:57:03.687693Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['<sample>', 'chr_name', 'region_start', 'region_end', 'mpmat_index', 'region_site_num', 'region_block_site_num', 'region_mut_site_num', 'region_site_index', 'region_block_state',\n",
       "       'region_highest_site_index', 'region_highest_site_mut_num', 'region_highest_site_cover_num', 'region_highest_site_mut_ratio', 'ctrl_count', 'treat_count', 'ctrl_mut_count', 'treat_mut_count',\n",
       "       'ctrl_count.norm', 'treat_count.norm', 'ctrl_mut_count.norm', 'treat_mut_count.norm', 'count_info', 'log2_FC', 'log2_FC_mut', 'test_state', 'p_value', 'FDR'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "Index(['<sample>', 'chr_name', 'region_start', 'region_end', 'mpmat_index', 'region_site_num', 'region_block_site_num', 'region_mut_site_num', 'region_site_index', 'region_block_state',\n",
       "       'region_highest_site_index', 'region_highest_site_mut_num', 'region_highest_site_cover_num', 'region_highest_site_mut_ratio', 'ctrl_count', 'treat_count', 'ctrl_mut_count', 'treat_mut_count',\n",
       "       'ctrl_count.norm', 'treat_count.norm', 'ctrl_mut_count.norm', 'treat_mut_count.norm', 'count_info', 'log2_FC', 'log2_FC_mut', 'test_state', 'p_value', 'FDR'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8849a931-9ea2-49f9-94aa-72fc10ec3f9e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:57:06.957253Z",
     "iopub.status.busy": "2022-11-29T14:57:06.956456Z",
     "iopub.status.idle": "2022-11-29T14:57:08.302016Z",
     "shell.execute_reply": "2022-11-29T14:57:08.301155Z",
     "shell.execute_reply.started": "2022-11-29T14:57:06.957200Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df.drop(columns=['region_site_index', 'region_block_state'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "237c76d2-4067-4844-bbd3-c49306b39e64",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-29T14:57:09.259650Z",
     "iopub.status.busy": "2022-11-29T14:57:09.258979Z",
     "iopub.status.idle": "2022-11-29T14:57:21.858857Z",
     "shell.execute_reply": "2022-11-29T14:57:21.858118Z",
     "shell.execute_reply.started": "2022-11-29T14:57:09.259570Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 12279920 entries, 0 to 12279919\n",
      "Data columns (total 26 columns):\n",
      " #   Column                         Dtype  \n",
      "---  ------                         -----  \n",
      " 0   <sample>                       object \n",
      " 1   chr_name                       object \n",
      " 2   region_start                   int64  \n",
      " 3   region_end                     int64  \n",
      " 4   mpmat_index                    object \n",
      " 5   region_site_num                int64  \n",
      " 6   region_block_site_num          int64  \n",
      " 7   region_mut_site_num            int64  \n",
      " 8   region_highest_site_index      object \n",
      " 9   region_highest_site_mut_num    int64  \n",
      " 10  region_highest_site_cover_num  int64  \n",
      " 11  region_highest_site_mut_ratio  float64\n",
      " 12  ctrl_count                     int64  \n",
      " 13  treat_count                    int64  \n",
      " 14  ctrl_mut_count                 int64  \n",
      " 15  treat_mut_count                int64  \n",
      " 16  ctrl_count.norm                float64\n",
      " 17  treat_count.norm               float64\n",
      " 18  ctrl_mut_count.norm            float64\n",
      " 19  treat_mut_count.norm           float64\n",
      " 20  count_info                     object \n",
      " 21  log2_FC                        float64\n",
      " 22  log2_FC_mut                    float64\n",
      " 23  test_state                     object \n",
      " 24  p_value                        float64\n",
      " 25  FDR                            float64\n",
      "dtypes: float64(9), int64(11), object(6)\n",
      "memory usage: 2.4+ GB\n",
      "None\n",
      "<sample>                               0\n",
      "chr_name                               0\n",
      "region_start                           0\n",
      "region_end                             0\n",
      "mpmat_index                            0\n",
      "region_site_num                        0\n",
      "region_block_site_num                  0\n",
      "region_mut_site_num                    0\n",
      "region_highest_site_index          25696\n",
      "region_highest_site_mut_num            0\n",
      "region_highest_site_cover_num          0\n",
      "region_highest_site_mut_ratio          0\n",
      "ctrl_count                             0\n",
      "treat_count                            0\n",
      "ctrl_mut_count                         0\n",
      "treat_mut_count                        0\n",
      "ctrl_count.norm                        0\n",
      "treat_count.norm                       0\n",
      "ctrl_mut_count.norm                    0\n",
      "treat_mut_count.norm                   0\n",
      "count_info                             0\n",
      "log2_FC                          1545839\n",
      "log2_FC_mut                      8948384\n",
      "test_state                             0\n",
      "p_value                                0\n",
      "FDR                                    0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df.info())\n",
    "print(df.isna().sum())\n",
    "# NAN应该是 test sample 中的，一会过滤一下"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7cc697a3-9bc3-4bdd-8201-0ddd55dc739c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['DetectSeq_ATP8-DddA11_REP-1' 'DetectSeq_ATP8-DddA6_REP-1'\n",
      " 'DetectSeq_ATP8-DddAwt_REP-1' 'DetectSeq_JAK2-DddA11_REP-1'\n",
      " 'DetectSeq_JAK2-DddA11_REP-2' 'DetectSeq_SIRT6-DddA11_REP-1'\n",
      " 'DetectSeq_SIRT6-DddA11_REP-2' 'test']\n"
     ]
    }
   ],
   "source": [
    "print(df['<sample>'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "63bf63e6-4868-4c9b-a5ab-1e829676792a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16395</td>\n",
       "      <td>16418</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_16395_CT</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>57</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.144296</td>\n",
       "      <td>0.113256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 53,4,...</td>\n",
       "      <td>-0.349446</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_16395_16418_highest_chr1_1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>16540</td>\n",
       "      <td>chr1_16534_16540</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_16538_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091135</td>\n",
       "      <td>0.047687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 36,0,0,0,0 8,0,0,0,0</td>\n",
       "      <td>-0.934408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_16534_16540_highest_chr1_1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20272</td>\n",
       "      <td>20284</td>\n",
       "      <td>chr1_20272_20284</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1_20283_GA</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.053648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...</td>\n",
       "      <td>0.820479</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_20272_20284_highest_chr1_2...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31029</td>\n",
       "      <td>31037</td>\n",
       "      <td>chr1_31029_31037</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_31031_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>0.035765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 2,0,0,0,0 5,0,1,0,0</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_31029_31037_highest_chr1_3...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54043</td>\n",
       "      <td>54047</td>\n",
       "      <td>chr1_54043_54047</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_54043_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 0,0,0,0 0,4,0,0</td>\n",
       "      <td>-1.456561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_54043_54047_highest_chr1_5...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744925</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155984850</td>\n",
       "      <td>155984857</td>\n",
       "      <td>chrX_155984850_155984857</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>chrX_155984850_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.032510</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2 0,0,0 2,5,0</td>\n",
       "      <td>-0.830575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155984850_155984857_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744926</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155984970</td>\n",
       "      <td>155984999</td>\n",
       "      <td>chrX_155984970_155984999</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_155984970_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5,6,7 0,0,0,0,0,0,0,0...</td>\n",
       "      <td>-0.316002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155984970_155984999_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744927</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155985095</td>\n",
       "      <td>155985105</td>\n",
       "      <td>chrX_155985095_155985105</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrX_155985095_GA</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018577</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009288</td>\n",
       "      <td>0,1,2,3 0,0,0,0 1,1,2,0</td>\n",
       "      <td>-1.637930</td>\n",
       "      <td>0.465867</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.372391</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155985095_155985105_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744928</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155988525</td>\n",
       "      <td>155988531</td>\n",
       "      <td>chrX_155988525_155988531</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chrX_155988525_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.069663</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 0,0,0,0,0 15,0,0,0,0</td>\n",
       "      <td>0.268960</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155988525_155988531_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744929</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155988614</td>\n",
       "      <td>155988629</td>\n",
       "      <td>chrX_155988614_155988629</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrX_155988627_GA</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.060375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5 0,0,0,0,0,0 2,11,0,...</td>\n",
       "      <td>0.062510</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155988614_155988629_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10744930 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              <sample> chr_name  region_start  region_end               mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "0          DetectSeq_ATP8-DddA11_REP-1     chr1         16395       16418          chr1_16395_16418               11                      0                   11             chr1_16395_CT   \n",
       "1          DetectSeq_ATP8-DddA11_REP-1     chr1         16534       16540          chr1_16534_16540                4                      1                    3             chr1_16538_CT   \n",
       "2          DetectSeq_ATP8-DddA11_REP-1     chr1         20272       20284          chr1_20272_20284                5                      0                    5             chr1_20283_GA   \n",
       "3          DetectSeq_ATP8-DddA11_REP-1     chr1         31029       31037          chr1_31029_31037                4                      1                    3             chr1_31031_GA   \n",
       "4          DetectSeq_ATP8-DddA11_REP-1     chr1         54043       54047          chr1_54043_54047                3                      0                    3             chr1_54043_CT   \n",
       "...                                ...      ...           ...         ...                       ...              ...                    ...                  ...                       ...   \n",
       "10744925  DetectSeq_SIRT6-DddA11_REP-2     chrX     155984850   155984857  chrX_155984850_155984857                2                      0                    2         chrX_155984850_CT   \n",
       "10744926  DetectSeq_SIRT6-DddA11_REP-2     chrX     155984970   155984999  chrX_155984970_155984999                7                      0                    7         chrX_155984970_CT   \n",
       "10744927  DetectSeq_SIRT6-DddA11_REP-2     chrX     155985095   155985105  chrX_155985095_155985105                3                      0                    3         chrX_155985095_GA   \n",
       "10744928  DetectSeq_SIRT6-DddA11_REP-2     chrX     155988525   155988531  chrX_155988525_155988531                4                      0                    4         chrX_155988525_CT   \n",
       "10744929  DetectSeq_SIRT6-DddA11_REP-2     chrX     155988614   155988629  chrX_155988614_155988629                5                      0                    5         chrX_155988627_GA   \n",
       "\n",
       "          region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "0                                   3                             16                       0.187500          57           19               0                2         0.144296          0.113256   \n",
       "1                                   0                              7                       0.000000          36            8               0                0         0.091135          0.047687   \n",
       "2                                   2                              7                       0.285714          12            9               0                2         0.030378          0.053648   \n",
       "3                                   1                              6                       0.166667           2            6               0                1         0.005063          0.035765   \n",
       "4                                   4                              4                       1.000000           0            4               0                0         0.000000          0.023843   \n",
       "...                               ...                            ...                            ...         ...          ...             ...              ...              ...               ...   \n",
       "10744925                            4                              6                       0.666667           0            7               0                0         0.000000          0.032510   \n",
       "10744926                            0                              8                       0.000000           0           10               0                0         0.000000          0.046442   \n",
       "10744927                            4                              4                       1.000000           0            4               0                2         0.000000          0.018577   \n",
       "10744928                            0                             14                       0.000000           0           15               0                0         0.000000          0.069663   \n",
       "10744929                           10                             10                       1.000000           0           13               0                0         0.000000          0.060375   \n",
       "\n",
       "          ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state   p_value       FDR                            bed_name strand  \n",
       "0                         0.0              0.011922  0,1,2,3,4,5,6,7,8,9,10,11 53,4,... -0.349446     0.557555     TestOK  0.377620  0.842234  chr1_16395_16418_highest_chr1_1...      .  \n",
       "1                         0.0              0.000000      0,1,2,3,4 36,0,0,0,0 8,0,0,0,0 -0.934408          NaN     TestOK  0.841752  0.842234  chr1_16534_16540_highest_chr1_1...      .  \n",
       "2                         0.0              0.011922  0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...  0.820479     0.557555     TestOK  0.377620  0.842234  chr1_20272_20284_highest_chr1_2...      .  \n",
       "3                         0.0              0.005961       0,1,2,3,4 2,0,0,0,0 5,0,1,0,0  2.820479    -0.442445     TestOK  0.582932  0.842234  chr1_31029_31037_highest_chr1_3...      .  \n",
       "4                         0.0              0.000000             0,1,2,3 0,0,0,0 0,4,0,0 -1.456561          NaN     TestOK  0.841752  0.842234  chr1_54043_54047_highest_chr1_5...      .  \n",
       "...                       ...                   ...                                 ...       ...          ...        ...       ...       ...                                 ...    ...  \n",
       "10744925                  0.0              0.000000                   0,1,2 0,0,0 2,5,0 -0.830575          NaN     TestOK  0.838398  0.897853  chrX_155984850_155984857_highes...      .  \n",
       "10744926                  0.0              0.000000  0,1,2,3,4,5,6,7 0,0,0,0,0,0,0,0... -0.316002          NaN     TestOK  0.838398  0.897853  chrX_155984970_155984999_highes...      .  \n",
       "10744927                  0.0              0.009288             0,1,2,3 0,0,0,0 1,1,2,0 -1.637930     0.465867     TestOK  0.372391  0.897853  chrX_155985095_155985105_highes...      .  \n",
       "10744928                  0.0              0.000000      0,1,2,3,4 0,0,0,0,0 15,0,0,0,0  0.268960          NaN     TestOK  0.838398  0.897853  chrX_155988525_155988531_highes...      .  \n",
       "10744929                  0.0              0.000000  0,1,2,3,4,5 0,0,0,0,0,0 2,11,0,...  0.062510          NaN     TestOK  0.838398  0.897853  chrX_155988614_155988629_highes...      .  \n",
       "\n",
       "[10744930 rows x 28 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(\n",
    "    bed_name=df.mpmat_index + '_highest_' + df.region_highest_site_index,\n",
    "    strand='.'\n",
    ")\n",
    "df = df[df['<sample>'] != 'test']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "672c5966-29d2-420b-9c9e-adb7fe12a2d8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 10744930 entries, 0 to 10744929\n",
      "Data columns (total 28 columns):\n",
      " #   Column                         Dtype  \n",
      "---  ------                         -----  \n",
      " 0   <sample>                       object \n",
      " 1   chr_name                       object \n",
      " 2   region_start                   int64  \n",
      " 3   region_end                     int64  \n",
      " 4   mpmat_index                    object \n",
      " 5   region_site_num                int64  \n",
      " 6   region_block_site_num          int64  \n",
      " 7   region_mut_site_num            int64  \n",
      " 8   region_highest_site_index      object \n",
      " 9   region_highest_site_mut_num    int64  \n",
      " 10  region_highest_site_cover_num  int64  \n",
      " 11  region_highest_site_mut_ratio  float64\n",
      " 12  ctrl_count                     int64  \n",
      " 13  treat_count                    int64  \n",
      " 14  ctrl_mut_count                 int64  \n",
      " 15  treat_mut_count                int64  \n",
      " 16  ctrl_count.norm                float64\n",
      " 17  treat_count.norm               float64\n",
      " 18  ctrl_mut_count.norm            float64\n",
      " 19  treat_mut_count.norm           float64\n",
      " 20  count_info                     object \n",
      " 21  log2_FC                        float64\n",
      " 22  log2_FC_mut                    float64\n",
      " 23  test_state                     object \n",
      " 24  p_value                        float64\n",
      " 25  FDR                            float64\n",
      " 26  bed_name                       object \n",
      " 27  strand                         object \n",
      "dtypes: float64(9), int64(11), object(8)\n",
      "memory usage: 2.3+ GB\n",
      "None\n",
      "<sample>                               0\n",
      "chr_name                               0\n",
      "region_start                           0\n",
      "region_end                             0\n",
      "mpmat_index                            0\n",
      "region_site_num                        0\n",
      "region_block_site_num                  0\n",
      "region_mut_site_num                    0\n",
      "region_highest_site_index          22484\n",
      "region_highest_site_mut_num            0\n",
      "region_highest_site_cover_num          0\n",
      "region_highest_site_mut_ratio          0\n",
      "ctrl_count                             0\n",
      "treat_count                            0\n",
      "ctrl_mut_count                         0\n",
      "treat_mut_count                        0\n",
      "ctrl_count.norm                        0\n",
      "treat_count.norm                       0\n",
      "ctrl_mut_count.norm                    0\n",
      "treat_mut_count.norm                   0\n",
      "count_info                             0\n",
      "log2_FC                            13944\n",
      "log2_FC_mut                      7413509\n",
      "test_state                             0\n",
      "p_value                                0\n",
      "FDR                                    0\n",
      "bed_name                           22484\n",
      "strand                                 0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df.info())\n",
    "print(df.isna().sum())\n",
    "# 过滤掉 test sample 之后 NAN 少了很多"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "69232a03-8fe7-4cc3-97b4-ca465a68c2cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA11_REP-1</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA6_REP-1</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddAwt_REP-1</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_JAK2-DddA11_REP-1</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_JAK2-DddA11_REP-2</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_SIRT6-DddA11_REP-1</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_SIRT6-DddA11_REP-2</th>\n",
       "      <td>1534990</td>\n",
       "      <td>1534990</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                count   unique               top freq\n",
       "<sample>                                                             \n",
       "DetectSeq_ATP8-DddA11_REP-1   1534990  1534990  chr1_16395_16418    1\n",
       "DetectSeq_ATP8-DddA6_REP-1    1534990  1534990  chr1_16395_16418    1\n",
       "DetectSeq_ATP8-DddAwt_REP-1   1534990  1534990  chr1_16395_16418    1\n",
       "DetectSeq_JAK2-DddA11_REP-1   1534990  1534990  chr1_16395_16418    1\n",
       "DetectSeq_JAK2-DddA11_REP-2   1534990  1534990  chr1_16395_16418    1\n",
       "DetectSeq_SIRT6-DddA11_REP-1  1534990  1534990  chr1_16395_16418    1\n",
       "DetectSeq_SIRT6-DddA11_REP-2  1534990  1534990  chr1_16395_16418    1"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 这里mpmat_index unique，即 frequency == 1 才合理，因为前期 call 点的时候，把所有的 region 放在一起考虑的\n",
    "df.groupby('<sample>').mpmat_index.describe()\n",
    "# call 点没问题\n",
    "# 继续后续分析"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b713bd65-1efc-4ccb-8f05-cfa6d83e7da8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## find significant region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a4cc2710-dfa7-4f61-b667-f5ce19fcc02d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16395</td>\n",
       "      <td>16418</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_16395_CT</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>57</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.144296</td>\n",
       "      <td>0.113256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 53,4,...</td>\n",
       "      <td>-0.349446</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_16395_16418_highest_chr1_1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>16540</td>\n",
       "      <td>chr1_16534_16540</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_16538_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>36</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091135</td>\n",
       "      <td>0.047687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 36,0,0,0,0 8,0,0,0,0</td>\n",
       "      <td>-0.934408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_16534_16540_highest_chr1_1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>20272</td>\n",
       "      <td>20284</td>\n",
       "      <td>chr1_20272_20284</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1_20283_GA</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.053648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...</td>\n",
       "      <td>0.820479</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_20272_20284_highest_chr1_2...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31029</td>\n",
       "      <td>31037</td>\n",
       "      <td>chr1_31029_31037</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_31031_GA</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>0.035765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 2,0,0,0,0 5,0,1,0,0</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_31029_31037_highest_chr1_3...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>54043</td>\n",
       "      <td>54047</td>\n",
       "      <td>chr1_54043_54047</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_54043_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023843</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 0,0,0,0 0,4,0,0</td>\n",
       "      <td>-1.456561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842234</td>\n",
       "      <td>chr1_54043_54047_highest_chr1_5...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744925</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155984850</td>\n",
       "      <td>155984857</td>\n",
       "      <td>chrX_155984850_155984857</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>chrX_155984850_CT</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.032510</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2 0,0,0 2,5,0</td>\n",
       "      <td>-0.830575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155984850_155984857_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744926</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155984970</td>\n",
       "      <td>155984999</td>\n",
       "      <td>chrX_155984970_155984999</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_155984970_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5,6,7 0,0,0,0,0,0,0,0...</td>\n",
       "      <td>-0.316002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155984970_155984999_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744927</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155985095</td>\n",
       "      <td>155985105</td>\n",
       "      <td>chrX_155985095_155985105</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrX_155985095_GA</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018577</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.009288</td>\n",
       "      <td>0,1,2,3 0,0,0,0 1,1,2,0</td>\n",
       "      <td>-1.637930</td>\n",
       "      <td>0.465867</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.372391</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155985095_155985105_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744928</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155988525</td>\n",
       "      <td>155988531</td>\n",
       "      <td>chrX_155988525_155988531</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chrX_155988525_CT</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.069663</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 0,0,0,0,0 15,0,0,0,0</td>\n",
       "      <td>0.268960</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155988525_155988531_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10744929</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155988614</td>\n",
       "      <td>155988629</td>\n",
       "      <td>chrX_155988614_155988629</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrX_155988627_GA</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.060375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5 0,0,0,0,0,0 2,11,0,...</td>\n",
       "      <td>0.062510</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.838398</td>\n",
       "      <td>0.897853</td>\n",
       "      <td>chrX_155988614_155988629_highes...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10744930 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              <sample> chr_name  region_start  region_end               mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "0          DetectSeq_ATP8-DddA11_REP-1     chr1         16395       16418          chr1_16395_16418               11                      0                   11             chr1_16395_CT   \n",
       "1          DetectSeq_ATP8-DddA11_REP-1     chr1         16534       16540          chr1_16534_16540                4                      1                    3             chr1_16538_CT   \n",
       "2          DetectSeq_ATP8-DddA11_REP-1     chr1         20272       20284          chr1_20272_20284                5                      0                    5             chr1_20283_GA   \n",
       "3          DetectSeq_ATP8-DddA11_REP-1     chr1         31029       31037          chr1_31029_31037                4                      1                    3             chr1_31031_GA   \n",
       "4          DetectSeq_ATP8-DddA11_REP-1     chr1         54043       54047          chr1_54043_54047                3                      0                    3             chr1_54043_CT   \n",
       "...                                ...      ...           ...         ...                       ...              ...                    ...                  ...                       ...   \n",
       "10744925  DetectSeq_SIRT6-DddA11_REP-2     chrX     155984850   155984857  chrX_155984850_155984857                2                      0                    2         chrX_155984850_CT   \n",
       "10744926  DetectSeq_SIRT6-DddA11_REP-2     chrX     155984970   155984999  chrX_155984970_155984999                7                      0                    7         chrX_155984970_CT   \n",
       "10744927  DetectSeq_SIRT6-DddA11_REP-2     chrX     155985095   155985105  chrX_155985095_155985105                3                      0                    3         chrX_155985095_GA   \n",
       "10744928  DetectSeq_SIRT6-DddA11_REP-2     chrX     155988525   155988531  chrX_155988525_155988531                4                      0                    4         chrX_155988525_CT   \n",
       "10744929  DetectSeq_SIRT6-DddA11_REP-2     chrX     155988614   155988629  chrX_155988614_155988629                5                      0                    5         chrX_155988627_GA   \n",
       "\n",
       "          region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "0                                   3                             16                       0.187500          57           19               0                2         0.144296          0.113256   \n",
       "1                                   0                              7                       0.000000          36            8               0                0         0.091135          0.047687   \n",
       "2                                   2                              7                       0.285714          12            9               0                2         0.030378          0.053648   \n",
       "3                                   1                              6                       0.166667           2            6               0                1         0.005063          0.035765   \n",
       "4                                   4                              4                       1.000000           0            4               0                0         0.000000          0.023843   \n",
       "...                               ...                            ...                            ...         ...          ...             ...              ...              ...               ...   \n",
       "10744925                            4                              6                       0.666667           0            7               0                0         0.000000          0.032510   \n",
       "10744926                            0                              8                       0.000000           0           10               0                0         0.000000          0.046442   \n",
       "10744927                            4                              4                       1.000000           0            4               0                2         0.000000          0.018577   \n",
       "10744928                            0                             14                       0.000000           0           15               0                0         0.000000          0.069663   \n",
       "10744929                           10                             10                       1.000000           0           13               0                0         0.000000          0.060375   \n",
       "\n",
       "          ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state   p_value       FDR                            bed_name strand  \n",
       "0                         0.0              0.011922  0,1,2,3,4,5,6,7,8,9,10,11 53,4,... -0.349446     0.557555     TestOK  0.377620  0.842234  chr1_16395_16418_highest_chr1_1...      .  \n",
       "1                         0.0              0.000000      0,1,2,3,4 36,0,0,0,0 8,0,0,0,0 -0.934408          NaN     TestOK  0.841752  0.842234  chr1_16534_16540_highest_chr1_1...      .  \n",
       "2                         0.0              0.011922  0,1,2,3,4,5 11,1,0,0,0,0 5,2,0,...  0.820479     0.557555     TestOK  0.377620  0.842234  chr1_20272_20284_highest_chr1_2...      .  \n",
       "3                         0.0              0.005961       0,1,2,3,4 2,0,0,0,0 5,0,1,0,0  2.820479    -0.442445     TestOK  0.582932  0.842234  chr1_31029_31037_highest_chr1_3...      .  \n",
       "4                         0.0              0.000000             0,1,2,3 0,0,0,0 0,4,0,0 -1.456561          NaN     TestOK  0.841752  0.842234  chr1_54043_54047_highest_chr1_5...      .  \n",
       "...                       ...                   ...                                 ...       ...          ...        ...       ...       ...                                 ...    ...  \n",
       "10744925                  0.0              0.000000                   0,1,2 0,0,0 2,5,0 -0.830575          NaN     TestOK  0.838398  0.897853  chrX_155984850_155984857_highes...      .  \n",
       "10744926                  0.0              0.000000  0,1,2,3,4,5,6,7 0,0,0,0,0,0,0,0... -0.316002          NaN     TestOK  0.838398  0.897853  chrX_155984970_155984999_highes...      .  \n",
       "10744927                  0.0              0.009288             0,1,2,3 0,0,0,0 1,1,2,0 -1.637930     0.465867     TestOK  0.372391  0.897853  chrX_155985095_155985105_highes...      .  \n",
       "10744928                  0.0              0.000000      0,1,2,3,4 0,0,0,0,0 15,0,0,0,0  0.268960          NaN     TestOK  0.838398  0.897853  chrX_155988525_155988531_highes...      .  \n",
       "10744929                  0.0              0.000000  0,1,2,3,4,5 0,0,0,0,0,0 2,11,0,...  0.062510          NaN     TestOK  0.838398  0.897853  chrX_155988614_155988629_highes...      .  \n",
       "\n",
       "[10744930 rows x 28 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67cc6c18-e573-4537-a3d5-7fae99249bc2",
   "metadata": {},
   "source": [
    "### strict selection "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "d83a38b2-8097-4762-8439-567659a7ddd4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_sign_strict = (\n",
    "#     df\n",
    "#     .query('FDR <= 0.0001')\n",
    "#     .query('log2_FC_mut >= 2')\n",
    "#     .query('ctrl_mut_count <= 1')\n",
    "#     .query('`treat_mut_count.norm` * 100 >= 10')\n",
    "#     .query('treat_mut_count >= 20')\n",
    "#     .query('treat_mut_count / treat_count >= 0.15')\n",
    "#     .query('region_block_site_num <= 1')\n",
    "#     .query('region_highest_site_mut_ratio >= 0.35')  # 1892\n",
    "# )\n",
    "\n",
    "# print(df_sign_strict.shape[0])\n",
    "# df_sign_strict.groupby('<sample>').mpmat_index.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "a426c781-54c8-405a-bdfd-a2dc6f101608",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_sign_strict.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad92d39-8ef1-44fe-a4c5-2a655139ae8c",
   "metadata": {},
   "source": [
    "### lenient selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "65bd4318-6622-4da7-a5fe-87db7ec6d3e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12349\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<sample>\n",
       "DetectSeq_ATP8-DddA11_REP-1      177\n",
       "DetectSeq_ATP8-DddA6_REP-1       589\n",
       "DetectSeq_ATP8-DddAwt_REP-1       57\n",
       "DetectSeq_JAK2-DddA11_REP-1      628\n",
       "DetectSeq_JAK2-DddA11_REP-2      667\n",
       "DetectSeq_SIRT6-DddA11_REP-1    2801\n",
       "DetectSeq_SIRT6-DddA11_REP-2    7430\n",
       "Name: mpmat_index, dtype: int64"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nature condicitons\n",
    "# df_sign_lenient = (\n",
    "#     df\n",
    "#     .query('FDR <= 0.01')\n",
    "#     .query('log2_FC_mut >= 2')\n",
    "#     .query('ctrl_mut_count <= 1')\n",
    "#     .query('`treat_mut_count.norm` * 100 >= 5')\n",
    "#     .query('treat_mut_count >= 10')\n",
    "#     .query('treat_mut_count / treat_count >= 0.15')\n",
    "#     .query('region_block_site_num <= 1')\n",
    "#     .query('region_highest_site_mut_ratio >= 0.30')  # 4007\n",
    "# )\n",
    "\n",
    "\n",
    "df_sign_lenient = (\n",
    "    df\n",
    "    .query('FDR <= 0.01')\n",
    "    .query('log2_FC_mut >= 2')\n",
    "    .query('ctrl_mut_count <= 1')\n",
    "    .query('`treat_mut_count.norm` * 100 >= 5')\n",
    "    .query('treat_mut_count >= 10')\n",
    "    .query('treat_mut_count / treat_count >= 0.10')  # TODO\n",
    "    .query('region_block_site_num <= 1')\n",
    "    .query('region_highest_site_mut_ratio >= 0.25')  # TODO\n",
    ")\n",
    "\n",
    "print(df_sign_lenient.shape[0])\n",
    "df_sign_lenient.groupby('<sample>').mpmat_index.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "3f2b9f6a-fc86-4bda-bcd3-7d75d7506a56",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_lenient.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "052bbbad-08fa-4671-8502-418befd285b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12349, 28)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_630992_GA</td>\n",
       "      <td>2975</td>\n",
       "      <td>4635</td>\n",
       "      <td>0.641855</td>\n",
       "      <td>38</td>\n",
       "      <td>6608</td>\n",
       "      <td>1</td>\n",
       "      <td>780</td>\n",
       "      <td>0.096198</td>\n",
       "      <td>39.389272</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>4.649460</td>\n",
       "      <td>0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...</td>\n",
       "      <td>8.677587</td>\n",
       "      <td>10.842847</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.511695e-310</td>\n",
       "      <td>5.507753e-305</td>\n",
       "      <td>chr1_630967_631005_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1732</td>\n",
       "      <td>4662</td>\n",
       "      <td>0.371514</td>\n",
       "      <td>33</td>\n",
       "      <td>6236</td>\n",
       "      <td>0</td>\n",
       "      <td>1189</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>37.171837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.087446</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...</td>\n",
       "      <td>8.797528</td>\n",
       "      <td>9.773088</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631791_CT</td>\n",
       "      <td>733</td>\n",
       "      <td>1147</td>\n",
       "      <td>0.639058</td>\n",
       "      <td>40</td>\n",
       "      <td>1483</td>\n",
       "      <td>0</td>\n",
       "      <td>547</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>8.839935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>4.972421e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>632382</td>\n",
       "      <td>632387</td>\n",
       "      <td>chr1_632382_632387</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_632387_GA</td>\n",
       "      <td>4739</td>\n",
       "      <td>6544</td>\n",
       "      <td>0.724175</td>\n",
       "      <td>25</td>\n",
       "      <td>8609</td>\n",
       "      <td>0</td>\n",
       "      <td>1032</td>\n",
       "      <td>0.063288</td>\n",
       "      <td>51.316925</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.151593</td>\n",
       "      <td>0,1,2 24,1,0 1601,5305,1703</td>\n",
       "      <td>9.663290</td>\n",
       "      <td>9.568782</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr1_632382_632387_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "      <td>chr1_633067_633147</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_633100_GA</td>\n",
       "      <td>696</td>\n",
       "      <td>1232</td>\n",
       "      <td>0.564935</td>\n",
       "      <td>33</td>\n",
       "      <td>2479</td>\n",
       "      <td>0</td>\n",
       "      <td>572</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>14.776938</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.409604</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 33,0,...</td>\n",
       "      <td>7.466665</td>\n",
       "      <td>8.717426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.596361e-224</td>\n",
       "      <td>3.062998e-219</td>\n",
       "      <td>chr1_633067_633147_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        <sample> chr_name  region_start  region_end         mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "83   DetectSeq_ATP8-DddA11_REP-1     chr1        630967      631005  chr1_630967_631005                7                      1                    6            chr1_630992_GA   \n",
       "84   DetectSeq_ATP8-DddA11_REP-1     chr1        631016      631037  chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "89   DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791  chr1_631778_631791                3                      0                    3            chr1_631791_CT   \n",
       "95   DetectSeq_ATP8-DddA11_REP-1     chr1        632382      632387  chr1_632382_632387                2                      0                    2            chr1_632387_GA   \n",
       "100  DetectSeq_ATP8-DddA11_REP-1     chr1        633067      633147  chr1_633067_633147               11                      0                   11            chr1_633100_GA   \n",
       "\n",
       "     region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "83                          2975                           4635                       0.641855          38         6608               1              780         0.096198         39.389272   \n",
       "84                          1732                           4662                       0.371514          33         6236               0             1189         0.083540         37.171837   \n",
       "89                           733                           1147                       0.639058          40         1483               0              547         0.101261          8.839935   \n",
       "95                          4739                           6544                       0.724175          25         8609               0             1032         0.063288         51.316925   \n",
       "100                          696                           1232                       0.564935          33         2479               0              572         0.083540         14.776938   \n",
       "\n",
       "     ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state        p_value            FDR                            bed_name strand  \n",
       "83              0.002532              4.649460  0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...  8.677587    10.842847     TestOK  2.511695e-310  5.507753e-305  chr1_630967_631005_highest_chr1...      .  \n",
       "84              0.000000              7.087446  0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...  8.797528     9.773088     TestOK   0.000000e+00   0.000000e+00  chr1_631016_631037_highest_chr1...      .  \n",
       "89              0.000000              3.260583    0,1,2,3 40,0,0,0 534,278,413,258  6.447891     8.652952     TestOK  2.915445e-214  4.972421e-209  chr1_631778_631791_highest_chr1...      .  \n",
       "95              0.000000              6.151593         0,1,2 24,1,0 1601,5305,1703  9.663290     9.568782     TestOK   0.000000e+00   0.000000e+00  chr1_632382_632387_highest_chr1...      .  \n",
       "100             0.000000              3.409604  0,1,2,3,4,5,6,7,8,9,10,11 33,0,...  7.466665     8.717426     TestOK  1.596361e-224  3.062998e-219  chr1_633067_633147_highest_chr1...      .  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = df_sign_lenient\n",
    "print(df_sign_lenient.shape)\n",
    "df_sign_lenient.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab1dca4-cf2d-4b0e-a073-d9ea4ce66f75",
   "metadata": {},
   "source": [
    "### 把 Nature 中的 region 对应过来，使用 Nature 中的 idx, fix mpmat_index (add old into it) -> df_pois"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "3b10cfa4-162b-409d-8274-ff6eea6cecae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count                     6881\n",
       "unique                    6881\n",
       "top       chr1_1471366_1471410\n",
       "freq                         1\n",
       "Name: region_id, dtype: object"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_id = pd.read_csv(\n",
    "    \"../tables/20220312-DdCBE-off_target_type.FinallistV4.CheckPrimer.AddV4ID.tsv\",\n",
    "    sep='\\t', header=0, usecols=['region_id'],)\n",
    "df_old_id.region_id.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "00ba7c78-3957-4acc-a527-d92225bb020e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 6881 entries, 0 to 6880\n",
      "Data columns (total 3 columns):\n",
      " #   Column  Non-Null Count  Dtype \n",
      "---  ------  --------------  ----- \n",
      " 0   chrom   6881 non-null   object\n",
      " 1   start   6881 non-null   int64 \n",
      " 2   end     6881 non-null   int64 \n",
      "dtypes: int64(2), object(1)\n",
      "memory usage: 161.4+ KB\n"
     ]
    }
   ],
   "source": [
    "df_old_id[['chrom', 'start', 'end'\n",
    "           ]] = pd.Series(df_old_id.region_id.unique()).str.split('_',\n",
    "                                                                  expand=True)\n",
    "df_old_id = df_old_id.iloc[:, 1:4]\n",
    "df_old_id[['start', 'end']] = df_old_id[['start', 'end']].astype(int)\n",
    "df_old_id.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "9846a3db-008a-47e0-b927-63b8eb5ff6f0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1471366</td>\n",
       "      <td>1471410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1693068</td>\n",
       "      <td>1693084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end\n",
       "0  chr1  1471366  1471410\n",
       "1  chr1  1693068  1693084"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_id.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "240143bf-5b70-403e-a51d-0466797e4fbb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>632382</td>\n",
       "      <td>632387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9961</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154031921</td>\n",
       "      <td>154031945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9962</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154313332</td>\n",
       "      <td>154313399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9963</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154679199</td>\n",
       "      <td>154679238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9964</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154817582</td>\n",
       "      <td>154817611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9965</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155216206</td>\n",
       "      <td>155216224</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9966 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0          1          2\n",
       "0     chr1     630967     631005\n",
       "1     chr1     631016     631037\n",
       "2     chr1     631778     631791\n",
       "3     chr1     632382     632387\n",
       "4     chr1     633067     633147\n",
       "...    ...        ...        ...\n",
       "9961  chrX  154031921  154031945\n",
       "9962  chrX  154313332  154313399\n",
       "9963  chrX  154679199  154679238\n",
       "9964  chrX  154817582  154817611\n",
       "9965  chrX  155216206  155216224\n",
       "\n",
       "[9966 rows x 3 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new_id = pd.Series(df_sign_lenient.mpmat_index.unique()).str.split('_', expand=True)\n",
    "df_new_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "8d80e92b-401f-4f1b-8d96-462f647b58a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9966 entries, 0 to 9965\n",
      "Data columns (total 3 columns):\n",
      " #   Column  Non-Null Count  Dtype \n",
      "---  ------  --------------  ----- \n",
      " 0   chrom   9966 non-null   object\n",
      " 1   start   9966 non-null   int64 \n",
      " 2   end     9966 non-null   int64 \n",
      "dtypes: int64(2), object(1)\n",
      "memory usage: 233.7+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end\n",
       "0  chr1  630967  631005\n",
       "1  chr1  631016  631037"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new_id.columns = ['chrom', 'start', 'end']\n",
    "df_new_id[['start', 'end']] = df_new_id[['start', 'end']].astype(int)\n",
    "\n",
    "print(df_new_id.info())\n",
    "df_new_id.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "d5019725-fad9-4f0d-9b29-0bbd64878f57",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "bed_new = BedTool.from_dataframe(df_new_id)\n",
    "bed_nat = BedTool.from_dataframe(df_old_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c0a9a3e-5224-40bf-82a9-eb7a15dc8afb",
   "metadata": {},
   "source": [
    "> https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html\n",
    "\n",
    "![](https://tva1.sinaimg.cn/large/008vxvgGly1h7jy4v629sj30ld0j9wg7.jpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "006c2927-342b-43e5-b42e-1505f0952001",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>632382</td>\n",
       "      <td>632387</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9963</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154031921</td>\n",
       "      <td>154031945</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9964</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154313332</td>\n",
       "      <td>154313399</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9965</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154679199</td>\n",
       "      <td>154679238</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9966</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154817582</td>\n",
       "      <td>154817611</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9967</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155216206</td>\n",
       "      <td>155216224</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9968 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end chrom2  start2  end2\n",
       "0     chr1     630967     631005      .      -1    -1\n",
       "1     chr1     631016     631037      .      -1    -1\n",
       "2     chr1     631778     631791      .      -1    -1\n",
       "3     chr1     632382     632387      .      -1    -1\n",
       "4     chr1     633067     633147      .      -1    -1\n",
       "...    ...        ...        ...    ...     ...   ...\n",
       "9963  chrX  154031921  154031945      .      -1    -1\n",
       "9964  chrX  154313332  154313399      .      -1    -1\n",
       "9965  chrX  154679199  154679238      .      -1    -1\n",
       "9966  chrX  154817582  154817611      .      -1    -1\n",
       "9967  chrX  155216206  155216224      .      -1    -1\n",
       "\n",
       "[9968 rows x 6 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bed_to_fix = bed_new.intersect(bed_nat, loj=True).to_dataframe()\n",
    "df_bed_to_fix.columns = ['chrom', 'start', 'end', 'chrom2', 'start2', 'end2']\n",
    "\n",
    "df_bed_to_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "2185b913-2afa-4382-b5b6-133f06017305",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18907073</td>\n",
       "      <td>18907133</td>\n",
       "      <td>chr1</td>\n",
       "      <td>18907111</td>\n",
       "      <td>18907133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29332389</td>\n",
       "      <td>29332405</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29332374</td>\n",
       "      <td>29332405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67532549</td>\n",
       "      <td>67532593</td>\n",
       "      <td>chr1</td>\n",
       "      <td>67532557</td>\n",
       "      <td>67532593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70133031</td>\n",
       "      <td>70133115</td>\n",
       "      <td>chr1</td>\n",
       "      <td>70133094</td>\n",
       "      <td>70133130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>78617681</td>\n",
       "      <td>78617723</td>\n",
       "      <td>chr1</td>\n",
       "      <td>78617702</td>\n",
       "      <td>78617723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8156</th>\n",
       "      <td>chr12</td>\n",
       "      <td>117022857</td>\n",
       "      <td>117022923</td>\n",
       "      <td>chr12</td>\n",
       "      <td>117022835</td>\n",
       "      <td>117022888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8255</th>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219857</td>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8933</th>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770746</td>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8995</th>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576657</td>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9444</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516831</td>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>434 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end chrom2     start2       end2\n",
       "9      chr1   18907073   18907133   chr1   18907111   18907133\n",
       "11     chr1   29332389   29332405   chr1   29332374   29332405\n",
       "12     chr1   67532549   67532593   chr1   67532557   67532593\n",
       "13     chr1   70133031   70133115   chr1   70133094   70133130\n",
       "14     chr1   78617681   78617723   chr1   78617702   78617723\n",
       "...     ...        ...        ...    ...        ...        ...\n",
       "8156  chr12  117022857  117022923  chr12  117022835  117022888\n",
       "8255  chr13   49219804   49219857  chr13   49219804   49219836\n",
       "8933  chr16   78770741   78770746  chr16   78770741   78770752\n",
       "8995  chr17    9576625    9576657  chr17    9576625    9576648\n",
       "9444  chr19   55516812   55516831  chr19   55516812   55516824\n",
       "\n",
       "[434 rows x 6 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 新点中有 434 个点和老点有 overlap\n",
    "df_bed_to_fix[df_bed_to_fix.end2 != -1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "9c14fc81-4806-454a-9730-5b96ee6d8bd7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>id_ND4</th>\n",
       "      <th>id_ND5.1</th>\n",
       "      <th>id_ND6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_2044988_2044998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_9702414_9702449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-2</td>\n",
       "      <td>ND6-TAS.IND-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_12618165_12618174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_13019998_13020019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_13354705_13354738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>chrX_150683735_150683758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-452</td>\n",
       "      <td>ND6-TAS.IND-540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-453</td>\n",
       "      <td>ND6-TAS.IND-541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-454</td>\n",
       "      <td>ND6-TAS.IND-542</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>650 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mpmat_index id_ND4           id_ND5.1           id_ND6\n",
       "0        chr1_2044988_2044998    NaN    ND5.1-TAS.IND-1              NaN\n",
       "1        chr1_9702414_9702449    NaN    ND5.1-TAS.IND-2    ND6-TAS.IND-1\n",
       "2      chr1_12618165_12618174    NaN    ND5.1-TAS.IND-3              NaN\n",
       "3      chr1_13019998_13020019    NaN                NaN    ND6-TAS.IND-2\n",
       "4      chr1_13354705_13354738    NaN                NaN    ND6-TAS.IND-3\n",
       "..                        ...    ...                ...              ...\n",
       "645  chrX_136925330_136925338    NaN                NaN  ND6-TAS.IND-538\n",
       "646  chrX_138128211_138128237    NaN                NaN  ND6-TAS.IND-539\n",
       "647  chrX_150683735_150683758    NaN  ND5.1-TAS.IND-452  ND6-TAS.IND-540\n",
       "648  chrX_153669436_153669453    NaN  ND5.1-TAS.IND-453  ND6-TAS.IND-541\n",
       "649  chrX_155049205_155049215    NaN  ND5.1-TAS.IND-454  ND6-TAS.IND-542\n",
       "\n",
       "[650 rows x 4 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "cd42c5b5-209b-4477-b032-2f5fc8aeffd3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18907111</td>\n",
       "      <td>18907133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29332374</td>\n",
       "      <td>29332405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67532557</td>\n",
       "      <td>67532593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70133094</td>\n",
       "      <td>70133130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>78617702</td>\n",
       "      <td>78617723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8156</th>\n",
       "      <td>chr12</td>\n",
       "      <td>117022835</td>\n",
       "      <td>117022888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8255</th>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8933</th>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8995</th>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9444</th>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>433 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom2     start2       end2\n",
       "9      chr1   18907111   18907133\n",
       "11     chr1   29332374   29332405\n",
       "12     chr1   67532557   67532593\n",
       "13     chr1   70133094   70133130\n",
       "14     chr1   78617702   78617723\n",
       "...     ...        ...        ...\n",
       "8156  chr12  117022835  117022888\n",
       "8255  chr13   49219804   49219836\n",
       "8933  chr16   78770741   78770752\n",
       "8995  chr17    9576625    9576648\n",
       "9444  chr19   55516812   55516824\n",
       "\n",
       "[433 rows x 3 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 看一下这里面有多少个是4、5.1、6 的 IND\n",
    "# df_old_share.mpmat_index.unique()\n",
    "tmpdf = df_bed_to_fix[df_bed_to_fix.end2 != -1].iloc[:, -3:].copy()\n",
    "tmpdf.drop_duplicates(keep='first', inplace=True)\n",
    "tmpdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "7c510b00-90a9-4290-ba18-3b2c427e5ed7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "433\n"
     ]
    }
   ],
   "source": [
    "tmpls = tmpdf.parallel_apply(lambda x: f'{x[0]}_{x[1]}_{x[2]}', axis=1).tolist()\n",
    "tmpls = list(set(tmpls))\n",
    "print(len(tmpls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "d700388c-0e6a-4edc-a5a1-413bd221d72b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "650"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_old_share.mpmat_index.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "a13a4165-a53d-499e-bff2-599fa3b0b4c5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "362\n"
     ]
    }
   ],
   "source": [
    "print(len([i for i in tmpls if i in df_old_share.mpmat_index.unique()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "040a4628-b14c-4cac-8b35-0bc2d605c4e5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>id_ND4</th>\n",
       "      <th>id_ND5.1</th>\n",
       "      <th>id_ND6</th>\n",
       "      <th>id_merge</th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_2044988_2044998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan_ND5.1-TAS.IND-1_nan</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_9702414_9702449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-2</td>\n",
       "      <td>ND6-TAS.IND-1</td>\n",
       "      <td>nan_ND5.1-TAS.IND-2_ND6-TAS.IND-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_12618165_12618174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan_ND5.1-TAS.IND-3_nan</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_13019998_13020019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-2</td>\n",
       "      <td>nan_nan_ND6-TAS.IND-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_13354705_13354738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-3</td>\n",
       "      <td>nan_nan_ND6-TAS.IND-3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX_136925330_136925338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-538</td>\n",
       "      <td>nan_nan_ND6-TAS.IND-538</td>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>chrX_138128211_138128237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND6-TAS.IND-539</td>\n",
       "      <td>nan_nan_ND6-TAS.IND-539</td>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>chrX_150683735_150683758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-452</td>\n",
       "      <td>ND6-TAS.IND-540</td>\n",
       "      <td>nan_ND5.1-TAS.IND-452_ND6-TAS.I...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>150683735</td>\n",
       "      <td>150683758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>chrX_153669436_153669453</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-453</td>\n",
       "      <td>ND6-TAS.IND-541</td>\n",
       "      <td>nan_ND5.1-TAS.IND-453_ND6-TAS.I...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>chrX_155049205_155049215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND5.1-TAS.IND-454</td>\n",
       "      <td>ND6-TAS.IND-542</td>\n",
       "      <td>nan_ND5.1-TAS.IND-454_ND6-TAS.I...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>650 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  mpmat_index id_ND4           id_ND5.1           id_ND6                            id_merge     a          b          c\n",
       "0        chr1_2044988_2044998    NaN    ND5.1-TAS.IND-1              NaN             nan_ND5.1-TAS.IND-1_nan  chr1    2044988    2044998\n",
       "1        chr1_9702414_9702449    NaN    ND5.1-TAS.IND-2    ND6-TAS.IND-1   nan_ND5.1-TAS.IND-2_ND6-TAS.IND-1  chr1    9702414    9702449\n",
       "2      chr1_12618165_12618174    NaN    ND5.1-TAS.IND-3              NaN             nan_ND5.1-TAS.IND-3_nan  chr1   12618165   12618174\n",
       "3      chr1_13019998_13020019    NaN                NaN    ND6-TAS.IND-2               nan_nan_ND6-TAS.IND-2  chr1   13019998   13020019\n",
       "4      chr1_13354705_13354738    NaN                NaN    ND6-TAS.IND-3               nan_nan_ND6-TAS.IND-3  chr1   13354705   13354738\n",
       "..                        ...    ...                ...              ...                                 ...   ...        ...        ...\n",
       "645  chrX_136925330_136925338    NaN                NaN  ND6-TAS.IND-538             nan_nan_ND6-TAS.IND-538  chrX  136925330  136925338\n",
       "646  chrX_138128211_138128237    NaN                NaN  ND6-TAS.IND-539             nan_nan_ND6-TAS.IND-539  chrX  138128211  138128237\n",
       "647  chrX_150683735_150683758    NaN  ND5.1-TAS.IND-452  ND6-TAS.IND-540  nan_ND5.1-TAS.IND-452_ND6-TAS.I...  chrX  150683735  150683758\n",
       "648  chrX_153669436_153669453    NaN  ND5.1-TAS.IND-453  ND6-TAS.IND-541  nan_ND5.1-TAS.IND-453_ND6-TAS.I...  chrX  153669436  153669453\n",
       "649  chrX_155049205_155049215    NaN  ND5.1-TAS.IND-454  ND6-TAS.IND-542  nan_ND5.1-TAS.IND-454_ND6-TAS.I...  chrX  155049205  155049215\n",
       "\n",
       "[650 rows x 8 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# form Old IND456 bed\n",
    "df_old_share['id_merge'] = df_old_share.iloc[:, 1:4].astype(str).agg('_'.join, axis=1)\n",
    "df_old_share[list('abc')] = df_old_share['mpmat_index'].str.split('_', expand=True)\n",
    "df_old_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "a37c3c8f-605c-4c31-ac38-81d9e29072b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_old_share[['a', 'b', 'c', 'id_merge']].to_csv('../tables/2022-11-25_old_650_ND456.bed', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "1143d20b-274f-4c08-97e0-d1b9781f6d5f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_632382_632387</td>\n",
       "      <td>chr1</td>\n",
       "      <td>632382</td>\n",
       "      <td>632387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_633067_633147</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8156</th>\n",
       "      <td>chr12_117022857_117022923</td>\n",
       "      <td>chr12</td>\n",
       "      <td>117022835</td>\n",
       "      <td>117022888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8255</th>\n",
       "      <td>chr13_49219804_49219857</td>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8933</th>\n",
       "      <td>chr16_78770741_78770746</td>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8995</th>\n",
       "      <td>chr17_9576625_9576657</td>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9444</th>\n",
       "      <td>chr19_55516812_55516831</td>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516824</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9968 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mpmat_index chr_name  region_start  region_end\n",
       "0            chr1_630967_631005     chr1        630967      631005\n",
       "1            chr1_631016_631037     chr1        631016      631037\n",
       "2            chr1_631778_631791     chr1        631778      631791\n",
       "3            chr1_632382_632387     chr1        632382      632387\n",
       "4            chr1_633067_633147     chr1        633067      633147\n",
       "...                         ...      ...           ...         ...\n",
       "8156  chr12_117022857_117022923    chr12     117022835   117022888\n",
       "8255    chr13_49219804_49219857    chr13      49219804    49219836\n",
       "8933    chr16_78770741_78770746    chr16      78770741    78770752\n",
       "8995      chr17_9576625_9576657    chr17       9576625     9576648\n",
       "9444    chr19_55516812_55516831    chr19      55516812    55516824\n",
       "\n",
       "[9968 rows x 4 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# end2 为-1 代表老点不存在，\n",
    "# end2 有值说明，新老点重合，以老点 region 坐标为准进行替换\n",
    "\n",
    "# 唯一标识mpmat_index\n",
    "df_bed_to_fix['mpmat_index'] = (\n",
    "    df_bed_to_fix['chrom'] +\n",
    "    '_' +\n",
    "    df_bed_to_fix['start'].astype(str) +\n",
    "    '_' +\n",
    "    df_bed_to_fix['end'].astype(str)\n",
    ")\n",
    "\n",
    "\n",
    "df_bed_to_fix_part1 = (\n",
    "    df_bed_to_fix\n",
    "    .query('end2 == -1')\n",
    "    [['mpmat_index', 'chrom', 'start', 'end']]\n",
    "    .copy()\n",
    ")\n",
    "\n",
    "df_bed_to_fix_part2 = (\n",
    "    df_bed_to_fix\n",
    "    .query('end2 != -1')\n",
    "    [['mpmat_index', 'chrom2', 'start2', 'end2']]\n",
    "    .copy()\n",
    ")\n",
    "\n",
    "df_bed_to_fix_part1.columns = ['mpmat_index', 'chr_name', 'region_start', 'region_end']\n",
    "df_bed_to_fix_part2.columns = ['mpmat_index', 'chr_name', 'region_start', 'region_end']\n",
    "\n",
    "df_bed_fixed_coordinate = pd.concat(\n",
    "    [df_bed_to_fix_part1, df_bed_to_fix_part2],\n",
    "    axis=0\n",
    ")\n",
    "df_bed_fixed_coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "b13675cc-83aa-4644-ad6d-170737691697",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 mpmat_index chr_name  region_start  region_end\n",
      "947  chr10_22989043_22989070    chr10      22989049    22989070\n",
      "948  chr10_22989043_22989070    chr10      22989052    22989071\n",
      "                  mpmat_index chr_name  region_start  region_end\n",
      "3491  chr17_67963461_67963466    chr17      67963444    67963466\n",
      "3492  chr17_67963461_67963466    chr17      67963458    67963466\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          mpmat_index chr_name  region_start  region_end\n",
       "0  chr1_630967_631005     chr1        630967      631005\n",
       "1  chr1_631016_631037     chr1        631016      631037"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 发现两个 duplicated 的 mpmat_index，check 一下为什么\n",
    "# print(df_bed_fixed_coordinate[df_bed_fixed_coordinate.mpmat_index.duplicated()])\n",
    "\n",
    "print(df_bed_fixed_coordinate.query('mpmat_index==\"chr10_22989043_22989070\"'))\n",
    "print(df_bed_fixed_coordinate.query('mpmat_index==\"chr17_67963461_67963466\"'))\n",
    "\n",
    "\n",
    "#                  mpmat_index chr_name  region_start  region_end\n",
    "# 864  chr10_22989043_22989070    chr10      22989049    22989070\n",
    "# 865  chr10_22989043_22989070    chr10      22989052    22989071\n",
    "#                   mpmat_index chr_name  region_start  region_end\n",
    "# 2931  chr17_67963461_67963466    chr17      67963444    67963466\n",
    "# 2932  chr17_67963461_67963466    chr17      67963458    67963466\n",
    "\n",
    "# 发现其实就是没定下来最终的导致了有多出来的点，图方便取第一个，注意以后 debug！\n",
    "# DEBUG\n",
    "\n",
    "df_bed_fixed_coordinate.drop_duplicates(subset='mpmat_index',\n",
    "                                        keep='first',\n",
    "                                        inplace=True)\n",
    "\n",
    "df_bed_fixed_coordinate.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "4e7e51f8-e29e-4ba1-b0fa-4c4354d6f7b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_630992_GA</td>\n",
       "      <td>2975</td>\n",
       "      <td>4635</td>\n",
       "      <td>0.641855</td>\n",
       "      <td>38</td>\n",
       "      <td>6608</td>\n",
       "      <td>1</td>\n",
       "      <td>780</td>\n",
       "      <td>0.096198</td>\n",
       "      <td>39.389272</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>4.649460</td>\n",
       "      <td>0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...</td>\n",
       "      <td>8.677587</td>\n",
       "      <td>10.842847</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.511695e-310</td>\n",
       "      <td>5.507753e-305</td>\n",
       "      <td>chr1_630967_631005_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1732</td>\n",
       "      <td>4662</td>\n",
       "      <td>0.371514</td>\n",
       "      <td>33</td>\n",
       "      <td>6236</td>\n",
       "      <td>0</td>\n",
       "      <td>1189</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>37.171837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.087446</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...</td>\n",
       "      <td>8.797528</td>\n",
       "      <td>9.773088</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1546</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.433782</td>\n",
       "      <td>33</td>\n",
       "      <td>4901</td>\n",
       "      <td>0</td>\n",
       "      <td>770</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>31.497951</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.948668</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...</td>\n",
       "      <td>8.558576</td>\n",
       "      <td>9.254862</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.024876e-307</td>\n",
       "      <td>1.029691e-301</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631791_CT</td>\n",
       "      <td>733</td>\n",
       "      <td>1147</td>\n",
       "      <td>0.639058</td>\n",
       "      <td>40</td>\n",
       "      <td>1483</td>\n",
       "      <td>0</td>\n",
       "      <td>547</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>8.839935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>4.972421e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631787_CT</td>\n",
       "      <td>731</td>\n",
       "      <td>1055</td>\n",
       "      <td>0.692891</td>\n",
       "      <td>40</td>\n",
       "      <td>1447</td>\n",
       "      <td>0</td>\n",
       "      <td>467</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>9.299640</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.001335</td>\n",
       "      <td>0,1,2,3 40,0,0,0 530,320,528,69</td>\n",
       "      <td>6.521031</td>\n",
       "      <td>8.533426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.072057e-182</td>\n",
       "      <td>6.736538e-177</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12344</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>117022835</td>\n",
       "      <td>117022888</td>\n",
       "      <td>chr12_117022835_117022888</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr12_117022887_CT</td>\n",
       "      <td>47</td>\n",
       "      <td>96</td>\n",
       "      <td>0.489583</td>\n",
       "      <td>26</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>0.065819</td>\n",
       "      <td>0.733786</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.176480</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 25,1,...</td>\n",
       "      <td>3.478776</td>\n",
       "      <td>4.888107</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.478798e-12</td>\n",
       "      <td>2.231631e-09</td>\n",
       "      <td>chr12_117022857_117022923_highe...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12345</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219836</td>\n",
       "      <td>chr13_49219804_49219836</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>chr13_49219811_CT</td>\n",
       "      <td>15</td>\n",
       "      <td>55</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>13</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>0.032910</td>\n",
       "      <td>0.394758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.171836</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13...</td>\n",
       "      <td>3.584386</td>\n",
       "      <td>4.401052</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.131671e-11</td>\n",
       "      <td>1.643442e-08</td>\n",
       "      <td>chr13_49219804_49219857_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12346</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770752</td>\n",
       "      <td>chr16_78770741_78770752</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chr16_78770741_CT</td>\n",
       "      <td>31</td>\n",
       "      <td>44</td>\n",
       "      <td>0.704545</td>\n",
       "      <td>12</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.227567</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102173</td>\n",
       "      <td>0,1,2,3,4 12,0,0,0,0 5,22,19,3,0</td>\n",
       "      <td>2.905182</td>\n",
       "      <td>3.927556</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>7.988221e-07</td>\n",
       "      <td>2.144053e-04</td>\n",
       "      <td>chr16_78770741_78770746_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12347</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576648</td>\n",
       "      <td>chr17_9576625_9576648</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr17_9576625_CT</td>\n",
       "      <td>16</td>\n",
       "      <td>61</td>\n",
       "      <td>0.262295</td>\n",
       "      <td>23</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0.058225</td>\n",
       "      <td>0.394758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083596</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 20...</td>\n",
       "      <td>2.761264</td>\n",
       "      <td>3.655176</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.286750e-05</td>\n",
       "      <td>2.301233e-03</td>\n",
       "      <td>chr17_9576625_9576657_highest_c...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12348</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516824</td>\n",
       "      <td>chr19_55516812_55516824</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr19_55516814_CT</td>\n",
       "      <td>18</td>\n",
       "      <td>41</td>\n",
       "      <td>0.439024</td>\n",
       "      <td>11</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0.027847</td>\n",
       "      <td>0.246144</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.116105</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 11...</td>\n",
       "      <td>3.143924</td>\n",
       "      <td>4.310758</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>6.821898e-08</td>\n",
       "      <td>2.361647e-05</td>\n",
       "      <td>chr19_55516812_55516831_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12349 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           <sample> chr_name  region_start  region_end                mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "0       DetectSeq_ATP8-DddA11_REP-1     chr1        630967      631005         chr1_630967_631005                7                      1                    6            chr1_630992_GA   \n",
       "1       DetectSeq_ATP8-DddA11_REP-1     chr1        631016      631037         chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "2        DetectSeq_ATP8-DddA6_REP-1     chr1        631016      631037         chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "3       DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791         chr1_631778_631791                3                      0                    3            chr1_631791_CT   \n",
       "4        DetectSeq_ATP8-DddA6_REP-1     chr1        631778      631791         chr1_631778_631791                3                      0                    3            chr1_631787_CT   \n",
       "...                             ...      ...           ...         ...                        ...              ...                    ...                  ...                       ...   \n",
       "12344  DetectSeq_SIRT6-DddA11_REP-2    chr12     117022835   117022888  chr12_117022835_117022888               11                      0                   11        chr12_117022887_CT   \n",
       "12345  DetectSeq_SIRT6-DddA11_REP-2    chr13      49219804    49219836    chr13_49219804_49219836               15                      0                   15         chr13_49219811_CT   \n",
       "12346  DetectSeq_SIRT6-DddA11_REP-2    chr16      78770741    78770752    chr16_78770741_78770752                4                      0                    4         chr16_78770741_CT   \n",
       "12347  DetectSeq_SIRT6-DddA11_REP-2    chr17       9576625     9576648      chr17_9576625_9576648               12                      0                   12          chr17_9576625_CT   \n",
       "12348  DetectSeq_SIRT6-DddA11_REP-2    chr19      55516812    55516824    chr19_55516812_55516824               12                      0                   12         chr19_55516814_CT   \n",
       "\n",
       "       region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "0                             2975                           4635                       0.641855          38         6608               1              780         0.096198         39.389272   \n",
       "1                             1732                           4662                       0.371514          33         6236               0             1189         0.083540         37.171837   \n",
       "2                             1546                           3564                       0.433782          33         4901               0              770         0.083540         31.497951   \n",
       "3                              733                           1147                       0.639058          40         1483               0              547         0.101261          8.839935   \n",
       "4                              731                           1055                       0.692891          40         1447               0              467         0.101261          9.299640   \n",
       "...                            ...                            ...                            ...         ...          ...             ...              ...              ...               ...   \n",
       "12344                           47                             96                       0.489583          26          158               0               38         0.065819          0.733786   \n",
       "12345                           15                             55                       0.272727          13           85               0               37         0.032910          0.394758   \n",
       "12346                           31                             44                       0.704545          12           49               0               22         0.030378          0.227567   \n",
       "12347                           16                             61                       0.262295          23           85               0               18         0.058225          0.394758   \n",
       "12348                           18                             41                       0.439024          11           53               0               25         0.027847          0.246144   \n",
       "\n",
       "       ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state        p_value            FDR                            bed_name strand  \n",
       "0                 0.002532              4.649460  0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...  8.677587    10.842847     TestOK  2.511695e-310  5.507753e-305  chr1_630967_631005_highest_chr1...      .  \n",
       "1                 0.000000              7.087446  0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...  8.797528     9.773088     TestOK   0.000000e+00   0.000000e+00  chr1_631016_631037_highest_chr1...      .  \n",
       "2                 0.000000              4.948668  0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...  8.558576     9.254862     TestOK  4.024876e-307  1.029691e-301  chr1_631016_631037_highest_chr1...      .  \n",
       "3                 0.000000              3.260583    0,1,2,3 40,0,0,0 534,278,413,258  6.447891     8.652952     TestOK  2.915445e-214  4.972421e-209  chr1_631778_631791_highest_chr1...      .  \n",
       "4                 0.000000              3.001335     0,1,2,3 40,0,0,0 530,320,528,69  6.521031     8.533426     TestOK  3.072057e-182  6.736538e-177  chr1_631778_631791_highest_chr1...      .  \n",
       "...                    ...                   ...                                 ...       ...          ...        ...            ...            ...                                 ...    ...  \n",
       "12344             0.000000              0.176480  0,1,2,3,4,5,6,7,8,9,10,11 25,1,...  3.478776     4.888107     TestOK   2.478798e-12   2.231631e-09  chr12_117022857_117022923_highe...      .  \n",
       "12345             0.000000              0.171836  0,1,2,3,4,5,6,7,8,9,10,11,12,13...  3.584386     4.401052     TestOK   2.131671e-11   1.643442e-08  chr13_49219804_49219857_highest...      .  \n",
       "12346             0.000000              0.102173    0,1,2,3,4 12,0,0,0,0 5,22,19,3,0  2.905182     3.927556     TestOK   7.988221e-07   2.144053e-04  chr16_78770741_78770746_highest...      .  \n",
       "12347             0.000000              0.083596  0,1,2,3,4,5,6,7,8,9,10,11,12 20...  2.761264     3.655176     TestOK   1.286750e-05   2.301233e-03  chr17_9576625_9576657_highest_c...      .  \n",
       "12348             0.000000              0.116105  0,1,2,3,4,5,6,7,8,9,10,11,12 11...  3.143924     4.310758     TestOK   6.821898e-08   2.361647e-05  chr19_55516812_55516831_highest...      .  \n",
       "\n",
       "[12349 rows x 28 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bed_fixed_coordinate2 = (\n",
    "    df_bed_fixed_coordinate\n",
    "    .merge(df_sign_lenient, on=['mpmat_index'], how='left')\n",
    "    .drop(\n",
    "        columns=['chr_name_y', 'region_start_y', 'region_end_y']\n",
    "    )\n",
    "    .rename(\n",
    "        columns={\n",
    "            'chr_name_x': 'chr_name',\n",
    "            'region_start_x': 'region_start',\n",
    "            'region_end_x': 'region_end'}\n",
    "    )\n",
    ")\n",
    "\n",
    "sample_names = df_bed_fixed_coordinate2.pop('<sample>')\n",
    "mpmat_indexes = df_bed_fixed_coordinate2.pop('mpmat_index')\n",
    "# 利用insert方法插入取出的数据列到指定位置\n",
    "\n",
    "df_bed_fixed_coordinate2.insert(0, '<sample>', sample_names)\n",
    "df_bed_fixed_coordinate2.insert(4, 'mpmat_index', mpmat_indexes)\n",
    "\n",
    "# fix mpmat_index\n",
    "df_bed_fixed_coordinate2['mpmat_index'] = df_bed_fixed_coordinate2['chr_name'] + '_' + \\\n",
    "    df_bed_fixed_coordinate2['region_start'].astype(str) + '_' + df_bed_fixed_coordinate2['region_end'].astype(str)\n",
    "\n",
    "df_bed_fixed_coordinate2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "71fe1440-1f26-488f-a1f2-b215a7a526b6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "0\n",
      "1\n",
      "sample=DetectSeq_JAK2-DddA11_REP-1\tdup_num=1\n",
      "                          <sample> chr_name  region_start  region_end              mpmat_index  region_site_num\n",
      "11921  DetectSeq_JAK2-DddA11_REP-1    chr12      81451345    81451359  chr12_81451345_81451359                7\n",
      "\n",
      "0\n",
      "1\n",
      "sample=DetectSeq_SIRT6-DddA11_REP-1\tdup_num=1\n",
      "                           <sample> chr_name  region_start  region_end              mpmat_index  region_site_num\n",
      "11922  DetectSeq_SIRT6-DddA11_REP-1    chr12      81451345    81451359  chr12_81451345_81451359                7\n",
      "\n",
      "1\n",
      "sample=DetectSeq_SIRT6-DddA11_REP-2\tdup_num=1\n",
      "                           <sample> chr_name  region_start  region_end              mpmat_index  region_site_num\n",
      "11923  DetectSeq_SIRT6-DddA11_REP-2    chr12      81451345    81451359  chr12_81451345_81451359                7\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# sample 内部无重合\n",
    "# 若有重合，仔细检查原因\n",
    "for sample, _df in df_bed_fixed_coordinate2.groupby('<sample>'):\n",
    "    dup_num = _df.mpmat_index.duplicated().sum()\n",
    "    print(dup_num)\n",
    "\n",
    "    if dup_num:\n",
    "        print(f'sample={sample}\\tdup_num={dup_num}')\n",
    "        print(_df[_df.mpmat_index.duplicated()].iloc[:, 0:6])\n",
    "        print()\n",
    "    else:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "d8e2ed3b-428d-4bd4-bca7-42cbb3fc6169",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                           <sample> chr_name  region_start  region_end           mpmat_index  region_site_num  region_block_site_num\n",
      "11324   DetectSeq_JAK2-DddA11_REP-1     chr1       9702414     9702449  chr1_9702414_9702449                8                      0\n",
      "11325   DetectSeq_JAK2-DddA11_REP-2     chr1       9702414     9702449  chr1_9702414_9702449                8                      0\n",
      "11326  DetectSeq_SIRT6-DddA11_REP-1     chr1       9702414     9702449  chr1_9702414_9702449                8                      0\n",
      "11327  DetectSeq_SIRT6-DddA11_REP-2     chr1       9702414     9702449  chr1_9702414_9702449                8                      0\n",
      "                           <sample> chr_name  region_start  region_end              mpmat_index  region_site_num  region_block_site_num\n",
      "11918   DetectSeq_JAK2-DddA11_REP-1    chr12      81451345    81451359  chr12_81451345_81451359               16                      0\n",
      "11919  DetectSeq_SIRT6-DddA11_REP-1    chr12      81451345    81451359  chr12_81451345_81451359               16                      0\n",
      "11920  DetectSeq_SIRT6-DddA11_REP-2    chr12      81451345    81451359  chr12_81451345_81451359               16                      0\n",
      "11921   DetectSeq_JAK2-DddA11_REP-1    chr12      81451345    81451359  chr12_81451345_81451359                7                      0\n",
      "11922  DetectSeq_SIRT6-DddA11_REP-1    chr12      81451345    81451359  chr12_81451345_81451359                7                      0\n",
      "11923  DetectSeq_SIRT6-DddA11_REP-2    chr12      81451345    81451359  chr12_81451345_81451359                7                      0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_630992_GA</td>\n",
       "      <td>2975</td>\n",
       "      <td>4635</td>\n",
       "      <td>0.641855</td>\n",
       "      <td>38</td>\n",
       "      <td>6608</td>\n",
       "      <td>1</td>\n",
       "      <td>780</td>\n",
       "      <td>0.096198</td>\n",
       "      <td>39.389272</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>4.649460</td>\n",
       "      <td>0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...</td>\n",
       "      <td>8.677587</td>\n",
       "      <td>10.842847</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.511695e-310</td>\n",
       "      <td>5.507753e-305</td>\n",
       "      <td>chr1_630967_631005_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1732</td>\n",
       "      <td>4662</td>\n",
       "      <td>0.371514</td>\n",
       "      <td>33</td>\n",
       "      <td>6236</td>\n",
       "      <td>0</td>\n",
       "      <td>1189</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>37.171837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.087446</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...</td>\n",
       "      <td>8.797528</td>\n",
       "      <td>9.773088</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1546</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.433782</td>\n",
       "      <td>33</td>\n",
       "      <td>4901</td>\n",
       "      <td>0</td>\n",
       "      <td>770</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>31.497951</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.948668</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...</td>\n",
       "      <td>8.558576</td>\n",
       "      <td>9.254862</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.024876e-307</td>\n",
       "      <td>1.029691e-301</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631791_CT</td>\n",
       "      <td>733</td>\n",
       "      <td>1147</td>\n",
       "      <td>0.639058</td>\n",
       "      <td>40</td>\n",
       "      <td>1483</td>\n",
       "      <td>0</td>\n",
       "      <td>547</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>8.839935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>4.972421e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631787_CT</td>\n",
       "      <td>731</td>\n",
       "      <td>1055</td>\n",
       "      <td>0.692891</td>\n",
       "      <td>40</td>\n",
       "      <td>1447</td>\n",
       "      <td>0</td>\n",
       "      <td>467</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>9.299640</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.001335</td>\n",
       "      <td>0,1,2,3 40,0,0,0 530,320,528,69</td>\n",
       "      <td>6.521031</td>\n",
       "      <td>8.533426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.072057e-182</td>\n",
       "      <td>6.736538e-177</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12344</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>117022835</td>\n",
       "      <td>117022888</td>\n",
       "      <td>chr12_117022835_117022888</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr12_117022887_CT</td>\n",
       "      <td>47</td>\n",
       "      <td>96</td>\n",
       "      <td>0.489583</td>\n",
       "      <td>26</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>0.065819</td>\n",
       "      <td>0.733786</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.176480</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 25,1,...</td>\n",
       "      <td>3.478776</td>\n",
       "      <td>4.888107</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.478798e-12</td>\n",
       "      <td>2.231631e-09</td>\n",
       "      <td>chr12_117022857_117022923_highe...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12345</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219836</td>\n",
       "      <td>chr13_49219804_49219836</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>chr13_49219811_CT</td>\n",
       "      <td>15</td>\n",
       "      <td>55</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>13</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>0.032910</td>\n",
       "      <td>0.394758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.171836</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13...</td>\n",
       "      <td>3.584386</td>\n",
       "      <td>4.401052</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.131671e-11</td>\n",
       "      <td>1.643442e-08</td>\n",
       "      <td>chr13_49219804_49219857_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12346</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770752</td>\n",
       "      <td>chr16_78770741_78770752</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chr16_78770741_CT</td>\n",
       "      <td>31</td>\n",
       "      <td>44</td>\n",
       "      <td>0.704545</td>\n",
       "      <td>12</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.227567</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102173</td>\n",
       "      <td>0,1,2,3,4 12,0,0,0,0 5,22,19,3,0</td>\n",
       "      <td>2.905182</td>\n",
       "      <td>3.927556</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>7.988221e-07</td>\n",
       "      <td>2.144053e-04</td>\n",
       "      <td>chr16_78770741_78770746_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12347</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576648</td>\n",
       "      <td>chr17_9576625_9576648</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr17_9576625_CT</td>\n",
       "      <td>16</td>\n",
       "      <td>61</td>\n",
       "      <td>0.262295</td>\n",
       "      <td>23</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0.058225</td>\n",
       "      <td>0.394758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083596</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 20...</td>\n",
       "      <td>2.761264</td>\n",
       "      <td>3.655176</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.286750e-05</td>\n",
       "      <td>2.301233e-03</td>\n",
       "      <td>chr17_9576625_9576657_highest_c...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12348</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516824</td>\n",
       "      <td>chr19_55516812_55516824</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr19_55516814_CT</td>\n",
       "      <td>18</td>\n",
       "      <td>41</td>\n",
       "      <td>0.439024</td>\n",
       "      <td>11</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0.027847</td>\n",
       "      <td>0.246144</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.116105</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 11...</td>\n",
       "      <td>3.143924</td>\n",
       "      <td>4.310758</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>6.821898e-08</td>\n",
       "      <td>2.361647e-05</td>\n",
       "      <td>chr19_55516812_55516831_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12349 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           <sample> chr_name  region_start  region_end                mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "0       DetectSeq_ATP8-DddA11_REP-1     chr1        630967      631005         chr1_630967_631005                7                      1                    6            chr1_630992_GA   \n",
       "1       DetectSeq_ATP8-DddA11_REP-1     chr1        631016      631037         chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "2        DetectSeq_ATP8-DddA6_REP-1     chr1        631016      631037         chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "3       DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791         chr1_631778_631791                3                      0                    3            chr1_631791_CT   \n",
       "4        DetectSeq_ATP8-DddA6_REP-1     chr1        631778      631791         chr1_631778_631791                3                      0                    3            chr1_631787_CT   \n",
       "...                             ...      ...           ...         ...                        ...              ...                    ...                  ...                       ...   \n",
       "12344  DetectSeq_SIRT6-DddA11_REP-2    chr12     117022835   117022888  chr12_117022835_117022888               11                      0                   11        chr12_117022887_CT   \n",
       "12345  DetectSeq_SIRT6-DddA11_REP-2    chr13      49219804    49219836    chr13_49219804_49219836               15                      0                   15         chr13_49219811_CT   \n",
       "12346  DetectSeq_SIRT6-DddA11_REP-2    chr16      78770741    78770752    chr16_78770741_78770752                4                      0                    4         chr16_78770741_CT   \n",
       "12347  DetectSeq_SIRT6-DddA11_REP-2    chr17       9576625     9576648      chr17_9576625_9576648               12                      0                   12          chr17_9576625_CT   \n",
       "12348  DetectSeq_SIRT6-DddA11_REP-2    chr19      55516812    55516824    chr19_55516812_55516824               12                      0                   12         chr19_55516814_CT   \n",
       "\n",
       "       region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "0                             2975                           4635                       0.641855          38         6608               1              780         0.096198         39.389272   \n",
       "1                             1732                           4662                       0.371514          33         6236               0             1189         0.083540         37.171837   \n",
       "2                             1546                           3564                       0.433782          33         4901               0              770         0.083540         31.497951   \n",
       "3                              733                           1147                       0.639058          40         1483               0              547         0.101261          8.839935   \n",
       "4                              731                           1055                       0.692891          40         1447               0              467         0.101261          9.299640   \n",
       "...                            ...                            ...                            ...         ...          ...             ...              ...              ...               ...   \n",
       "12344                           47                             96                       0.489583          26          158               0               38         0.065819          0.733786   \n",
       "12345                           15                             55                       0.272727          13           85               0               37         0.032910          0.394758   \n",
       "12346                           31                             44                       0.704545          12           49               0               22         0.030378          0.227567   \n",
       "12347                           16                             61                       0.262295          23           85               0               18         0.058225          0.394758   \n",
       "12348                           18                             41                       0.439024          11           53               0               25         0.027847          0.246144   \n",
       "\n",
       "       ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state        p_value            FDR                            bed_name strand  \n",
       "0                 0.002532              4.649460  0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...  8.677587    10.842847     TestOK  2.511695e-310  5.507753e-305  chr1_630967_631005_highest_chr1...      .  \n",
       "1                 0.000000              7.087446  0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...  8.797528     9.773088     TestOK   0.000000e+00   0.000000e+00  chr1_631016_631037_highest_chr1...      .  \n",
       "2                 0.000000              4.948668  0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...  8.558576     9.254862     TestOK  4.024876e-307  1.029691e-301  chr1_631016_631037_highest_chr1...      .  \n",
       "3                 0.000000              3.260583    0,1,2,3 40,0,0,0 534,278,413,258  6.447891     8.652952     TestOK  2.915445e-214  4.972421e-209  chr1_631778_631791_highest_chr1...      .  \n",
       "4                 0.000000              3.001335     0,1,2,3 40,0,0,0 530,320,528,69  6.521031     8.533426     TestOK  3.072057e-182  6.736538e-177  chr1_631778_631791_highest_chr1...      .  \n",
       "...                    ...                   ...                                 ...       ...          ...        ...            ...            ...                                 ...    ...  \n",
       "12344             0.000000              0.176480  0,1,2,3,4,5,6,7,8,9,10,11 25,1,...  3.478776     4.888107     TestOK   2.478798e-12   2.231631e-09  chr12_117022857_117022923_highe...      .  \n",
       "12345             0.000000              0.171836  0,1,2,3,4,5,6,7,8,9,10,11,12,13...  3.584386     4.401052     TestOK   2.131671e-11   1.643442e-08  chr13_49219804_49219857_highest...      .  \n",
       "12346             0.000000              0.102173    0,1,2,3,4 12,0,0,0,0 5,22,19,3,0  2.905182     3.927556     TestOK   7.988221e-07   2.144053e-04  chr16_78770741_78770746_highest...      .  \n",
       "12347             0.000000              0.083596  0,1,2,3,4,5,6,7,8,9,10,11,12 20...  2.761264     3.655176     TestOK   1.286750e-05   2.301233e-03  chr17_9576625_9576657_highest_c...      .  \n",
       "12348             0.000000              0.116105  0,1,2,3,4,5,6,7,8,9,10,11,12 11...  3.143924     4.310758     TestOK   6.821898e-08   2.361647e-05  chr19_55516812_55516831_highest...      .  \n",
       "\n",
       "[12349 rows x 28 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 这个点被 call 到了两次，因为视为了两个点: chr12_81451345_81451359\n",
    "# 后续处理中保留这个点，不再进行去重\n",
    "print(df_bed_fixed_coordinate2.query('mpmat_index==\"chr1_9702414_9702449\"').iloc[:, 0:7])\n",
    "print(df_bed_fixed_coordinate2.query('mpmat_index==\"chr12_81451345_81451359\"').iloc[:, 0:7])\n",
    "\n",
    "\n",
    "df_bed_fixed_coordinate2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "5d1c58e8-7752-4e47-bc8c-5afbeeda04f0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_630992_GA</td>\n",
       "      <td>2975</td>\n",
       "      <td>4635</td>\n",
       "      <td>0.641855</td>\n",
       "      <td>38</td>\n",
       "      <td>6608</td>\n",
       "      <td>1</td>\n",
       "      <td>780</td>\n",
       "      <td>0.096198</td>\n",
       "      <td>39.389272</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>4.649460</td>\n",
       "      <td>0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...</td>\n",
       "      <td>8.677587</td>\n",
       "      <td>10.842847</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.511695e-310</td>\n",
       "      <td>5.507753e-305</td>\n",
       "      <td>chr1_630967_631005_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1732</td>\n",
       "      <td>4662</td>\n",
       "      <td>0.371514</td>\n",
       "      <td>33</td>\n",
       "      <td>6236</td>\n",
       "      <td>0</td>\n",
       "      <td>1189</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>37.171837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.087446</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...</td>\n",
       "      <td>8.797528</td>\n",
       "      <td>9.773088</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1546</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.433782</td>\n",
       "      <td>33</td>\n",
       "      <td>4901</td>\n",
       "      <td>0</td>\n",
       "      <td>770</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>31.497951</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.948668</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...</td>\n",
       "      <td>8.558576</td>\n",
       "      <td>9.254862</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.024876e-307</td>\n",
       "      <td>1.029691e-301</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631791_CT</td>\n",
       "      <td>733</td>\n",
       "      <td>1147</td>\n",
       "      <td>0.639058</td>\n",
       "      <td>40</td>\n",
       "      <td>1483</td>\n",
       "      <td>0</td>\n",
       "      <td>547</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>8.839935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>4.972421e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631787_CT</td>\n",
       "      <td>731</td>\n",
       "      <td>1055</td>\n",
       "      <td>0.692891</td>\n",
       "      <td>40</td>\n",
       "      <td>1447</td>\n",
       "      <td>0</td>\n",
       "      <td>467</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>9.299640</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.001335</td>\n",
       "      <td>0,1,2,3 40,0,0,0 530,320,528,69</td>\n",
       "      <td>6.521031</td>\n",
       "      <td>8.533426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.072057e-182</td>\n",
       "      <td>6.736538e-177</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      <sample> chr_name  region_start  region_end         mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "0  DetectSeq_ATP8-DddA11_REP-1     chr1        630967      631005  chr1_630967_631005                7                      1                    6            chr1_630992_GA   \n",
       "1  DetectSeq_ATP8-DddA11_REP-1     chr1        631016      631037  chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "2   DetectSeq_ATP8-DddA6_REP-1     chr1        631016      631037  chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "3  DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791  chr1_631778_631791                3                      0                    3            chr1_631791_CT   \n",
       "4   DetectSeq_ATP8-DddA6_REP-1     chr1        631778      631791  chr1_631778_631791                3                      0                    3            chr1_631787_CT   \n",
       "\n",
       "   region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "0                         2975                           4635                       0.641855          38         6608               1              780         0.096198         39.389272   \n",
       "1                         1732                           4662                       0.371514          33         6236               0             1189         0.083540         37.171837   \n",
       "2                         1546                           3564                       0.433782          33         4901               0              770         0.083540         31.497951   \n",
       "3                          733                           1147                       0.639058          40         1483               0              547         0.101261          8.839935   \n",
       "4                          731                           1055                       0.692891          40         1447               0              467         0.101261          9.299640   \n",
       "\n",
       "   ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state        p_value            FDR                            bed_name strand  \n",
       "0             0.002532              4.649460  0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...  8.677587    10.842847     TestOK  2.511695e-310  5.507753e-305  chr1_630967_631005_highest_chr1...      .  \n",
       "1             0.000000              7.087446  0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...  8.797528     9.773088     TestOK   0.000000e+00   0.000000e+00  chr1_631016_631037_highest_chr1...      .  \n",
       "2             0.000000              4.948668  0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...  8.558576     9.254862     TestOK  4.024876e-307  1.029691e-301  chr1_631016_631037_highest_chr1...      .  \n",
       "3             0.000000              3.260583    0,1,2,3 40,0,0,0 534,278,413,258  6.447891     8.652952     TestOK  2.915445e-214  4.972421e-209  chr1_631778_631791_highest_chr1...      .  \n",
       "4             0.000000              3.001335     0,1,2,3 40,0,0,0 530,320,528,69  6.521031     8.533426     TestOK  3.072057e-182  6.736538e-177  chr1_631778_631791_highest_chr1...      .  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pois = df_bed_fixed_coordinate2.copy()\n",
    "df_pois.head()\n",
    "\n",
    "# 现在只要是有 nature 当中的 region，都和 nature 的 coordinate\n",
    "# 和mpmat_index保持一致了，后面如果回溯命名，直接和 v4 的 table merge一下即可"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "552f2639-ffda-4bda-ad5b-551a4af539f0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pois.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "c35ea202-8afa-4e3f-b1a4-81f7d045b84f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_pois.to_csv('/Users/zhaohuanan/Downloads/2022-11-25_for_home_use.csv.gz', index=None)\n",
    "df_pois.to_csv('../tables/2022-11-25_for_home_use.csv.gz', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "48816244-5424-4092-baca-a2272e33737a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# del df, df_mpmat, df_bed_fixed_coordinate2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "8dd02cd2-8fc2-491a-a29a-630eccb28385",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %reset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00da264c-82b8-4f18-b3d2-dfbaf670e59c",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 重载预处理好的数据 reload data -> df_pois"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "58914210-d3e9-45c0-b778-865982612169",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_pois = pd.read_csv('/Users/zhaohuanan/Downloads/2022-11-25_for_home_use.csv.gz', header=0, index_col=None)\n",
    "df_pois = pd.read_csv('../tables/2022-11-25_for_home_use.csv.gz', header=0, index_col=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "fc55aa6d",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_highest_site_index</th>\n",
       "      <th>region_highest_site_mut_num</th>\n",
       "      <th>region_highest_site_cover_num</th>\n",
       "      <th>region_highest_site_mut_ratio</th>\n",
       "      <th>ctrl_count</th>\n",
       "      <th>treat_count</th>\n",
       "      <th>ctrl_mut_count</th>\n",
       "      <th>treat_mut_count</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630967</td>\n",
       "      <td>631005</td>\n",
       "      <td>chr1_630967_631005</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_630992_GA</td>\n",
       "      <td>2975</td>\n",
       "      <td>4635</td>\n",
       "      <td>0.641855</td>\n",
       "      <td>38</td>\n",
       "      <td>6608</td>\n",
       "      <td>1</td>\n",
       "      <td>780</td>\n",
       "      <td>0.096198</td>\n",
       "      <td>39.389272</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>4.649460</td>\n",
       "      <td>0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...</td>\n",
       "      <td>8.677587</td>\n",
       "      <td>10.842847</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.511695e-310</td>\n",
       "      <td>5.507753e-305</td>\n",
       "      <td>chr1_630967_631005_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1732</td>\n",
       "      <td>4662</td>\n",
       "      <td>0.371514</td>\n",
       "      <td>33</td>\n",
       "      <td>6236</td>\n",
       "      <td>0</td>\n",
       "      <td>1189</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>37.171837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.087446</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...</td>\n",
       "      <td>8.797528</td>\n",
       "      <td>9.773088</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631016</td>\n",
       "      <td>631037</td>\n",
       "      <td>chr1_631016_631037</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1_631036_CT</td>\n",
       "      <td>1546</td>\n",
       "      <td>3564</td>\n",
       "      <td>0.433782</td>\n",
       "      <td>33</td>\n",
       "      <td>4901</td>\n",
       "      <td>0</td>\n",
       "      <td>770</td>\n",
       "      <td>0.083540</td>\n",
       "      <td>31.497951</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.948668</td>\n",
       "      <td>0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...</td>\n",
       "      <td>8.558576</td>\n",
       "      <td>9.254862</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.024876e-307</td>\n",
       "      <td>1.029691e-301</td>\n",
       "      <td>chr1_631016_631037_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631791_CT</td>\n",
       "      <td>733</td>\n",
       "      <td>1147</td>\n",
       "      <td>0.639058</td>\n",
       "      <td>40</td>\n",
       "      <td>1483</td>\n",
       "      <td>0</td>\n",
       "      <td>547</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>8.839935</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>4.972421e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631787_CT</td>\n",
       "      <td>731</td>\n",
       "      <td>1055</td>\n",
       "      <td>0.692891</td>\n",
       "      <td>40</td>\n",
       "      <td>1447</td>\n",
       "      <td>0</td>\n",
       "      <td>467</td>\n",
       "      <td>0.101261</td>\n",
       "      <td>9.299640</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.001335</td>\n",
       "      <td>0,1,2,3 40,0,0,0 530,320,528,69</td>\n",
       "      <td>6.521031</td>\n",
       "      <td>8.533426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.072057e-182</td>\n",
       "      <td>6.736538e-177</td>\n",
       "      <td>chr1_631778_631791_highest_chr1...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12344</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>117022835</td>\n",
       "      <td>117022888</td>\n",
       "      <td>chr12_117022835_117022888</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr12_117022887_CT</td>\n",
       "      <td>47</td>\n",
       "      <td>96</td>\n",
       "      <td>0.489583</td>\n",
       "      <td>26</td>\n",
       "      <td>158</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>0.065819</td>\n",
       "      <td>0.733786</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.176480</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 25,1,...</td>\n",
       "      <td>3.478776</td>\n",
       "      <td>4.888107</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.478798e-12</td>\n",
       "      <td>2.231631e-09</td>\n",
       "      <td>chr12_117022857_117022923_highe...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12345</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr13</td>\n",
       "      <td>49219804</td>\n",
       "      <td>49219836</td>\n",
       "      <td>chr13_49219804_49219836</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>chr13_49219811_CT</td>\n",
       "      <td>15</td>\n",
       "      <td>55</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>13</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>0.032910</td>\n",
       "      <td>0.394758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.171836</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13...</td>\n",
       "      <td>3.584386</td>\n",
       "      <td>4.401052</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.131671e-11</td>\n",
       "      <td>1.643442e-08</td>\n",
       "      <td>chr13_49219804_49219857_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12346</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr16</td>\n",
       "      <td>78770741</td>\n",
       "      <td>78770752</td>\n",
       "      <td>chr16_78770741_78770752</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chr16_78770741_CT</td>\n",
       "      <td>31</td>\n",
       "      <td>44</td>\n",
       "      <td>0.704545</td>\n",
       "      <td>12</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0.030378</td>\n",
       "      <td>0.227567</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102173</td>\n",
       "      <td>0,1,2,3,4 12,0,0,0,0 5,22,19,3,0</td>\n",
       "      <td>2.905182</td>\n",
       "      <td>3.927556</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>7.988221e-07</td>\n",
       "      <td>2.144053e-04</td>\n",
       "      <td>chr16_78770741_78770746_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12347</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>9576625</td>\n",
       "      <td>9576648</td>\n",
       "      <td>chr17_9576625_9576648</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr17_9576625_CT</td>\n",
       "      <td>16</td>\n",
       "      <td>61</td>\n",
       "      <td>0.262295</td>\n",
       "      <td>23</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0.058225</td>\n",
       "      <td>0.394758</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.083596</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 20...</td>\n",
       "      <td>2.761264</td>\n",
       "      <td>3.655176</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.286750e-05</td>\n",
       "      <td>2.301233e-03</td>\n",
       "      <td>chr17_9576625_9576657_highest_c...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12348</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>55516812</td>\n",
       "      <td>55516824</td>\n",
       "      <td>chr19_55516812_55516824</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>chr19_55516814_CT</td>\n",
       "      <td>18</td>\n",
       "      <td>41</td>\n",
       "      <td>0.439024</td>\n",
       "      <td>11</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0.027847</td>\n",
       "      <td>0.246144</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.116105</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12 11...</td>\n",
       "      <td>3.143924</td>\n",
       "      <td>4.310758</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>6.821898e-08</td>\n",
       "      <td>2.361647e-05</td>\n",
       "      <td>chr19_55516812_55516831_highest...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12349 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           <sample> chr_name  region_start  region_end                mpmat_index  region_site_num  region_block_site_num  region_mut_site_num region_highest_site_index  \\\n",
       "0       DetectSeq_ATP8-DddA11_REP-1     chr1        630967      631005         chr1_630967_631005                7                      1                    6            chr1_630992_GA   \n",
       "1       DetectSeq_ATP8-DddA11_REP-1     chr1        631016      631037         chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "2        DetectSeq_ATP8-DddA6_REP-1     chr1        631016      631037         chr1_631016_631037                6                      0                    6            chr1_631036_CT   \n",
       "3       DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791         chr1_631778_631791                3                      0                    3            chr1_631791_CT   \n",
       "4        DetectSeq_ATP8-DddA6_REP-1     chr1        631778      631791         chr1_631778_631791                3                      0                    3            chr1_631787_CT   \n",
       "...                             ...      ...           ...         ...                        ...              ...                    ...                  ...                       ...   \n",
       "12344  DetectSeq_SIRT6-DddA11_REP-2    chr12     117022835   117022888  chr12_117022835_117022888               11                      0                   11        chr12_117022887_CT   \n",
       "12345  DetectSeq_SIRT6-DddA11_REP-2    chr13      49219804    49219836    chr13_49219804_49219836               15                      0                   15         chr13_49219811_CT   \n",
       "12346  DetectSeq_SIRT6-DddA11_REP-2    chr16      78770741    78770752    chr16_78770741_78770752                4                      0                    4         chr16_78770741_CT   \n",
       "12347  DetectSeq_SIRT6-DddA11_REP-2    chr17       9576625     9576648      chr17_9576625_9576648               12                      0                   12          chr17_9576625_CT   \n",
       "12348  DetectSeq_SIRT6-DddA11_REP-2    chr19      55516812    55516824    chr19_55516812_55516824               12                      0                   12         chr19_55516814_CT   \n",
       "\n",
       "       region_highest_site_mut_num  region_highest_site_cover_num  region_highest_site_mut_ratio  ctrl_count  treat_count  ctrl_mut_count  treat_mut_count  ctrl_count.norm  treat_count.norm  \\\n",
       "0                             2975                           4635                       0.641855          38         6608               1              780         0.096198         39.389272   \n",
       "1                             1732                           4662                       0.371514          33         6236               0             1189         0.083540         37.171837   \n",
       "2                             1546                           3564                       0.433782          33         4901               0              770         0.083540         31.497951   \n",
       "3                              733                           1147                       0.639058          40         1483               0              547         0.101261          8.839935   \n",
       "4                              731                           1055                       0.692891          40         1447               0              467         0.101261          9.299640   \n",
       "...                            ...                            ...                            ...         ...          ...             ...              ...              ...               ...   \n",
       "12344                           47                             96                       0.489583          26          158               0               38         0.065819          0.733786   \n",
       "12345                           15                             55                       0.272727          13           85               0               37         0.032910          0.394758   \n",
       "12346                           31                             44                       0.704545          12           49               0               22         0.030378          0.227567   \n",
       "12347                           16                             61                       0.262295          23           85               0               18         0.058225          0.394758   \n",
       "12348                           18                             41                       0.439024          11           53               0               25         0.027847          0.246144   \n",
       "\n",
       "       ctrl_mut_count.norm  treat_mut_count.norm                          count_info   log2_FC  log2_FC_mut test_state        p_value            FDR                            bed_name strand  \n",
       "0                 0.002532              4.649460  0,1,2,3,4,5,6,7 35,2,1,0,0,0,0,...  8.677587    10.842847     TestOK  2.511695e-310  5.507753e-305  chr1_630967_631005_highest_chr1...      .  \n",
       "1                 0.000000              7.087446  0,1,2,3,4,5,6 33,0,0,0,0,0,0 31...  8.797528     9.773088     TestOK   0.000000e+00   0.000000e+00  chr1_631016_631037_highest_chr1...      .  \n",
       "2                 0.000000              4.948668  0,1,2,3,4,5,6 33,0,0,0,0,0,0 24...  8.558576     9.254862     TestOK  4.024876e-307  1.029691e-301  chr1_631016_631037_highest_chr1...      .  \n",
       "3                 0.000000              3.260583    0,1,2,3 40,0,0,0 534,278,413,258  6.447891     8.652952     TestOK  2.915445e-214  4.972421e-209  chr1_631778_631791_highest_chr1...      .  \n",
       "4                 0.000000              3.001335     0,1,2,3 40,0,0,0 530,320,528,69  6.521031     8.533426     TestOK  3.072057e-182  6.736538e-177  chr1_631778_631791_highest_chr1...      .  \n",
       "...                    ...                   ...                                 ...       ...          ...        ...            ...            ...                                 ...    ...  \n",
       "12344             0.000000              0.176480  0,1,2,3,4,5,6,7,8,9,10,11 25,1,...  3.478776     4.888107     TestOK   2.478798e-12   2.231631e-09  chr12_117022857_117022923_highe...      .  \n",
       "12345             0.000000              0.171836  0,1,2,3,4,5,6,7,8,9,10,11,12,13...  3.584386     4.401052     TestOK   2.131671e-11   1.643442e-08  chr13_49219804_49219857_highest...      .  \n",
       "12346             0.000000              0.102173    0,1,2,3,4 12,0,0,0,0 5,22,19,3,0  2.905182     3.927556     TestOK   7.988221e-07   2.144053e-04  chr16_78770741_78770746_highest...      .  \n",
       "12347             0.000000              0.083596  0,1,2,3,4,5,6,7,8,9,10,11,12 20...  2.761264     3.655176     TestOK   1.286750e-05   2.301233e-03  chr17_9576625_9576657_highest_c...      .  \n",
       "12348             0.000000              0.116105  0,1,2,3,4,5,6,7,8,9,10,11,12 11...  3.143924     4.310758     TestOK   6.821898e-08   2.361647e-05  chr19_55516812_55516831_highest...      .  \n",
       "\n",
       "[12349 rows x 28 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pois"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "d4fcd124",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sample>\n",
       "DetectSeq_ATP8-DddA11_REP-1      177\n",
       "DetectSeq_ATP8-DddA6_REP-1       589\n",
       "DetectSeq_ATP8-DddAwt_REP-1       57\n",
       "DetectSeq_JAK2-DddA11_REP-1      628\n",
       "DetectSeq_JAK2-DddA11_REP-2      667\n",
       "DetectSeq_SIRT6-DddA11_REP-1    2801\n",
       "DetectSeq_SIRT6-DddA11_REP-2    7430\n",
       "Name: mpmat_index, dtype: int64"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pois.groupby('<sample>').mpmat_index.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72eb92d4-39ae-45d1-be45-9e3e9002a847",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 看一下有没有重合的\n",
    "# 有的话去上一步 fix 掉\n",
    "df_pois.groupby('<sample>').mpmat_index.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3674919-0c5c-4238-ac80-09ef45f6a98c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 这里 fix 一个连起来的点chr12_81451345_81451359\n",
    "df_pois[df_pois['mpmat_index']!='chr12_81451345_81451359'].groupby('<sample>').mpmat_index.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2344f9e-0756-4c55-a023-b7b14923edcb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# [TODO] share analysis of different DdCBE treatment -> upset plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d23ee8c-5573-42a8-b8bd-5b8a8a70cd81",
   "metadata": {},
   "source": [
    "## [TODO]plot upset-plot: share info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c000aace-405d-47c5-975a-da994ebd0995",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# 使用ggpubr包中的ggarrange()函数来排版多个图形：\n",
    "# ggarrange(bxp, dp, bp + rremove(\"x.text\"), labels=c('A', 'B', 'C'), ncol=2, nrow=2)\n",
    "\n",
    "# 这个简书写的太好了！\n",
    "# https://www.jianshu.com/p/c154ca35530b\n",
    "# 注释排版的图形\n",
    "# 对齐绘图区\n",
    "# 更改图形的行列跨度\n",
    "# 使用cowplot包中的draw_plot_label()函数注释图形\n",
    "\n",
    "\n",
    "# %%R -i df2r\n",
    "# df = as.data.frame(df2r)\n",
    "# # head(df, 10)\n",
    "\n",
    "# g_a = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=ATP8_DddAwt_1,\n",
    "#             y=ATP8_DddA6_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_continuous(\n",
    "#         name=\"Detect-seq_signals of ATP8_DddAwt_1\",\n",
    "#         limits=c(0, 0.5),\n",
    "#     ) +\n",
    "#     scale_y_continuous(\n",
    "#         name=\"Detect-seq_signals of ATP8_DddA6_1\",\n",
    "#         limits=c(0, 0.5),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"ATP8_DddAwt_1 v.s. ATP8_DddA6_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_ATP8_DddAwt_1v.s.ATP8_DddA6_1.pdf', plot=g_a, width=4.6, height=4)\n",
    "\n",
    "\n",
    "# df = df %>%\n",
    "#     group_by(color) %>% summarise(count=n()) %>%\n",
    "#     mutate(\n",
    "#         group=color,\n",
    "#         csum=rev(cumsum(rev(count))),\n",
    "#         total=sum(count),\n",
    "#         ratio=count/total*100,\n",
    "#         pos=count/2 + lead(csum, 1),\n",
    "#         pos=if_else(is.na(pos), count/2, pos)\n",
    "#     )\n",
    "# g_b = ggplot(data=df, mapping=aes(x=\"\", y=count, fill=color, group=fct_inorder(color))) +\n",
    "#     geom_col(color=\"white\") +\n",
    "#     geom_text(\n",
    "#         aes(label=sprintf(\"%i\\n%.0f%%\", count, ratio)),  # 标注\n",
    "#         position=position_stack(vjust=0.5),\n",
    "#         size=3\n",
    "#     ) +\n",
    "#     coord_polar(theta=\"y\") +\n",
    "#     scale_fill_brewer(palette=\"Pastel2\") +\n",
    "#     ggtitle(\"Pie: ATP8_DddAwt_1 v.s. ATP8_DddA6_1\") +\n",
    "#     theme_void()\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_pie-plot_ATP8_DddAwt_1v.s.ATP8_DddA6_1.pdf', plot=g_b, width=5.2, height=4)\n",
    "# print(g_a)\n",
    "# print(g_b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da1d3983-5df1-402a-8679-dbf6966e733f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_upset = df_wide_score.set_index('mpmat_index').applymap(lambda x: True if x >= 0.001 else False)\n",
    "# # .T\n",
    "# df_upset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f118434b-0dba-47fc-b941-275068713e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_upset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2353d71-26f7-42ba-8500-7e0551f5925a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R -i df_upset\n",
    "\n",
    "# df_upset = tibble(df_upset)\n",
    "\n",
    "# g = df_upset %>%\n",
    "#     group_by(OTs) %>%\n",
    "#     summarize(Treatments=list(Treatment)) %>%\n",
    "#     ggplot(aes(x=Treatments)) +\n",
    "#         geom_bar() +\n",
    "#         scale_x_upset() +\n",
    "#         scale_y_log10(\n",
    "#             breaks=c(10, 20, 30, 40, 50, 100, 200, 300, 400, 500, 2000, 2500)\n",
    "#         ) +\n",
    "#         ylab('log10_OTs\\' count') +\n",
    "#         ggtitle('Off-target sharing of different DdCBEs')\n",
    "#         theme_bw()\n",
    "# ggsave('2022-10-30_Detect-seq_upset-plot_all.pdf', plot=g, width=8, height=9)\n",
    "# g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d14976e-3dff-45b2-8d7b-668a50a2812b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# g = ggplot(df_upset) +\n",
    "#     geom_bar(\n",
    "#         aes(x=Treatment, fill=Treatment, group=Treatment),\n",
    "#         stat='count',\n",
    "#         width=0.8\n",
    "#     ) +\n",
    "#     scale_y_continuous(\n",
    "#         limits=c(0, 2800),\n",
    "#         breaks=c(0, 50, 100, 200, 500, 1000, 2000, 2650)\n",
    "#     ) +\n",
    "#     ylab(\"total counts\") +\n",
    "#     ggtitle(\"Total OTs induced by DdCBEs\") +\n",
    "#     coord_flip() +\n",
    "#     theme_classic() +\n",
    "#     theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1, size=6))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_bar-plot_all.pdf', plot=g, width=8, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b7f2b3-1fc1-4d1a-b390-901fd179870f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# df_upset %>%\n",
    "#     group_by(OTs) %>%\n",
    "#     filter(grepl(\"ATP8\", Treatment))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6d3b661-dc21-425b-bf5d-969004abf85c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "# g = df_upset %>%\n",
    "#     group_by(OTs) %>%\n",
    "#     filter(grepl(\"ATP8\", Treatment)) %>%\n",
    "#     summarize(Treatments=list(Treatment)) %>%\n",
    "#     ggplot(aes(x=Treatments)) +\n",
    "#         geom_bar() +\n",
    "#         scale_x_upset() +\n",
    "#         scale_y_continuous(\n",
    "#             breaks=seq(0, 400, 50)\n",
    "#         ) +\n",
    "#         ggtitle('Off-target sharing of different ATP8-DdCBEs')\n",
    "#         theme_bw()\n",
    "# ggsave('2022-10-30_Detect-seq_upset-plot_all_ATP8.pdf', plot=g, width=8, height=9)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa5df120-f08d-4115-aeee-bcdf43524902",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Detect-seq signal comparation -> scatter plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfeca4fb-215a-4119-889d-5cdeb5765435",
   "metadata": {},
   "source": [
    "## data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f116a34-81d6-4970-b3f2-a878bffba130",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pois"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fd78b5c-c61a-4c5a-a9cb-49f1ead8e638",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pois.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5898580c-ac85-4246-98af-8e14a614da0c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pois['test_state'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61070182-1d78-4acf-9876-5883cd3898ce",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pois_sel = df_pois[['<sample>', 'mpmat_index',\n",
    "                       'treat_mut_count.norm', 'treat_count.norm']]\n",
    "df_pois_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f63d52a9-2b1d-4be6-bbf4-1d2b315cac3c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 计算detect_seq_score\n",
    "def calculate_detect_seq_score(x):\n",
    "    treat_mut_count_norm = x['treat_mut_count.norm']\n",
    "    treat_count_norm = x['treat_count.norm']\n",
    "    treat_mut_ratio = treat_mut_count_norm / treat_count_norm\n",
    "\n",
    "    if treat_count_norm != 0:\n",
    "        score = (treat_mut_ratio ** 2) * treat_mut_count_norm * 100\n",
    "    else:\n",
    "        print(x)\n",
    "        raise ValueError\n",
    "    return score\n",
    "\n",
    "\n",
    "df_pois_sel = df_pois_sel.assign(\n",
    "    sample=df_pois_sel['<sample>'],\n",
    "    detect_seq_score=df_pois_sel.parallel_apply(\n",
    "        calculate_detect_seq_score, axis=1\n",
    "    ),\n",
    ")\n",
    "\n",
    "\n",
    "df_pois_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9b90689-d795-4ed1-9b09-126259b1f056",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_detect_seq_score = df_pois_sel[['mpmat_index', 'sample', 'detect_seq_score']].copy()\n",
    "df_detect_seq_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d14464d-ab7a-4251-88db-58cadb7a9fe5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_detect_seq_score.detect_seq_score.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12c5badc-64c3-4640-985b-b9e9c52dec6b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_detect_seq_score['cut'] = pd.cut(df_detect_seq_score.detect_seq_score, bins=range(0,180, 10))\n",
    "df_detect_seq_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07b5183-e730-4813-8fad-d090cff42de8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_detect_seq_score.groupby(['sample', 'cut']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f45fd1-93cf-42fd-9356-4a982dbb4584",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 画个图了解一下 Detect-seq signal 的分布情况\n",
    "for sample, _df in df_detect_seq_score.groupby(['sample', 'cut']).count().reset_index().groupby('sample'):\n",
    "    # print(_df.reset_index())\n",
    "    _df = _df.reset_index()\n",
    "    _df['cut'] = _df['cut'].map(lambda x: str(x).split(',')[-1].replace(']', '')).astype(int)\n",
    "    ax = _df.plot(kind='bar', x='cut', y='detect_seq_score', title=sample)\n",
    "    ax.set_xlabel('score')\n",
    "    ax.set_ylabel('count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "309a8f20-030e-4bed-8ef8-52bb572ec75e",
   "metadata": {},
   "outputs": [],
   "source": [
    "_df.plot?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1749830a-bf51-4dac-ab09-dccfd93a0b44",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_wide = pd.DataFrame(\n",
    "    df_detect_seq_score.mpmat_index.unique(),\n",
    "    columns=['mpmat_index'])\n",
    "\n",
    "\n",
    "for _dfn, _df in df_detect_seq_score.groupby('sample'):\n",
    "    print(_dfn)\n",
    "    _df.columns = ['mpmat_index', _dfn, f'score_{_dfn}']\n",
    "    df_wide = pd.merge(\n",
    "        left=df_wide, right=_df.iloc[:, [0, 2]],\n",
    "        on='mpmat_index', how='left')\n",
    "\n",
    "# df_wide_score = df_wide.fillna(0)\n",
    "df_wide_score = df_wide\n",
    "df_wide_score\n",
    "# df_wide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abc4edee-1874-44cb-9410-6fcb74ab198d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 重命名列名\n",
    "df_wide_score.columns = [\n",
    "    \"mpmat_index\",\n",
    "    \"ATP8_DddA11_1\",\n",
    "    \"ATP8_DddA6_1\",\n",
    "    \"ATP8_DddAwt_1\",\n",
    "    \"JAK2_DddA11_1\",\n",
    "    \"JAK2_DddA11_2\",\n",
    "    \"SIRT6_DddA11_1\",\n",
    "    \"SIRT6_DddA11_2\",\n",
    "]\n",
    "\n",
    "df_wide_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ead90502-b98d-4d3e-a5c8-c6e561ae0b19",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_wide_score = df_wide_score[df_wide_score.iloc[:,1:].sum(axis=1) > 0.1].copy()\n",
    "# df_wide_score\n",
    "df_wide_score.query('mpmat_index==\"chr2_59759874_59759954\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42cfb8cb-b5bc-44b9-8493-b3c55d675323",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pois.query('mpmat_index==\"chr2_59759874_59759954\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd22f6a5-a1b6-4394-b699-ee1051a86a66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df.query('mpmat_index==\"chr2_59759874_59759954\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "269a6825-c75d-4b6b-a27a-59d264eef60f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def map_color(x):\n",
    "    color = ''\n",
    "    if x.iloc[1] >= 0.01:\n",
    "        color += 'a'\n",
    "    if x.iloc[2] >= 0.01:\n",
    "        color += 'b'\n",
    "    # a, b, ab, ''\n",
    "    if len(color) == 0:\n",
    "        color = np.NaN\n",
    "\n",
    "    return color"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbf7de05-6a1a-4991-9250-9b1ead6e2ccf",
   "metadata": {},
   "source": [
    "## plot comparison between different DddA variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bdacdc5-df74-46e0-b152-1895c82e97bb",
   "metadata": {},
   "source": [
    "### {scatter plot} ATP8_DddAwt_1 v.s. ATP8_DddA6_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "032f8862-a603-4ef6-a7b8-501f35706ca5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load data in R env\n",
    "df2r = df_wide_score[['mpmat_index', 'ATP8_DddAwt_1', 'ATP8_DddA6_1']].copy()\n",
    "df2r['color'] = df2r.parallel_apply(map_color, axis=1)\n",
    "df2r = df2r.query('color.notnull()')\n",
    "df2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c5b9335-1e04-4a84-9628-a09bd7b9fdde",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df2r.color.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb4c535d-d748-4362-acec-c0439a0edb61",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df2r.query('mpmat_index==\"chr2_88803448_88803454\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ab1eaa1-7da8-4755-b5ba-26183693e59f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 有可疑的点在这里 check\n",
    "# df_pois.query('mpmat_index==\"chr22_13878035_13878043\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b9cccd7-862e-490b-bbd5-9dad7293ab44",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 因为图放不下，所以临时 fix 以下\n",
    "# df2r.iloc[5, 1:3] = df2r.iloc[5, 1:3] / 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "653dca3a-08f4-4fca-9181-b09b658de36d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df2r.iloc[5, 1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08cd08d4-2c1b-4ad5-aafc-b89b731b0a7c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df2r -w 500 -h 500\n",
    "df = tibble(df2r)\n",
    "# head(df, 10)\n",
    "\n",
    "g = ggplot(data=df) +\n",
    "    geom_point(\n",
    "        mapping=aes(\n",
    "            x=ATP8_DddAwt_1,\n",
    "            y=ATP8_DddA6_1,\n",
    "            alpha=0.99,\n",
    "            # size=1,\n",
    "            color=color,\n",
    "            fill=color,\n",
    "        )\n",
    "    ) +\n",
    "    scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "    scale_x_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddAwt_1\",\n",
    "        # limits=c(0, 66),\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddA6_1\",\n",
    "        # limits=c(0, 66),\n",
    "    ) +\n",
    "    geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "    annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "    ggtitle(\"ATP8_DddAwt_1 v.s. ATP8_DddA6_1\") +\n",
    "    theme_classic() +\n",
    "    theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "ggsave('2022-10-30_Detect-seq_scatter-plot_ATP8_DddAwt_1v.s.ATP8_DddA6_1.pdf', plot=g, width=4.6, height=4)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e6d3b5b-143b-47cb-bff8-22be334a0bce",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# g = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=ATP8_DddAwt_1,\n",
    "#             y=ATP8_DddA6_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#             # shape=color,\n",
    "#             # group=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddAwt_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "#     ) +\n",
    "#     scale_y_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddA6_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"log10_ATP8_DddAwt_1 v.s. ATP8_DddA6_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_log10_ATP8_DddAwt_1v.s.ATP8_DddA6_1.pdf', plot=g, width=4.6, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a58e15a7-0ac2-4d94-87d3-7d169e5bc831",
   "metadata": {},
   "source": [
    "#### 问题：关于 wt 独立的点需要 check 一下 IGV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1fbd15f-352d-42d0-9c11-b69581a08233",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_igv = df2r.query('color==\"a\"').copy()\n",
    "df_igv[list('abc')] = df_igv.mpmat_index.str.split('_', expand=True)\n",
    "print(df_igv.shape[0])\n",
    "df_igv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa0828c7-95e1-4b71-81a1-40a445d0efa9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_pois.query('mpmat_index==\"chr2_59759874_59759954\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06258293-3d93-46a8-a878-f721ea27f93d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_igv = df_igv.merge(df_old_share, how='left')\n",
    "df_igv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37e69362-a63e-4bf9-92fc-eb7244f0f50f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_igv = df_igv[['a', 'b', 'c', 'id_merge']].fillna('nan').copy()\n",
    "df_igv[['b', 'c']] = df_igv[['b', 'c']].astype(int)\n",
    "df_igv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2cbbd3a-edbf-4e8a-be5b-8c908a69f460",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_igv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a8503fc-943a-4ac4-82db-055e29cb02b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 填写相关信息\n",
    "\n",
    "path_out = '/Volumes/Data-a/Bio/3.project/2022_DdCBE-3D-Genome_topic/2022-09-30_Detect-seq_batch-1/igv'\n",
    "date = 20221125\n",
    "format_ = \"png\"\n",
    "height = 1500\n",
    "\n",
    "# 格式化脚本\n",
    "text = f\"maxPanelHeight {height}\\nsnapshotDirectory {path_out}/off-targets_{date}\\n\\n\"\n",
    "# print(text)\n",
    "\n",
    "\n",
    "df_snapshot = df_igv.iloc[:, 0:4]\n",
    "\n",
    "for index, row_info in df_snapshot.iterrows():\n",
    "    chrom, start, stop, bed_name = row_info\n",
    "\n",
    "    path_out_png = f'{chrom}_{start}_{stop}_{bed_name}.snapshot.{format_}'\n",
    "    middle = int((start + stop) / 2)\n",
    "\n",
    "    text += f\"goto {chrom}:{middle - 100}-{middle + 100}\\nsort position\\nexpand\\nviewaspairs\\nsnapshot {path_out_png}\\n\\n\"\n",
    "print(text[:1000])\n",
    "\n",
    "\n",
    "with open(f'{path_out}/{date}_off-targets_snapshot.igv_shot_script', 'wt') as f:\n",
    "    f.write(text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f8c1afc-0ac9-4121-b8d3-c3cf1caa02b7",
   "metadata": {},
   "source": [
    "### {scatter plot} ATP8_DddAwt_1 vs ATP8_DddA11_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b0bbe26-9e50-472e-988c-afa113863cb3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load data in R env\n",
    "df2r = df_wide_score[['mpmat_index', 'ATP8_DddAwt_1', 'ATP8_DddA11_1']].copy()\n",
    "df2r['color'] = df2r.apply(map_color, axis=1)\n",
    "df2r = df2r.query('color.notnull()')\n",
    "df2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adf9fa69-eebc-4f68-bd38-a970e60c6898",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df2r -w 500 -h 500\n",
    "df = as.data.frame(df2r)\n",
    "# head(df, 10)\n",
    "\n",
    "g = ggplot(data=df) +\n",
    "    geom_point(\n",
    "        mapping=aes(\n",
    "            x=ATP8_DddAwt_1,\n",
    "            y=ATP8_DddA11_1,\n",
    "            # alpha=0.99,\n",
    "            # size=1,\n",
    "            color=color,\n",
    "            fill=color,\n",
    "            # shape=color,\n",
    "            # group=color,\n",
    "        )\n",
    "    ) +\n",
    "    scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "    scale_x_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddAwt_1\",\n",
    "        limits=c(0, 0.5),\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddA11_1\",\n",
    "        limits=c(0, 0.5),\n",
    "    ) +\n",
    "    geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "    annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "    ggtitle(\"ATP8_DddAwt_1 v.s. ATP8_DddA11_1\") +\n",
    "    theme_classic() +\n",
    "    theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "ggsave('2022-10-30_Detect-seq_scatter-plot_ATP8_DddAwt_1v.s.ATP8_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "120e333f-eb40-4cdc-9362-def8836f52a8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# g = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=ATP8_DddAwt_1,\n",
    "#             y=ATP8_DddA11_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#             # shape=color,\n",
    "#             # group=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddAwt_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "#     ) +\n",
    "#     scale_y_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"log10_ATP8_DddAwt_1 v.s. ATP8_DddA11_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_log10_ATP8_DddAwt_1v.s.ATP8_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "918e368b-9bda-4b94-b926-2bd5336b87b3",
   "metadata": {
    "tags": []
   },
   "source": [
    "### {scatter plot} ATP8_DddA6_1 vs ATP8_DddA11_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2b42eab-18cb-453b-8e1e-6fb4e36d09ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load data in R env\n",
    "df2r = df_wide_score[['mpmat_index', 'ATP8_DddA6_1', 'ATP8_DddA11_1']].copy()\n",
    "df2r['color'] = df2r.apply(map_color, axis=1)\n",
    "df2r = df2r.query('color.notnull()')\n",
    "df2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c564a0-271a-41ab-bd3f-f384928a7ad1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df2r\n",
    "df = as.data.frame(df2r)\n",
    "# head(df, 10)\n",
    "\n",
    "g = ggplot(data=df) +\n",
    "    geom_point(\n",
    "        mapping=aes(\n",
    "            x=ATP8_DddA6_1,\n",
    "            y=ATP8_DddA11_1,\n",
    "            # alpha=0.99,\n",
    "            # size=1,\n",
    "            color=color,\n",
    "            fill=color,\n",
    "            # shape=color,\n",
    "            # group=color,\n",
    "        )\n",
    "    ) +\n",
    "    scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "    scale_x_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddA6_1\",\n",
    "        limits=c(0, 0.5),\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddA11_1\",\n",
    "        limits=c(0, 0.5),\n",
    "    ) +\n",
    "    geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "    annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "    ggtitle(\"ATP8_DddA6_1 v.s. ATP8_DddA11_1\") +\n",
    "    theme_classic() +\n",
    "    theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "ggsave('2022-10-30_Detect-seq_scatter-plot_ATP8_DddA6_1v.s.ATP8_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49f41413-b852-4ef7-9b79-177844c90e4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# g = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=ATP8_DddA6_1,\n",
    "#             y=ATP8_DddA11_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#             # shape=color,\n",
    "#             # group=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddA6_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "#     ) +\n",
    "#     scale_y_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"log10_ATP8_DddA6_1 v.s. ATP8_DddA11_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_log10_ATP8_DddA6_1v.s.ATP8_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01de76d1-5841-45e3-8ce2-a49154ed251a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### plot ATP8_DddA11_1 vs JAK2_DddA11_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21e8c0ff-22d1-43c2-afa1-5d2bdcc0abf3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load data in R env\n",
    "df2r = df_wide_score[['mpmat_index', 'ATP8_DddA11_1', 'JAK2_DddA11_1']].copy()\n",
    "df2r['color'] = df2r.apply(map_color, axis=1)\n",
    "df2r = df2r.query('color.notnull()')\n",
    "df2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8bc4647-9041-4250-9e3c-ced7bc909b06",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df2r\n",
    "df = as.data.frame(df2r)\n",
    "# head(df, 10)\n",
    "\n",
    "g = ggplot(data=df) +\n",
    "    geom_point(\n",
    "        mapping=aes(\n",
    "            x=ATP8_DddA11_1,\n",
    "            y=JAK2_DddA11_1,\n",
    "            # alpha=0.99,\n",
    "            # size=1,\n",
    "            color=color,\n",
    "            fill=color,\n",
    "            # shape=color,\n",
    "            # group=color,\n",
    "        )\n",
    "    ) +\n",
    "    scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "    scale_x_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddA11_1\",\n",
    "        limits=c(0, 0.8),\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        name=\"Detect-seq_signals of JAK2_DddA11_1\",\n",
    "        limits=c(0, 0.8),\n",
    "    ) +\n",
    "    geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "    annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "    ggtitle(\"ATP8_DddA11_1 v.s. JAK2_DddA11_1\") +\n",
    "    theme_classic() +\n",
    "    theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "ggsave('2022-10-30_Detect-seq_scatter-plot_ATP8_DddA11_1v.s.JAK2_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b31fc673-6ba8-45e0-bc36-dda42f6b4fb7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# g = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=ATP8_DddA11_1,\n",
    "#             y=JAK2_DddA11_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#             # shape=color,\n",
    "#             # group=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "#     ) +\n",
    "#     scale_y_log10(\n",
    "#         name=\"log10_Detect-seq_signals of JAK2_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"log10_ATP8_DddA11_1 v.s. JAK2_DddA11_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_log10_ATP8_DddA11_1v.s.JAK2_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca74a59d-3971-4c7e-9680-601d46917769",
   "metadata": {
    "tags": []
   },
   "source": [
    "### plot ATP8_DddA11_1 vs SIRT6_DddA11_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "117352e5-b995-4127-a3de-7c88530377a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load data in R env\n",
    "df2r = df_wide_score[['mpmat_index', 'ATP8_DddA11_1', 'SIRT6_DddA11_1']].copy()\n",
    "df2r['color'] = df2r.apply(map_color, axis=1)\n",
    "df2r = df2r.query('color.notnull()')\n",
    "df2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbe657fa-1113-4894-a2bf-f15d2460d027",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df2r\n",
    "df = as.data.frame(df2r)\n",
    "# head(df, 10)\n",
    "\n",
    "g = ggplot(data=df) +\n",
    "    geom_point(\n",
    "        mapping=aes(\n",
    "            x=ATP8_DddA11_1,\n",
    "            y=SIRT6_DddA11_1,\n",
    "            # alpha=0.99,\n",
    "            # size=1,\n",
    "            color=color,\n",
    "            fill=color,\n",
    "            # shape=color,\n",
    "            # group=color,\n",
    "        )\n",
    "    ) +\n",
    "    scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "    scale_x_continuous(\n",
    "        name=\"Detect-seq_signals of ATP8_DddA11_1\",\n",
    "        limits=c(0, 0.8),\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        name=\"Detect-seq_signals of SIRT6_DddA11_1\",\n",
    "        limits=c(0, 0.8),\n",
    "    ) +\n",
    "    geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "    annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "    ggtitle(\"ATP8_DddA11_1 v.s. SIRT6_DddA11_1\") +\n",
    "    theme_classic() +\n",
    "    theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "ggsave('2022-10-30_Detect-seq_scatter-plot_ATP8_DddA11_1v.s.SIRT6_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f272edf-ff3a-43a0-95b9-73c9948be3a0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# g = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=ATP8_DddA11_1,\n",
    "#             y=SIRT6_DddA11_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#             # shape=color,\n",
    "#             # group=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_log10(\n",
    "#         name=\"log10_Detect-seq_signals of ATP8_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "#     ) +\n",
    "#     scale_y_log10(\n",
    "#         name=\"log10_Detect-seq_signals of SIRT6_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"log10_ATP8_DddA11_1 v.s. SIRT6_DddA11_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_log10_ATP8_DddA11_1v.s.SIRT6_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a5719b3-72d3-4134-95f2-284d02120a81",
   "metadata": {
    "tags": []
   },
   "source": [
    "### plot JAK2_DddA11_1 vs SIRT6_DddA11_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9edc4f7-2e02-4a1e-93da-9b9465593428",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# load data in R env\n",
    "df2r = df_wide_score[['mpmat_index', 'JAK2_DddA11_1', 'SIRT6_DddA11_1']].copy()\n",
    "df2r['color'] = df2r.apply(map_color, axis=1)\n",
    "df2r = df2r.query('color.notnull()')\n",
    "df2r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fe172ad-846c-473e-a610-1bd78d6a53f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df2r\n",
    "df = as.data.frame(df2r)\n",
    "# head(df, 10)\n",
    "\n",
    "g = ggplot(data=df) +\n",
    "    geom_point(\n",
    "        mapping=aes(\n",
    "            x=JAK2_DddA11_1,\n",
    "            y=SIRT6_DddA11_1,\n",
    "            # alpha=0.99,\n",
    "            # size=1,\n",
    "            color=color,\n",
    "            fill=color,\n",
    "            # shape=color,\n",
    "            # group=color,\n",
    "        )\n",
    "    ) +\n",
    "    scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "    scale_x_continuous(\n",
    "        name=\"Detect-seq_signals of JAK2_DddA11_1\",\n",
    "        limits=c(0, 0.9),\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        name=\"Detect-seq_signals of SIRT6_DddA11_1\",\n",
    "        limits=c(0, 0.9),\n",
    "    ) +\n",
    "    geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "    annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "    ggtitle(\"JAK2_DddA11_1 v.s. SIRT6_DddA11_1\") +\n",
    "    theme_classic() +\n",
    "    theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "ggsave('2022-10-30_Detect-seq_scatter-plot_JAK2_DddA11_1v.s.SIRT6_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0e5dae4-bbf4-46c2-812b-c0546afe3e27",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %%R\n",
    "\n",
    "# g = ggplot(data=df) +\n",
    "#     geom_point(\n",
    "#         mapping=aes(\n",
    "#             x=JAK2_DddA11_1,\n",
    "#             y=SIRT6_DddA11_1,\n",
    "#             # alpha=0.99,\n",
    "#             # size=1,\n",
    "#             color=color,\n",
    "#             fill=color,\n",
    "#             # shape=color,\n",
    "#             # group=color,\n",
    "#         )\n",
    "#     ) +\n",
    "#     scale_color_manual(values=c(\"#ff9f1c\", \"#447B9D\", \"#e36414\")) +\n",
    "#     scale_x_log10(\n",
    "#         name=\"log10_Detect-seq_signals of JAK2_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "#     ) +\n",
    "#     scale_y_log10(\n",
    "#         name=\"log10_Detect-seq_signals of SIRT6_DddA11_1\",\n",
    "#         limits=c(0.01, 1.5),\n",
    "#         # breaks=c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "#     ) +\n",
    "#     geom_abline(intercept=0, slope=1, linetype=\"dashed\", color=\"#333333\") +\n",
    "#     annotate(\"text\", x=40, y=9, hjust=0, parse=F, colour=\"black\", label=sprintf(\"N = %i\", dim(df)[1])) +\n",
    "#     ggtitle(\"log10_JAK2_DddA11_1 v.s. SIRT6_DddA11_1\") +\n",
    "#     theme_classic() +\n",
    "#     theme(plot.title=element_text(hjust=0.5))\n",
    "\n",
    "# ggsave('2022-10-30_Detect-seq_scatter-plot_log10_JAK2_DddA11_1v.s.SIRT6_DddA11_1.pdf', plot=g, width=4.6, height=4)\n",
    "# g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "929c974d-c968-4e32-877d-a4116cf75292",
   "metadata": {},
   "source": [
    "### new share list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "710fade2-d18d-4f66-81b2-11695211a41c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_share_old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08dbe00f-7100-4fd4-8b67-35feb1b27674",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_if_share = df_upset.T\n",
    "df_share_new_id = pd.DataFrame(df_if_share[df_if_share.sum(axis=1).map(lambda x: x >= 2)].index)\n",
    "df_share_new_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fabc4bff-bf4f-4c20-bb46-8d2a4dfabb27",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls_share_old_id = list(set(df_share_old.mpmat_index.values.tolist()))\n",
    "ls_share_new_id = list(set(df_share_new_id.mpmat_index.values.tolist()))\n",
    "print(len(ls_share_new_id))\n",
    "print(len(ls_share_old_id))\n",
    "print(len(set(ls_share_old_id + ls_share_new_id)))\n",
    "\n",
    "print(len(set(ls_share_old_id + ls_share_new_id) - set(ls_share_old_id)))\n",
    "print(len(set(ls_share_old_id + ls_share_new_id) - set(ls_share_new_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60bd1e84-1284-4346-ae00-8147388d3c56",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# venn3(\n",
    "#     subsets,\n",
    "#     set_labels=('A', 'B', 'C'),\n",
    "#     set_colors=('r', 'g', 'b'),\n",
    "#     alpha=0.4,\n",
    "#     normalize_to=1.0,\n",
    "#     ax=None,\n",
    "#     subset_label_formatter=None,\n",
    "# )\n",
    "\n",
    "# TODO ?\n",
    "v = venn2([set(ls_share_new_id), set(ls_share_old_id)], set_labels=(\"share\", \"share-reported\"))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de02609a-dd58-489e-a3e9-c063946e0084",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_new_share_bed = pd.DataFrame([i.split('_') for i in ls_share_new_id], columns=['chrom', 'start', 'stop'])\n",
    "df_new_share_bed['start'] = df_new_share_bed['start'].astype(int)\n",
    "df_new_share_bed['stop'] = df_new_share_bed['stop'].astype(int)\n",
    "# df_new_share_bed.info()\n",
    "df_new_share_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf9d52d3-f285-4045-bd66-a87aec177817",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_ctcf = pd.read_csv('../bed/ENCFF285QVL_CTCF_binding_sites.bed', sep='\\t', header=None)\n",
    "df_ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c2e8996-775a-4f12-bb35-f0666727be05",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "bed_ctcf_like = BedTool.from_dataframe(df_new_share_bed)\n",
    "bed_ctcf_true = BedTool.from_dataframe(df_ctcf.iloc[:, 0:3])\n",
    "\n",
    "df_bed_to_fix = bed_ctcf_like.intersect(bed_ctcf_true, loj=True).to_dataframe()\n",
    "df_bed_to_fix.columns = ['chrom', 'start', 'end', 'chrom2', 'start2', 'end2']\n",
    "\n",
    "df_bed_to_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37820257-77cc-4f3b-9302-46f45a4a0b85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "bed_ctcf_true_in_new = df_bed_to_fix.query('end2!=-1')\n",
    "bed_ctcf_true_in_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e104b1-11a4-4cf2-936c-c68711fd35ee",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls_share_new_ctcf_true = (bed_ctcf_true_in_new['chrom'] + '_' + bed_ctcf_true_in_new['start'].astype(str) + '_' + bed_ctcf_true_in_new['end'].astype(str)).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac87152-c523-40f5-a08c-7111cf74c595",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_boxplot = df_detect_seq_score[df_detect_seq_score.mpmat_index.map(lambda x: x in ls_share_new_ctcf_true)].copy()\n",
    "df_boxplot\n",
    "df_boxplot.groupby('sample').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3f1d54e-1d13-4340-bd7f-041fbd0c1651",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_boxplot['sample'] = df_boxplot['sample'].str.replace('DetectSeq_', '')\n",
    "df_boxplot = df_boxplot.query('detect_seq_score>=0.001')\n",
    "df_boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8471810c-55fe-466c-a577-8a06aecac9e6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "order = [\n",
    "    'ATP8-DddAwt_REP-1',\n",
    "    'ATP8-DddA6_REP-1',\n",
    "    'ATP8-DddA11_REP-1',\n",
    "    'SIRT6-DddA11_REP-1',\n",
    "    'JAK2-DddA11_REP-1']\n",
    "\n",
    "ax = sns.boxplot(data=df_boxplot, x=\"sample\", y=\"detect_seq_score\", order=order)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=30)\n",
    "plt.ylim(0, 0.4)\n",
    "\n",
    "\n",
    "plt.savefig(\"2022-11-02_Detect-seq_signal_strength.pdf\", dpi=200, bbox_inches='tight')\n",
    "ax.get_xticklabels(),"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c2a225-2a67-49e3-bf6f-e041506f04ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2c7513fe-d7d2-441d-9388-edbace632cdf",
   "metadata": {},
   "source": [
    "## 2022-10-21 co-localization between all CTCF peaks and \u000b",
    "Detect-seq signals from different DdCBE treatments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffb02e19-328b-4a65-9833-574ab6794ffa",
   "metadata": {
    "tags": []
   },
   "source": [
    "# IGV 截图脚本\n",
    "> https://www.jianshu.com/p/d77f2d34b7fd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b1364e4-cf2f-4c83-9e7b-352eae6592a4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_igv = df_pois[[\n",
    "    'chr_name', 'region_start', 'region_end', 'bed_name', 'log2_FC_mut'\n",
    "]]\n",
    "df_igv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b62590f5-903f-40ee-a15a-8dc726fa420e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 理论上这里得到的是所有 samples 的 candidate list\n",
    "# 经过先 merge mpmat 的处理，再 call 点，应该不存在能 overlap 到一起的 region，只能是样本间 share\n",
    "# 相同位置或者 not share，不存在 overlap 又不相同的情况了"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8efbe775-5ee6-4101-96ba-3855c529b97b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_igv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "516cb100-b792-49da-af69-5826e8732052",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 对call 到的所有点进行随机采样\n",
    "np.random.seed = 2022\n",
    "df_igv = df_igv.sample(n=300)\n",
    "df_igv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c312b1-da76-48a0-9167-01a3857aaeb2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 整理df格式为bed文件格式\n",
    "print(df_igv)\n",
    "#        0          1          2\n",
    "# 0   chr5   69093805   69093830\n",
    "# 1   chr8   37153384   37153424\n",
    "# 2  chr15   57559994   57560017\n",
    "# 3  chr15   68651256   68651277\n",
    "# 4  chr10  119445511  119445546\n",
    "# 5   chr8   20184990   20185028\n",
    "# 6  chr19   45187694   45187712\n",
    "# 7  chr15   81265992   81266016\n",
    "# 8   chr2  201232409  201232430\n",
    "# 9   chr9   98034893   98034930"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e590a105-62bd-4e7a-b775-3f69dc8cc0a6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 填写相关信息\n",
    "\n",
    "path_out = '/Volumes/Data-a/Bio/3.project/2022_DdCBE-3D-Genome_topic/2022-09-30_Detect-seq_batch-1/igv'\n",
    "# path_out = '/Volumes/zhaohn_HD/3.project/2022_DdCBE-3D-Genome_topic/2022-09-30_Detect-seq_batch-1/igv'\n",
    "date = 20221125\n",
    "format_ = \"png\"\n",
    "height = 1500\n",
    "\n",
    "# 格式化脚本\n",
    "text = f\"maxPanelHeight {height}\\nsnapshotDirectory {path_out}/off-targets_{date}\\n\\n\"\n",
    "# print(text)\n",
    "\n",
    "\n",
    "df_snapshot = df_igv.iloc[:, 0:5]\n",
    "\n",
    "for index, row_info in df_snapshot.iterrows():\n",
    "    chrom, start, stop, bed_name, score = row_info\n",
    "\n",
    "    path_out_png = f'{score}_{bed_name}.snapshot.{format_}'\n",
    "    middle = int((start + stop) / 2)\n",
    "\n",
    "    text += f\"goto {chrom}:{middle - 100}-{middle + 100}\\nsort position\\ncollapsesnapshot {path_out_png}\\n\\n\"\n",
    "print(text[:1000])\n",
    "\n",
    "\n",
    "# with open(f'{path_out}/off-targets_{date}_snapshot.igv_shot_script', 'wt') as f:\n",
    "#     f.write(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "619960a3",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# DetectSeq_JAK2\n",
    "# DetectSeq_SIRT6\n",
    "# IND share?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70263df9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# off-target analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c6057e9",
   "metadata": {},
   "source": [
    "## off-target list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e0477dc",
   "metadata": {},
   "source": [
    "### circos plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9635b7b",
   "metadata": {},
   "source": [
    "#### learn circos plot\n",
    "- [tutorial1](https://colab.research.google.com/drive/1xmAnv7AHWUTA2HWfjqV1lFWkFMSLJHG0?usp=sharing)\n",
    "- [tutorial2](https://colab.research.google.com/drive/1RYSo4aXpDIZlSQ9EhO2kPCeF8FOwyvXv?usp=sharing)\n",
    "- [tutorial3](https://colab.research.google.com/drive/1EPxCQCgOouVxtXcGyxu2ZqQvfucVnOJ-?usp=sharing)\n",
    "- [tutorial4(Drawing pylogenetic tree)](https://colab.research.google.com/drive/140m2jpQpgSZwSlP-3u3Oj8IcJUbP2NGD?usp=sharing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec7308ac-b380-4ad3-aeea-b6562797d3b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pip install python-circos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2ff1e4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# !mkdir -p../ temp_file\n",
    "# % cd../ temp_file\n",
    "# !mkdir -p pycircos\n",
    "# % cd pycircos\n",
    "# !mkdir -p sample_data\n",
    "# % cd sample_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "969a1dec",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# #The following example data was downloaded from https://venyao.xyz/shinyCircos/.\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_barplot.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_chromosome_cytoband.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_chromosome_general.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_links.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_point.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_rect_gradual.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a9976e4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# % cd / Volumes / zhaohn_HD / 3.project / 2022_DdCBE-3D-Genome_topic / 2022-09-30_Detect-seq_batch-1 / snakepipes_detect-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d0f69e",
   "metadata": {},
   "source": [
    "#### circos plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6c604bb-54e3-4599-8dcb-ecbc88866c84",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-21T07:16:22.990392Z",
     "iopub.status.busy": "2022-10-21T07:16:22.990109Z",
     "iopub.status.idle": "2022-10-21T07:16:22.994832Z",
     "shell.execute_reply": "2022-10-21T07:16:22.993948Z",
     "shell.execute_reply.started": "2022-10-21T07:16:22.990374Z"
    }
   },
   "source": [
    "##### 准备 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14871568-295c-4bf8-ac25-86b88397d3d4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 基因组 fai 文件（染色体长度信息），区分 hg19 和 hg38\n",
    "path_genome_length = '/Users/zhaohuanan/Bio/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa.fai'\n",
    "df_genome_length = pd.read_csv(\n",
    "    path_genome_length,\n",
    "    header=None,\n",
    "    sep='\\t',\n",
    "    usecols=[0, 1],\n",
    "    names=['chrom', 'end'],\n",
    ")\n",
    "df_genome_length.insert(loc=1, column='start', value=1)\n",
    "\n",
    "df_genome_length = df_genome_length\\\n",
    "    .query(\"chrom != 'chrY'\")\\\n",
    "    .query(\"chrom != 'chrM'\")\n",
    "\n",
    "print(df_genome_length.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4335f60c-d17c-4218-aa64-016fcd1ed106",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# cytoband 信息，可以在 ucsc 下载，区分 hg19 和 hg38\n",
    "df_genome_cytoband = pd.read_csv(\n",
    "    \"http://hgdownload.cse.ucsc.edu/goldenpath/hg38/database/cytoBand.txt.gz\",\n",
    "    header=None,\n",
    "    sep='\\t',\n",
    "    usecols=range(5),\n",
    "    names=['chrom', 'start', 'end', 'value1', 'value2'])\n",
    "# 对 chromosome 进行过滤，保留标准染色体\n",
    "df_genome_cytoband = df_genome_cytoband\\\n",
    "    .query(\"not chrom.str.endswith('fix')\")\\\n",
    "    .query(\"not chrom.str.endswith('alt')\")\\\n",
    "    .query(\"not chrom.str.endswith('random')\")\\\n",
    "    .query(\"not chrom.str.startswith('chrUn')\")\\\n",
    "    .query(\"chrom != 'chrY'\")\\\n",
    "    .query(\"chrom != 'chrM'\")\n",
    "print(sorted(df_genome_cytoband.chrom.unique()))\n",
    "print()\n",
    "print(df_genome_cytoband.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d7bf9d9-2e9f-4378-a4bf-bde136c0f717",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_circos_point = df[[\n",
    "    '<sample>', 'chr_name', 'region_start', 'region_end', 'log2_FC_mut'\n",
    "]].copy()\n",
    "df_circos_point.columns = ['sample', 'chrom', 'start', 'end', 'score']\n",
    "df_circos_point.reset_index(inplace=True, drop=True)\n",
    "df_circos_point"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11dafed5-60a8-468d-9b8a-e4d9b6cf923c",
   "metadata": {},
   "source": [
    "##### 设置比例尺"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66418fd2-90c1-49d8-aeb8-25b021e005f2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "scale = 1.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7b3a9f-007e-4827-961b-3cb91b809c5a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "arc_range_i = df_circos_point['sample'].unique().__len__() + 1\n",
    "arc_raixs_range = ((arc_range_i * 100.0) * scale,\n",
    "                   (arc_range_i * 100.0 + 20) * scale)\n",
    "arc_raixs_range"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f266bf8-28df-45c2-bbeb-6e67349ea8a9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-21T06:17:24.104466Z",
     "iopub.status.busy": "2022-10-21T06:17:24.104224Z",
     "iopub.status.idle": "2022-10-21T06:17:24.111202Z",
     "shell.execute_reply": "2022-10-21T06:17:24.110435Z",
     "shell.execute_reply.started": "2022-10-21T06:17:24.104449Z"
    },
    "tags": []
   },
   "source": [
    "##### 设置染色体信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "875fc0e9",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# set chromeosomes\n",
    "circle = Gcircle()\n",
    "\n",
    "for idx, row in df_genome_length.iterrows():\n",
    "    chrom, start, end = row\n",
    "    length = end - start + 1\n",
    "\n",
    "    arc = Garc(\n",
    "        arc_id=chrom,  # 染色体名称\n",
    "        size=length,  # 染色体长度\n",
    "        interspace=1,  # 间隔距离\n",
    "        raxis_range=arc_raixs_range,  # 内外半径长度\n",
    "        labelposition=60,  # 越往里越小，越往外越大\n",
    "        label_visible=True,  # 是否展示 label，也就是染色体名称\n",
    "    )\n",
    "    circle.add_garc(arc)  # 添加一个染色体信息\n",
    "\n",
    "circle.set_garcs()  # 整合设置所有添加过的染色体信息"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7bc261d-ec42-4c4b-833e-0568d23caeb2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-21T06:22:00.236412Z",
     "iopub.status.busy": "2022-10-21T06:22:00.236161Z",
     "iopub.status.idle": "2022-10-21T06:22:00.283305Z",
     "shell.execute_reply": "2022-10-21T06:22:00.282835Z",
     "shell.execute_reply.started": "2022-10-21T06:22:00.236395Z"
    }
   },
   "source": [
    "##### 整理 cytoband 信息到arcdata\\_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97912204-81f4-45c1-82e1-430dbbd6ab4f",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 定义不同 cytoband 的颜色\n",
    "color_dict = {\n",
    "    \"gneg\": \"#FFFFFF00\",\n",
    "    \"gpos25\": \"#EEEEEE\",\n",
    "    \"gpos50\": \"#BBBBBB\",\n",
    "    \"gpos75\": \"#777777\",\n",
    "    \"gpos100\": \"#000000\",\n",
    "    \"gvar\": \"#FFFFFF00\",\n",
    "    \"stalk\": \"#C01E27\",\n",
    "    \"acen\": \"#D82322\"\n",
    "}\n",
    "\n",
    "# defaultdict的说明\n",
    "# https://zhuanlan.zhihu.com/p/46476348\n",
    "# 其实就是当引用的 key 不存在时返回默认值,这里是默认返回一个 dict 而不抛出 KeyError\n",
    "arcdata_dict = collections.defaultdict(dict)  # 传入 dict 函数，调用时返回空 dict\n",
    "# or\n",
    "# arcdata_dict = collections.defaultdict(lambda: {})\n",
    "# arcdata_dict\n",
    "\n",
    "for idx, row in df_genome_cytoband.iterrows():\n",
    "    chrom, start, end, value1, value2 = row\n",
    "    width = end - start + 1\n",
    "    # 在进行arcdata_dict对键chrom 取值取不到的时候\n",
    "    # 默认创建空 dict 而不是 raise KeyError\n",
    "    if chrom not in arcdata_dict:\n",
    "        arcdata_dict[chrom]['positions'] = []\n",
    "        arcdata_dict[chrom]['widths'] = []\n",
    "        arcdata_dict[chrom]['colors'] = []\n",
    "    else:\n",
    "        arcdata_dict[chrom]['positions'].append(start)\n",
    "        arcdata_dict[chrom]['widths'].append(width)\n",
    "        arcdata_dict[chrom]['colors'].append(color_dict[value2])\n",
    "\n",
    "print(arcdata_dict.__str__()[:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98f18958-28f9-4051-83ec-ef2bd3c366e3",
   "metadata": {},
   "source": [
    "##### 将 cytoband 信息加到 circle 对象中去"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e6cdb7f-b523-4f3e-9221-1a002bd4477d",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for chrom in arcdata_dict:\n",
    "    circle.barplot(chrom,\n",
    "                   data=[1] * len(arcdata_dict[chrom]['positions']),\n",
    "                   positions=arcdata_dict[chrom]['positions'],\n",
    "                   width=arcdata_dict[chrom]['widths'],\n",
    "                   raxis_range=arc_raixs_range,\n",
    "                   facecolor=arcdata_dict[chrom]['colors'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e852695c-ed96-4368-b40b-0732ff8d1e7b",
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "##### 查看绘制的circos plot骨架"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72dd14a0-7daa-4cd4-a695-3172329342b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# circle.figure?\n",
    "circle.figure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d4d8318-5ad4-44cd-be3c-85de6fff653b",
   "metadata": {},
   "source": [
    "##### 添加每个样本中的 off-target sites 信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b392d24-6d83-4bdc-8d6b-5262bcb1c452",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# TODO\n",
    "# sample lable\n",
    "# background color\n",
    "# point color\n",
    "# edge color of point\n",
    "\n",
    "# scatter plot\n",
    "counter_circle = 0\n",
    "# color\n",
    "\n",
    "for sample, sample_df in df_circos_point.groupby('sample'):\n",
    "    print(sample, sample_df.shape[0])\n",
    "    counter_circle += 1\n",
    "\n",
    "    values_all = []\n",
    "    arcdata_dict = collections.defaultdict(dict)\n",
    "\n",
    "    for idx, row in sample_df.iterrows():\n",
    "        _, chrom, start, end, score = row\n",
    "        middle = (start + end) / 2\n",
    "        values_all.append(score)\n",
    "\n",
    "        if chrom not in arcdata_dict:\n",
    "            arcdata_dict[chrom][\"positions\"] = []\n",
    "            arcdata_dict[chrom][\"values\"] = []\n",
    "        else:\n",
    "            arcdata_dict[chrom][\"positions\"].append(middle)\n",
    "            arcdata_dict[chrom][\"values\"].append(score)\n",
    "\n",
    "    vmin, vmax = min(values_all), max(values_all)\n",
    "\n",
    "    arc_raixs_range_sample = ((counter_circle * 100.0) * scale,\n",
    "                              (counter_circle * 100.0 + 80) * scale)\n",
    "\n",
    "    for chrom in arcdata_dict:\n",
    "        circle.scatterplot(\n",
    "            chrom,\n",
    "            data=arcdata_dict[chrom][\"values\"],\n",
    "            positions=arcdata_dict[chrom][\"positions\"],\n",
    "            rlim=[vmin - 0.05 * abs(vmin), vmax + 0.05 * abs(vmax)],\n",
    "            markershape='o',\n",
    "            markersize=1,\n",
    "            raxis_range=arc_raixs_range_sample,\n",
    "            facecolor=\"#468FBE\",\n",
    "            edgecolor=\"#000000\",\n",
    "            linewidth=0.03,\n",
    "            spine=True)\n",
    "\n",
    "circle.figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe612234-5480-4ba7-80dd-cde9396898e9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "circle.figure?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a1f113e",
   "metadata": {},
   "source": [
    "\n",
    "### upset plot\n",
    "\n",
    "## signal strength\n",
    "\n",
    "### scatter plot\n",
    "\n",
    "## alignment\n",
    "### art plot\n",
    "\n",
    "## editing window\n",
    "### indel comparation\n",
    "\n",
    "## ctcf analysis\n",
    "### shared off-target motif\n",
    "### DddAwt,6,11 co-localization with ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48bd2502-2b5a-42df-a259-8c897aeca2af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f096e6e",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
