{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "834e5693-79fe-4e0d-b00f-927245ecfc4a",
   "metadata": {},
   "source": [
    "# packages and settings "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b880507-7d5a-432f-ba5f-5cacb058a8d4",
   "metadata": {},
   "source": [
    "import python packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2ccbf30b-41ed-4ca8-b6d6-296308872dd3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:30:33.849769Z",
     "iopub.status.busy": "2022-10-28T07:30:33.849517Z",
     "iopub.status.idle": "2022-10-28T07:30:33.853096Z",
     "shell.execute_reply": "2022-10-28T07:30:33.852564Z",
     "shell.execute_reply.started": "2022-10-28T07:30:33.849751Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import collections\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pycircos\n",
    "from pybedtools import BedTool\n",
    "from pycircos import Garc, Gcircle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e8e66af-b297-4ee4-90eb-07ae2eec3b93",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-27T08:20:10.504290Z",
     "iopub.status.busy": "2022-10-27T08:20:10.504004Z",
     "iopub.status.idle": "2022-10-27T08:20:10.509189Z",
     "shell.execute_reply": "2022-10-27T08:20:10.508390Z",
     "shell.execute_reply.started": "2022-10-27T08:20:10.504271Z"
    }
   },
   "source": [
    "enable rpy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2625a77e-0c70-494d-80e5-924b6b108702",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:30:40.354455Z",
     "iopub.status.busy": "2022-10-28T07:30:40.354186Z",
     "iopub.status.idle": "2022-10-28T07:30:41.108183Z",
     "shell.execute_reply": "2022-10-28T07:30:41.107603Z",
     "shell.execute_reply.started": "2022-10-28T07:30:40.354438Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# enables the gaR magic,  not neceasary  if you’ve already done this\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0bd5e40-c367-4566-af46-e7a64198cf2d",
   "metadata": {},
   "source": [
    "import R libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b771c8c3-77fd-4b5b-b194-2c69022fa58e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:32:20.365084Z",
     "iopub.status.busy": "2022-10-28T07:32:20.364747Z",
     "iopub.status.idle": "2022-10-28T07:32:21.508408Z",
     "shell.execute_reply": "2022-10-28T07:32:21.507895Z",
     "shell.execute_reply.started": "2022-10-28T07:32:20.365062Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "── Attaching packages ─────────────────────────────────────── tidyverse 1.3.2 ──\n",
      "✔ ggplot2 3.3.6      ✔ purrr   0.3.5 \n",
      "✔ tibble  3.1.8      ✔ dplyr   1.0.10\n",
      "✔ tidyr   1.2.1      ✔ stringr 1.4.1 \n",
      "✔ readr   2.1.3      ✔ forcats 0.5.2 \n",
      "── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "✖ dplyr::filter() masks stats::filter()\n",
      "✖ dplyr::lag()    masks stats::lag()\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# 可以再%%R后面放光标 cmd + i\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec1b0974-6822-45c1-93f4-168b41c928f4",
   "metadata": {},
   "source": [
    "# 初步探索poisson table\n",
    "- log2_FC: log2(treat_count.norm/ctrl_count.norm), 如果ctrl_count.norm不存在，就用 chr 突变背景\n",
    "- log2_FC_mut: log2(treat_mut_count.norm/ctrl_mut_count.norm), 如果ctrl_mut_count.norm不存在，就用 chr 突变背景\n",
    "- region_block_state: B-Blocked, S-SNV, N-Non-SNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "013f4366-febf-4dff-b7de-171b51bc2099",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:32:29.181934Z",
     "iopub.status.busy": "2022-10-28T07:32:29.181395Z",
     "iopub.status.idle": "2022-10-28T07:32:29.185111Z",
     "shell.execute_reply": "2022-10-28T07:32:29.184416Z",
     "shell.execute_reply.started": "2022-10-28T07:32:29.181914Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "file = '../poisson_res/poisson_res_all.tsv.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cba03e93-ccb4-447f-bacd-58b2f19216be",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:32:30.230594Z",
     "iopub.status.busy": "2022-10-28T07:32:30.230339Z",
     "iopub.status.idle": "2022-10-28T07:33:11.592090Z",
     "shell.execute_reply": "2022-10-28T07:33:11.591380Z",
     "shell.execute_reply.started": "2022-10-28T07:32:30.230577Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>...</th>\n",
       "      <th>ctrl_count.norm</th>\n",
       "      <th>treat_count.norm</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16395</td>\n",
       "      <td>16418</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_16395_CT,chr1_16397_C.,chr1_16403_CT,chr1...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.144296</td>\n",
       "      <td>0.113256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 53,4,0,0,0,0,0,0,0,0...</td>\n",
       "      <td>-0.349446</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>16540</td>\n",
       "      <td>chr1_16534_16540</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_16534_CT,chr1_16538_CT,chr1_16539_C.,chr1...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.091135</td>\n",
       "      <td>0.047687</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 36,0,0,0,0 8,0,0,0,0</td>\n",
       "      <td>-0.934408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31029</td>\n",
       "      <td>31037</td>\n",
       "      <td>chr1_31029_31037</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_31029_GA,chr1_31030_G.,chr1_31031_GA,chr1...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005063</td>\n",
       "      <td>0.035765</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 2,0,0,0,0 5,0,1,0,0</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>56485</td>\n",
       "      <td>56490</td>\n",
       "      <td>chr1_56485_56490</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_56485_CT,chr1_56487_CT,chr1_56488_C.,chr1...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.025315</td>\n",
       "      <td>0.041726</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 10,0,0,0,0 6,0,1,0,0</td>\n",
       "      <td>0.720943</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>56638</td>\n",
       "      <td>56647</td>\n",
       "      <td>chr1_56638_56647</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_56638_CT,chr1_56641_C.,chr1_56647_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040504</td>\n",
       "      <td>0.077491</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 16,0,0,0 3,10,0,0</td>\n",
       "      <td>0.935956</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842215</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      <sample> chr_name  region_start  region_end  \\\n",
       "0  DetectSeq_ATP8-DddA11_REP-1     chr1         16395       16418   \n",
       "1  DetectSeq_ATP8-DddA11_REP-1     chr1         16534       16540   \n",
       "2  DetectSeq_ATP8-DddA11_REP-1     chr1         31029       31037   \n",
       "3  DetectSeq_ATP8-DddA11_REP-1     chr1         56485       56490   \n",
       "4  DetectSeq_ATP8-DddA11_REP-1     chr1         56638       56647   \n",
       "\n",
       "        mpmat_index  region_site_num  region_block_site_num  \\\n",
       "0  chr1_16395_16418               11                      0   \n",
       "1  chr1_16534_16540                4                      1   \n",
       "2  chr1_31029_31037                4                      1   \n",
       "3  chr1_56485_56490                4                      1   \n",
       "4  chr1_56638_56647                3                      0   \n",
       "\n",
       "   region_mut_site_num                                  region_site_index  \\\n",
       "0                   11  chr1_16395_CT,chr1_16397_C.,chr1_16403_CT,chr1...   \n",
       "1                    3  chr1_16534_CT,chr1_16538_CT,chr1_16539_C.,chr1...   \n",
       "2                    3  chr1_31029_GA,chr1_31030_G.,chr1_31031_GA,chr1...   \n",
       "3                    3  chr1_56485_CT,chr1_56487_CT,chr1_56488_C.,chr1...   \n",
       "4                    3          chr1_56638_CT,chr1_56641_C.,chr1_56647_CT   \n",
       "\n",
       "      region_block_state  ... ctrl_count.norm  treat_count.norm  \\\n",
       "0  N-N-N-N-N-N-N-N-N-N-N  ...        0.144296          0.113256   \n",
       "1                B-N-N-N  ...        0.091135          0.047687   \n",
       "2                B-N-N-N  ...        0.005063          0.035765   \n",
       "3                B-N-N-N  ...        0.025315          0.041726   \n",
       "4                  N-N-N  ...        0.040504          0.077491   \n",
       "\n",
       "   ctrl_mut_count.norm  treat_mut_count.norm  \\\n",
       "0                  0.0              0.011922   \n",
       "1                  0.0              0.000000   \n",
       "2                  0.0              0.005961   \n",
       "3                  0.0              0.005961   \n",
       "4                  0.0              0.000000   \n",
       "\n",
       "                                          count_info   log2_FC  log2_FC_mut  \\\n",
       "0  0,1,2,3,4,5,6,7,8,9,10,11 53,4,0,0,0,0,0,0,0,0... -0.349446     0.557555   \n",
       "1                     0,1,2,3,4 36,0,0,0,0 8,0,0,0,0 -0.934408          NaN   \n",
       "2                      0,1,2,3,4 2,0,0,0,0 5,0,1,0,0  2.820479    -0.442445   \n",
       "3                     0,1,2,3,4 10,0,0,0,0 6,0,1,0,0  0.720943    -0.442445   \n",
       "4                          0,1,2,3 16,0,0,0 3,10,0,0  0.935956          NaN   \n",
       "\n",
       "   test_state   p_value       FDR  \n",
       "0      TestOK  0.377620  0.842215  \n",
       "1      TestOK  0.841752  0.842215  \n",
       "2      TestOK  0.582932  0.842215  \n",
       "3      TestOK  0.582932  0.842215  \n",
       "4      TestOK  0.841752  0.842215  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(file, header=0, index_col=None, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "237c76d2-4067-4844-bbd3-c49306b39e64",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:24.130454Z",
     "iopub.status.busy": "2022-10-28T07:33:24.130184Z",
     "iopub.status.idle": "2022-10-28T07:33:24.133429Z",
     "shell.execute_reply": "2022-10-28T07:33:24.132717Z",
     "shell.execute_reply.started": "2022-10-28T07:33:24.130435Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ed147f02-9e4d-401c-a4e8-4d1dd42e88b8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:24.535632Z",
     "iopub.status.busy": "2022-10-28T07:33:24.535300Z",
     "iopub.status.idle": "2022-10-28T07:33:24.538498Z",
     "shell.execute_reply": "2022-10-28T07:33:24.537921Z",
     "shell.execute_reply.started": "2022-10-28T07:33:24.535610Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1981853d-4cde-4fe6-ac0b-63fe0e34cfd7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:24.928458Z",
     "iopub.status.busy": "2022-10-28T07:33:24.928219Z",
     "iopub.status.idle": "2022-10-28T07:33:24.931181Z",
     "shell.execute_reply": "2022-10-28T07:33:24.930678Z",
     "shell.execute_reply.started": "2022-10-28T07:33:24.928441Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # 查看一下两列 NA 是什么\n",
    "# test_df = df.query('region_highest_site_index.isna()')\n",
    "# test2_df = df.query('log2_FC_mut.isna()')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "63bf63e6-4868-4c9b-a5ab-1e829676792a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:28.655451Z",
     "iopub.status.busy": "2022-10-28T07:33:28.655162Z",
     "iopub.status.idle": "2022-10-28T07:33:37.364323Z",
     "shell.execute_reply": "2022-10-28T07:33:37.363825Z",
     "shell.execute_reply.started": "2022-10-28T07:33:28.655434Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>...</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16395</td>\n",
       "      <td>16418</td>\n",
       "      <td>chr1_16395_16418</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_16395_CT,chr1_16397_C.,chr1_16403_CT,chr1...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.011922</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 53,4,0,0,0,0,0,0,0,0...</td>\n",
       "      <td>-0.349446</td>\n",
       "      <td>0.557555</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.377620</td>\n",
       "      <td>0.842215</td>\n",
       "      <td>chr1_16395_16418_highest_chr1_16395_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16534</td>\n",
       "      <td>16540</td>\n",
       "      <td>chr1_16534_16540</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_16534_CT,chr1_16538_CT,chr1_16539_C.,chr1...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 36,0,0,0,0 8,0,0,0,0</td>\n",
       "      <td>-0.934408</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842215</td>\n",
       "      <td>chr1_16534_16540_highest_chr1_16538_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31029</td>\n",
       "      <td>31037</td>\n",
       "      <td>chr1_31029_31037</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_31029_GA,chr1_31030_G.,chr1_31031_GA,chr1...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 2,0,0,0,0 5,0,1,0,0</td>\n",
       "      <td>2.820479</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842215</td>\n",
       "      <td>chr1_31029_31037_highest_chr1_31031_GA</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>56485</td>\n",
       "      <td>56490</td>\n",
       "      <td>chr1_56485_56490</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_56485_CT,chr1_56487_CT,chr1_56488_C.,chr1...</td>\n",
       "      <td>B-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005961</td>\n",
       "      <td>0,1,2,3,4 10,0,0,0,0 6,0,1,0,0</td>\n",
       "      <td>0.720943</td>\n",
       "      <td>-0.442445</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.582932</td>\n",
       "      <td>0.842215</td>\n",
       "      <td>chr1_56485_56490_highest_chr1_56487_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>56638</td>\n",
       "      <td>56647</td>\n",
       "      <td>chr1_56638_56647</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_56638_CT,chr1_56641_C.,chr1_56647_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 16,0,0,0 3,10,0,0</td>\n",
       "      <td>0.935956</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.841752</td>\n",
       "      <td>0.842215</td>\n",
       "      <td>chr1_56638_56647_highest_chr1_56638_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6421000</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155984850</td>\n",
       "      <td>155984857</td>\n",
       "      <td>chrX_155984850_155984857</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>chrX_155984850_CT,chrX_155984857_CT</td>\n",
       "      <td>N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2 0,0,0 5,7,0</td>\n",
       "      <td>0.207009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.806282</td>\n",
       "      <td>0.857628</td>\n",
       "      <td>chrX_155984850_155984857_highest_chrX_15598485...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6421001</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155984970</td>\n",
       "      <td>155984999</td>\n",
       "      <td>chrX_155984970_155984999</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>chrX_155984970_CT,chrX_155984977_C.,chrX_15598...</td>\n",
       "      <td>N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005561</td>\n",
       "      <td>0,1,2,3,4,5,6,7 0,0,0,0,0,0,0,0 11,2,1,0,0,0,0,0</td>\n",
       "      <td>0.429401</td>\n",
       "      <td>-0.274156</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.528695</td>\n",
       "      <td>0.857628</td>\n",
       "      <td>chrX_155984970_155984999_highest_chrX_15598497...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6421002</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155985095</td>\n",
       "      <td>155985105</td>\n",
       "      <td>chrX_155985095_155985105</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chrX_155985095_GA,chrX_155985099_G.,chrX_15598...</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3 0,0,0,0 1,0,0,0</td>\n",
       "      <td>-3.377953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.806282</td>\n",
       "      <td>0.857628</td>\n",
       "      <td>chrX_155985095_155985105_highest_chrX_15598509...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6421003</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155988525</td>\n",
       "      <td>155988531</td>\n",
       "      <td>chrX_155988525_155988531</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>chrX_155988525_CT,chrX_155988529_C.,chrX_15598...</td>\n",
       "      <td>N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4 0,0,0,0,0 2,0,0,0,0</td>\n",
       "      <td>-2.377953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.806282</td>\n",
       "      <td>0.857628</td>\n",
       "      <td>chrX_155988525_155988531_highest_chrX_15598852...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6421004</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155988614</td>\n",
       "      <td>155988629</td>\n",
       "      <td>chrX_155988614_155988629</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>chrX_155988614_GA,chrX_155988615_G.,chrX_15598...</td>\n",
       "      <td>N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0,1,2,3,4,5 0,0,0,0,0,0 1,2,0,0,0,0</td>\n",
       "      <td>-1.792991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>0.806282</td>\n",
       "      <td>0.857628</td>\n",
       "      <td>chrX_155988614_155988629_highest_chrX_15598862...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6421005 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             <sample> chr_name  region_start  region_end  \\\n",
       "0         DetectSeq_ATP8-DddA11_REP-1     chr1         16395       16418   \n",
       "1         DetectSeq_ATP8-DddA11_REP-1     chr1         16534       16540   \n",
       "2         DetectSeq_ATP8-DddA11_REP-1     chr1         31029       31037   \n",
       "3         DetectSeq_ATP8-DddA11_REP-1     chr1         56485       56490   \n",
       "4         DetectSeq_ATP8-DddA11_REP-1     chr1         56638       56647   \n",
       "...                               ...      ...           ...         ...   \n",
       "6421000  DetectSeq_SIRT6-DddA11_REP-1     chrX     155984850   155984857   \n",
       "6421001  DetectSeq_SIRT6-DddA11_REP-1     chrX     155984970   155984999   \n",
       "6421002  DetectSeq_SIRT6-DddA11_REP-1     chrX     155985095   155985105   \n",
       "6421003  DetectSeq_SIRT6-DddA11_REP-1     chrX     155988525   155988531   \n",
       "6421004  DetectSeq_SIRT6-DddA11_REP-1     chrX     155988614   155988629   \n",
       "\n",
       "                      mpmat_index  region_site_num  region_block_site_num  \\\n",
       "0                chr1_16395_16418               11                      0   \n",
       "1                chr1_16534_16540                4                      1   \n",
       "2                chr1_31029_31037                4                      1   \n",
       "3                chr1_56485_56490                4                      1   \n",
       "4                chr1_56638_56647                3                      0   \n",
       "...                           ...              ...                    ...   \n",
       "6421000  chrX_155984850_155984857                2                      0   \n",
       "6421001  chrX_155984970_155984999                7                      0   \n",
       "6421002  chrX_155985095_155985105                3                      0   \n",
       "6421003  chrX_155988525_155988531                4                      0   \n",
       "6421004  chrX_155988614_155988629                5                      0   \n",
       "\n",
       "         region_mut_site_num  \\\n",
       "0                         11   \n",
       "1                          3   \n",
       "2                          3   \n",
       "3                          3   \n",
       "4                          3   \n",
       "...                      ...   \n",
       "6421000                    2   \n",
       "6421001                    7   \n",
       "6421002                    3   \n",
       "6421003                    4   \n",
       "6421004                    5   \n",
       "\n",
       "                                         region_site_index  \\\n",
       "0        chr1_16395_CT,chr1_16397_C.,chr1_16403_CT,chr1...   \n",
       "1        chr1_16534_CT,chr1_16538_CT,chr1_16539_C.,chr1...   \n",
       "2        chr1_31029_GA,chr1_31030_G.,chr1_31031_GA,chr1...   \n",
       "3        chr1_56485_CT,chr1_56487_CT,chr1_56488_C.,chr1...   \n",
       "4                chr1_56638_CT,chr1_56641_C.,chr1_56647_CT   \n",
       "...                                                    ...   \n",
       "6421000                chrX_155984850_CT,chrX_155984857_CT   \n",
       "6421001  chrX_155984970_CT,chrX_155984977_C.,chrX_15598...   \n",
       "6421002  chrX_155985095_GA,chrX_155985099_G.,chrX_15598...   \n",
       "6421003  chrX_155988525_CT,chrX_155988529_C.,chrX_15598...   \n",
       "6421004  chrX_155988614_GA,chrX_155988615_G.,chrX_15598...   \n",
       "\n",
       "            region_block_state  ... ctrl_mut_count.norm  treat_mut_count.norm  \\\n",
       "0        N-N-N-N-N-N-N-N-N-N-N  ...                 0.0              0.011922   \n",
       "1                      B-N-N-N  ...                 0.0              0.000000   \n",
       "2                      B-N-N-N  ...                 0.0              0.005961   \n",
       "3                      B-N-N-N  ...                 0.0              0.005961   \n",
       "4                        N-N-N  ...                 0.0              0.000000   \n",
       "...                        ...  ...                 ...                   ...   \n",
       "6421000                    N-N  ...                 0.0              0.000000   \n",
       "6421001          N-N-N-N-N-N-N  ...                 0.0              0.005561   \n",
       "6421002                  N-N-N  ...                 0.0              0.000000   \n",
       "6421003                N-N-N-N  ...                 0.0              0.000000   \n",
       "6421004              N-N-N-N-N  ...                 0.0              0.000000   \n",
       "\n",
       "                                                count_info   log2_FC  \\\n",
       "0        0,1,2,3,4,5,6,7,8,9,10,11 53,4,0,0,0,0,0,0,0,0... -0.349446   \n",
       "1                           0,1,2,3,4 36,0,0,0,0 8,0,0,0,0 -0.934408   \n",
       "2                            0,1,2,3,4 2,0,0,0,0 5,0,1,0,0  2.820479   \n",
       "3                           0,1,2,3,4 10,0,0,0,0 6,0,1,0,0  0.720943   \n",
       "4                                0,1,2,3 16,0,0,0 3,10,0,0  0.935956   \n",
       "...                                                    ...       ...   \n",
       "6421000                                  0,1,2 0,0,0 5,7,0  0.207009   \n",
       "6421001   0,1,2,3,4,5,6,7 0,0,0,0,0,0,0,0 11,2,1,0,0,0,0,0  0.429401   \n",
       "6421002                            0,1,2,3 0,0,0,0 1,0,0,0 -3.377953   \n",
       "6421003                      0,1,2,3,4 0,0,0,0,0 2,0,0,0,0 -2.377953   \n",
       "6421004                0,1,2,3,4,5 0,0,0,0,0,0 1,2,0,0,0,0 -1.792991   \n",
       "\n",
       "         log2_FC_mut  test_state   p_value       FDR  \\\n",
       "0           0.557555      TestOK  0.377620  0.842215   \n",
       "1                NaN      TestOK  0.841752  0.842215   \n",
       "2          -0.442445      TestOK  0.582932  0.842215   \n",
       "3          -0.442445      TestOK  0.582932  0.842215   \n",
       "4                NaN      TestOK  0.841752  0.842215   \n",
       "...              ...         ...       ...       ...   \n",
       "6421000          NaN      TestOK  0.806282  0.857628   \n",
       "6421001    -0.274156      TestOK  0.528695  0.857628   \n",
       "6421002          NaN      TestOK  0.806282  0.857628   \n",
       "6421003          NaN      TestOK  0.806282  0.857628   \n",
       "6421004          NaN      TestOK  0.806282  0.857628   \n",
       "\n",
       "                                                  bed_name  strand  \n",
       "0                   chr1_16395_16418_highest_chr1_16395_CT       .  \n",
       "1                   chr1_16534_16540_highest_chr1_16538_CT       .  \n",
       "2                   chr1_31029_31037_highest_chr1_31031_GA       .  \n",
       "3                   chr1_56485_56490_highest_chr1_56487_CT       .  \n",
       "4                   chr1_56638_56647_highest_chr1_56638_CT       .  \n",
       "...                                                    ...     ...  \n",
       "6421000  chrX_155984850_155984857_highest_chrX_15598485...       .  \n",
       "6421001  chrX_155984970_155984999_highest_chrX_15598497...       .  \n",
       "6421002  chrX_155985095_155985105_highest_chrX_15598509...       .  \n",
       "6421003  chrX_155988525_155988531_highest_chrX_15598852...       .  \n",
       "6421004  chrX_155988614_155988629_highest_chrX_15598862...       .  \n",
       "\n",
       "[6421005 rows x 30 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.assign(\n",
    "    bed_name=df.mpmat_index + '_highest_' + df.region_highest_site_index,\n",
    "    strand='.'\n",
    ")\n",
    "df = df[df['<sample>'] != 'test']\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b713bd65-1efc-4ccb-8f05-cfa6d83e7da8",
   "metadata": {},
   "source": [
    "# find significant region"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67cc6c18-e573-4537-a3d5-7fae99249bc2",
   "metadata": {},
   "source": [
    "## strict selection "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d83a38b2-8097-4762-8439-567659a7ddd4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:39.520263Z",
     "iopub.status.busy": "2022-10-28T07:33:39.520007Z",
     "iopub.status.idle": "2022-10-28T07:33:39.757693Z",
     "shell.execute_reply": "2022-10-28T07:33:39.757207Z",
     "shell.execute_reply.started": "2022-10-28T07:33:39.520243Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1892\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_start</th>\n",
       "      <th colspan=\"2\" halign=\"left\">region_end</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">p_value</th>\n",
       "      <th colspan=\"8\" halign=\"left\">FDR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>...</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA11_REP-1</th>\n",
       "      <td>72.0</td>\n",
       "      <td>7.232027e+07</td>\n",
       "      <td>5.418565e+07</td>\n",
       "      <td>631391.0</td>\n",
       "      <td>26704962.00</td>\n",
       "      <td>62013707.5</td>\n",
       "      <td>1.122798e+08</td>\n",
       "      <td>238396642.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>7.232032e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>2.953279e-10</td>\n",
       "      <td>1.226002e-08</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.609762e-17</td>\n",
       "      <td>6.514572e-11</td>\n",
       "      <td>3.036639e-06</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA6_REP-1</th>\n",
       "      <td>189.0</td>\n",
       "      <td>7.886078e+07</td>\n",
       "      <td>5.397664e+07</td>\n",
       "      <td>631391.0</td>\n",
       "      <td>35339095.00</td>\n",
       "      <td>72688097.0</td>\n",
       "      <td>1.148110e+08</td>\n",
       "      <td>226406759.0</td>\n",
       "      <td>189.0</td>\n",
       "      <td>7.886083e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>4.238782e-09</td>\n",
       "      <td>3.022062e-08</td>\n",
       "      <td>189.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.511549e-12</td>\n",
       "      <td>9.258856e-08</td>\n",
       "      <td>1.664663e-05</td>\n",
       "      <td>0.000096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddAwt_REP-1</th>\n",
       "      <td>27.0</td>\n",
       "      <td>4.491637e+07</td>\n",
       "      <td>4.291506e+07</td>\n",
       "      <td>631391.0</td>\n",
       "      <td>17145676.50</td>\n",
       "      <td>27147712.0</td>\n",
       "      <td>5.676308e+07</td>\n",
       "      <td>173839060.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>4.491641e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.872361e-10</td>\n",
       "      <td>2.584274e-09</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>8.475808e-08</td>\n",
       "      <td>4.466082e-07</td>\n",
       "      <td>1.031558e-05</td>\n",
       "      <td>0.000056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_JAK2-DddA11_REP-1</th>\n",
       "      <td>400.0</td>\n",
       "      <td>7.720246e+07</td>\n",
       "      <td>5.466300e+07</td>\n",
       "      <td>1279842.0</td>\n",
       "      <td>33854748.25</td>\n",
       "      <td>66652709.0</td>\n",
       "      <td>1.163821e+08</td>\n",
       "      <td>246689645.0</td>\n",
       "      <td>400.0</td>\n",
       "      <td>7.720249e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>2.309763e-10</td>\n",
       "      <td>3.415243e-08</td>\n",
       "      <td>400.0</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>7.206266e-102</td>\n",
       "      <td>9.844659e-19</td>\n",
       "      <td>4.976689e-11</td>\n",
       "      <td>6.804493e-07</td>\n",
       "      <td>0.000073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_SIRT6-DddA11_REP-1</th>\n",
       "      <td>1204.0</td>\n",
       "      <td>7.768406e+07</td>\n",
       "      <td>5.564547e+07</td>\n",
       "      <td>242415.0</td>\n",
       "      <td>33904291.75</td>\n",
       "      <td>67106716.5</td>\n",
       "      <td>1.123429e+08</td>\n",
       "      <td>248905893.0</td>\n",
       "      <td>1204.0</td>\n",
       "      <td>7.768411e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.257744e-09</td>\n",
       "      <td>2.089933e-07</td>\n",
       "      <td>1204.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>8.602823e-153</td>\n",
       "      <td>7.967822e-17</td>\n",
       "      <td>1.283779e-09</td>\n",
       "      <td>2.256525e-06</td>\n",
       "      <td>0.000097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             region_start                              \\\n",
       "                                    count          mean           std   \n",
       "<sample>                                                                \n",
       "DetectSeq_ATP8-DddA11_REP-1          72.0  7.232027e+07  5.418565e+07   \n",
       "DetectSeq_ATP8-DddA6_REP-1          189.0  7.886078e+07  5.397664e+07   \n",
       "DetectSeq_ATP8-DddAwt_REP-1          27.0  4.491637e+07  4.291506e+07   \n",
       "DetectSeq_JAK2-DddA11_REP-1         400.0  7.720246e+07  5.466300e+07   \n",
       "DetectSeq_SIRT6-DddA11_REP-1       1204.0  7.768406e+07  5.564547e+07   \n",
       "\n",
       "                                                                  \\\n",
       "                                    min          25%         50%   \n",
       "<sample>                                                           \n",
       "DetectSeq_ATP8-DddA11_REP-1    631391.0  26704962.00  62013707.5   \n",
       "DetectSeq_ATP8-DddA6_REP-1     631391.0  35339095.00  72688097.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1    631391.0  17145676.50  27147712.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   1279842.0  33854748.25  66652709.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1   242415.0  33904291.75  67106716.5   \n",
       "\n",
       "                                                        region_end  \\\n",
       "                                       75%          max      count   \n",
       "<sample>                                                             \n",
       "DetectSeq_ATP8-DddA11_REP-1   1.122798e+08  238396642.0       72.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    1.148110e+08  226406759.0      189.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   5.676308e+07  173839060.0       27.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   1.163821e+08  246689645.0      400.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1.123429e+08  248905893.0     1204.0   \n",
       "\n",
       "                                            ...       p_value                \\\n",
       "                                      mean  ...           75%           max   \n",
       "<sample>                                    ...                               \n",
       "DetectSeq_ATP8-DddA11_REP-1   7.232032e+07  ...  2.953279e-10  1.226002e-08   \n",
       "DetectSeq_ATP8-DddA6_REP-1    7.886083e+07  ...  4.238782e-09  3.022062e-08   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   4.491641e+07  ...  3.872361e-10  2.584274e-09   \n",
       "DetectSeq_JAK2-DddA11_REP-1   7.720249e+07  ...  2.309763e-10  3.415243e-08   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  7.768411e+07  ...  3.257744e-09  2.089933e-07   \n",
       "\n",
       "                                 FDR                                     \\\n",
       "                               count      mean       std            min   \n",
       "<sample>                                                                  \n",
       "DetectSeq_ATP8-DddA11_REP-1     72.0  0.000008  0.000019   0.000000e+00   \n",
       "DetectSeq_ATP8-DddA6_REP-1     189.0  0.000012  0.000022   0.000000e+00   \n",
       "DetectSeq_ATP8-DddAwt_REP-1     27.0  0.000009  0.000017   0.000000e+00   \n",
       "DetectSeq_JAK2-DddA11_REP-1    400.0  0.000005  0.000014  7.206266e-102   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1204.0  0.000007  0.000018  8.602823e-153   \n",
       "\n",
       "                                                                        \\\n",
       "                                       25%           50%           75%   \n",
       "<sample>                                                                 \n",
       "DetectSeq_ATP8-DddA11_REP-1   3.609762e-17  6.514572e-11  3.036639e-06   \n",
       "DetectSeq_ATP8-DddA6_REP-1    1.511549e-12  9.258856e-08  1.664663e-05   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   8.475808e-08  4.466082e-07  1.031558e-05   \n",
       "DetectSeq_JAK2-DddA11_REP-1   9.844659e-19  4.976689e-11  6.804493e-07   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  7.967822e-17  1.283779e-09  2.256525e-06   \n",
       "\n",
       "                                        \n",
       "                                   max  \n",
       "<sample>                                \n",
       "DetectSeq_ATP8-DddA11_REP-1   0.000098  \n",
       "DetectSeq_ATP8-DddA6_REP-1    0.000096  \n",
       "DetectSeq_ATP8-DddAwt_REP-1   0.000056  \n",
       "DetectSeq_JAK2-DddA11_REP-1   0.000073  \n",
       "DetectSeq_SIRT6-DddA11_REP-1  0.000097  \n",
       "\n",
       "[5 rows x 160 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_strict = (df\n",
    "                  .query('FDR <= 0.0001')\n",
    "                  .query('log2_FC_mut >= 2')\n",
    "                  .query('ctrl_mut_count <= 1')\n",
    "                  .query('`treat_mut_count.norm` * 100 >= 10')\n",
    "                  .query('treat_mut_count >= 20')\n",
    "                  .query('treat_mut_count / treat_count >= 0.15')\n",
    "                  .query('region_block_site_num <= 1')\n",
    "                  .query('region_highest_site_mut_ratio >= 0.35')  # 1892\n",
    "                  )\n",
    "\n",
    "print(df_sign_strict.shape[0])\n",
    "df_sign_strict.groupby('<sample>').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a426c781-54c8-405a-bdfd-a2dc6f101608",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:41.340487Z",
     "iopub.status.busy": "2022-10-28T07:33:41.340228Z",
     "iopub.status.idle": "2022-10-28T07:33:41.350489Z",
     "shell.execute_reply": "2022-10-28T07:33:41.349957Z",
     "shell.execute_reply.started": "2022-10-28T07:33:41.340467Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_strict.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad92d39-8ef1-44fe-a4c5-2a655139ae8c",
   "metadata": {},
   "source": [
    "## lenient selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "65bd4318-6622-4da7-a5fe-87db7ec6d3e4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:42.495352Z",
     "iopub.status.busy": "2022-10-28T07:33:42.495093Z",
     "iopub.status.idle": "2022-10-28T07:33:42.729780Z",
     "shell.execute_reply": "2022-10-28T07:33:42.729213Z",
     "shell.execute_reply.started": "2022-10-28T07:33:42.495329Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4007\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">region_start</th>\n",
       "      <th colspan=\"2\" halign=\"left\">region_end</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">p_value</th>\n",
       "      <th colspan=\"8\" halign=\"left\">FDR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>...</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA11_REP-1</th>\n",
       "      <td>136.0</td>\n",
       "      <td>7.806753e+07</td>\n",
       "      <td>5.564656e+07</td>\n",
       "      <td>631391.0</td>\n",
       "      <td>33692937.75</td>\n",
       "      <td>69108378.0</td>\n",
       "      <td>1.176057e+08</td>\n",
       "      <td>242043276.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>7.806758e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>5.923322e-08</td>\n",
       "      <td>1.847701e-06</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0.001004</td>\n",
       "      <td>0.002092</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>8.085970e-12</td>\n",
       "      <td>5.023110e-06</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.009126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddA6_REP-1</th>\n",
       "      <td>490.0</td>\n",
       "      <td>8.236542e+07</td>\n",
       "      <td>5.517979e+07</td>\n",
       "      <td>58741.0</td>\n",
       "      <td>36676013.50</td>\n",
       "      <td>74671918.5</td>\n",
       "      <td>1.189550e+08</td>\n",
       "      <td>245995442.0</td>\n",
       "      <td>490.0</td>\n",
       "      <td>8.236547e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>5.322003e-07</td>\n",
       "      <td>6.243530e-06</td>\n",
       "      <td>490.0</td>\n",
       "      <td>0.001365</td>\n",
       "      <td>0.002426</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.017237e-08</td>\n",
       "      <td>6.874875e-05</td>\n",
       "      <td>0.001158</td>\n",
       "      <td>0.009637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_ATP8-DddAwt_REP-1</th>\n",
       "      <td>56.0</td>\n",
       "      <td>5.362142e+07</td>\n",
       "      <td>4.577483e+07</td>\n",
       "      <td>631391.0</td>\n",
       "      <td>17355850.00</td>\n",
       "      <td>44585004.5</td>\n",
       "      <td>7.035096e+07</td>\n",
       "      <td>201232409.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>5.362146e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.484702e-07</td>\n",
       "      <td>8.404122e-07</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.001748</td>\n",
       "      <td>0.002776</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.466082e-07</td>\n",
       "      <td>4.359292e-05</td>\n",
       "      <td>0.002067</td>\n",
       "      <td>0.009224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_JAK2-DddA11_REP-1</th>\n",
       "      <td>658.0</td>\n",
       "      <td>7.490521e+07</td>\n",
       "      <td>5.412918e+07</td>\n",
       "      <td>1056219.0</td>\n",
       "      <td>32369589.25</td>\n",
       "      <td>63261637.0</td>\n",
       "      <td>1.094002e+08</td>\n",
       "      <td>246696632.0</td>\n",
       "      <td>658.0</td>\n",
       "      <td>7.490524e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>1.170807e-07</td>\n",
       "      <td>7.326257e-06</td>\n",
       "      <td>658.0</td>\n",
       "      <td>0.000832</td>\n",
       "      <td>0.002095</td>\n",
       "      <td>7.206266e-102</td>\n",
       "      <td>7.891722e-14</td>\n",
       "      <td>2.819972e-07</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.009935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DetectSeq_SIRT6-DddA11_REP-1</th>\n",
       "      <td>2667.0</td>\n",
       "      <td>7.902458e+07</td>\n",
       "      <td>5.575653e+07</td>\n",
       "      <td>79700.0</td>\n",
       "      <td>34749176.00</td>\n",
       "      <td>68550649.0</td>\n",
       "      <td>1.163118e+08</td>\n",
       "      <td>248905893.0</td>\n",
       "      <td>2667.0</td>\n",
       "      <td>7.902463e+07</td>\n",
       "      <td>...</td>\n",
       "      <td>3.484243e-06</td>\n",
       "      <td>4.180053e-05</td>\n",
       "      <td>2667.0</td>\n",
       "      <td>0.001171</td>\n",
       "      <td>0.002324</td>\n",
       "      <td>8.602823e-153</td>\n",
       "      <td>1.249246e-10</td>\n",
       "      <td>9.457196e-06</td>\n",
       "      <td>0.001175</td>\n",
       "      <td>0.009690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             region_start                              \\\n",
       "                                    count          mean           std   \n",
       "<sample>                                                                \n",
       "DetectSeq_ATP8-DddA11_REP-1         136.0  7.806753e+07  5.564656e+07   \n",
       "DetectSeq_ATP8-DddA6_REP-1          490.0  8.236542e+07  5.517979e+07   \n",
       "DetectSeq_ATP8-DddAwt_REP-1          56.0  5.362142e+07  4.577483e+07   \n",
       "DetectSeq_JAK2-DddA11_REP-1         658.0  7.490521e+07  5.412918e+07   \n",
       "DetectSeq_SIRT6-DddA11_REP-1       2667.0  7.902458e+07  5.575653e+07   \n",
       "\n",
       "                                                                  \\\n",
       "                                    min          25%         50%   \n",
       "<sample>                                                           \n",
       "DetectSeq_ATP8-DddA11_REP-1    631391.0  33692937.75  69108378.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1      58741.0  36676013.50  74671918.5   \n",
       "DetectSeq_ATP8-DddAwt_REP-1    631391.0  17355850.00  44585004.5   \n",
       "DetectSeq_JAK2-DddA11_REP-1   1056219.0  32369589.25  63261637.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1    79700.0  34749176.00  68550649.0   \n",
       "\n",
       "                                                        region_end  \\\n",
       "                                       75%          max      count   \n",
       "<sample>                                                             \n",
       "DetectSeq_ATP8-DddA11_REP-1   1.176057e+08  242043276.0      136.0   \n",
       "DetectSeq_ATP8-DddA6_REP-1    1.189550e+08  245995442.0      490.0   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   7.035096e+07  201232409.0       56.0   \n",
       "DetectSeq_JAK2-DddA11_REP-1   1.094002e+08  246696632.0      658.0   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1.163118e+08  248905893.0     2667.0   \n",
       "\n",
       "                                            ...       p_value                \\\n",
       "                                      mean  ...           75%           max   \n",
       "<sample>                                    ...                               \n",
       "DetectSeq_ATP8-DddA11_REP-1   7.806758e+07  ...  5.923322e-08  1.847701e-06   \n",
       "DetectSeq_ATP8-DddA6_REP-1    8.236547e+07  ...  5.322003e-07  6.243530e-06   \n",
       "DetectSeq_ATP8-DddAwt_REP-1   5.362146e+07  ...  1.484702e-07  8.404122e-07   \n",
       "DetectSeq_JAK2-DddA11_REP-1   7.490524e+07  ...  1.170807e-07  7.326257e-06   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  7.902463e+07  ...  3.484243e-06  4.180053e-05   \n",
       "\n",
       "                                 FDR                                     \\\n",
       "                               count      mean       std            min   \n",
       "<sample>                                                                  \n",
       "DetectSeq_ATP8-DddA11_REP-1    136.0  0.001004  0.002092   0.000000e+00   \n",
       "DetectSeq_ATP8-DddA6_REP-1     490.0  0.001365  0.002426   0.000000e+00   \n",
       "DetectSeq_ATP8-DddAwt_REP-1     56.0  0.001748  0.002776   0.000000e+00   \n",
       "DetectSeq_JAK2-DddA11_REP-1    658.0  0.000832  0.002095  7.206266e-102   \n",
       "DetectSeq_SIRT6-DddA11_REP-1  2667.0  0.001171  0.002324  8.602823e-153   \n",
       "\n",
       "                                                                              \n",
       "                                       25%           50%       75%       max  \n",
       "<sample>                                                                      \n",
       "DetectSeq_ATP8-DddA11_REP-1   8.085970e-12  5.023110e-06  0.000407  0.009126  \n",
       "DetectSeq_ATP8-DddA6_REP-1    2.017237e-08  6.874875e-05  0.001158  0.009637  \n",
       "DetectSeq_ATP8-DddAwt_REP-1   4.466082e-07  4.359292e-05  0.002067  0.009224  \n",
       "DetectSeq_JAK2-DddA11_REP-1   7.891722e-14  2.819972e-07  0.000230  0.009935  \n",
       "DetectSeq_SIRT6-DddA11_REP-1  1.249246e-10  9.457196e-06  0.001175  0.009690  \n",
       "\n",
       "[5 rows x 160 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nature protocol standard\n",
    "# df_sign_lenient = (df\n",
    "#                    .query('FDR <= 0.01')\n",
    "#                    .query('log2_FC_mut >= 2')\n",
    "#                    .query('ctrl_mut_count <= 1')\n",
    "#                    .query('`treat_mut_count.norm` * 100 >= 5')\n",
    "#                    .query('treat_mut_count >= 10')\n",
    "#                    .query('treat_mut_count / treat_count >= 0.15')\n",
    "#                    .query('region_block_site_num <= 1')\n",
    "#                    .query('region_highest_site_mut_ratio >= 0.30')  # 4007\n",
    "#                    )\n",
    "\n",
    "# print(df_sign_lenient.shape[0])\n",
    "# df_sign_lenient.groupby('<sample>').describe()\n",
    "\n",
    "\n",
    "# 2022-10-27 update\n",
    "df_sign_lenient = (df\n",
    "                   .query('FDR <= 0.01')\n",
    "                   .query('log2_FC_mut >= 2')\n",
    "                   .query('ctrl_mut_count <= 1')\n",
    "                   .query('`treat_mut_count.norm` * 100 >= 5')\n",
    "                   .query('treat_mut_count >= 10')\n",
    "                   .query('treat_mut_count / treat_count >= 0.15')\n",
    "                   .query('region_block_site_num <= 1')\n",
    "                   .query('region_highest_site_mut_ratio >= 0.30')  # 4007\n",
    "                   )\n",
    "\n",
    "print(df_sign_lenient.shape[0])\n",
    "df_sign_lenient.groupby('<sample>').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3f2b9f6a-fc86-4bda-bcd3-7d75d7506a56",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:43.568237Z",
     "iopub.status.busy": "2022-10-28T07:33:43.567941Z",
     "iopub.status.idle": "2022-10-28T07:33:43.578614Z",
     "shell.execute_reply": "2022-10-28T07:33:43.578150Z",
     "shell.execute_reply.started": "2022-10-28T07:33:43.568219Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sign_lenient.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "052bbbad-08fa-4671-8502-418befd285b8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:47.113640Z",
     "iopub.status.busy": "2022-10-28T07:33:47.113330Z",
     "iopub.status.idle": "2022-10-28T07:33:47.128995Z",
     "shell.execute_reply": "2022-10-28T07:33:47.128337Z",
     "shell.execute_reply.started": "2022-10-28T07:33:47.113618Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4007, 30)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>...</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.801598</td>\n",
       "      <td>0,1,2,3 22,0,0,0 689,406,500,0</td>\n",
       "      <td>7.415426</td>\n",
       "      <td>8.434072</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>9.573874e-183</td>\n",
       "      <td>1.536847e-177</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631778_CT,chr1_631787_CT,chr1_631791_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>5.348597e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631791_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "      <td>chr1_633067_633147</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>chr1_633067_GA,chr1_633068_GA,chr1_633073_GA,c...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.409604</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11 33,0,0,0,0,0,0,0,0,0...</td>\n",
       "      <td>7.466665</td>\n",
       "      <td>8.717426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.596361e-224</td>\n",
       "      <td>3.416748e-219</td>\n",
       "      <td>chr1_633067_633147_highest_chr1_633100_GA</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633244</td>\n",
       "      <td>633286</td>\n",
       "      <td>chr1_633244_633286</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>chr1_633244_GA,chr1_633246_GA,chr1_633249_GA,c...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.001422</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 19,3,0,0,0,0,0,0,0,0 564,1...</td>\n",
       "      <td>6.655668</td>\n",
       "      <td>6.949872</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.429064e-61</td>\n",
       "      <td>1.733004e-56</td>\n",
       "      <td>chr1_633244_633286_highest_chr1_633286_GA</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633376</td>\n",
       "      <td>633442</td>\n",
       "      <td>chr1_633376_633442</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>chr1_633376_CT,chr1_633377_CT,chr1_633378_C.,c...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>1.591546</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...</td>\n",
       "      <td>6.225823</td>\n",
       "      <td>9.296213</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.421498e-100</td>\n",
       "      <td>1.404222e-95</td>\n",
       "      <td>chr1_633376_633442_highest_chr1_633384_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       <sample> chr_name  region_start  region_end  \\\n",
       "67  DetectSeq_ATP8-DddA11_REP-1     chr1        631391      631393   \n",
       "69  DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791   \n",
       "74  DetectSeq_ATP8-DddA11_REP-1     chr1        633067      633147   \n",
       "75  DetectSeq_ATP8-DddA11_REP-1     chr1        633244      633286   \n",
       "78  DetectSeq_ATP8-DddA11_REP-1     chr1        633376      633442   \n",
       "\n",
       "           mpmat_index  region_site_num  region_block_site_num  \\\n",
       "67  chr1_631391_631393                3                      1   \n",
       "69  chr1_631778_631791                3                      0   \n",
       "74  chr1_633067_633147               11                      0   \n",
       "75  chr1_633244_633286                9                      0   \n",
       "78  chr1_633376_633442               20                      0   \n",
       "\n",
       "    region_mut_site_num                                  region_site_index  \\\n",
       "67                    2       chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "69                    3       chr1_631778_CT,chr1_631787_CT,chr1_631791_CT   \n",
       "74                   11  chr1_633067_GA,chr1_633068_GA,chr1_633073_GA,c...   \n",
       "75                    9  chr1_633244_GA,chr1_633246_GA,chr1_633249_GA,c...   \n",
       "78                   20  chr1_633376_CT,chr1_633377_CT,chr1_633378_C.,c...   \n",
       "\n",
       "                         region_block_state  ... ctrl_mut_count.norm  \\\n",
       "67                                    B-N-N  ...            0.000000   \n",
       "69                                    N-N-N  ...            0.000000   \n",
       "74                    N-N-N-N-N-N-N-N-N-N-N  ...            0.000000   \n",
       "75                        N-N-N-N-N-N-N-N-N  ...            0.000000   \n",
       "78  N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N  ...            0.002532   \n",
       "\n",
       "    treat_mut_count.norm                                         count_info  \\\n",
       "67              2.801598                     0,1,2,3 22,0,0,0 689,406,500,0   \n",
       "69              3.260583                   0,1,2,3 40,0,0,0 534,278,413,258   \n",
       "74              3.409604  0,1,2,3,4,5,6,7,8,9,10,11 33,0,0,0,0,0,0,0,0,0...   \n",
       "75              1.001422  0,1,2,3,4,5,6,7,8,9 19,3,0,0,0,0,0,0,0,0 564,1...   \n",
       "78              1.591546  0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...   \n",
       "\n",
       "     log2_FC  log2_FC_mut  test_state        p_value            FDR  \\\n",
       "67  7.415426     8.434072      TestOK  9.573874e-183  1.536847e-177   \n",
       "69  6.447891     8.652952      TestOK  2.915445e-214  5.348597e-209   \n",
       "74  7.466665     8.717426      TestOK  1.596361e-224  3.416748e-219   \n",
       "75  6.655668     6.949872      TestOK   2.429064e-61   1.733004e-56   \n",
       "78  6.225823     9.296213      TestOK  1.421498e-100   1.404222e-95   \n",
       "\n",
       "                                     bed_name  strand  \n",
       "67  chr1_631391_631393_highest_chr1_631392_CT       .  \n",
       "69  chr1_631778_631791_highest_chr1_631791_CT       .  \n",
       "74  chr1_633067_633147_highest_chr1_633100_GA       .  \n",
       "75  chr1_633244_633286_highest_chr1_633286_GA       .  \n",
       "78  chr1_633376_633442_highest_chr1_633384_CT       .  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_sign_lenient\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab1dca4-cf2d-4b0e-a073-d9ea4ce66f75",
   "metadata": {},
   "source": [
    "## fix mpmat_index (add old into it)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3b10cfa4-162b-409d-8274-ff6eea6cecae",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:47.142891Z",
     "iopub.status.busy": "2022-10-28T07:33:47.142656Z",
     "iopub.status.idle": "2022-10-28T07:33:47.189026Z",
     "shell.execute_reply": "2022-10-28T07:33:47.188356Z",
     "shell.execute_reply.started": "2022-10-28T07:33:47.142873Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count                     6881\n",
       "unique                    6881\n",
       "top       chr1_1471366_1471410\n",
       "freq                         1\n",
       "Name: region_id, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_id = pd.read_csv(\n",
    "    \"../tables/20220312-DdCBE-off_target_type.FinallistV4.CheckPrimer.AddV4ID.tsv\",\n",
    "    sep='\\t',\n",
    "    header=0,\n",
    "    usecols=['region_id'],\n",
    ")\n",
    "df_old_id.region_id.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "00ba7c78-3957-4acc-a527-d92225bb020e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:48.029233Z",
     "iopub.status.busy": "2022-10-28T07:33:48.028952Z",
     "iopub.status.idle": "2022-10-28T07:33:48.049375Z",
     "shell.execute_reply": "2022-10-28T07:33:48.048859Z",
     "shell.execute_reply.started": "2022-10-28T07:33:48.029215Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1471366</td>\n",
       "      <td>1471410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1693068</td>\n",
       "      <td>1693084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5806517</td>\n",
       "      <td>5806533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5983069</td>\n",
       "      <td>5983085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6876</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6877</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154373542</td>\n",
       "      <td>154373556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6878</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6879</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155068321</td>\n",
       "      <td>155068358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6880</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155682866</td>\n",
       "      <td>155682903</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6881 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end\n",
       "0     chr1    1471366    1471410\n",
       "1     chr1    1693068    1693084\n",
       "2     chr1    2044988    2044998\n",
       "3     chr1    5806517    5806533\n",
       "4     chr1    5983069    5983085\n",
       "...    ...        ...        ...\n",
       "6876  chrX  153669436  153669453\n",
       "6877  chrX  154373542  154373556\n",
       "6878  chrX  155049205  155049215\n",
       "6879  chrX  155068321  155068358\n",
       "6880  chrX  155682866  155682903\n",
       "\n",
       "[6881 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_id[['chrom', 'start', 'end']] = pd.Series(df_old_id.region_id.unique()).str.split('_', expand=True)\n",
    "df_old_id = df_old_id.iloc[:,1:4]\n",
    "df_old_id['start'] = df_old_id['start'].astype(int)\n",
    "df_old_id['end'] = df_old_id['end'].astype(int)\n",
    "df_old_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9846a3db-008a-47e0-b927-63b8eb5ff6f0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:48.883178Z",
     "iopub.status.busy": "2022-10-28T07:33:48.883011Z",
     "iopub.status.idle": "2022-10-28T07:33:48.891144Z",
     "shell.execute_reply": "2022-10-28T07:33:48.890603Z",
     "shell.execute_reply.started": "2022-10-28T07:33:48.883162Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 6881 entries, 0 to 6880\n",
      "Data columns (total 3 columns):\n",
      " #   Column  Non-Null Count  Dtype \n",
      "---  ------  --------------  ----- \n",
      " 0   chrom   6881 non-null   object\n",
      " 1   start   6881 non-null   int64 \n",
      " 2   end     6881 non-null   int64 \n",
      "dtypes: int64(2), object(1)\n",
      "memory usage: 161.4+ KB\n"
     ]
    }
   ],
   "source": [
    "df_old_id.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "38ec28e0-0c34-46cd-b266-b6a6bfd464e8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:48.893829Z",
     "iopub.status.busy": "2022-10-28T07:33:48.893611Z",
     "iopub.status.idle": "2022-10-28T07:33:48.910024Z",
     "shell.execute_reply": "2022-10-28T07:33:48.909467Z",
     "shell.execute_reply.started": "2022-10-28T07:33:48.893812Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633244</td>\n",
       "      <td>633286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633376</td>\n",
       "      <td>633442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3227</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154448849</td>\n",
       "      <td>154448867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3228</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154457770</td>\n",
       "      <td>154457792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3229</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154457793</td>\n",
       "      <td>154457822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3230</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154924102</td>\n",
       "      <td>154924198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3231</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155439898</td>\n",
       "      <td>155439954</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3232 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end\n",
       "0     chr1     631391     631393\n",
       "1     chr1     631778     631791\n",
       "2     chr1     633067     633147\n",
       "3     chr1     633244     633286\n",
       "4     chr1     633376     633442\n",
       "...    ...        ...        ...\n",
       "3227  chrX  154448849  154448867\n",
       "3228  chrX  154457770  154457792\n",
       "3229  chrX  154457793  154457822\n",
       "3230  chrX  154924102  154924198\n",
       "3231  chrX  155439898  155439954\n",
       "\n",
       "[3232 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new_id = pd.Series(df.mpmat_index.unique()).str.split('_', expand=True)\n",
    "df_new_id.columns = ['chrom', 'start', 'end']\n",
    "df_new_id['start'] = df_new_id['start'].astype(int)\n",
    "df_new_id['end'] = df_new_id['end'].astype(int)\n",
    "\n",
    "df_new_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d5019725-fad9-4f0d-9b29-0bbd64878f57",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:49.671177Z",
     "iopub.status.busy": "2022-10-28T07:33:49.670881Z",
     "iopub.status.idle": "2022-10-28T07:33:49.699765Z",
     "shell.execute_reply": "2022-10-28T07:33:49.699166Z",
     "shell.execute_reply.started": "2022-10-28T07:33:49.671156Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "bed_new = BedTool.from_dataframe(df_new_id)\n",
    "bed_nat = BedTool.from_dataframe(df_old_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c0a9a3e-5224-40bf-82a9-eb7a15dc8afb",
   "metadata": {},
   "source": [
    "> https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html\n",
    "\n",
    "![](https://tva1.sinaimg.cn/large/008vxvgGly1h7jy4v629sj30ld0j9wg7.jpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "006c2927-342b-43e5-b42e-1505f0952001",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:51.868348Z",
     "iopub.status.busy": "2022-10-28T07:33:51.868072Z",
     "iopub.status.idle": "2022-10-28T07:33:51.960478Z",
     "shell.execute_reply": "2022-10-28T07:33:51.959845Z",
     "shell.execute_reply.started": "2022-10-28T07:33:51.868330Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633244</td>\n",
       "      <td>633286</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>633376</td>\n",
       "      <td>633442</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3229</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154448849</td>\n",
       "      <td>154448867</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3230</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154457770</td>\n",
       "      <td>154457792</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3231</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154457793</td>\n",
       "      <td>154457822</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3232</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154924102</td>\n",
       "      <td>154924198</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3233</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155439898</td>\n",
       "      <td>155439954</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3234 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end chrom2  start2  end2\n",
       "0     chr1     631391     631393      .      -1    -1\n",
       "1     chr1     631778     631791      .      -1    -1\n",
       "2     chr1     633067     633147      .      -1    -1\n",
       "3     chr1     633244     633286      .      -1    -1\n",
       "4     chr1     633376     633442      .      -1    -1\n",
       "...    ...        ...        ...    ...     ...   ...\n",
       "3229  chrX  154448849  154448867      .      -1    -1\n",
       "3230  chrX  154457770  154457792      .      -1    -1\n",
       "3231  chrX  154457793  154457822      .      -1    -1\n",
       "3232  chrX  154924102  154924198      .      -1    -1\n",
       "3233  chrX  155439898  155439954      .      -1    -1\n",
       "\n",
       "[3234 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bed_to_fix = bed_new.intersect(bed_nat, loj=True).to_dataframe()\n",
    "df_bed_to_fix.columns = ['chrom', 'start', 'end', 'chrom2', 'start2', 'end2']\n",
    "\n",
    "df_bed_to_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1143d20b-274f-4c08-97e0-d1b9781f6d5f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:53.748797Z",
     "iopub.status.busy": "2022-10-28T07:33:53.748546Z",
     "iopub.status.idle": "2022-10-28T07:33:53.769170Z",
     "shell.execute_reply": "2022-10-28T07:33:53.768672Z",
     "shell.execute_reply.started": "2022-10-28T07:33:53.748779Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_633067_633147</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_633244_633286</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633244</td>\n",
       "      <td>633286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_633376_633442</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633376</td>\n",
       "      <td>633442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3170</th>\n",
       "      <td>chrX_41024494_41024536</td>\n",
       "      <td>chrX</td>\n",
       "      <td>41024523</td>\n",
       "      <td>41024536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3186</th>\n",
       "      <td>chrX_72126743_72126762</td>\n",
       "      <td>chrX</td>\n",
       "      <td>72126743</td>\n",
       "      <td>72126762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3199</th>\n",
       "      <td>chrX_103211187_103211215</td>\n",
       "      <td>chrX</td>\n",
       "      <td>103211203</td>\n",
       "      <td>103211249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3220</th>\n",
       "      <td>chrX_132292956_132292983</td>\n",
       "      <td>chrX</td>\n",
       "      <td>132292961</td>\n",
       "      <td>132292983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3221</th>\n",
       "      <td>chrX_133005157_133005236</td>\n",
       "      <td>chrX</td>\n",
       "      <td>133005219</td>\n",
       "      <td>133005236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3234 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   mpmat_index chr_name  region_start  region_end\n",
       "0           chr1_631391_631393     chr1        631391      631393\n",
       "1           chr1_631778_631791     chr1        631778      631791\n",
       "2           chr1_633067_633147     chr1        633067      633147\n",
       "3           chr1_633244_633286     chr1        633244      633286\n",
       "4           chr1_633376_633442     chr1        633376      633442\n",
       "...                        ...      ...           ...         ...\n",
       "3170    chrX_41024494_41024536     chrX      41024523    41024536\n",
       "3186    chrX_72126743_72126762     chrX      72126743    72126762\n",
       "3199  chrX_103211187_103211215     chrX     103211203   103211249\n",
       "3220  chrX_132292956_132292983     chrX     132292961   132292983\n",
       "3221  chrX_133005157_133005236     chrX     133005219   133005236\n",
       "\n",
       "[3234 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# end2 为-1 代表老点不存在，\n",
    "# end2 有值说明，新老点重合，以老点 region 坐标为准进行替换\n",
    "# df_test = df_bed_to_fix.query('end2 != -1')\n",
    "# abs((df_test.end2 - df_test.start2 - df_test.end + df_test.start)).describe()\n",
    "\n",
    "\n",
    "# 唯一标识mpmat_index\n",
    "\n",
    "df_bed_to_fix['mpmat_index'] = df_bed_to_fix['chrom'] + '_' + df_bed_to_fix['start'].astype(str) + '_' + df_bed_to_fix['end'].astype(str)\n",
    "\n",
    "\n",
    "df_bed_to_fix_part1 = df_bed_to_fix.query('end2 == -1')[['mpmat_index', 'chrom', 'start', 'end']].copy()\n",
    "df_bed_to_fix_part2 = df_bed_to_fix.query('end2 != -1')[['mpmat_index', 'chrom2', 'start2', 'end2']].copy()\n",
    "\n",
    "\n",
    "df_bed_to_fix_part1.columns = ['mpmat_index', 'chr_name', 'region_start', 'region_end']\n",
    "df_bed_to_fix_part2.columns = ['mpmat_index', 'chr_name', 'region_start', 'region_end']\n",
    "\n",
    "df_bed_fixed_coordinate = pd.concat([df_bed_to_fix_part1, df_bed_to_fix_part2], axis=0)\n",
    "df_bed_fixed_coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b13675cc-83aa-4644-ad6d-170737691697",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:33:57.179999Z",
     "iopub.status.busy": "2022-10-28T07:33:57.179725Z",
     "iopub.status.idle": "2022-10-28T07:33:57.195352Z",
     "shell.execute_reply": "2022-10-28T07:33:57.194559Z",
     "shell.execute_reply.started": "2022-10-28T07:33:57.179979Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 mpmat_index chr_name  region_start  region_end\n",
      "864  chr10_22989043_22989070    chr10      22989049    22989070\n",
      "865  chr10_22989043_22989070    chr10      22989052    22989071\n",
      "                  mpmat_index chr_name  region_start  region_end\n",
      "2931  chr17_67963461_67963466    chr17      67963444    67963466\n",
      "2932  chr17_67963461_67963466    chr17      67963458    67963466\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_633067_633147</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633067</td>\n",
       "      <td>633147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_633244_633286</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633244</td>\n",
       "      <td>633286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_633376_633442</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633376</td>\n",
       "      <td>633442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3170</th>\n",
       "      <td>chrX_41024494_41024536</td>\n",
       "      <td>chrX</td>\n",
       "      <td>41024523</td>\n",
       "      <td>41024536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3186</th>\n",
       "      <td>chrX_72126743_72126762</td>\n",
       "      <td>chrX</td>\n",
       "      <td>72126743</td>\n",
       "      <td>72126762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3199</th>\n",
       "      <td>chrX_103211187_103211215</td>\n",
       "      <td>chrX</td>\n",
       "      <td>103211203</td>\n",
       "      <td>103211249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3220</th>\n",
       "      <td>chrX_132292956_132292983</td>\n",
       "      <td>chrX</td>\n",
       "      <td>132292961</td>\n",
       "      <td>132292983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3221</th>\n",
       "      <td>chrX_133005157_133005236</td>\n",
       "      <td>chrX</td>\n",
       "      <td>133005219</td>\n",
       "      <td>133005236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3232 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   mpmat_index chr_name  region_start  region_end\n",
       "0           chr1_631391_631393     chr1        631391      631393\n",
       "1           chr1_631778_631791     chr1        631778      631791\n",
       "2           chr1_633067_633147     chr1        633067      633147\n",
       "3           chr1_633244_633286     chr1        633244      633286\n",
       "4           chr1_633376_633442     chr1        633376      633442\n",
       "...                        ...      ...           ...         ...\n",
       "3170    chrX_41024494_41024536     chrX      41024523    41024536\n",
       "3186    chrX_72126743_72126762     chrX      72126743    72126762\n",
       "3199  chrX_103211187_103211215     chrX     103211203   103211249\n",
       "3220  chrX_132292956_132292983     chrX     132292961   132292983\n",
       "3221  chrX_133005157_133005236     chrX     133005219   133005236\n",
       "\n",
       "[3232 rows x 4 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 发现两个 duplicated 的 mpmat_index，check 一下为什么\n",
    "# df_bed_fixed_coordinate[df_bed_fixed_coordinate.mpmat_index.duplicated()]\n",
    "\n",
    "print(df_bed_fixed_coordinate.query('mpmat_index==\"chr10_22989043_22989070\"'))\n",
    "print(df_bed_fixed_coordinate.query('mpmat_index==\"chr17_67963461_67963466\"'))\n",
    "#                  mpmat_index chr_name  region_start  region_end\n",
    "# 864  chr10_22989043_22989070    chr10      22989049    22989070\n",
    "# 865  chr10_22989043_22989070    chr10      22989052    22989071\n",
    "#                   mpmat_index chr_name  region_start  region_end\n",
    "# 2931  chr17_67963461_67963466    chr17      67963444    67963466\n",
    "# 2932  chr17_67963461_67963466    chr17      67963458    67963466\n",
    "\n",
    "# 发现其实就是没定下来最终的导致了有多出来的点，图方便取第一个，注意以后 debug！\n",
    "# DEBUG\n",
    "\n",
    "df_bed_fixed_coordinate.drop_duplicates(subset='mpmat_index', keep='first', inplace=True)\n",
    "\n",
    "df_bed_fixed_coordinate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4e7e51f8-e29e-4ba1-b0fa-4c4354d6f7b6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:00.350455Z",
     "iopub.status.busy": "2022-10-28T07:34:00.350172Z",
     "iopub.status.idle": "2022-10-28T07:34:00.411156Z",
     "shell.execute_reply": "2022-10-28T07:34:00.410481Z",
     "shell.execute_reply.started": "2022-10-28T07:34:00.350434Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>...</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.801598</td>\n",
       "      <td>0,1,2,3 22,0,0,0 689,406,500,0</td>\n",
       "      <td>7.415426</td>\n",
       "      <td>8.434072</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>9.573874e-183</td>\n",
       "      <td>1.536847e-177</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.750688</td>\n",
       "      <td>0,1,2,3 22,0,0,0 511,169,485,0</td>\n",
       "      <td>7.070792</td>\n",
       "      <td>8.407614</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.632264e-166</td>\n",
       "      <td>4.829081e-161</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.181582</td>\n",
       "      <td>0,1,2,3 22,0,0,0 157,60,43,0</td>\n",
       "      <td>4.442013</td>\n",
       "      <td>4.486517</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.391279e-12</td>\n",
       "      <td>2.014030e-07</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631778_CT,chr1_631787_CT,chr1_631791_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>5.348597e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631791_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631778_CT,chr1_631787_CT,chr1_631791_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.001335</td>\n",
       "      <td>0,1,2,3 40,0,0,0 530,320,528,69</td>\n",
       "      <td>6.521031</td>\n",
       "      <td>8.533426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.072057e-182</td>\n",
       "      <td>6.575230e-177</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631787_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4002</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>41024523</td>\n",
       "      <td>41024536</td>\n",
       "      <td>chrX_41024523_41024536</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>chrX_41024494_GA,chrX_41024498_GA,chrX_4102450...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.311431</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13 22,0,0,0,0,0,0...</td>\n",
       "      <td>3.245836</td>\n",
       "      <td>5.533199</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.108715e-19</td>\n",
       "      <td>8.180768e-16</td>\n",
       "      <td>chrX_41024494_41024536_highest_chrX_41024528_GA</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4003</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>72126743</td>\n",
       "      <td>72126762</td>\n",
       "      <td>chrX_72126743_72126762</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>chrX_72126743_CT,chrX_72126744_CT,chrX_7212674...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.088980</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9 28,0,0,0,0,0,0,0,0,0 9,0,3...</td>\n",
       "      <td>0.971913</td>\n",
       "      <td>3.725844</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.269560e-05</td>\n",
       "      <td>7.913254e-03</td>\n",
       "      <td>chrX_72126743_72126762_highest_chrX_72126750_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4004</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>103211203</td>\n",
       "      <td>103211249</td>\n",
       "      <td>chrX_103211203_103211249</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>chrX_103211187_GA,chrX_103211197_GA,chrX_10321...</td>\n",
       "      <td>N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.172399</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8 34,0,0,0,0,0,0,0,0 27,10,12,...</td>\n",
       "      <td>2.135412</td>\n",
       "      <td>4.680040</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.985607e-10</td>\n",
       "      <td>2.539152e-07</td>\n",
       "      <td>chrX_103211187_103211215_highest_chrX_10321121...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4005</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>132292961</td>\n",
       "      <td>132292983</td>\n",
       "      <td>chrX_132292961_132292983</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>chrX_132292956_GA,chrX_132292961_GA,chrX_13229...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002532</td>\n",
       "      <td>0.088980</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10 40,0,1,0,0,0,0,0,0,0,0 ...</td>\n",
       "      <td>1.135412</td>\n",
       "      <td>5.135412</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.269560e-05</td>\n",
       "      <td>7.913254e-03</td>\n",
       "      <td>chrX_132292956_132292983_highest_chrX_13229298...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4006</th>\n",
       "      <td>DetectSeq_SIRT6-DddA11_REP-1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>133005219</td>\n",
       "      <td>133005236</td>\n",
       "      <td>chrX_133005219_133005236</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>chrX_133005157_GA,chrX_133005162_G.,chrX_13300...</td>\n",
       "      <td>N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127909</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...</td>\n",
       "      <td>2.015117</td>\n",
       "      <td>4.249406</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>1.635736e-07</td>\n",
       "      <td>7.739919e-05</td>\n",
       "      <td>chrX_133005157_133005236_highest_chrX_13300523...</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4007 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          <sample> chr_name  region_start  region_end  \\\n",
       "0      DetectSeq_ATP8-DddA11_REP-1     chr1        631391      631393   \n",
       "1       DetectSeq_ATP8-DddA6_REP-1     chr1        631391      631393   \n",
       "2      DetectSeq_ATP8-DddAwt_REP-1     chr1        631391      631393   \n",
       "3      DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791   \n",
       "4       DetectSeq_ATP8-DddA6_REP-1     chr1        631778      631791   \n",
       "...                            ...      ...           ...         ...   \n",
       "4002  DetectSeq_SIRT6-DddA11_REP-1     chrX      41024523    41024536   \n",
       "4003  DetectSeq_SIRT6-DddA11_REP-1     chrX      72126743    72126762   \n",
       "4004  DetectSeq_SIRT6-DddA11_REP-1     chrX     103211203   103211249   \n",
       "4005  DetectSeq_SIRT6-DddA11_REP-1     chrX     132292961   132292983   \n",
       "4006  DetectSeq_SIRT6-DddA11_REP-1     chrX     133005219   133005236   \n",
       "\n",
       "                   mpmat_index  region_site_num  region_block_site_num  \\\n",
       "0           chr1_631391_631393                3                      1   \n",
       "1           chr1_631391_631393                3                      1   \n",
       "2           chr1_631391_631393                3                      1   \n",
       "3           chr1_631778_631791                3                      0   \n",
       "4           chr1_631778_631791                3                      0   \n",
       "...                        ...              ...                    ...   \n",
       "4002    chrX_41024523_41024536               13                      0   \n",
       "4003    chrX_72126743_72126762                9                      0   \n",
       "4004  chrX_103211203_103211249                8                      0   \n",
       "4005  chrX_132292961_132292983               10                      0   \n",
       "4006  chrX_133005219_133005236               19                      0   \n",
       "\n",
       "      region_mut_site_num                                  region_site_index  \\\n",
       "0                       2       chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "1                       2       chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "2                       2       chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "3                       3       chr1_631778_CT,chr1_631787_CT,chr1_631791_CT   \n",
       "4                       3       chr1_631778_CT,chr1_631787_CT,chr1_631791_CT   \n",
       "...                   ...                                                ...   \n",
       "4002                   13  chrX_41024494_GA,chrX_41024498_GA,chrX_4102450...   \n",
       "4003                    9  chrX_72126743_CT,chrX_72126744_CT,chrX_7212674...   \n",
       "4004                    8  chrX_103211187_GA,chrX_103211197_GA,chrX_10321...   \n",
       "4005                   10  chrX_132292956_GA,chrX_132292961_GA,chrX_13229...   \n",
       "4006                   19  chrX_133005157_GA,chrX_133005162_G.,chrX_13300...   \n",
       "\n",
       "                         region_block_state  ... ctrl_mut_count.norm  \\\n",
       "0                                     B-N-N  ...            0.000000   \n",
       "1                                     B-N-N  ...            0.000000   \n",
       "2                                     B-N-N  ...            0.000000   \n",
       "3                                     N-N-N  ...            0.000000   \n",
       "4                                     N-N-N  ...            0.000000   \n",
       "...                                     ...  ...                 ...   \n",
       "4002              N-N-N-N-N-N-N-N-N-N-N-N-N  ...            0.000000   \n",
       "4003                      N-N-N-N-N-N-N-N-N  ...            0.000000   \n",
       "4004                        N-N-N-N-N-N-N-N  ...            0.000000   \n",
       "4005                    N-N-N-N-N-N-N-N-N-N  ...            0.002532   \n",
       "4006  N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N-N  ...            0.000000   \n",
       "\n",
       "      treat_mut_count.norm                                         count_info  \\\n",
       "0                 2.801598                     0,1,2,3 22,0,0,0 689,406,500,0   \n",
       "1                 2.750688                     0,1,2,3 22,0,0,0 511,169,485,0   \n",
       "2                 0.181582                       0,1,2,3 22,0,0,0 157,60,43,0   \n",
       "3                 3.260583                   0,1,2,3 40,0,0,0 534,278,413,258   \n",
       "4                 3.001335                    0,1,2,3 40,0,0,0 530,320,528,69   \n",
       "...                    ...                                                ...   \n",
       "4002              0.311431  0,1,2,3,4,5,6,7,8,9,10,11,12,13 22,0,0,0,0,0,0...   \n",
       "4003              0.088980  0,1,2,3,4,5,6,7,8,9 28,0,0,0,0,0,0,0,0,0 9,0,3...   \n",
       "4004              0.172399  0,1,2,3,4,5,6,7,8 34,0,0,0,0,0,0,0,0 27,10,12,...   \n",
       "4005              0.088980  0,1,2,3,4,5,6,7,8,9,10 40,0,1,0,0,0,0,0,0,0,0 ...   \n",
       "4006              0.127909  0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...   \n",
       "\n",
       "       log2_FC  log2_FC_mut  test_state        p_value            FDR  \\\n",
       "0     7.415426     8.434072      TestOK  9.573874e-183  1.536847e-177   \n",
       "1     7.070792     8.407614      TestOK  2.632264e-166  4.829081e-161   \n",
       "2     4.442013     4.486517      TestOK   4.391279e-12   2.014030e-07   \n",
       "3     6.447891     8.652952      TestOK  2.915445e-214  5.348597e-209   \n",
       "4     6.521031     8.533426      TestOK  3.072057e-182  6.575230e-177   \n",
       "...        ...          ...         ...            ...            ...   \n",
       "4002  3.245836     5.533199      TestOK   3.108715e-19   8.180768e-16   \n",
       "4003  0.971913     3.725844      TestOK   3.269560e-05   7.913254e-03   \n",
       "4004  2.135412     4.680040      TestOK   2.985607e-10   2.539152e-07   \n",
       "4005  1.135412     5.135412      TestOK   3.269560e-05   7.913254e-03   \n",
       "4006  2.015117     4.249406      TestOK   1.635736e-07   7.739919e-05   \n",
       "\n",
       "                                               bed_name  strand  \n",
       "0             chr1_631391_631393_highest_chr1_631392_CT       .  \n",
       "1             chr1_631391_631393_highest_chr1_631392_CT       .  \n",
       "2             chr1_631391_631393_highest_chr1_631392_CT       .  \n",
       "3             chr1_631778_631791_highest_chr1_631791_CT       .  \n",
       "4             chr1_631778_631791_highest_chr1_631787_CT       .  \n",
       "...                                                 ...     ...  \n",
       "4002    chrX_41024494_41024536_highest_chrX_41024528_GA       .  \n",
       "4003    chrX_72126743_72126762_highest_chrX_72126750_CT       .  \n",
       "4004  chrX_103211187_103211215_highest_chrX_10321121...       .  \n",
       "4005  chrX_132292956_132292983_highest_chrX_13229298...       .  \n",
       "4006  chrX_133005157_133005236_highest_chrX_13300523...       .  \n",
       "\n",
       "[4007 rows x 30 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bed_fixed_coordinate2 = df_bed_fixed_coordinate.merge(df, on=['mpmat_index'], how='left').drop(columns=['chr_name_y', 'region_start_y', 'region_end_y']).rename(\n",
    "    columns={\n",
    "        'chr_name_x': 'chr_name', \n",
    "        'region_start_x': 'region_start', \n",
    "        'region_end_x': 'region_end'\n",
    "    }\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "sample_names = df_bed_fixed_coordinate2.pop('<sample>')\n",
    "mpmat_indexes = df_bed_fixed_coordinate2.pop('mpmat_index')\n",
    "#利用insert方法插入取出的数据列到指定位置\n",
    "\n",
    "df_bed_fixed_coordinate2.insert(0, '<sample>', sample_names)\n",
    "df_bed_fixed_coordinate2.insert(4, 'mpmat_index', mpmat_indexes)\n",
    "\n",
    "# fix mpmat_index\n",
    "df_bed_fixed_coordinate2['mpmat_index'] = df_bed_fixed_coordinate2['chr_name'] + '_' + df_bed_fixed_coordinate2['region_start'].astype(str) + '_' + df_bed_fixed_coordinate2['region_end'].astype(str)\n",
    "\n",
    "df_bed_fixed_coordinate2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5d1c58e8-7752-4e47-bc8c-5afbeeda04f0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:01.245383Z",
     "iopub.status.busy": "2022-10-28T07:34:01.245017Z",
     "iopub.status.idle": "2022-10-28T07:34:01.264033Z",
     "shell.execute_reply": "2022-10-28T07:34:01.263338Z",
     "shell.execute_reply.started": "2022-10-28T07:34:01.245348Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>&lt;sample&gt;</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>mpmat_index</th>\n",
       "      <th>region_site_num</th>\n",
       "      <th>region_block_site_num</th>\n",
       "      <th>region_mut_site_num</th>\n",
       "      <th>region_site_index</th>\n",
       "      <th>region_block_state</th>\n",
       "      <th>...</th>\n",
       "      <th>ctrl_mut_count.norm</th>\n",
       "      <th>treat_mut_count.norm</th>\n",
       "      <th>count_info</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "      <th>test_state</th>\n",
       "      <th>p_value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.801598</td>\n",
       "      <td>0,1,2,3 22,0,0,0 689,406,500,0</td>\n",
       "      <td>7.415426</td>\n",
       "      <td>8.434072</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>9.573874e-183</td>\n",
       "      <td>1.536847e-177</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.750688</td>\n",
       "      <td>0,1,2,3 22,0,0,0 511,169,485,0</td>\n",
       "      <td>7.070792</td>\n",
       "      <td>8.407614</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.632264e-166</td>\n",
       "      <td>4.829081e-161</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DetectSeq_ATP8-DddAwt_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1_631391_CT,chr1_631392_CT,chr1_631393_CT</td>\n",
       "      <td>B-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.181582</td>\n",
       "      <td>0,1,2,3 22,0,0,0 157,60,43,0</td>\n",
       "      <td>4.442013</td>\n",
       "      <td>4.486517</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>4.391279e-12</td>\n",
       "      <td>2.014030e-07</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DetectSeq_ATP8-DddA11_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631778_CT,chr1_631787_CT,chr1_631791_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.260583</td>\n",
       "      <td>0,1,2,3 40,0,0,0 534,278,413,258</td>\n",
       "      <td>6.447891</td>\n",
       "      <td>8.652952</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>2.915445e-214</td>\n",
       "      <td>5.348597e-209</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631791_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DetectSeq_ATP8-DddA6_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1_631778_CT,chr1_631787_CT,chr1_631791_CT</td>\n",
       "      <td>N-N-N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.001335</td>\n",
       "      <td>0,1,2,3 40,0,0,0 530,320,528,69</td>\n",
       "      <td>6.521031</td>\n",
       "      <td>8.533426</td>\n",
       "      <td>TestOK</td>\n",
       "      <td>3.072057e-182</td>\n",
       "      <td>6.575230e-177</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631787_CT</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      <sample> chr_name  region_start  region_end  \\\n",
       "0  DetectSeq_ATP8-DddA11_REP-1     chr1        631391      631393   \n",
       "1   DetectSeq_ATP8-DddA6_REP-1     chr1        631391      631393   \n",
       "2  DetectSeq_ATP8-DddAwt_REP-1     chr1        631391      631393   \n",
       "3  DetectSeq_ATP8-DddA11_REP-1     chr1        631778      631791   \n",
       "4   DetectSeq_ATP8-DddA6_REP-1     chr1        631778      631791   \n",
       "\n",
       "          mpmat_index  region_site_num  region_block_site_num  \\\n",
       "0  chr1_631391_631393                3                      1   \n",
       "1  chr1_631391_631393                3                      1   \n",
       "2  chr1_631391_631393                3                      1   \n",
       "3  chr1_631778_631791                3                      0   \n",
       "4  chr1_631778_631791                3                      0   \n",
       "\n",
       "   region_mut_site_num                             region_site_index  \\\n",
       "0                    2  chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "1                    2  chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "2                    2  chr1_631391_CT,chr1_631392_CT,chr1_631393_CT   \n",
       "3                    3  chr1_631778_CT,chr1_631787_CT,chr1_631791_CT   \n",
       "4                    3  chr1_631778_CT,chr1_631787_CT,chr1_631791_CT   \n",
       "\n",
       "  region_block_state  ... ctrl_mut_count.norm  treat_mut_count.norm  \\\n",
       "0              B-N-N  ...                 0.0              2.801598   \n",
       "1              B-N-N  ...                 0.0              2.750688   \n",
       "2              B-N-N  ...                 0.0              0.181582   \n",
       "3              N-N-N  ...                 0.0              3.260583   \n",
       "4              N-N-N  ...                 0.0              3.001335   \n",
       "\n",
       "                         count_info   log2_FC  log2_FC_mut  test_state  \\\n",
       "0    0,1,2,3 22,0,0,0 689,406,500,0  7.415426     8.434072      TestOK   \n",
       "1    0,1,2,3 22,0,0,0 511,169,485,0  7.070792     8.407614      TestOK   \n",
       "2      0,1,2,3 22,0,0,0 157,60,43,0  4.442013     4.486517      TestOK   \n",
       "3  0,1,2,3 40,0,0,0 534,278,413,258  6.447891     8.652952      TestOK   \n",
       "4   0,1,2,3 40,0,0,0 530,320,528,69  6.521031     8.533426      TestOK   \n",
       "\n",
       "         p_value            FDR                                   bed_name  \\\n",
       "0  9.573874e-183  1.536847e-177  chr1_631391_631393_highest_chr1_631392_CT   \n",
       "1  2.632264e-166  4.829081e-161  chr1_631391_631393_highest_chr1_631392_CT   \n",
       "2   4.391279e-12   2.014030e-07  chr1_631391_631393_highest_chr1_631392_CT   \n",
       "3  2.915445e-214  5.348597e-209  chr1_631778_631791_highest_chr1_631791_CT   \n",
       "4  3.072057e-182  6.575230e-177  chr1_631778_631791_highest_chr1_631787_CT   \n",
       "\n",
       "   strand  \n",
       "0       .  \n",
       "1       .  \n",
       "2       .  \n",
       "3       .  \n",
       "4       .  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pois = df_bed_fixed_coordinate2.copy()\n",
    "df_pois.head()\n",
    "\n",
    "# 现在只要是有 nature 当中的 region，都和 nature 的 coordinate 和mpmat_index保持一致了，后面如果回溯命名，直接和 v4 的 table merge一下即可"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "552f2639-ffda-4bda-ad5b-551a4af539f0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:02.876685Z",
     "iopub.status.busy": "2022-10-28T07:34:02.876408Z",
     "iopub.status.idle": "2022-10-28T07:34:02.887450Z",
     "shell.execute_reply": "2022-10-28T07:34:02.886898Z",
     "shell.execute_reply.started": "2022-10-28T07:34:02.876664Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pois.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9171c757-054d-417b-838f-3747b7dfe1d6",
   "metadata": {
    "tags": []
   },
   "source": [
    "# IGV 截图脚本"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5b1364e4-cf2f-4c83-9e7b-352eae6592a4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:25.543555Z",
     "iopub.status.busy": "2022-10-28T07:34:25.543272Z",
     "iopub.status.idle": "2022-10-28T07:34:25.552357Z",
     "shell.execute_reply": "2022-10-28T07:34:25.551780Z",
     "shell.execute_reply.started": "2022-10-28T07:34:25.543536Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>8.434072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>8.407614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>4.486517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631791_CT</td>\n",
       "      <td>8.652952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631787_CT</td>\n",
       "      <td>8.533426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_name  region_start  region_end  \\\n",
       "0     chr1        631391      631393   \n",
       "1     chr1        631391      631393   \n",
       "2     chr1        631391      631393   \n",
       "3     chr1        631778      631791   \n",
       "4     chr1        631778      631791   \n",
       "\n",
       "                                    bed_name  log2_FC_mut  \n",
       "0  chr1_631391_631393_highest_chr1_631392_CT     8.434072  \n",
       "1  chr1_631391_631393_highest_chr1_631392_CT     8.407614  \n",
       "2  chr1_631391_631393_highest_chr1_631392_CT     4.486517  \n",
       "3  chr1_631778_631791_highest_chr1_631791_CT     8.652952  \n",
       "4  chr1_631778_631791_highest_chr1_631787_CT     8.533426  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_igv = df_pois[['chr_name', 'region_start', 'region_end', 'bed_name', 'log2_FC_mut']]\n",
    "df_igv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b62590f5-903f-40ee-a15a-8dc726fa420e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:27.432316Z",
     "iopub.status.busy": "2022-10-28T07:34:27.432060Z",
     "iopub.status.idle": "2022-10-28T07:34:27.434859Z",
     "shell.execute_reply": "2022-10-28T07:34:27.434320Z",
     "shell.execute_reply.started": "2022-10-28T07:34:27.432298Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 理论上这里得到的是所有 samples 的 candidate list\n",
    "# 经过先 merge mpmat 的处理，再 call 点，应该不存在能 overlap 到一起的 region，只能是样本间 share 相同位置或者 not share，不存在 overlap 又不相同的情况了"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8efbe775-5ee6-4101-96ba-3855c529b97b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:29.243123Z",
     "iopub.status.busy": "2022-10-28T07:34:29.242811Z",
     "iopub.status.idle": "2022-10-28T07:34:29.250083Z",
     "shell.execute_reply": "2022-10-28T07:34:29.249571Z",
     "shell.execute_reply.started": "2022-10-28T07:34:29.243103Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>8.434072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>8.407614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631391</td>\n",
       "      <td>631393</td>\n",
       "      <td>chr1_631391_631393_highest_chr1_631392_CT</td>\n",
       "      <td>4.486517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631791_CT</td>\n",
       "      <td>8.652952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>631778</td>\n",
       "      <td>631791</td>\n",
       "      <td>chr1_631778_631791_highest_chr1_631787_CT</td>\n",
       "      <td>8.533426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_name  region_start  region_end  \\\n",
       "0     chr1        631391      631393   \n",
       "1     chr1        631391      631393   \n",
       "2     chr1        631391      631393   \n",
       "3     chr1        631778      631791   \n",
       "4     chr1        631778      631791   \n",
       "\n",
       "                                    bed_name  log2_FC_mut  \n",
       "0  chr1_631391_631393_highest_chr1_631392_CT     8.434072  \n",
       "1  chr1_631391_631393_highest_chr1_631392_CT     8.407614  \n",
       "2  chr1_631391_631393_highest_chr1_631392_CT     4.486517  \n",
       "3  chr1_631778_631791_highest_chr1_631791_CT     8.652952  \n",
       "4  chr1_631778_631791_highest_chr1_631787_CT     8.533426  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_igv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "516cb100-b792-49da-af69-5826e8732052",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:31.971447Z",
     "iopub.status.busy": "2022-10-28T07:34:31.970690Z",
     "iopub.status.idle": "2022-10-28T07:34:31.981870Z",
     "shell.execute_reply": "2022-10-28T07:34:31.981234Z",
     "shell.execute_reply.started": "2022-10-28T07:34:31.971427Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_name</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>bed_name</th>\n",
       "      <th>log2_FC_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3649</th>\n",
       "      <td>chr9</td>\n",
       "      <td>98776015</td>\n",
       "      <td>98776036</td>\n",
       "      <td>chr9_98776018_98776031_highest_chr9_98776018_CT</td>\n",
       "      <td>5.758135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3705</th>\n",
       "      <td>chr11</td>\n",
       "      <td>72434248</td>\n",
       "      <td>72434259</td>\n",
       "      <td>chr11_72434248_72434264_highest_chr11_72434248_CT</td>\n",
       "      <td>5.084848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091</th>\n",
       "      <td>chr15</td>\n",
       "      <td>89579202</td>\n",
       "      <td>89579271</td>\n",
       "      <td>chr15_89579202_89579271_highest_chr15_89579202_CT</td>\n",
       "      <td>7.089608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>chr19</td>\n",
       "      <td>44552466</td>\n",
       "      <td>44552502</td>\n",
       "      <td>chr19_44552466_44552502_highest_chr19_44552502_GA</td>\n",
       "      <td>6.313897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3565</th>\n",
       "      <td>chr6</td>\n",
       "      <td>40824476</td>\n",
       "      <td>40824499</td>\n",
       "      <td>chr6_40824472_40824490_highest_chr6_40824478_CT</td>\n",
       "      <td>5.906919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1036</th>\n",
       "      <td>chr13</td>\n",
       "      <td>42969744</td>\n",
       "      <td>42969778</td>\n",
       "      <td>chr13_42969744_42969778_highest_chr13_42969778_GA</td>\n",
       "      <td>4.448647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>chr22</td>\n",
       "      <td>13961600</td>\n",
       "      <td>13961640</td>\n",
       "      <td>chr22_13961600_13961640_highest_chr22_13961608_CT</td>\n",
       "      <td>5.363628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>chr5</td>\n",
       "      <td>134924072</td>\n",
       "      <td>134924111</td>\n",
       "      <td>chr5_134924072_134924111_highest_chr5_13492410...</td>\n",
       "      <td>5.942922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2970</th>\n",
       "      <td>chr18</td>\n",
       "      <td>40654461</td>\n",
       "      <td>40654519</td>\n",
       "      <td>chr18_40654461_40654519_highest_chr18_40654463_CT</td>\n",
       "      <td>4.243971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2587</th>\n",
       "      <td>chr12</td>\n",
       "      <td>16621227</td>\n",
       "      <td>16621312</td>\n",
       "      <td>chr12_16621227_16621312_highest_chr12_16621239_CT</td>\n",
       "      <td>5.720374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr_name  region_start  region_end  \\\n",
       "3649     chr9      98776015    98776036   \n",
       "3705    chr11      72434248    72434259   \n",
       "1091    chr15      89579202    89579271   \n",
       "1142    chr19      44552466    44552502   \n",
       "3565     chr6      40824476    40824499   \n",
       "...       ...           ...         ...   \n",
       "1036    chr13      42969744    42969778   \n",
       "210     chr22      13961600    13961640   \n",
       "80       chr5     134924072   134924111   \n",
       "2970    chr18      40654461    40654519   \n",
       "2587    chr12      16621227    16621312   \n",
       "\n",
       "                                               bed_name  log2_FC_mut  \n",
       "3649    chr9_98776018_98776031_highest_chr9_98776018_CT     5.758135  \n",
       "3705  chr11_72434248_72434264_highest_chr11_72434248_CT     5.084848  \n",
       "1091  chr15_89579202_89579271_highest_chr15_89579202_CT     7.089608  \n",
       "1142  chr19_44552466_44552502_highest_chr19_44552502_GA     6.313897  \n",
       "3565    chr6_40824472_40824490_highest_chr6_40824478_CT     5.906919  \n",
       "...                                                 ...          ...  \n",
       "1036  chr13_42969744_42969778_highest_chr13_42969778_GA     4.448647  \n",
       "210   chr22_13961600_13961640_highest_chr22_13961608_CT     5.363628  \n",
       "80    chr5_134924072_134924111_highest_chr5_13492410...     5.942922  \n",
       "2970  chr18_40654461_40654519_highest_chr18_40654463_CT     4.243971  \n",
       "2587  chr12_16621227_16621312_highest_chr12_16621239_CT     5.720374  \n",
       "\n",
       "[300 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 对call 到的所有点进行随机采样\n",
    "np.random.seed = 2022\n",
    "df_igv = df_igv.sample(n=300)\n",
    "df_igv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b9c312b1-da76-48a0-9167-01a3857aaeb2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:34:54.899291Z",
     "iopub.status.busy": "2022-10-28T07:34:54.898968Z",
     "iopub.status.idle": "2022-10-28T07:34:54.905684Z",
     "shell.execute_reply": "2022-10-28T07:34:54.905101Z",
     "shell.execute_reply.started": "2022-10-28T07:34:54.899271Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chr_name  region_start  region_end  \\\n",
      "3649     chr9      98776015    98776036   \n",
      "3705    chr11      72434248    72434259   \n",
      "1091    chr15      89579202    89579271   \n",
      "1142    chr19      44552466    44552502   \n",
      "3565     chr6      40824476    40824499   \n",
      "...       ...           ...         ...   \n",
      "1036    chr13      42969744    42969778   \n",
      "210     chr22      13961600    13961640   \n",
      "80       chr5     134924072   134924111   \n",
      "2970    chr18      40654461    40654519   \n",
      "2587    chr12      16621227    16621312   \n",
      "\n",
      "                                               bed_name  log2_FC_mut  \n",
      "3649    chr9_98776018_98776031_highest_chr9_98776018_CT     5.758135  \n",
      "3705  chr11_72434248_72434264_highest_chr11_72434248_CT     5.084848  \n",
      "1091  chr15_89579202_89579271_highest_chr15_89579202_CT     7.089608  \n",
      "1142  chr19_44552466_44552502_highest_chr19_44552502_GA     6.313897  \n",
      "3565    chr6_40824472_40824490_highest_chr6_40824478_CT     5.906919  \n",
      "...                                                 ...          ...  \n",
      "1036  chr13_42969744_42969778_highest_chr13_42969778_GA     4.448647  \n",
      "210   chr22_13961600_13961640_highest_chr22_13961608_CT     5.363628  \n",
      "80    chr5_134924072_134924111_highest_chr5_13492410...     5.942922  \n",
      "2970  chr18_40654461_40654519_highest_chr18_40654463_CT     4.243971  \n",
      "2587  chr12_16621227_16621312_highest_chr12_16621239_CT     5.720374  \n",
      "\n",
      "[300 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "# 整理df格式为bed文件格式\n",
    "print(df_igv)\n",
    "#        0          1          2\n",
    "# 0   chr5   69093805   69093830\n",
    "# 1   chr8   37153384   37153424\n",
    "# 2  chr15   57559994   57560017\n",
    "# 3  chr15   68651256   68651277\n",
    "# 4  chr10  119445511  119445546\n",
    "# 5   chr8   20184990   20185028\n",
    "# 6  chr19   45187694   45187712\n",
    "# 7  chr15   81265992   81266016\n",
    "# 8   chr2  201232409  201232430\n",
    "# 9   chr9   98034893   98034930"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e590a105-62bd-4e7a-b775-3f69dc8cc0a6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:35:07.041255Z",
     "iopub.status.busy": "2022-10-28T07:35:07.041000Z",
     "iopub.status.idle": "2022-10-28T07:35:07.045189Z",
     "shell.execute_reply": "2022-10-28T07:35:07.044561Z",
     "shell.execute_reply.started": "2022-10-28T07:35:07.041236Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "maxPanelHeight 1500\n",
      "snapshotDirectory /Volumes/zhaohn_HD/3.project/2022_DdCBE-3D-Genome_topic/2022-09-30_Detect-seq_batch-1/igv/off-targets_20221027\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 填写相关信息\n",
    "\n",
    "path_out = '/Volumes/zhaohn_HD/3.project/2022_DdCBE-3D-Genome_topic/2022-09-30_Detect-seq_batch-1/igv'\n",
    "date = 20221027\n",
    "format_ = \"png\"\n",
    "height = 1500\n",
    "\n",
    "# 格式化脚本\n",
    "text = f\"maxPanelHeight {height}\\nsnapshotDirectory {path_out}/off-targets_{date}\\n\"\n",
    "print(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e2503d2e-244a-4114-9fa9-07f795627641",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:35:15.448465Z",
     "iopub.status.busy": "2022-10-28T07:35:15.448217Z",
     "iopub.status.idle": "2022-10-28T07:35:15.466297Z",
     "shell.execute_reply": "2022-10-28T07:35:15.465813Z",
     "shell.execute_reply.started": "2022-10-28T07:35:15.448446Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "maxPanelHeight 1500\n",
      "snapshotDirectory /Volumes/zhaohn_HD/3.project/2022_DdCBE-3D-Genome_topic/2022-09-30_Detect-seq_batch-1/igv/off-targets_20221027\n",
      "goto chr9:98775925-98776125\n",
      "snapshot 5.758135307099849_chr9_98776018_98776031_highest_chr9_98776018_CT.snapshot.png\n",
      "goto chr11:72434153-72434353\n",
      "snapshot 5.084847541736642_chr11_72434248_72434264_highest_chr11_72434248_CT.snapshot.png\n",
      "goto chr15:89579136-89579336\n",
      "snapshot 7.089607954133245_chr15_89579202_89579271_highest_chr15_89579202_CT.snapshot.png\n",
      "goto chr19:44552384-44552584\n",
      "snapshot 6.313897020377333_chr19_44552466_44552502_highest_chr19_44552502_GA.snapshot.png\n",
      "goto chr6:40824387-40824587\n",
      "snapshot 5.9069188586267325_chr6_40824472_40824490_highest_chr6_40824478_CT.snapshot.png\n",
      "goto chr5:47749710-47749910\n",
      "snapshot 4.275638913448643_chr5_47749810_47749811_highest_chr5_47749810_GA.snapshot.png\n",
      "goto chrX:136304519-136304719\n",
      "snapshot 4.118161092089268_chrX_136304603_136304635_highest_chrX_136304616_CT.snapshot.png\n",
      "goto chr8:99549939-99550\n"
     ]
    }
   ],
   "source": [
    "df_snapshot = df_igv.iloc[:, 0:5]\n",
    "\n",
    "for index, row_info in df_snapshot.iterrows():\n",
    "    chrom, start, stop, bed_name, score = row_info\n",
    "\n",
    "    path_out_png = f'{score}_{bed_name}.snapshot.{format_}'\n",
    "    middle = int((start + stop) / 2)\n",
    "\n",
    "    text += f\"goto {chrom}:{middle - 100}-{middle + 100}\\nsnapshot {path_out_png}\\n\"\n",
    "print(text[:1000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6a5415c2-0226-49e4-8c86-50922aec8d49",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-28T07:35:16.937048Z",
     "iopub.status.busy": "2022-10-28T07:35:16.936786Z",
     "iopub.status.idle": "2022-10-28T07:35:16.941121Z",
     "shell.execute_reply": "2022-10-28T07:35:16.940409Z",
     "shell.execute_reply.started": "2022-10-28T07:35:16.937028Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "with open(f'{path_out}/off-targets_{date}_snapshot.igv_shot_script', 'wt') as f:\n",
    "    f.write(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "619960a3",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2022-10-28T07:35:29.529449Z",
     "iopub.status.busy": "2022-10-28T07:35:29.529122Z",
     "iopub.status.idle": "2022-10-28T07:35:29.532155Z",
     "shell.execute_reply": "2022-10-28T07:35:29.531613Z",
     "shell.execute_reply.started": "2022-10-28T07:35:29.529429Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# DetectSeq_JAK2\n",
    "# DetectSeq_SIRT6\n",
    "# IND share?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af136a7-2450-44e6-915a-6eb635fe76c1",
   "metadata": {},
   "source": [
    "# [TODO] share analysis of different DdCBE treatment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fa2f882-f27e-41a3-8027-19be52de15dd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 找到每个样本中 call 到 region 的mpmat_index\n",
    "dt_mpmat_index = {}\n",
    "\n",
    "for sample, _df in df_sign_lenient.groupby('<sample>'):\n",
    "    dt_mpmat_index[sample] = _df.mpmat_index.tolist()\n",
    "\n",
    "    \n",
    "# dt_mpmat_index\n",
    "dt_mpmat_index['DetectSeq_ATP8-DddA11_REP-1'][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fc40f87-81be-4345-a60e-c582957e9ee1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# TODO\n",
    "def check_mpmat_index_in_others(x):\n",
    "    dt_mpmat_index_isin = dict()\n",
    "\n",
    "    for sample, sample_mpmat_index in dt_mpmat_index.items():\n",
    "        if x['mpmat_index'] in sample_mpmat_index:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e15b4a3b-0857-4c10-8831-640728cfc840",
   "metadata": {},
   "source": [
    "# analysis with old TAS-independent list (650 sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2261ff8-5150-4993-849d-0efe514f9e76",
   "metadata": {},
   "source": [
    "## generate old TAS-independent off-target mpmat to call regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "613f27e9-52a2-4d35-a3c2-e83b1ea6587f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_old_final_list = pd.read_csv(\n",
    "    \"../tables/20220312-DdCBE-off_target_type.FinallistV4.CheckPrimer.AddV4ID.tsv\",\n",
    "    sep='\\t',\n",
    "    header=0,\n",
    "    usecols=['region_id', 'off_target_id.V4.ND4', 'off_target_id.V4.ND5.1', 'off_target_id.V4.ND6'],\n",
    ")\n",
    "df_old_final_list.columns = ['mpmat_index', 'id_ND4', 'id_ND5.1', 'id_ND6']\n",
    "df_old_final_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1795a1cc-3fcf-4bfa-9d47-c873da3e8138",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def query_ind(x):\n",
    "    x = x.fillna('')\n",
    "    if 'IND' in x['id_ND4']: return True\n",
    "    elif 'IND' in x['id_ND5.1']: return True\n",
    "    elif 'IND' in x['id_ND6']: return True\n",
    "    else: return False\n",
    "\n",
    "df_old_final_list.apply(query_ind, axis=1).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d160c059-b463-4f2a-a1ef-002732b9dfa8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_old_share = df_old_final_list[df_old_final_list.apply(query_ind, axis=1)].reset_index(drop=True)\n",
    "df_old_share"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1e6097-5822-44ae-9dec-c7c72aee1e35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 拿 650 个 old share 的点做一个 mpmat 出来去对着这五个 sample call 一下相关位置\n",
    "df_mpmat = pd.read_csv(\n",
    "    '../tables/293T-DdCBE-merge_hg38.MAPQ20.C6_M4_R1_T10.sort.V4.mpmat.gz',\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    ")\n",
    "df_mpmat['mpmat_index'] = df_mpmat[0] + '_' + df_mpmat[1].astype(str) + '_' + df_mpmat[2].astype(str)\n",
    "df_mpmat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a41a3dd-bab6-493f-9cd7-d7d709bf1b6c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_merge = pd.merge(left=df_mpmat, right=df_old_share, on='mpmat_index', how='right')\n",
    "\n",
    "df_aim_mpmat = df_merge.iloc[:,:-4]\n",
    "df_aim_mpmat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4f7dbc2-9619-4dad-a39e-08c25036742a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_aim_mpmat.to_csv('../mpmat/2022-10-21_nature_4_5.1_6_share_650-off-targets.mpmat', header=None, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5aab32b2-bf6b-44ad-a38e-ecbc9fb9fa31",
   "metadata": {},
   "source": [
    "## Detect-seq signal comparation for DddA\\_wt/6/11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d5f2bdb-e1ff-44b3-9900-64be6a9fc2f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 650 mpmat to get 650 poisson_res\n",
    "df_old_site_new_signal = pd.read_csv(\n",
    "    '../poisson_res_use_old650/poisson_res_all_use650.tsv.gz',\n",
    "    header=0, \n",
    "    index_col=None, \n",
    "    sep='\\t'\n",
    ")\n",
    "df_old_site_new_signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3d166bf-bfcb-45b9-a996-0c748ab32203",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df_old_site_new_signal.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ec1531-4398-40fe-88bd-be0eeabc8675",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_old_site_new_signal.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e57c3672-6095-45b6-946b-ba63b4c1eb54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_old_site_new_signal = df_old_site_new_signal.assign(\n",
    "    bed_name=df_old_site_new_signal.mpmat_index + '_highest_' + df_old_site_new_signal.region_highest_site_index,\n",
    "    strand='.'\n",
    ")\n",
    "df_old_site_new_signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4cbfb36-a9fb-44c0-a0ee-6057733fd9b5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_sign_strict = (df_old_site_new_signal\n",
    "                  .query('FDR <= 0.0001')  # 10911\n",
    "                  .query('log2_FC_mut >= 2')  # 10911\n",
    "                  .query('ctrl_mut_count <= 1')  # 9551\n",
    "                  .query('`treat_mut_count.norm` * 100 >= 10')  # 9551\n",
    "                  .query('treat_mut_count >= 20')  # 9551\n",
    "                  .query('treat_mut_count / treat_count >= 0.15')  # 6080\n",
    "                  .query('region_block_site_num <= 1')  # 5720\n",
    "                  .query('region_highest_site_mut_ratio >= 0.35')  # 2929\n",
    "                  )\n",
    "\n",
    "print(df_sign_strict.shape[0])\n",
    "df_sign_strict.groupby('<sample>').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2cffc45-6f3a-41f8-a08c-b053caeab06a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_sign_lenient = (df_old_site_new_signal\n",
    "                   .query('FDR <= 0.01')  # 25118\n",
    "                   .query('log2_FC_mut >= 2')  # 25118\n",
    "                   .query('ctrl_mut_count <= 1')  # 22009\n",
    "                   .query('`treat_mut_count.norm` * 100 >= 5')  # 22009\n",
    "                   .query('treat_mut_count >= 10')  # 22009\n",
    "                   .query('treat_mut_count / treat_count >= 0.15')  # 14476\n",
    "                   .query('region_block_site_num <= 1')  # 13522\n",
    "                   .query('region_highest_site_mut_ratio >= 0.30')  # 7001\n",
    "                   )\n",
    "\n",
    "print(df_sign_lenient.shape[0])\n",
    "df_sign_lenient.groupby('<sample>').describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b0fc442-f6ab-4c97-b03c-383c59f42b30",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_sign_lenient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "448d2e0c-84fe-4f4d-b5da-b09e30110689",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_sign_lenient.mpmat_index.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa5df120-f08d-4115-aeee-bcdf43524902",
   "metadata": {},
   "source": [
    "## 2022-10-21 Detect-seq signal comparation for different localization of DdCBE_DddA11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "032f8862-a603-4ef6-a7b8-501f35706ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data in R env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a35f942-ea08-49ef-a31b-b46f0b60654f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R -i df_sign_lenient\n",
    "\n",
    "# head(df_sign_lenient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c4f0104-bc4f-40c4-bb53-e970bea7e538",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "df = df_sign_lenient\n",
    "colnames(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39606851-ef26-4a1e-82f7-9b7602821321",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "df %>% select(mpmat_index, ctrl_count, treat_count) %>% head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "816e33be-6b68-486a-a269-f34b2cb8268b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "info.df.select_dist <- df %>% \n",
    "  select(sgRNA_name, distance_to_sapce, site_name, rep_name, genome_base, mut_percent, Detect.A, Detect.T, Detect.total)\n",
    "\n",
    "info.df.select_dist <- info.df.select_dist %>% filter(\n",
    "  genome_base==\"C\" & Detect.T >= 10 | genome_base==\"G\" & Detect.A >= 10\n",
    ")\n",
    "\n",
    "\n",
    "info.ratio.df <- info.df.select_dist %>% mutate(\n",
    "  target_seq.ratio = mut_percent,\n",
    "  detect_seq.score = ifelse(\n",
    "    genome_base==\"C\", \n",
    "    (Detect.T/Detect.total)^2 * Detect.T, \n",
    "    (Detect.A/Detect.total)^2 * Detect.A\n",
    "  )\n",
    "  ) %>% group_by(sgRNA_name, distance_to_sapce, site_name) %>% summarise(\n",
    "  target_seq.ratio.mean = mean(target_seq.ratio),\n",
    "  detect_seq.score.mean = mean(detect_seq.score)\n",
    ")\n",
    "\n",
    "df <- info.ratio.df\n",
    "df$target_seq.ratio.mean <- df$target_seq.ratio.mean / 100\n",
    "\n",
    "\n",
    "pearson_r = cor(log10(df$target_seq.ratio.mean), log10(df$detect_seq.score.mean), method = \"pearson\")\n",
    "# pvalue:\n",
    "# cor.test(log10(df$target_seq.ratio.mean), log10(df$detect_seq.score.mean), method = \"pearson\")\n",
    "info.ratio.df.select <- df\n",
    "# Merge TargetSeq\n",
    "ggplot(data = info.ratio.df.select) +\n",
    "  geom_point(mapping = aes(x = target_seq.ratio.mean, y=detect_seq.score.mean, label=site_name, colour = sgRNA_name)) +\n",
    "  geom_smooth(mapping = aes(x = target_seq.ratio.mean, y=detect_seq.score.mean), method = \"lm\") +\n",
    "  scale_color_manual(values=c(\"#457b9d\",\"#ff9f1c\",\"#e36414\")) +\n",
    "  scale_x_log10(\n",
    "    name=\"Mutation ratio by targeted deep sequencing\",\n",
    "    limits=c(0.0001,1),\n",
    "    breaks = c(0.0001, 0.001, 0.01, 0.1, 1),\n",
    "    labels = scales::percent,\n",
    "    ) +\n",
    "  scale_y_log10(\n",
    "    name=\"Detect-seq Score\",\n",
    "    limits=c(0.01, 1200),\n",
    "    breaks = c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "    labels = c(0.01, 0.1, 1, 10, 100, 1000),\n",
    "  ) +\n",
    "  ggtitle(\"Targeted deep sequencing v.s. Detect-seq\") +\n",
    "  theme_classic() + \n",
    "  # geom_abline(intercept = 0, slope = 1,linetype=\"dashed\",color=\"#333333\") +\n",
    "  theme(plot.title = element_text(hjust = 0.5)) +\n",
    "  annotate(\"text\",x=0.0002,y=500,hjust = 0,parse=TRUE,colour=\"black\", label=sprintf(\"r == %.3f\",pearson_r)) +\n",
    "  annotate(\"text\",x=0.0002,y=300,hjust = 0,parse=TRUE,colour=\"black\", label=\"p-value < 2.2e-16\")\n",
    "\n",
    "ggsave(filename = \"2021-10-14-293T-Merge-Target_seq__vs__Detect_seq.ScatterPlot.pdf\",width = 5.4,height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a5fe77f-53ae-4ecf-b2ce-452f628ac85f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73b099ea-d587-4d49-b223-042658c0ef4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2c7513fe-d7d2-441d-9388-edbace632cdf",
   "metadata": {},
   "source": [
    "## 2022-10-21 co-localization between all CTCF peaks and \u000b",
    "Detect-seq signals from different DdCBE treatments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70263df9",
   "metadata": {},
   "source": [
    "# off-target analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c6057e9",
   "metadata": {},
   "source": [
    "## off-target list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e0477dc",
   "metadata": {},
   "source": [
    "### circos plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9635b7b",
   "metadata": {},
   "source": [
    "#### learn circos plot\n",
    "- [tutorial1](https://colab.research.google.com/drive/1xmAnv7AHWUTA2HWfjqV1lFWkFMSLJHG0?usp=sharing)\n",
    "- [tutorial2](https://colab.research.google.com/drive/1RYSo4aXpDIZlSQ9EhO2kPCeF8FOwyvXv?usp=sharing)\n",
    "- [tutorial3](https://colab.research.google.com/drive/1EPxCQCgOouVxtXcGyxu2ZqQvfucVnOJ-?usp=sharing)\n",
    "- [tutorial4(Drawing pylogenetic tree)](https://colab.research.google.com/drive/140m2jpQpgSZwSlP-3u3Oj8IcJUbP2NGD?usp=sharing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec7308ac-b380-4ad3-aeea-b6562797d3b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# pip install python-circos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2ff1e4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# !mkdir -p../ temp_file\n",
    "# % cd../ temp_file\n",
    "# !mkdir -p pycircos\n",
    "# % cd pycircos\n",
    "# !mkdir -p sample_data\n",
    "# % cd sample_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "969a1dec",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# #The following example data was downloaded from https://venyao.xyz/shinyCircos/.\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_barplot.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_chromosome_cytoband.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_chromosome_general.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_links.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_point.csv\n",
    "# !wget https: // github.com / ponnhide / pyCircos-examples / raw / main / example_notebooks / sample_data / example_data_rect_gradual.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a9976e4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# % cd / Volumes / zhaohn_HD / 3.project / 2022_DdCBE-3D-Genome_topic / 2022-09-30_Detect-seq_batch-1 / snakepipes_detect-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d0f69e",
   "metadata": {},
   "source": [
    "#### circos plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6c604bb-54e3-4599-8dcb-ecbc88866c84",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-21T07:16:22.990392Z",
     "iopub.status.busy": "2022-10-21T07:16:22.990109Z",
     "iopub.status.idle": "2022-10-21T07:16:22.994832Z",
     "shell.execute_reply": "2022-10-21T07:16:22.993948Z",
     "shell.execute_reply.started": "2022-10-21T07:16:22.990374Z"
    }
   },
   "source": [
    "##### 准备 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14871568-295c-4bf8-ac25-86b88397d3d4",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 基因组 fai 文件（染色体长度信息），区分 hg19 和 hg38\n",
    "path_genome_length = '/Users/zhaohuanan/Bio/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa.fai'\n",
    "df_genome_length = pd.read_csv(\n",
    "    path_genome_length,\n",
    "    header=None,\n",
    "    sep='\\t',\n",
    "    usecols=[0, 1],\n",
    "    names=['chrom', 'end'],\n",
    ")\n",
    "df_genome_length.insert(loc=1, column='start', value=1)\n",
    "\n",
    "df_genome_length = df_genome_length\\\n",
    "    .query(\"chrom != 'chrY'\")\\\n",
    "    .query(\"chrom != 'chrM'\")\n",
    "\n",
    "print(df_genome_length.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4335f60c-d17c-4218-aa64-016fcd1ed106",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# cytoband 信息，可以在 ucsc 下载，区分 hg19 和 hg38\n",
    "df_genome_cytoband = pd.read_csv(\n",
    "    \"http://hgdownload.cse.ucsc.edu/goldenpath/hg38/database/cytoBand.txt.gz\",\n",
    "    header=None,\n",
    "    sep='\\t',\n",
    "    usecols=range(5),\n",
    "    names=['chrom', 'start', 'end', 'value1', 'value2']\n",
    ")\n",
    "# 对 chromosome 进行过滤，保留标准染色体\n",
    "df_genome_cytoband = df_genome_cytoband\\\n",
    "    .query(\"not chrom.str.endswith('fix')\")\\\n",
    "    .query(\"not chrom.str.endswith('alt')\")\\\n",
    "    .query(\"not chrom.str.endswith('random')\")\\\n",
    "    .query(\"not chrom.str.startswith('chrUn')\")\\\n",
    "    .query(\"chrom != 'chrY'\")\\\n",
    "    .query(\"chrom != 'chrM'\")\n",
    "print(sorted(df_genome_cytoband.chrom.unique()))\n",
    "print()\n",
    "print(df_genome_cytoband.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d7bf9d9-2e9f-4378-a4bf-bde136c0f717",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_circos_point = df_sign_lenient[['<sample>','chr_name', 'region_start', 'region_end', 'log2_FC_mut']].copy()\n",
    "df_circos_point.columns = ['sample', 'chrom', 'start', 'end', 'score']\n",
    "df_circos_point.reset_index(inplace=True, drop=True)\n",
    "df_circos_point"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11dafed5-60a8-468d-9b8a-e4d9b6cf923c",
   "metadata": {},
   "source": [
    "##### 设置比例尺"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66418fd2-90c1-49d8-aeb8-25b021e005f2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "scale = 1.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a7b3a9f-007e-4827-961b-3cb91b809c5a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "arc_range_i = df_circos_point['sample'].unique().__len__() + 1\n",
    "arc_raixs_range = (\n",
    "    (arc_range_i * 100.0) * scale, \n",
    "    (arc_range_i * 100.0 + 20) * scale\n",
    ") \n",
    "arc_raixs_range"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f266bf8-28df-45c2-bbeb-6e67349ea8a9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-21T06:17:24.104466Z",
     "iopub.status.busy": "2022-10-21T06:17:24.104224Z",
     "iopub.status.idle": "2022-10-21T06:17:24.111202Z",
     "shell.execute_reply": "2022-10-21T06:17:24.110435Z",
     "shell.execute_reply.started": "2022-10-21T06:17:24.104449Z"
    },
    "tags": []
   },
   "source": [
    "##### 设置染色体信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "875fc0e9",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# set chromeosomes\n",
    "circle = Gcircle()\n",
    "\n",
    "\n",
    "\n",
    "for idx, row in df_genome_length.iterrows():\n",
    "    chrom, start, end = row\n",
    "    length = end - start + 1\n",
    "    \n",
    "    arc = Garc(\n",
    "        arc_id=chrom,  # 染色体名称\n",
    "        size=length,  # 染色体长度\n",
    "        interspace=1,  # 间隔距离\n",
    "        raxis_range=arc_raixs_range,  # 内外半径长度\n",
    "        labelposition=60,  # 越往里越小，越往外越大\n",
    "        label_visible=True,  # 是否展示 label，也就是染色体名称\n",
    "    )\n",
    "    circle.add_garc(arc)  # 添加一个染色体信息\n",
    "\n",
    "circle.set_garcs()  # 整合设置所有添加过的染色体信息"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7bc261d-ec42-4c4b-833e-0568d23caeb2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-21T06:22:00.236412Z",
     "iopub.status.busy": "2022-10-21T06:22:00.236161Z",
     "iopub.status.idle": "2022-10-21T06:22:00.283305Z",
     "shell.execute_reply": "2022-10-21T06:22:00.282835Z",
     "shell.execute_reply.started": "2022-10-21T06:22:00.236395Z"
    }
   },
   "source": [
    "##### 整理 cytoband 信息到arcdata\\_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97912204-81f4-45c1-82e1-430dbbd6ab4f",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 定义不同 cytoband 的颜色\n",
    "color_dict = {\n",
    "    \"gneg\": \"#FFFFFF00\",\n",
    "    \"gpos25\": \"#EEEEEE\",\n",
    "    \"gpos50\": \"#BBBBBB\",\n",
    "    \"gpos75\": \"#777777\",\n",
    "    \"gpos100\": \"#000000\",\n",
    "    \"gvar\": \"#FFFFFF00\",\n",
    "    \"stalk\": \"#C01E27\",\n",
    "    \"acen\": \"#D82322\"\n",
    "}\n",
    "\n",
    "# defaultdict的说明\n",
    "# https://zhuanlan.zhihu.com/p/46476348\n",
    "# 其实就是当引用的 key 不存在时返回默认值,这里是默认返回一个 dict 而不抛出 KeyError\n",
    "arcdata_dict = collections.defaultdict(dict)  # 传入 dict 函数，调用时返回空 dict\n",
    "# or\n",
    "# arcdata_dict = collections.defaultdict(lambda: {})\n",
    "# arcdata_dict\n",
    "\n",
    "for idx, row in df_genome_cytoband.iterrows():\n",
    "    chrom, start, end, value1, value2 = row\n",
    "    width = end - start + 1\n",
    "    # 在进行arcdata_dict对键chrom 取值取不到的时候\n",
    "    # 默认创建空 dict 而不是 raise KeyError\n",
    "    if chrom not in arcdata_dict:\n",
    "        arcdata_dict[chrom]['positions'] = []\n",
    "        arcdata_dict[chrom]['widths'] = []\n",
    "        arcdata_dict[chrom]['colors'] = []\n",
    "    else:\n",
    "        arcdata_dict[chrom]['positions'].append(start)\n",
    "        arcdata_dict[chrom]['widths'].append(width)\n",
    "        arcdata_dict[chrom]['colors'].append(color_dict[value2])\n",
    "\n",
    "print(arcdata_dict.__str__()[:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98f18958-28f9-4051-83ec-ef2bd3c366e3",
   "metadata": {},
   "source": [
    "##### 将 cytoband 信息加到 circle 对象中去"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e6cdb7f-b523-4f3e-9221-1a002bd4477d",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for chrom in arcdata_dict:\n",
    "    circle.barplot(\n",
    "        chrom,\n",
    "        data=[1] * len(arcdata_dict[chrom]['positions']),\n",
    "        positions=arcdata_dict[chrom]['positions'],\n",
    "        width=arcdata_dict[chrom]['widths'],\n",
    "        raxis_range=arc_raixs_range,\n",
    "        facecolor=arcdata_dict[chrom]['colors']\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e852695c-ed96-4368-b40b-0732ff8d1e7b",
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "source": [
    "##### 查看绘制的circos plot骨架"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72dd14a0-7daa-4cd4-a695-3172329342b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# circle.figure?\n",
    "circle.figure"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d4d8318-5ad4-44cd-be3c-85de6fff653b",
   "metadata": {},
   "source": [
    "##### 添加每个样本中的 off-target sites 信息"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b392d24-6d83-4bdc-8d6b-5262bcb1c452",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# TODO\n",
    "# sample lable\n",
    "# background color\n",
    "# point color\n",
    "# edge color of point\n",
    "\n",
    "#scatter plot\n",
    "counter_circle = 0\n",
    "# color\n",
    "\n",
    "for sample, sample_df in df_circos_point.groupby('sample'):\n",
    "    print(sample, sample_df.shape[0])\n",
    "    counter_circle += 1\n",
    "    \n",
    "    values_all   = [] \n",
    "    arcdata_dict = collections.defaultdict(dict)\n",
    "    \n",
    "    for idx, row in sample_df.iterrows():\n",
    "        _, chrom, start, end, score = row\n",
    "        middle = (start + end) / 2\n",
    "        values_all.append(score)\n",
    "        \n",
    "        if chrom not in arcdata_dict:\n",
    "            arcdata_dict[chrom][\"positions\"] = []\n",
    "            arcdata_dict[chrom][\"values\"] = []\n",
    "        else:\n",
    "            arcdata_dict[chrom][\"positions\"].append(middle) \n",
    "            arcdata_dict[chrom][\"values\"].append(score)\n",
    "    \n",
    "    vmin, vmax = min(values_all), max(values_all)\n",
    "    \n",
    "    arc_raixs_range_sample = (\n",
    "        (counter_circle * 100.0) * scale, \n",
    "        (counter_circle * 100.0 + 80) * scale\n",
    "    ) \n",
    "\n",
    "    for chrom in arcdata_dict:\n",
    "        circle.scatterplot(\n",
    "            chrom, \n",
    "            data=arcdata_dict[chrom][\"values\"], \n",
    "            positions=arcdata_dict[chrom][\"positions\"], \n",
    "            rlim=[vmin - 0.05 * abs(vmin), vmax + 0.05 * abs(vmax)],\n",
    "            markershape='o',\n",
    "            markersize=1,\n",
    "            raxis_range=arc_raixs_range_sample, \n",
    "            facecolor=\"#468FBE\",\n",
    "            edgecolor=\"#000000\",\n",
    "            linewidth=0.03,\n",
    "            spine=True\n",
    "        )\n",
    "\n",
    "circle.figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe612234-5480-4ba7-80dd-cde9396898e9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "circle.figure?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a1f113e",
   "metadata": {},
   "source": [
    "\n",
    "### upset plot\n",
    "\n",
    "## signal strength\n",
    "\n",
    "### scatter plot\n",
    "\n",
    "## alignment\n",
    "### art plot\n",
    "\n",
    "## editing window\n",
    "### indel comparation\n",
    "\n",
    "## ctcf analysis\n",
    "### shared off-target motif\n",
    "### DddAwt,6,11 co-localization with ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48bd2502-2b5a-42df-a259-8c897aeca2af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f096e6e",
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
